Index: iproute2-2.6.33/Makefile
===================================================================
--- iproute2-2.6.33.orig/Makefile	2018-11-06 23:03:55.342477521 +0800
+++ iproute2-2.6.33/Makefile	2018-11-06 23:06:00.891395528 +0800
@@ -11,6 +11,8 @@
 else
 #KERNEL_INCLUDES:=$(KERNEL_DIR)/include $(KERNEL_DIR)/arch/$(ARCH)/include
 #LIBC_INCLUDE=$(KERNEL_DIR)/include/linux
+KERNEL_INCLUDES:=$(KERNEL_DIR)/user_headers/include
+LIBC_INCLUDE=$(KERNEL_DIR)/user_headers/include/linux
 endif
 
 # Path to db_185.h include
