Index: clinkc/src/cybergarage/http/chttp_request.c
===================================================================
--- clinkc.orig/src/cybergarage/http/chttp_request.c	2019-06-06 10:59:22.099396559 +0800
+++ clinkc/src/cybergarage/http/chttp_request.c	2019-06-06 10:52:09.445450700 +0800
@@ -1185,8 +1185,8 @@
 {
 	char *uri = NULL;
 	char *pch = NULL;
-	char tempStr[128];
-	int idx, i, check_ascii_0, check_ascii;
+	char tempStr[128] = {0};
+	int len, i, check_ascii_0, check_ascii;
 	BOOL bFail = FALSE;
 
 	if (!httpReq) {
@@ -1222,21 +1222,14 @@
 	}
 
 	/********* case 2 *********/
-	pch = strchr(uri, '?');	
-	while (pch != NULL){
-		idx = pch-uri+1;
-		//printf ("found at %d\n",idx);
-		tempStr[idx-1] = '\0';
-		strncpy(tempStr, uri , idx-1);
-		break;
-	}
-	//printf("Get tempStr string = %s \n", tempStr);	
-	for (i=0; tempStr[i]!='\0'; i++){
-		check_ascii = tempStr[i];
-		if (check_ascii == 61){ //checjk '='
-			//printf("%d: %c\n",i , tempStr[i]);
+	pch = strchr(uri, '?');  //check there is  no '=' before '?'
+	
+	if(pch != NULL){
+		len = pch-uri;
+		memset(tempStr, '\0', sizeof(tempStr));
+		strncpy(tempStr, uri , len);
+		if(strchr(tempStr, '=')){
 			bFail = TRUE;
-			break;
 		}
 	}
 	if (bFail == TRUE){ //return bFail;
