diff -urN clinkc.org/src/cybergarage/http/chttp_request.c clinkc/src/cybergarage/http/chttp_request.c
--- clinkc.org/src/cybergarage/http/chttp_request.c	2019-05-07 17:22:22.834286800 +0800
+++ clinkc/src/cybergarage/http/chttp_request.c	2019-05-08 10:43:48.846650880 +0800
@@ -796,7 +796,7 @@
 ****************************************/
 BOOL cg_http_request_read(CgHttpRequest *httpReq, CgSocket *sock)
 {
-	char lineBuf[CG_HTTP_READLINE_BUFSIZE];
+	char lineBuf[CG_HTTP_READLINE_BUFSIZE] = "";
 	CgStringTokenizer *strTok;
 	int readLen;
 	CgNetURI *uri = NULL;
diff -urN clinkc.org/src/cybergarage/http/chttp_server_list.c clinkc/src/cybergarage/http/chttp_server_list.c
--- clinkc.org/src/cybergarage/http/chttp_server_list.c	2019-05-07 17:22:22.814286800 +0800
+++ clinkc/src/cybergarage/http/chttp_server_list.c	2019-05-08 10:44:08.871945920 +0800
@@ -81,7 +81,7 @@
 {
 	CgNetworkInterfaceList *netIfList;
 	CgNetworkInterface *netIf;
-	CgHttpServer *httpServer;
+	CgHttpServer *httpServer = NULL;
 	char *bindAddr;
 	BOOL result =FALSE;
 	
