Index: clinkc/include/cybergarage/http/chttp_authentication.h
===================================================================
--- clinkc.orig/include/cybergarage/http/chttp_authentication.h	2019-12-12 10:37:05.197480098 +0800
+++ clinkc/include/cybergarage/http/chttp_authentication.h	2019-12-12 10:37:07.333480037 +0800
@@ -42,7 +42,10 @@
 #define CG_HTTP_ALGORITHM "algorithm"
 #define CG_HTTP_USERNAME "username"
 
-#define CG_HTTP_DEFAULT_OPAQUE "WnlYRUwuY29t"
+
+#define CG_HTTP_DEFAULT_OPAQUE "WllYRUwgb3V0cHV0IHRoZSBiZXN0IGhlbHA="
+
+
 #define CG_HTTP_DEFAULT_REALM "ZyXEL.com"
 #define CG_HTTP_DEFAULT_QOP "auth"
 #define CG_HTTP_DEFAULT_ALGORITHM "MD5"
Index: clinkc/src/cybergarage/http/chttp_authentication.c
===================================================================
--- clinkc.orig/src/cybergarage/http/chttp_authentication.c	2019-12-12 10:37:07.321480037 +0800
+++ clinkc/src/cybergarage/http/chttp_authentication.c	2019-12-12 10:37:07.333480037 +0800
@@ -811,11 +811,13 @@
 		if(now->opaque){
 			tmp 				= cg_http_request_getopaque(httpReq);
 			if(!tmp){
+#if 0
 				auth->opaque	= NULL;
 				printf("error: can not get opaque");
 				ZOS_FREE(auth);
 				cg_http_server_unlock(httpServer);
 				return HTTP_AUTH_ERROR;
+#endif
 			}
 			else{
 				auth->opaque	= tmp;
