diff -urN procd-2015-10-29.1_org/CMakeLists.txt procd-2015-10-29.1/CMakeLists.txt
--- procd-2015-10-29.1_org/CMakeLists.txt	2018-11-15 11:20:03.360719098 +0800
+++ procd-2015-10-29.1/CMakeLists.txt	2018-11-16 16:11:42.346035000 +0800
@@ -36,6 +36,14 @@
   add_subdirectory(upgraded)
 ENDIF()
 
+IF(SHOW_BOOT_ON_CONSOLE)
+  ADD_DEFINITIONS(-DSHOW_BOOT_ON_CONSOLE)
+ENDIF()
+
+IF(IPQ_807X)
+  ADD_DEFINITIONS(-DIPQ_807X)
+ENDIF()
+
 ADD_EXECUTABLE(procd ${SOURCES})
 TARGET_LINK_LIBRARIES(procd ${LIBS})
 INSTALL(TARGETS procd
diff -urN procd-2015-10-29.1_org/rcS.c procd-2015-10-29.1/rcS.c
--- procd-2015-10-29.1_org/rcS.c	2018-11-15 11:20:03.360719098 +0800
+++ procd-2015-10-29.1/rcS.c	2018-11-16 16:14:19.288454518 +0800
@@ -21,7 +21,9 @@
 #include <stdlib.h>
 #include <stdio.h>
 #include <unistd.h>
-
+#ifdef IPQ_807X
+#include <fcntl.h>
+#endif
 #include <sys/types.h>
 #include <sys/stat.h>
 #include <glob.h>
@@ -87,8 +89,22 @@
 		return;
 	}
 	close(pipefd[0]);
+#ifdef IPQ_807X
+	{
+		//direct stdout stderr to /dev/console
+		int fd = open("/dev/console", O_RDWR);
+
+		if (fd > -1) {
+			dup2(fd, STDOUT_FILENO);
+			dup2(fd, STDERR_FILENO);
+			if (fd > STDERR_FILENO)
+				close(fd);
+		}
+	}
+#else
 	dup2(pipefd[1], STDOUT_FILENO);
 	dup2(pipefd[1], STDERR_FILENO);
+#endif
 
 	execlp(s->file, s->file, s->param, NULL);
 	exit(1);
