Index: igmpproxy-0.2.1/src/config.c
===================================================================
--- igmpproxy-0.2.1.orig/src/config.c	2018-02-14 03:21:42.000000000 +0800
+++ igmpproxy-0.2.1/src/config.c	2019-05-02 09:11:56.808937950 +0800
@@ -78,7 +78,7 @@
     commonConfig.startupQueryCount = DEFAULT_ROBUSTNESS;
 
     // Default values for leave intervals...
-    commonConfig.lastMemberQueryInterval = INTERVAL_QUERY_RESPONSE;
+    commonConfig.lastMemberQueryInterval = LASTMEMBERQUERYINTERVAL;
     commonConfig.lastMemberQueryCount    = DEFAULT_ROBUSTNESS;
 
     // If 1, a leave message is sent upstream on leave messages from downstream.
Index: igmpproxy-0.2.1/src/igmpproxy.c
===================================================================
--- igmpproxy-0.2.1.orig/src/igmpproxy.c	2018-02-14 03:21:42.000000000 +0800
+++ igmpproxy-0.2.1/src/igmpproxy.c	2019-05-02 09:11:49.249350997 +0800
@@ -298,6 +298,7 @@
         } else {
             timeout->tv_nsec = 0;
             timeout->tv_sec = (secs > 3) ? 3 : secs; // aimwang: set max timeout
+			secs = timeout->tv_sec;
         }
 
         // Prepare for select.
Index: igmpproxy-0.2.1/src/igmpproxy.h
===================================================================
--- igmpproxy-0.2.1.orig/src/igmpproxy.h	2018-02-14 03:21:42.000000000 +0800
+++ igmpproxy-0.2.1/src/igmpproxy.h	2019-05-02 09:11:53.280484073 +0800
@@ -128,9 +128,9 @@
 #define DEFAULT_RATELIMIT      0
 
 // Define timer constants (in seconds...)
-#define INTERVAL_QUERY          125
+#define INTERVAL_QUERY          60
 #define INTERVAL_QUERY_RESPONSE  10
-//#define INTERVAL_QUERY_RESPONSE  10
+#define LASTMEMBERQUERYINTERVAL  2
 
 #define ROUTESTATE_NOTJOINED            0   // The group corresponding to route is not joined
 #define ROUTESTATE_JOINED               1   // The group corresponding to route is joined
