Index: igmpproxy-0.2.1/src/igmp.c
===================================================================
--- igmpproxy-0.2.1.orig/src/igmp.c	2018-02-14 03:21:42.000000000 +0800
+++ igmpproxy-0.2.1/src/igmp.c	2019-06-27 16:50:45.617775551 +0800
@@ -155,6 +155,7 @@
                             inetFmt(src, s1), inetFmt(dst, s2));
                         return;
                     }
+#if 0
                     else if(!isAdressValidForIf(checkVIF, src)) {
                         struct IfDesc *downVIF = getIfByAddress(src);
                         if (downVIF && downVIF->state & IF_STATE_DOWNSTREAM) {
@@ -164,7 +165,9 @@
                             my_log(LOG_WARNING, 0, "The source address %s for group %s, is not in any valid net for upstream VIF[%d].",
                                 inetFmt(src, s1), inetFmt(dst, s2), i);
                         }
-                    } else {
+                    } 
+#endif
+                    else {
                         // Activate the route.
                         int vifindex = checkVIF->index;
                         my_log(LOG_DEBUG, 0, "Route activate request from %s to %s on VIF[%d]",
