Index: dnsmasq-2.78/src/extern.c
===================================================================
--- dnsmasq-2.78.orig/src/extern.c	2019-04-08 16:58:40.000000000 +0800
+++ dnsmasq-2.78/src/extern.c	2019-04-08 17:02:36.280891251 +0800
@@ -164,18 +164,27 @@
   FILE *fp = NULL;
 
   phyifname[0] = 0;
-  sprintf(tmp, "brctl showmacs %s | grep %s", ifname, macaddr);
+  //sprintf(tmp, "brctl showmacs %s | grep %s", ifname, macaddr);
+  sprintf(tmp, "brctl showmacs %s | grep %s | awk '{print$NF}'", ifname, macaddr);
+
+  my_syslog(MS_DHCP | LOG_DEBUG, "%s command = %s", __func__, tmp);
+
   if ((fp = popen(tmp, "r"))) {
     tmp[0] = '\0';
     fgets(tmp, sizeof(tmp) - 1, fp);
     pclose(fp);
-    if (sscanf(tmp, "%*s %*s %*s %*s %s", phyifname) == 1) {
+    //if (sscanf(tmp, "%*s %*s %*s %*s %s", phyifname) == 1) {
+      if(sscanf(tmp, "%s", phyifname) == 1){
       strcpy(reply, phyifname);
       if (reply)
         return reply;
       return phyifname;
     }
   }
+  else
+  {
+      my_syslog(MS_DHCP | LOG_DEBUG, "%s Cannot get Phy Ifname = %s", __func__);
+  }
 
   return "";
 }
