Index: libzyutil-1.0/common.c
===================================================================
--- libzyutil-1.0.orig/common.c	2019-01-15 19:05:19.353273246 +0800
+++ libzyutil-1.0/common.c	2019-01-16 16:59:55.597897321 +0800
@@ -1533,6 +1533,19 @@
 }
 
 /*!
+ *  Check if it is Hex. value.
+ *
+ *  @param[in] unsigned char c  Input character.
+ *  @return                     1:hex; 0:non-hex
+ */
+static int _common_isHex(unsigned char ch);
+static int _common_isHex(unsigned char ch)
+{
+    /* refer to ui_zycommands.c */
+	return (((ch >= '0') && (ch <= '9')) || ((ch >= 'A') && (ch <= 'F')));
+}
+
+/*!
  *  Check serial number format.
  *
  *  @param[in] char *str        Input string.
@@ -1559,4 +1572,65 @@
     }
 
     return ret;
+}
+
+/*!
+ *  Check GPON serial number format.
+ *
+ *  @param[in] char *str        Input string.
+ *  @param[in] int inputLen     Lenth of input string.
+ *  @param[in] int chkLen       Lenth want to check.
+ *  @return                     0:ok; -1:error.
+ */
+int zyUtilChkGPONSnFormat(char *str, int inputLen, int chkLen)
+{
+    int i = 0, ret = -1;
+
+    if (inputLen == chkLen)
+    {
+        for (i = 0; i < 4; i++)
+        {
+            if ( ! _common_isalnum(str[i]) )
+            {
+                return ret;
+            }
+        }
+        for (i = 4; i < chkLen; i++) /* refer to ui_zycommands.c */
+        {
+            if ( ! _common_isHex(str[i]) )
+            {
+                return ret;
+            }
+        }
+        ret = 0;
+    }
+
+    return ret;
+}
+
+/*!
+ *  Check password format.
+ *
+ *  @param[in] char *str        Input string.
+ *  @param[in] int inputLen     Lenth of input string.
+ *  @param[in] int chkLen       Lenth want to check.
+ *  @return                     0:ok; -1:error.
+ */
+int zyUtilChkPwdFormat(char *str, int inputLen, int chkLen)
+{
+    int i = 0, ret = -1;
+
+    if (inputLen == chkLen)
+    {
+        for (i = 0; i < chkLen; i++)
+        {
+            if ( ! _common_isalnum(str[i]) )
+            {
+                return ret;
+            }
+        }
+        ret = 0;
+    }
+
+    return ret;
 }
Index: libzyutil-1.0/zyutil.h
===================================================================
--- libzyutil-1.0.orig/zyutil.h	2019-01-15 19:05:19.353273246 +0800
+++ libzyutil-1.0/zyutil.h	2019-01-15 19:05:19.421273240 +0800
@@ -142,4 +142,24 @@
  */
 int zyUtilChkSnFormat(char *str, int len);
 
+/*!
+ *  Check GPON serial number format.
+ *
+ *  @param[in] char *str        Input string.
+ *  @param[in] int inputLen     Lenth of input string.
+ *  @param[in] int chkLen       Lenth want to check.
+ *  @return                     0:ok; -1:error.
+ */
+int zyUtilChkGPONSnFormat(char *str, int inputLen, int chkLen);
+
+/*!
+ *  Check password format.
+ *
+ *  @param[in] char *str        Input string.
+ *  @param[in] int inputLen     Lenth of input string.
+ *  @param[in] int chkLen       Lenth want to check.
+ *  @return                     0:ok; -1:error.
+ */
+int zyUtilChkPwdFormat(char *str, int inputLen, int chkLen);
+
 #endif
