Index: libzyutil-1.0/BROADCOM/libzyutil_brcm_wrapper.h
===================================================================
--- libzyutil-1.0.orig/BROADCOM/libzyutil_brcm_wrapper.h	2019-03-22 15:09:56.949275471 +0800
+++ libzyutil-1.0/BROADCOM/libzyutil_brcm_wrapper.h	2019-03-22 15:13:26.698566797 +0800
@@ -12,7 +12,11 @@
 #include "bcm_hwdefs.h"
 #include "board.h"
 
+#if defined(TARGET_ROOTFS_UBIFS)
+#define MagicNubmer 4255
+#else
 #define MagicNubmer 1985
+#endif
 
 int boardIoctl(int board_ioctl, BOARD_IOCTL_ACTION action, char *string, int strLen, int offset, char *buf);
 int nvramDataWrite(NVRAM_DATA *nvramData);
