Index: libzyutil-1.0/zyutil_cofw.c
===================================================================
--- libzyutil-1.0.orig/zyutil_cofw.c	2019-06-21 16:29:06.520318430 +0800
+++ libzyutil-1.0/zyutil_cofw.c	2019-06-21 16:29:06.604318430 +0800
@@ -40,6 +40,10 @@
 			return MODEL_VMG9827_B50A;
 		}else if (strstr(modelName,"EMG6726-B10A") != NULL){
 			return MODEL_EMG6726_B10A;
+		}else if (strstr(modelName,"PX7501-B0") != NULL){
+			return MODEL_PX7501_B0;
+		}else if (strstr(modelName,"AX7501-B0") != NULL){
+			return MODEL_AX7501_B0;
 		}
 	}
 
Index: libzyutil-1.0/zyutil_cofw.h
===================================================================
--- libzyutil-1.0.orig/zyutil_cofw.h	2019-06-21 16:29:06.520318430 +0800
+++ libzyutil-1.0/zyutil_cofw.h	2019-06-21 16:29:06.604318430 +0800
@@ -14,7 +14,9 @@
     MODEL_EMG8726_B10A,
     MODEL_VMG4927_B50A,
     MODEL_VMG9827_B50A,
-    MODEL_EMG6726_B10A
+    MODEL_EMG6726_B10A,
+    MODEL_PX7501_B0,
+    MODEL_AX7501_B0
 } model_name_t;
 
 /*!
Index: libzyutil-1.0/BROADCOM/libzyutil_wrapper.c
===================================================================
--- libzyutil-1.0.orig/BROADCOM/libzyutil_wrapper.c	2019-06-21 16:29:06.584318430 +0800
+++ libzyutil-1.0/BROADCOM/libzyutil_wrapper.c	2019-06-21 17:20:56.423041865 +0800
@@ -8,6 +8,8 @@
 #include "bcm_hwdefs.h"
 #include "bcmTag.h"
 
+
+
 #define BOARD_DEVICE_NAME  "/dev/brcmboard"
 
 #if defined(ZYPRJ_VMG3926) && defined(SAME_FW_FOR_TWO_BOARD) && defined(BCM_SDK_416L05)
@@ -435,6 +437,37 @@
 	return ZCFG_SUCCESS;
 }
 
+#if defined(ZYXEL_BOARD_DETECT_BY_GPIO)
+zcfgRet_t zyUtilIGetScratchPadSetting(char *psp_wantype, char *psp_wanrate)
+{
+    int rc;
+
+    if (psp_wantype == NULL || psp_wanrate == NULL)
+    {
+        return ZCFG_INTERNAL_ERROR;
+    }
+
+    rc = boardIoctl(BOARD_IOCTL_FLASH_READ, SCRATCH_PAD, (char*)RDPA_WAN_TYPE_PSP_KEY, 0, PSP_BUFLEN_16, psp_wantype);
+    if (rc < 0)
+    {
+        printf("Unable to get WAN type");
+        return ZCFG_INTERNAL_ERROR;
+    }
+#if !defined(BCM_SDK_416L05)
+// SDK 416L05 didn`t have RDPA_WAN_RATE_PSP_KEY.
+        rc = boardIoctl(BOARD_IOCTL_FLASH_READ, SCRATCH_PAD, (char*)RDPA_WAN_RATE_PSP_KEY, 0, PSP_BUFLEN_16, psp_wanrate);
+    if (rc < 0)
+    {
+        printf("Unable to get WAN rate");
+        return ZCFG_INTERNAL_ERROR;
+    }
+#endif
+
+    return ZCFG_SUCCESS;
+}
+#endif
+
+
 int boardIoctl(int board_ioctl, BOARD_IOCTL_ACTION action, char *string, int strLen, int offset, char *buf)
 {
     BOARD_IOCTL_PARMS IoctlParms;
Index: libzyutil-1.0/libzyutil_wrapper.h
===================================================================
--- libzyutil-1.0.orig/libzyutil_wrapper.h	2019-06-21 16:29:06.552318430 +0800
+++ libzyutil-1.0/libzyutil_wrapper.h	2019-06-21 17:14:00.187912690 +0800
@@ -24,6 +24,13 @@
 
 #ifdef BROADCOM_PLATFORM
 zcfgRet_t zyUtilIGetDefaultGponSnPw(char *serianumber, char *password);
+
+#if defined(ZYXEL_BOARD_DETECT_BY_GPIO)
+//Reference bcmdrivers/opensource/include/bcm963xx/board.h
+#define ZYXEL_RDPA_WAN_TYPE_VALUE_XGS     "XGS"	//simulate RDPA_WAN_TYPE_VALUE_XGS
+#define ZYXEL_RDPA_WAN_TYPE_VALUE_GBE     "GBE" //simulate RDPA_WAN_TYPE_VALUE_GBE
+zcfgRet_t zyUtilIGetScratchPadSetting(char *psp_wantype, char *psp_wanrate);
+#endif
 #endif
 
 #ifdef ECONET_PLATFORM
