Index: libzyutil-1.0/ECONET/libzyutil_wrapper.c
===================================================================
--- libzyutil-1.0.orig/ECONET/libzyutil_wrapper.c	2019-05-21 14:57:36.424495130 +0800
+++ libzyutil-1.0/ECONET/libzyutil_wrapper.c	2019-05-21 17:28:08.423810802 +0800
@@ -966,4 +966,27 @@
 }
 #endif
 
+#ifdef ZYXEL_AT_SET_KEY
+char *zyGetSupervisorPWD(void){
+	MRD_DATA mrd = {0};
+	static char fake[MRD_SUPER_PWD_LEN+1] = {0};
+	if (ZCFG_SUCCESS == zyUtilGetMrdInfo(&mrd))
+		memcpy(fake, mrd.supervisor, MRD_SUPER_PWD_LEN);
+    return fake;
+}
+char *zyGetAdminPWD(void){
+	MRD_DATA mrd = {0};
+	static char fake[MRD_ADMIN_PWD_LEN+1] = {0};
+	if (ZCFG_SUCCESS == zyUtilGetMrdInfo(&mrd))
+		memcpy(fake, mrd.admin, MRD_ADMIN_PWD_LEN);
+    return fake;
+}
+char *zyGetPSK(void){
+	MRD_DATA mrd = {0};
+	static char fake[MRD_KEYPASSPHRASE_LEN+1] = {0};
+	if (ZCFG_SUCCESS == zyUtilGetMrdInfo(&mrd))
+		memcpy(fake, mrd.WpaPskKey, MRD_KEYPASSPHRASE_LEN);
+    return fake;
+}
+#endif
 #endif
Index: libzyutil-1.0/zyutil.h
===================================================================
--- libzyutil-1.0.orig/zyutil.h	2019-05-21 14:57:36.420495130 +0800
+++ libzyutil-1.0/zyutil.h	2019-05-21 17:26:46.348672953 +0800
@@ -145,6 +145,12 @@
 void zyRemoveFile(int pid);
 #endif
 
+#ifdef ZYXEL_AT_SET_KEY
+char *zyGetSupervisorPWD(void);
+char *zyGetAdminPWD(void);
+char *zyGetPSK(void);
+#endif
+
 /*!
  *  Check serial number format.
  *
