Index: libzyutil-1.0/BROADCOM/libzyutil_wrapper.c
===================================================================
--- libzyutil-1.0.orig/BROADCOM/libzyutil_wrapper.c	2019-06-19 22:24:22.240770939 +0800
+++ libzyutil-1.0/BROADCOM/libzyutil_wrapper.c	2019-06-19 22:27:50.858531027 +0800
@@ -171,7 +171,7 @@
 
 int zyFwInfoNonBootGet(ZY_FW_INFO *zyFwInfo)
 {
-#if defined(BCM_SDK_502L05)
+#if defined(BCM_SDK_502L05) || defined(BCM_SDK_502L06)
 	int boardFd = 0;
 	int rc = 0;
 	unsigned int offset = 0;
@@ -199,9 +199,9 @@
 		return -1;
 	}
 
-#else //defined(BCM_SDK_502L05)
+#else //defined(BCM_SDK_502L05) || defined(BCM_SDK_502L06)
     printf("zyFwInfoNonBootGet() only support BRCM SDK 502L05, you need to add version check in this API\n");
-#endif //defined(BCM_SDK_502L05)
+#endif //defined(BCM_SDK_502L05) || defined(BCM_SDK_502L06)
 	return 0;
 }
 
