Index: libzyutil-1.0/BROADCOM/libzyutil_wrapper.c
===================================================================
--- libzyutil-1.0.orig/BROADCOM/libzyutil_wrapper.c	2019-08-02 14:49:44.688857692 +0800
+++ libzyutil-1.0/BROADCOM/libzyutil_wrapper.c	2019-08-02 15:00:41.099246300 +0800
@@ -332,7 +332,7 @@
 	strcpy(pdname, nvramData.ProductName);
 #if defined(ZYPRJ_VMG3926) && defined(SAME_FW_FOR_TWO_BOARD) && defined(BCM_SDK_416L05)
 {
-    FILE *fp;
+    FILE *fp = NULL;
     char str[NVRAM_BOARD_ID_STRING_LEN];
 
     fp = fopen(PROC_NVRAM_BOARDID, "r");
@@ -536,7 +536,7 @@
 {
 	int size=0;
 	FILE *fp = NULL;
-	char buffer[64];
+	char buffer[64] = {0};
 
 	fp = popen("cat /proc/mtd  | grep rootfs | awk '{print $2}'", "r");
 
@@ -574,8 +574,8 @@
 {
 	int magic_number=-1;
 	FILE *fp = NULL;
-	char buffer[64];
-	char cmd[256];
+	char buffer[64] = {0};
+	char cmd[256] = {0};
 
 	if (uploadfilename == NULL)
 	{
Index: libzyutil-1.0/ECONET/libzyutil_wrapper.c
===================================================================
--- libzyutil-1.0.orig/ECONET/libzyutil_wrapper.c	2019-08-02 14:49:44.680857652 +0800
+++ libzyutil-1.0/ECONET/libzyutil_wrapper.c	2019-08-02 14:53:57.629470398 +0800
@@ -53,8 +53,8 @@
 
 int mtd_open(const char *mtd, int flags)
 {
-	FILE *fp;
-	char dev[256];
+	FILE *fp = NULL;
+	char dev[256] = {0};
 	int i;
 
 	if ((fp = fopen("/proc/mtd", "r"))) {
@@ -163,9 +163,9 @@
 {
 	char sysCmd[128] = {0};
 	FILE *fptr = NULL;
-    char username[16];
+	char username[16] = {0};
 	void *sysCmd_reply=NULL;
-	char filename[32] = {0};	
+	char filename[32] = {0};
 
 	// clean the MRD info
 	memset(mrd, 0x0, sizeof(MRD_DATA));
@@ -475,7 +475,7 @@
 {
 	char sysCmd[128] = {0};
 	FILE *fptr = NULL;
-	char bootflag[2];
+	char bootflag[2] = {0};
 
 	sprintf(sysCmd,FLASH_READ_CMD,BOOT_FLAG_FILE,(long unsigned int)BOOT_FLAG_SIZE,(long unsigned int)BOOT_FLAG_OFFSET,RESERVEAREA_NAME);
 	system(sysCmd);
@@ -713,7 +713,7 @@
 {
 	int size=0;
 	FILE *fp = NULL;
-	char buffer[64];
+	char buffer[64] = {0};
 
 	fp = popen("cat /proc/mtd  | grep tclinux_slave | awk '{print $2}'", "r");
 
@@ -749,8 +749,8 @@
 int endian_check()
 {
 	FILE *fp = NULL;
-	char buffer[64];
-	char cpu[64];
+	char buffer[64] = {0};
+	char cpu[64] = {0};
 
 	fp = popen("uname -m", "r");
 	if (fp == NULL)
@@ -800,8 +800,8 @@
 	int header_number2=-1;
 	int cpu_type=-1;
 	FILE *fp = NULL;
-	char buffer[64];
-	char cmd[256];
+	char buffer[64] = {0};
+	char cmd[256] = {0};
 
 	if (uploadfilename == NULL)
 	{
@@ -892,6 +892,7 @@
 {
    FILE *pf = fopen(ECONET_BOARDFEATURES_PROC_PATH, "r");
    if(!pf || !bits){
+     if(pf != NULL) {fclose(pf); pf = NULL;}
      return -1;
    }
 
@@ -905,6 +906,7 @@
    FILE *pf = fopen(ECONET_BOARDFEATURES_PRNAME_PATH, "r");
 
    if(!pf || !prname){
+     if(pf != NULL) {fclose(pf); pf = NULL;}
      return -1;
    }
 
Index: libzyutil-1.0/QUALCOMM/libzyutil_wrapper.c
===================================================================
--- libzyutil-1.0.orig/QUALCOMM/libzyutil_wrapper.c	2019-08-02 14:49:44.580857156 +0800
+++ libzyutil-1.0/QUALCOMM/libzyutil_wrapper.c	2019-08-02 15:01:50.317132800 +0800
@@ -105,7 +105,7 @@
 {
 	int size=0;
 	FILE *fp = NULL;
-	char buffer[64];
+	char buffer[64] = {0};
 
 	fp = popen("cat /proc/mtd  | grep rootfs | awk '{print $2}'", "r");
 
@@ -143,8 +143,8 @@
 {
 	int magic_number=-1;
 	FILE *fp = NULL;
-	char buffer[64];
-	char cmd[256];
+	char buffer[64] = {0};
+	char cmd[256] = {0};
 
 	if (uploadfilename == NULL)
 	{
Index: libzyutil-1.0/zyutil_sfp.c
===================================================================
--- libzyutil-1.0.orig/zyutil_sfp.c	2019-08-02 14:49:44.608857295 +0800
+++ libzyutil-1.0/zyutil_sfp.c	2019-08-02 15:03:39.736907100 +0800
@@ -178,9 +178,9 @@
 
 int zyUtil_is_sfp_present(void)
 {
-    FILE *fp;
+    FILE *fp = NULL;
     int value=0;
-    char str[3];
+    char str[3] = {0};
 
     fp = fopen(PROC_SFP_PRESENT, "r");
     if ( !fp )
@@ -215,7 +215,7 @@
  */
 int zyUtil_is_sfp_linkup(void)
 {
-    FILE *fp;
+    FILE *fp = NULL;
     char str[256] ={0};
 	int value = 0;
     fp = popen("ethctl phy int 6 1 | awk '{print $5}'", "r");
Index: libzyutil-1.0/libzyutil_zyxelfeaturebits.c
===================================================================
--- libzyutil-1.0.orig/libzyutil_zyxelfeaturebits.c	2019-08-02 14:49:44.452856521 +0800
+++ libzyutil-1.0/libzyutil_zyxelfeaturebits.c	2019-08-02 15:04:15.305845700 +0800
@@ -20,8 +20,8 @@
 static zcfgRet_t _zyxelfeaturebits_allValueGet(unsigned int *value);
 static zcfgRet_t _zyxelfeaturebits_allValueGet(unsigned int *value)
 {
-    FILE *fp;
-    char str[9];
+    FILE *fp = NULL;
+    char str[9] = {0};
 
     *value = 0;
     fp = fopen(PROC_ZY_FEATURE_BITS, "r");
Index: libzyutil-1.0/common.c
===================================================================
--- libzyutil-1.0.orig/common.c	2019-08-02 14:49:44.700857751 +0800
+++ libzyutil-1.0/common.c	2019-08-02 15:05:40.936052700 +0800
@@ -54,7 +54,7 @@
 void zyUtilIfaceHwAddrGet(char *ifname, char *hwaddr)
 {
 	char buf[MID_SIZE_MAX+60];
-	char sysCmd[1024];
+	char sysCmd[1024] = {0};
 	int i = 0, j = 0;
 	char *p = NULL;
 	FILE *fp = NULL;
@@ -89,7 +89,7 @@
 
 bool zyUtilIsAppRunning(char *appName)
 {
-	char sysCmd[64];
+	char sysCmd[64] = {0};
 	int size = 0;
 	FILE *fp = NULL;
 	bool isRunning = false;
@@ -150,7 +150,7 @@
 
 void zyUtilAddIntfToBridge(char *ifName, char *bridgeName)
 {
-	char sysCmd[64];
+	char sysCmd[64] = {0};
 
 	if(ifName == NULL || ifName[0] == 0) return;
 	if(bridgeName == NULL || bridgeName[0] == 0) return;
@@ -171,7 +171,7 @@
 
 void zyUtilDelIntfFromBridge(char *ifName, char *bridgeName)
 {
-	char sysCmd[64];
+	char sysCmd[64] = {0};
 
 	sprintf(sysCmd, "brctl delif %s %s", bridgeName, ifName);
 	system(sysCmd);
@@ -262,43 +262,50 @@
 bool zyUtilRetrieveSystemDefaultRoutIface(char *interfaceName)
 {
 	system("echo > /var/tr143NullInterface");//If interface is null
-	FILE *fp;
+	FILE *fp = NULL;
 	char defaultGateway[256] = {0};
 
 	fp = popen("route", "r");
+	if(fp != NULL)
+	{
 	while ((fgets(defaultGateway, 255, fp)) != NULL)
 	{
-		if (strstr(defaultGateway, "default") != 0)
+		if (strstr(defaultGateway, "default") != NULL)
 		{
-			if(strstr(defaultGateway, "eth") != 0)
+			if(strstr(defaultGateway, "eth") != NULL)
 				strncpy(interfaceName, strstr(defaultGateway, "eth"), (strstr(defaultGateway, "\n")) - (strstr(defaultGateway, "eth")));
-			else if(strstr(defaultGateway, "ptm") != 0)
+			else if(strstr(defaultGateway, "ptm") != NULL)
 				strncpy(interfaceName, strstr(defaultGateway, "ptm"), (strstr(defaultGateway, "\n")) - (strstr(defaultGateway, "ptm")));
-			else if(strstr(defaultGateway, "atm") != 0)
+			else if(strstr(defaultGateway, "atm") != NULL)
 				strncpy(interfaceName, strstr(defaultGateway, "atm"), (strstr(defaultGateway, "\n")) - (strstr(defaultGateway, "atm")));
-			else if(strstr(defaultGateway, "ppp") != 0)
+			else if(strstr(defaultGateway, "ppp") != NULL)
 				strncpy(interfaceName, strstr(defaultGateway, "ppp"), (strstr(defaultGateway, "\n")) - (strstr(defaultGateway, "ppp")));
-			else if(strstr(defaultGateway, "br") != 0)
+			else if(strstr(defaultGateway, "br") != NULL)
 				strncpy(interfaceName, strstr(defaultGateway, "br"), (strstr(defaultGateway, "\n")) - (strstr(defaultGateway, "br")));
 //#ifdef ZYXEL_ECONET_WAN_INTERFACE_NAMING_ENABLE
-			else if(strstr(defaultGateway, "nas") != 0)
-            	strncpy(interfaceName, strstr(defaultGateway, "nas")
-                    	, (strstr(defaultGateway, "\n")) - (strstr(defaultGateway, "nas")));
-			else if(strstr(defaultGateway, "ppp") != 0)
-            	strncpy(interfaceName, strstr(defaultGateway, "ppp")
-                    	, (strstr(defaultGateway, "\n")) - (strstr(defaultGateway, "ppp")));
+			else if(strstr(defaultGateway, "nas") != NULL)
+				strncpy(interfaceName, strstr(defaultGateway, "nas"), (strstr(defaultGateway, "\n")) - (strstr(defaultGateway, "nas")));
+			else if(strstr(defaultGateway, "ppp") != NULL)
+				strncpy(interfaceName, strstr(defaultGateway, "ppp"), (strstr(defaultGateway, "\n")) - (strstr(defaultGateway, "ppp")));
 //#endif
 			else
+			{
+				pclose(fp);
+				fp = NULL;
 				return false;
+			}
 
+			pclose(fp);
+			fp = NULL;
 			return true;
 		}
 		else{
 			strcpy(interfaceName, "br0");
 		}
 	}
-
 	pclose(fp);
+	fp = NULL;
+	}
 
 	return false;
 }
@@ -1009,14 +1016,14 @@
 	struct json_object *value = NULL;
 	char sysCmd[2084] = {0};
 	char *bodypath = "/tmp/mailbody";
-	const char *eventType;
+	const char *eventType = NULL;
 	time_t time_stamp = time(NULL);
 	struct tm *tPtr = localtime(&time_stamp);
 	const char *filepath = NULL;
 	int files = 0, delfile = 0;
 	int rm_ret;
-	FILE *fp;
-	char buf[128];
+	FILE *fp = NULL;
+	char buf[128] = {0};
 	bool isSendSuccess = false;
 
 	zcfgLog(ZCFG_LOG_INFO, "%s : Enter\n", __FUNCTION__);
@@ -1030,9 +1037,13 @@
 	//printf("Got message: %s\n", msg);
 	sendObj = json_tokener_parse(msg);
 
-	fprintf(mailbody, "%s\n", json_object_get_string(json_object_object_get(sendObj, "body")));
-	fprintf(mailbody, "\nSend at\t%s", asctime(tPtr));
-	fclose(mailbody);
+	if(mailbody != NULL)
+	{
+		fprintf(mailbody, "%s\n", json_object_get_string(json_object_object_get(sendObj, "body")));
+		fprintf(mailbody, "\nSend at\t%s", asctime(tPtr));
+		fclose(mailbody);
+		mailbody = NULL;
+	}
 
 	json_object_object_foreach(sendObj, keys, val) {
 		if(!strcmp(keys,"body")) continue;
@@ -1078,17 +1089,20 @@
 			subject);
 
 	if(files) {
-		FILE *fp;
 		char attachCmd[2048] ={0};
 		char buff[1024] = {0};
 		fp = popen("cat /tmp/mailbody", "r");
-		fgets(buff, sizeof(buff), fp);
-		/* mailsend can choose option between -attach file or -mime-type.
-		 * It can attach multi files, but if it need mail body, use option -M only
-		 * And now mailsend can only attach one text file */
-		sprintf(attachCmd, " -attach \"%s,text/plain\" -M \"%s\"", filepath, buff);
-		strcat(sysCmd, attachCmd);
-		fclose(fp);
+		if(fp != NULL)
+		{
+			fgets(buff, sizeof(buff), fp);
+			/* mailsend can choose option between -attach file or -mime-type.
+			* It can attach multi files, but if it need mail body, use option -M only
+			* And now mailsend can only attach one text file */
+			sprintf(attachCmd, " -attach \"%s,text/plain\" -M \"%s\"", filepath, buff);
+			strcat(sysCmd, attachCmd);
+			fclose(fp);
+			fp = NULL;
+		}
 	} else {
 		char *normalCmd = " -mime-type \"text/plain\" -msg-body \"/tmp/mailbody\"";
 		strcat(sysCmd, normalCmd);
@@ -1101,19 +1115,20 @@
 	system(sysCmd);
 
 	chmod("/tmp/mailLog", strtol("0777", 0, 8));
-	if  ((fp = fopen("/tmp/mailLog", "r")) == NULL) {
-                printf("fopen error!\n");
-        }
-
-	while (fgets(buf, sizeof(buf), fp) != NULL){
-		if (strstr(buf, "Mail sent successfully") != NULL ){
-			isSendSuccess = true;
-			break;
+	if ((fp = fopen("/tmp/mailLog", "r")) == NULL) {
+		printf("fopen error!\n");
+	}
+	else
+	{
+		while (fgets(buf, sizeof(buf), fp) != NULL){
+			if (strstr(buf, "Mail sent successfully") != NULL ){
+				isSendSuccess = true;
+				break;
+			}
 		}
-        }
-
-	fclose(fp);
-
+		fclose(fp);
+		fp = NULL;
+	}
 	unlink("/tmp/mailLog");
 
 	if(delfile) {
@@ -1134,7 +1149,7 @@
 
 bool zyUtilFileExists(const char * filename)
 {
-	FILE *file;
+	FILE *file = NULL;
 	if (file = fopen(filename, "r"))
 	{
 		fclose(file);
@@ -1240,23 +1255,26 @@
 		strcpy(fileName, "/tmp/iptablesConfig-XXXXXX");
 		fd = mkstemp(fileName);
 		fp = fdopen(fd, "w+");
-		if(ipt->len_m){
-			fprintf(fp, "*mangle\n");
-			fprintf(fp, "%s", ipt->buf_m);
-			fprintf(fp, "COMMIT\n");
-		}
-		if(ipt->len_f){
-			fprintf(fp, "*filter\n");
-			fprintf(fp, "%s", ipt->buf_f);
-			fprintf(fp, "COMMIT\n");
-		}
-		if(ipt->len_n){
-			fprintf(fp, "*nat\n");
-			fprintf(fp, "%s", ipt->buf_n);
-			fprintf(fp, "COMMIT\n");
+		if(fp != NULL)
+		{
+			if(ipt->len_m){
+				fprintf(fp, "*mangle\n");
+				fprintf(fp, "%s", ipt->buf_m);
+				fprintf(fp, "COMMIT\n");
+			}
+			if(ipt->len_f){
+				fprintf(fp, "*filter\n");
+				fprintf(fp, "%s", ipt->buf_f);
+				fprintf(fp, "COMMIT\n");
+			}
+			if(ipt->len_n){
+				fprintf(fp, "*nat\n");
+				fprintf(fp, "%s", ipt->buf_n);
+				fprintf(fp, "COMMIT\n");
+			}
+			fclose(fp);
+			fp = NULL;
 		}
-		fclose(fp);
-
 		sprintf(cmd, "%s --noflush %s", IPTABLESRESTORE, fileName);
 		ret = system(cmd);
 		if(ret != 0){
@@ -1274,18 +1292,21 @@
 		strcpy(fileName, "/tmp/iptablesConfig-XXXXXX");
 		fd = mkstemp(fileName);
 		fp = fdopen(fd, "w+");
-		if(ipt->len_m6){
-			fprintf(fp, "*mangle\n");
-			fprintf(fp, "%s", ipt->buf_m6);
-			fprintf(fp, "COMMIT\n");
-		}
-		if(ipt->len_f6){
-			fprintf(fp, "*filter\n");
-			fprintf(fp, "%s", ipt->buf_f6);
-			fprintf(fp, "COMMIT\n");
+		if(fp != NULL)
+		{
+			if(ipt->len_m6){
+				fprintf(fp, "*mangle\n");
+				fprintf(fp, "%s", ipt->buf_m6);
+				fprintf(fp, "COMMIT\n");
+			}
+			if(ipt->len_f6){
+				fprintf(fp, "*filter\n");
+				fprintf(fp, "%s", ipt->buf_f6);
+				fprintf(fp, "COMMIT\n");
+			}
+			fclose(fp);
+			fp = NULL;
 		}
-		fclose(fp);
-
 		sprintf(cmd, "%s --noflush %s", IP6TABLESRESTORE, fileName);
 		ret = system(cmd);
 		if(ret != 0){
@@ -1363,7 +1384,7 @@
 
 char *zyGetAccountInfoFromFile(char *username, char *filePath){
     static char accountInfo[513] = {0};
-    FILE *fp;
+    FILE *fp = NULL;
 
     fp = fopen(filePath, "r");
     if(fp == NULL){
@@ -1388,8 +1409,8 @@
 {
     char    strBuff[513] = {0};
     char    *tmpFilePath = NULL;
-    FILE    *fp;
-    FILE    *tmpFp;
+    FILE    *fp = NULL;
+    FILE    *tmpFp = NULL;
 
     fp = fopen(filePath, "r");
     if (fp == NULL)
@@ -1399,11 +1420,12 @@
     }
 
     tmpFilePath = ZOS_MALLOC(sizeof(char) * (strlen(filePath) + 4 + 1));
-	if (tmpFilePath == NULL)
-	{
-	    ZLOG_ERROR("memory insufficient");
-	    return;
-	}
+    if (tmpFilePath == NULL)
+    {
+        ZLOG_ERROR("memory insufficient");
+        if(fp != NULL) {fclose(fp); fp = NULL;}
+        return;
+    }
 
     sprintf(tmpFilePath, "%s.tmp", filePath);
     tmpFp = fopen(tmpFilePath, "w");
@@ -1446,11 +1468,15 @@
 	fp = fopen("/tmp/isWANRoute.sh", "r");
 	if (fp != NULL)
 	{
+		fclose(fp);
+		fp = NULL;
 		fp = fopen("/tmp/isWANRoute.sh", "w");
 		if (fp == NULL)
 		{
-		    ZLOG_ERROR("fp == NULL");
+			ZLOG_ERROR("fp == NULL");
 		}
+		else
+		{
 		fprintf(fp, "route=`ip r get $1`\n");
 		fprintf(fp, "if expr \"$1\" : '[0-9][0-9]*\.[0-9][0-9]*\.[0-9][0-9]*\.[0-9][0-9]*$' >/dev/null; then\n");
 		fprintf(fp, "	iface=`echo $route | awk '{print substr($3,1,3)}'`\n");
@@ -1492,8 +1518,11 @@
 		//fprintf(fp, "	echo $1 comes from br side\n");
 		fprintf(fp, "exit 0\n");
 #endif
+		fclose(fp);
+		fp = NULL;
+		}
 	}
-	fclose(fp);
+
 	chmod("/tmp/isWANRoute.sh", S_IRUSR|S_IWUSR|S_IXUSR|S_IRGRP|S_IROTH);
 
 	if (strstr(ip, "."))
