Index: libzyutil-1.0/common.c
===================================================================
--- libzyutil-1.0.orig/common.c	2019-12-31 16:50:02.010233209 +0800
+++ libzyutil-1.0/common.c	2019-12-31 16:38:26.894424900 +0800
@@ -838,6 +838,31 @@
 	return new_Csv;
 }
 
+void zyUtilStrCsvAppend(char **csv, const char *string)
+{
+	int n = 0, csvLen;
+
+	if(!csv || !string || !strlen(string))
+		return;
+
+	if(*csv)
+	{
+		csvLen = strlen(*csv);
+		n = strlen(string) + 2;
+		char *newcsv = ZOS_MALLOC(csvLen+n);
+		strcpy(newcsv, *csv);
+		free(*csv);
+		char *ptr = newcsv + csvLen;
+		sprintf(ptr, ",%s", string);
+		*csv = newcsv;
+	}
+	else{
+		n = strlen(string) + 1;
+		*csv = ZOS_MALLOC(n);
+		strcpy(*csv, string);
+	}
+}
+
 int zyUtilValidIpAddr(const char *ipAddress)
 {
 	struct in6_addr ipv6addr;
Index: libzyutil-1.0/zyutil.h
===================================================================
--- libzyutil-1.0.orig/zyutil.h	2019-12-31 16:50:01.966233210 +0800
+++ libzyutil-1.0/zyutil.h	2019-12-31 16:50:02.026233208 +0800
@@ -56,6 +56,7 @@
 int zyUtilQueryUtilityEid(const char *name);
 
 char *zyUtilStrCsvString(const char *, const char *);
+void zyUtilStrCsvAppend(char **, const char *);
 char *zyUtilCsvStringStripTerm(const char *, const char *);
 
 #ifdef ZYXEL_DETECT_PRODUCT_TYPE
