Index: ppp-2.4.3/pppd/ipcp.c
===================================================================
--- ppp-2.4.3.orig/pppd/ipcp.c	2018-12-18 18:58:50.785393386 +0800
+++ ppp-2.4.3/pppd/ipcp.c	2018-12-19 10:52:59.763059500 +0800
@@ -1703,7 +1703,7 @@
 
     notice("local  IP address %I", wo->ouraddr);
     notice("remote IP address %I", wo->hisaddr);
-#if 1 /* ZyXEL, zcfg */	
+#if 0 /* ZyXEL, zcfg */	
     ppp_arg = json_object_new_object();
     json_object_object_add(ppp_arg, "ip", json_object_new_string(ip_ntoa(wo->ouraddr)));
 
@@ -1967,6 +1967,7 @@
 #endif
     }
 #if 1 /* ZyXEL, zcfg */
+    addr_obtained = 1;
     ppp_arg = json_object_new_object();
     json_object_object_add(ppp_arg, "ip", json_object_new_string(ip_ntoa(go->ouraddr)));
     json_object_object_add(ppp_arg, "remote", json_object_new_string(ip_ntoa(ho->hisaddr)));
Index: ppp-2.4.3/pppd/ipv6cp.c
===================================================================
--- ppp-2.4.3.orig/pppd/ipv6cp.c	2018-12-18 18:58:50.465393384 +0800
+++ ppp-2.4.3/pppd/ipv6cp.c	2018-12-19 10:53:15.137559500 +0800
@@ -1174,7 +1174,7 @@
     notice("local  LL address %s", llv6_ntoa(wo->ourid));
     notice("remote LL address %s", llv6_ntoa(wo->hisid));
 
-#if 1 /* ZyXEL zcfg */
+#if 0 /* ZyXEL zcfg */
     ppp_arg = json_object_new_object();
     json_object_object_add(ppp_arg, "local", json_object_new_string(llv6_ntoa(wo->ourid)));
     json_object_object_add(ppp_arg, "remote", json_object_new_string(llv6_ntoa(wo->hisid)));
@@ -1328,6 +1328,7 @@
 	ipv6cp_script(_PATH_IPV6UP);
     }
 #if 1 /* ZyXEL zcfg */
+    addr_obtained = 1;
     ppp_arg = json_object_new_object();
     json_object_object_add(ppp_arg, "local", json_object_new_string(llv6_ntoa(wo->ourid)));
     json_object_object_add(ppp_arg, "remote", json_object_new_string(llv6_ntoa(wo->hisid)));
Index: ppp-2.4.3/pppd/main.c
===================================================================
--- ppp-2.4.3.orig/pppd/main.c	2018-12-18 18:58:50.821393386 +0800
+++ ppp-2.4.3/pppd/main.c	2018-12-19 10:51:52.311059500 +0800
@@ -220,6 +220,9 @@
 #endif
 
 #if 1 /* ZyXEL zcfg */
+/* For demand-dial: don't enter dormant mode until obtain address successfully. */
+bool addr_obtained = 0;
+
 #include "zcfg_common.h"
 #include "zcfg_msg.h"
 #include "zcfg_eid.h"
@@ -597,7 +600,7 @@
 	doing_callback = do_callback;
 	do_callback = 0;
 
-	if (demand && !doing_callback) {
+	if (demand && !doing_callback && addr_obtained) {
 	    /*
 	     * Don't do anything until we see some activity.
 	     */
Index: ppp-2.4.3/pppd/pppd.h
===================================================================
--- ppp-2.4.3.orig/pppd/pppd.h	2018-12-18 18:58:50.677393385 +0800
+++ ppp-2.4.3/pppd/pppd.h	2018-12-19 10:53:24.430559500 +0800
@@ -267,6 +267,11 @@
 #define CALLBACK_DIALIN		1	/* we are expecting the call back */
 #define CALLBACK_DIALOUT	2	/* we are dialling out to call back */
 
+#if 1 /* Zyxel */
+/* For demand-dial: don't enter dormant mode until obtain address successfully. */
+extern bool addr_obtained;
+#endif
+
 #if 1 /* brcm patch, ZyXEL */
 #define FIELD0    4        /* ppp device number ppp0, ppp1, the third digit (max 16) */
 #define FIELD1    8        /* if 0, default mode, 1 vlan mux, 2 msc */
