diff -urN net-snmp-5.8.org/agent/mibgroup/if-mib/data_access/interface_linux.c net-snmp-5.8/agent/mibgroup/if-mib/data_access/interface_linux.c
--- net-snmp-5.8.org/agent/mibgroup/if-mib/data_access/interface_linux.c	2019-04-08 10:56:29.445078431 +0800
+++ net-snmp-5.8/agent/mibgroup/if-mib/data_access/interface_linux.c	2019-04-08 11:04:28.789332087 +0800
@@ -689,7 +689,12 @@
         return -1;
     }
 
-    if (!(devin = fopen("/proc/net/dev", "r"))) {
+#if defined(BROADCOM) //__ZYXEL__, Mark , for 416L01SDK
+    if (!(devin = fopen("/proc/net/dev_extstats", "r"))) 
+#else
+    if (!(devin = fopen("/proc/net/dev", "r"))) 
+#endif
+    {
         DEBUGMSGTL(("access:interface",
                     "Failed to load Interface Table (linux1)\n"));
         snmp_log_perror("interface_linux: cannot open /proc/net/dev");
