diff -urN net-snmp-5.8.org/agent/mibgroup/ZYXEL-PRESTIGE-MIB/prestigeSystem.c net-snmp-5.8/agent/mibgroup/ZYXEL-PRESTIGE-MIB/prestigeSystem.c
--- net-snmp-5.8.org/agent/mibgroup/ZYXEL-PRESTIGE-MIB/prestigeSystem.c	2019-04-08 17:34:58.217390500 +0800
+++ net-snmp-5.8/agent/mibgroup/ZYXEL-PRESTIGE-MIB/prestigeSystem.c	2019-04-09 10:44:43.265851469 +0800
@@ -517,6 +517,7 @@
 			snmp_log(LOG_ERR, "unknown mode (%d) in %s\n", reqinfo->mode,__FUNCTION__ );
 			return SNMP_ERR_GENERR;
 	}
+  *(requests->requestvb->val.integer) = *(requests->requestvb->val.integer) * 100;
 	return ret;
 }
 
