Index: libzyuitimer-1.0/uitimer.c
===================================================================
--- libzyuitimer-1.0.orig/uitimer.c	2019-01-22 10:32:38.000000000 +0800
+++ libzyuitimer-1.0/uitimer.c	2019-03-25 13:47:44.672675466 +0800
@@ -114,7 +114,7 @@
 			timer_list_iterator_remove(&iter);
 			pthread_mutex_unlock(&timerMutex);
 			if(uitimer->fn) {
-				uitimer->fn();
+				uitimer->fn(uitimer->data);
 			}
 			ZOS_FREE(uitimer);
 			uitimer =NULL;
@@ -137,7 +137,7 @@
 				continue;
 			}
 			if(uitimer->fn) {
-				uitimer->fn();
+				uitimer->fn(uitimer->data);
 			}
 			memset(&(uitimer->timeout), 0, sizeof(struct timeval));
 			memcpy(&(uitimer->timeout), &timer_val, sizeof(struct timeval));
@@ -252,7 +252,7 @@
 	return res;
 }
 
-int UIDelTimer(int id){
+int UIDelTimer(int id, void **dataptr){
 	CMTimer_t *uitimer=NULL;
 	timer_list_iterator_t iter;
 	int found = 0;
@@ -268,6 +268,9 @@
 			continue;
 		}
 		found = id;
+		if(NULL != dataptr){
+			*dataptr = uitimer->data;
+		}
 		timer_list_iterator_remove(&iter);
 		ZOS_FREE(uitimer);
 		uitimer = NULL;
@@ -280,7 +283,7 @@
 	return found;
 }
 
-int UIDelPeriodicTimer(int id){
+int UIDelPeriodicTimer(int id, void **dataptr){
 	CMTimer_t *uitimer=NULL;
 	timer_list_iterator_t iter;
 	pthread_mutex_lock(&timerMutex);
@@ -293,6 +296,9 @@
 			pthread_mutex_unlock(&timerMutex);
 			continue;
 		}
+		if(NULL != dataptr){
+			*dataptr = uitimer->data;
+		}
 		timer_list_iterator_remove(&iter);
 		ZOS_FREE(uitimer);
 		uitimer = NULL;
Index: libzyuitimer-1.0/uitimer.h
===================================================================
--- libzyuitimer-1.0.orig/uitimer.h	2019-01-22 10:32:38.000000000 +0800
+++ libzyuitimer-1.0/uitimer.h	2019-03-25 13:46:57.537954566 +0800
@@ -37,7 +37,7 @@
 #define SWITCH_TIMER 0x498
 #define RESTORE_TIMER 0x499
 
-typedef void (*TMRFUNC)(void);
+typedef void (*TMRFUNC)(void *);
 
 /*new a time structure for longer timeout, ZyXEL*/
 struct timeval_t {
@@ -57,10 +57,10 @@
 void UITimer_Destory(void);
 /*change "timerLen" type to uint32 for longer timeout, ZyXEL*/
 int UIAddTimer(int id,uint32_t timerLen,TMRFUNC timefn,void *privdata);
-int UIDelTimer(int id);
+int UIDelTimer(int id, void **dataptr);
 int UIDelAllTimer(void);
 int UIAddPeriodicTimer(int id,int timerLen,TMRFUNC timefn,void *privdata);
-int UIDelPeriodicTimer(int id);
+int UIDelPeriodicTimer(int id, void **dataptr);
 int getSystemUptime(struct timeval_t *timeout);
 void zyTms_get(struct timespec *tms);
 
