Index: libzcfg_fe_dal-1.0/network/broadband/zcfg_fe_dal_broadband.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/broadband/zcfg_fe_dal_broadband.c	2019-01-11 19:08:50.842900318 +0800
+++ libzcfg_fe_dal-1.0/network/broadband/zcfg_fe_dal_broadband.c	2019-01-11 19:09:50.386900318 +0800
@@ -84,7 +84,7 @@
 	{"ip6Static",			dalType_boolean,	0,	0,	NULL},
 	{"IP6Address",			dalType_v6AddrPrefix,	0,	0,	NULL},
 	{"NextHop",				dalType_v6Addr,	0,	0,	NULL},
-	{"CurIP6Address",		dalType_v6AddrPrefix,	0,	0,	NULL, NULL, dalcmd_Forbid}, //read only
+	{"CurIP6Address",		dalType_v6Addr,	0,	0,	NULL, NULL, dalcmd_Forbid}, //read only
      /* __ZyXEL__, Albert, 20180209,Support IPv6 option IA_NA and IA_PD  */
 	{"IPv6Manual_IANA",		dalType_boolean,	0,	0,	NULL},
 	{"IPv6Manual_IAPD",		dalType_boolean,	0,	0,	NULL},	
