Index: libzcfg_fe_dal-1.0/zcfg_fe_dal.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/zcfg_fe_dal.c	2019-01-22 16:53:24.902062208 +0800
+++ libzcfg_fe_dal-1.0/zcfg_fe_dal.c	2019-01-22 18:31:28.258841441 +0800
@@ -1915,7 +1915,7 @@
 	const char *band = NULL;
 	const char *stardard = NULL;
 	char buf[32] = {0}, mode[64] = {0};
-	char *tmp = NULL, *b = NULL, *g = NULL, *n = NULL, *a = NULL, *ac = NULL;
+	char *tmp = NULL, *tmp2 = NULL, *b = NULL, *g = NULL, *n = NULL, *a = NULL, *ac = NULL, *ax = NULL;
 	int len = 0;
 	*modeObj = json_object_new_array();
 	while(zcfgFeObjJsonGetNextWithoutUpdate(RDM_OID_WIFI_RADIO, &radioIid, &radioObj) == ZCFG_SUCCESS){
@@ -1926,38 +1926,47 @@
 		if(!strcmp(band, "2.4GHz")){
 			strcpy(buf, stardard);
 			b = strtok_r(buf, ",", &tmp);
-			g = strtok_r(tmp, ",", &n);
+			g = strtok_r(tmp, ",", &tmp2);
+			n = strtok_r(tmp2, ",", &ax);
 			if(!strcmp(b, "b")){
-				strcat(mode, "802.11");
+				strcat(mode, "11");
 				strcat(mode,b);
 				strcat(mode,",");
 			}
 			if(!strcmp(g, "g")){
-				strcat(mode, "802.11");
+				strcat(mode, "11");
 				strcat(mode,g);
 				strcat(mode,",");
 			}
 			if(!strcmp(n, "n")){
-				strcat(mode, "802.11");
+				strcat(mode, "11");
 				strcat(mode,n);
 				strcat(mode,",");
 			}
 			if(!strcmp(b, "b") && !strcmp(g, "g")){
-				strcat(mode, "802.11");
+				strcat(mode, "11");
 				strcat(mode,b);
-				strcat(mode,"/");
+				//strcat(mode,"/");
 				strcat(mode,g);
 				strcat(mode,",");
 			}
 			if(!strcmp(b, "b") && !strcmp(g, "g") && !strcmp(n, "n")){
-				strcat(mode, "802.11");
+				strcat(mode, "11");
 				strcat(mode,b);
-				strcat(mode,"/");
+				//strcat(mode,"/");
 				strcat(mode,g);
-				strcat(mode,"/");
+				//strcat(mode,"/");
 				strcat(mode,n);
 				strcat(mode,",");
 			}
+			if(!strcmp(b, "b") && !strcmp(g, "g") && !strcmp(n, "n") && !strcmp(ax, "ax")){
+				strcat(mode, "11");
+				strcat(mode,b);
+				strcat(mode,g);
+				strcat(mode,n);
+				strcat(mode,ax);
+				strcat(mode,",");
+			}
 			len = strlen(mode);
 			if(len > 0){
 				if(mode[len-1] == ',')
@@ -1968,36 +1977,45 @@
 		else if(!strcmp(band, "5GHz")){
 			strcpy(buf, stardard);
 			a = strtok_r(buf, ",", &tmp);
-			n = strtok_r(tmp, ",", &ac);
+			n = strtok_r(tmp, ",", &tmp2);
+			ac = strtok_r(tmp2, ",", &ax);
 			if(!strcmp(a, "a")){
-				strcat(mode, "802.11");
+				strcat(mode, "11");
 				strcat(mode,a);
 				strcat(mode,",");
 			}
 			if(!strcmp(n, "n")){
-				strcat(mode, "802.11");
+				strcat(mode, "11");
 				strcat(mode,n);
 				strcat(mode,",");
 			}
 			if(!strcmp(ac, "ac")){
-				strcat(mode, "802.11");
+				strcat(mode, "11");
 				strcat(mode,ac);
 				strcat(mode,",");
 			}
 			if(!strcmp(a, "a") && !strcmp(n, "n")){
-				strcat(mode, "802.11");
+				strcat(mode, "11");
 				strcat(mode,a);
-				strcat(mode,"/");
+				//strcat(mode,"/");
 				strcat(mode,n);
 				strcat(mode,",");
 			}
 			if(!strcmp(a, "a") && !strcmp(n, "n") && !strcmp(ac, "ac")){
-				strcat(mode, "802.11");
+				strcat(mode, "11");
+				strcat(mode,a);
+				//strcat(mode,"/");
+				strcat(mode,n);
+				//strcat(mode,"/");
+				strcat(mode,ac);
+				strcat(mode,",");
+			}
+			if(!strcmp(a, "a") && !strcmp(n, "n") && !strcmp(ac, "ac") && !strcmp(ax, "ax")){
+				strcat(mode, "11");
 				strcat(mode,a);
-				strcat(mode,"/");
 				strcat(mode,n);
-				strcat(mode,"/");
 				strcat(mode,ac);
+				strcat(mode,ax);
 				strcat(mode,",");
 			}
 			len = strlen(mode);
