Index: libzcfg_fe_dal-1.0/zcfg_fe_dal.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/zcfg_fe_dal.c	2019-02-01 09:01:14.188813291 +0800
+++ libzcfg_fe_dal-1.0/zcfg_fe_dal.c	2019-02-01 09:45:33.360828905 +0800
@@ -1018,7 +1018,14 @@
 		if(X_ZYXEL_SrvName==NULL || X_ZYXEL_ConnectionType==NULL || LowerLayers==NULL){
 			continue;
 		}
-			
+
+#ifndef ZyXEL_WEB_GUI_SHOW_WWAN
+        if(strcmp(X_ZYXEL_SrvName,"WWAN") == 0)
+        {
+            continue;
+        }
+#endif
+
 		Jobj = json_object_new_object();
 		json_object_object_add(Jobj, "IpIfacePath", json_object_new_string(ifacePath));
 		json_object_object_add(Jobj, "Name", json_object_new_string(X_ZYXEL_SrvName));
@@ -1394,7 +1401,7 @@
 		if(X_ZYXEL_SrvName==NULL || X_ZYXEL_ConnectionType==NULL || LowerLayers==NULL){
 			continue;
 		}
-			
+
 		Jobj = json_object_new_object();
 		json_object_object_add(Jobj, "IpIfacePath", json_object_new_string(ifacePath));
 		json_object_object_add(Jobj, "Name", json_object_new_string(X_ZYXEL_SrvName));
Index: libzcfg_fe_dal-1.0/Makefile
===================================================================
--- libzcfg_fe_dal-1.0.orig/Makefile	2019-02-01 09:01:14.120813292 +0800
+++ libzcfg_fe_dal-1.0/Makefile	2019-02-01 09:01:41.369299347 +0800
@@ -208,6 +208,11 @@
 CFLAGS += -DZYXEL_GUIVUE_USB_INFO_SHOW_SYSINFO
 endif
 
+#ZyXEL_WEB_GUI_SHOW_WWAN
+ifeq ($(CONFIG_ZyXEL_WEB_GUI_SHOW_WWAN),y)
+CFLAGS += -DZyXEL_WEB_GUI_SHOW_WWAN
+endif
+
 ifeq ($(CONFIG_PACKAGE_ZyIMS),y)
 CFLAGS += -DBUILD_PACKAGE_ZyIMS=1
 INC_PATH += -I$(VOIP_PATH)
