Index: libzcfg_fe_dal-1.0/maintenance/logsetting/zcfg_fe_dal_log_setting.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/maintenance/logsetting/zcfg_fe_dal_log_setting.c	2019-03-08 11:37:54.099970799 +0800
+++ libzcfg_fe_dal-1.0/maintenance/logsetting/zcfg_fe_dal_log_setting.c	2019-03-08 11:40:08.446914822 +0800
@@ -39,6 +39,7 @@
 	{"xDSL",								dalType_boolean,0,	0,	NULL},
 	{"ACL",									dalType_boolean,0,	0,	NULL},
 	{"Wireless",							dalType_boolean,0,	0,	NULL},
+    {"OMCI",							    dalType_boolean,0,	0,	NULL},
 	{"Voice",								dalType_boolean,0,	0,	NULL},
 	{"Account",								dalType_boolean,0,	0,	NULL},
 	{"Attack",								dalType_boolean,0,	0,	NULL},
@@ -673,6 +674,17 @@
 				zcfgFeObjJsonSetWithoutApply(RDM_OID_LOG_CATEGORY, &categoryIid, categoryObj, NULL);
 				zcfgFeJsonObjFree(categoryObj);
 			}
+			zcfgFeJsonObjFree(categoryObj);
+		}
+	}
+    IID_INIT(categoryIid);
+	if(json_object_object_get(Jobj, "OMCI")){
+		while(zcfgFeObjJsonGetNext(RDM_OID_LOG_CATEGORY, &categoryIid, &categoryObj) == ZCFG_SUCCESS){
+			if(!strcmp(json_object_get_string(json_object_object_get(categoryObj, "Describe")),"OMCI")){
+				json_object_object_add(categoryObj, "Enable", JSON_OBJ_COPY(json_object_object_get(Jobj, "OMCI")));
+				zcfgFeObjJsonSetWithoutApply(RDM_OID_LOG_CATEGORY, &categoryIid, categoryObj, NULL);
+				zcfgFeJsonObjFree(categoryObj);
+			}
 			zcfgFeJsonObjFree(categoryObj);
 		}
 	}
