diff -urN libzcfg_fe_dal-1.0.org/network/routing/zcfg_fe_dal_static_route.c libzcfg_fe_dal-1.0/network/routing/zcfg_fe_dal_static_route.c
--- libzcfg_fe_dal-1.0.org/network/routing/zcfg_fe_dal_static_route.c	2019-05-07 17:54:18.594315300 +0800
+++ libzcfg_fe_dal-1.0/network/routing/zcfg_fe_dal_static_route.c	2019-05-08 11:25:02.319141675 +0800
@@ -132,7 +132,7 @@
 	zcfgRet_t ret = ZCFG_SUCCESS;
 	struct json_object *StaticRouteObj = NULL, *obj = NULL, *objtmp = NULL;
 	objIndex_t StaticRouteIid = {0};
-	zcfg_offset_t oid;
+	zcfg_offset_t oid = 0;
 	const char *iptype = NULL;
 	const char *origiptype = NULL;
 	const char *Name = NULL;
@@ -272,7 +272,7 @@
 	zcfgRet_t ret = ZCFG_SUCCESS;
 	struct json_object *StaticRouteObj = NULL, *obj = NULL, *objtmp = NULL;
 	objIndex_t StaticRouteIid = {0};
-	zcfg_offset_t oid;
+	zcfg_offset_t oid = 0;
 	const char *iptype = NULL;
 	const char *Name = NULL;
 	const char *GatewayIPAddress = NULL;
@@ -373,7 +373,7 @@
 	struct json_object *obj = NULL, *objtmp = NULL;
 	obj = json_object_new_array();
 	objIndex_t StaticRouteIid = {0};
-	zcfg_offset_t oid;
+	zcfg_offset_t oid = 0;
 	const char *iptype = NULL, *idx = NULL;
 	int index = 0, len = 0, i;
 	index = json_object_get_int(json_object_object_get(Jobj, "Index"));
diff -urN libzcfg_fe_dal-1.0.org/zcfg_fe_dal.c libzcfg_fe_dal-1.0/zcfg_fe_dal.c
--- libzcfg_fe_dal-1.0.org/zcfg_fe_dal.c	2019-05-07 17:54:23.014315400 +0800
+++ libzcfg_fe_dal-1.0/zcfg_fe_dal.c	2019-05-08 11:26:05.877760467 +0800
@@ -962,7 +962,7 @@
  */
 zcfgRet_t getWanLanList(struct json_object **Jarray){
 	struct json_object *ipIfaceJarray, *pppIfaceJarray, *vlanTermJarray, *ethLinkJarray, *bridgeBrJarray;
-	struct json_object *ipIfaceObj, *pppIfaceObj, *vlanTermObj, *ethLinkObj, *bridgeBrObj, *routerObj;
+	struct json_object *ipIfaceObj, *pppIfaceObj = NULL, *vlanTermObj, *ethLinkObj = NULL, *bridgeBrObj, *routerObj;
 	struct json_object *v4AddrJarray, *v6AddrJarray, *v4AddrObj, *v6AddrObj;
 	struct json_object *natIntfJarray, *natIntfObj, *igmpObj, *mldObj;
 	struct json_object *atmLinkJarray, *atmLinkObj;
