Index: libzcfg_fe_dal-1.0/network/vlanGroup/zcfg_fe_dal_vlan_group.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/vlanGroup/zcfg_fe_dal_vlan_group.c	2019-05-17 15:27:59.454793606 +0800
+++ libzcfg_fe_dal-1.0/network/vlanGroup/zcfg_fe_dal_vlan_group.c	2019-05-17 16:14:09.308108497 +0800
@@ -189,14 +189,9 @@
 	const char *intfList = NULL;
 	char untag[32] = {0}, tag[32] = {0};
 
-	idx = json_object_get_int(json_object_object_get(Jobj, "Index"));
-
-	ret = convertIndextoIid(idx, &vlanGrpIid, RDM_OID_VLAN_GROUP, NULL, NULL, replyMsg);
-
-	if(ret != ZCFG_SUCCESS)
-	{
-		return ret;
-	}
+	IID_INIT(vlanGrpIid);
+	vlanGrpIid.level = 1;
+	vlanGrpIid.idx[0] = json_object_get_int(json_object_object_get(Jobj, "Index"));
 
 		vlanID = json_object_get_int(json_object_object_get(Jobj, "VlanId"));
 		tagList = json_object_get_string(json_object_object_get(Jobj, "TagList"));
@@ -244,33 +239,53 @@
 	zcfgRet_t ret = ZCFG_SUCCESS;
 	objIndex_t vlanGrpIid = {0};
 	int index = 0;
+	bool found = false;
+	struct json_object *VlanGroupJobj = NULL;
 
-	index = json_object_get_int(json_object_object_get(Jobj, "Index"));
-
-	ret = convertIndextoIid(index, &vlanGrpIid, RDM_OID_VLAN_GROUP, NULL, NULL, replyMsg);
-
-	if(ret != ZCFG_SUCCESS)
-	{
-		return ret;
+	IID_INIT(vlanGrpIid);
+	vlanGrpIid.level = 1;
+	vlanGrpIid.idx[0] = json_object_get_int(json_object_object_get(Jobj, "Index"));
+
+	if(zcfgFeObjJsonGetWithoutUpdate(RDM_OID_VLAN_GROUP, &vlanGrpIid, &VlanGroupJobj) == ZCFG_SUCCESS) {
+		if(json_object_get_boolean(json_object_object_get(VlanGroupJobj, "Enable")))
+			found = true;
+		zcfgFeJsonObjFree(VlanGroupJobj);
+	}
+
+	if (found) {
+		ret = zcfgFeObjJsonDel(RDM_OID_VLAN_GROUP, &vlanGrpIid, NULL);
+	} else {
+		if (replyMsg != NULL)
+			strcpy(replyMsg, "Can't find the group");
+		return ZCFG_NOT_FOUND;
 	}
 
-	ret = zcfgFeObjJsonDel(RDM_OID_VLAN_GROUP, &vlanGrpIid, NULL);
 	return ret;
-		}
+}
 zcfgRet_t zcfgFeDal_VlanGroup_Get(struct json_object *Jobj, struct json_object *Jarray, char *replyMsg){
 	zcfgRet_t ret = ZCFG_SUCCESS;
 	struct json_object *paramJobj = NULL;
 	struct json_object *VlanJobj = NULL;
+	struct json_object *indexObj = NULL;
 	objIndex_t VlanIid = {0};
+	int indexGet = -1;
 
+	if ((indexObj = json_object_object_get(Jobj, "Index")) != NULL){
+		indexGet = json_object_get_int(indexObj);
+	}
+
+	IID_INIT(VlanIid);
 	while(zcfgFeObjJsonGetNextWithoutUpdate(RDM_OID_VLAN_GROUP, &VlanIid, &VlanJobj) == ZCFG_SUCCESS){
-		paramJobj = json_object_new_object();
-		json_object_object_add(paramJobj, "Index", json_object_new_int(VlanIid.idx[0]));
-		json_object_object_add(paramJobj, "GroupName", JSON_OBJ_COPY(json_object_object_get(VlanJobj, "GroupName")));
-		json_object_object_add(paramJobj, "VlanId", JSON_OBJ_COPY(json_object_object_get(VlanJobj, "VlanId")));
-		json_object_object_add(paramJobj, "IntfList", JSON_OBJ_COPY(json_object_object_get(VlanJobj, "IntfList")));
-		json_object_object_add(paramJobj, "TagList", JSON_OBJ_COPY(json_object_object_get(VlanJobj, "TagList")));
-		json_object_array_add(Jarray, paramJobj);
+		if ((indexGet == -1) || (VlanIid.idx[0] == indexGet))
+		{
+			paramJobj = json_object_new_object();
+			json_object_object_add(paramJobj, "Index", json_object_new_int(VlanIid.idx[0]));
+			json_object_object_add(paramJobj, "GroupName", JSON_OBJ_COPY(json_object_object_get(VlanJobj, "GroupName")));
+			json_object_object_add(paramJobj, "VlanId", JSON_OBJ_COPY(json_object_object_get(VlanJobj, "VlanId")));
+			json_object_object_add(paramJobj, "IntfList", JSON_OBJ_COPY(json_object_object_get(VlanJobj, "IntfList")));
+			json_object_object_add(paramJobj, "TagList", JSON_OBJ_COPY(json_object_object_get(VlanJobj, "TagList")));
+			json_object_array_add(Jarray, paramJobj);
+		}
 		zcfgFeJsonObjFree(VlanJobj);
 	}
 	return ret;
