Index: libzcfg_fe_dal-1.0/network/homeNetworking/zcfg_fe_dal_lansetup.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/homeNetworking/zcfg_fe_dal_lansetup.c	2019-05-14 16:56:00.837763998 +0800
+++ libzcfg_fe_dal-1.0/network/homeNetworking/zcfg_fe_dal_lansetup.c	2019-05-14 16:58:07.377488990 +0800
@@ -360,7 +360,7 @@
 		while(zcfgFeObjJsonGetNext(RDM_OID_IP_IFACE, &iid, &obj) == ZCFG_SUCCESS) {
 			if(!strcmp(json_object_get_string(json_object_object_get(obj, "X_ZYXEL_SrvName")),brtmpname)){
 				brIfName = json_object_get_string(json_object_object_get(obj, "X_ZYXEL_IfName"));
-				break;
+				if(!strncmp(brIfName, "br", 2)) break;
 			}
 			zcfgFeJsonObjFree(obj);
 		}		
