Index: libzcfg_fe_dal-1.0/network/dns/zcfg_fe_dal_d_dns.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/dns/zcfg_fe_dal_d_dns.c	2019-05-27 10:35:20.026150562 +0800
+++ libzcfg_fe_dal-1.0/network/dns/zcfg_fe_dal_d_dns.c	2019-05-27 10:43:25.444557622 +0800
@@ -15,6 +15,8 @@
 	{"Enable", 			dalType_boolean,	0,	0,		NULL, NULL,		0},
 #ifdef ABOG_CUSTOMIZATION
 	{"ServiceProvider", 		dalType_string, 	1, 	32,		NULL, "dyndns|zoneedit|noip|dtdns|freedns|dynv6|userdefined",		0},
+#elif ZYXEL_USA_PRODUCT
+    {"ServiceProvider",         dalType_string,     1,  32,     NULL, "dyndns|zoneedit|noip|dtdns|freedns|userdefined|dnsomatic",       0},
 #else
 	{"ServiceProvider", 		dalType_string, 	1, 	32,		NULL, "dyndns|zoneedit|noip|dtdns|freedns|userdefined",		0},
 #endif
