Index: libzcfg_fe_dal-1.0/network/wireless/zcfg_fe_dal_wifi_general.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/wireless/zcfg_fe_dal_wifi_general.c	2019-05-24 16:59:27.903132400 +0800
+++ libzcfg_fe_dal-1.0/network/wireless/zcfg_fe_dal_wifi_general.c	2019-06-10 14:50:43.269705800 +0800
@@ -1792,7 +1792,14 @@
 
 	IID_INIT(wifiWPSIid);
 	wifiWPSIid.level = 1;
+#ifdef ZYXEL_SUPPORT_8021X
+	if(Index >= 1 && Index <= 4)
+		wifiWPSIid.idx[0] = 1;
+	else if(Index >= 5 && Index <= 8)
+		wifiWPSIid.idx[0] = 5;
+#else
 	wifiWPSIid.idx[0] = Index;
+#endif
 	if((ret = zcfgFeObjJsonGetWithoutUpdate(RDM_OID_WIFI_ACCESS_POINT_WPS, &wifiWPSIid, &wifiWPSObj)) == ZCFG_SUCCESS) {
 		if(json_object_get_boolean(json_object_object_get(wifiWPSObj, "Enable"))){
 			wpsEnabled = true;
