Index: libzcfg_fe_dal-1.0/Makefile
===================================================================
--- libzcfg_fe_dal-1.0.orig/Makefile	2019-07-08 15:38:04.243123000 +0800
+++ libzcfg_fe_dal-1.0/Makefile	2019-07-08 16:09:59.059284189 +0800
@@ -210,6 +210,10 @@
     CFLAGS += -DBUILD_SONIQ
 endif
 
+ifeq ($(CONFIG_ZYXEL_DHCPV6_OPTION64),y)
+	CFLAGS += -DZYXEL_DHCPV6_OPTION64
+endif
+
 ifeq ($(CONFIG_ZYXEL_ONESSID),y)
     CFLAGS += -DZYXEL_ONESSID
 endif
Index: libzcfg_fe_dal-1.0/network/broadband/zcfg_fe_dal_broadband.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/broadband/zcfg_fe_dal_broadband.c	2019-07-08 15:38:04.243123032 +0800
+++ libzcfg_fe_dal-1.0/network/broadband/zcfg_fe_dal_broadband.c	2019-07-08 16:16:29.152376100 +0800
@@ -71,7 +71,11 @@
 	{"BorderRelayIPv4Addresses",dalType_v4Addr,	0,	0,	NULL},
 	//DSLite
 	{"DSLiteEnable",		dalType_boolean,	0,	0,	NULL},
+#ifdef ZYXEL_DHCPV6_OPTION64
+	{"DSLiteType",			dalType_string,	0,	0,	NULL, "dhcp|static"},
+#else
 	{"DSLiteType",			dalType_string,	0,	0,	NULL, "static"},
+#endif
 	{"DSLiteRelayIPv6Addresses",	dalType_v6Addr,	0,	0,	NULL},
 	//ipIface or pppIface
 	{"MaxMTUSize",			dalType_int,	0,	0,	NULL},
