Index: libzcfg_fe_dal-1.0/maintenance/logsetting/zcfg_fe_dal_log_setting.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/maintenance/logsetting/zcfg_fe_dal_log_setting.c	2019-07-08 17:01:07.870994600 +0800
+++ libzcfg_fe_dal-1.0/maintenance/logsetting/zcfg_fe_dal_log_setting.c	2019-07-11 11:07:31.744865421 +0800
@@ -47,6 +47,7 @@
 	{"MACFilter",							dalType_boolean,0,	0,	NULL},
 	{"esmd",								dalType_boolean,0,	0,	NULL},
 	{"zcmd",								dalType_boolean,0,	0,	NULL},
+	{"IGMP",								dalType_boolean,0,	0,	NULL},
 	{NULL,									dalType_int,	0,	0,	NULL},
 
 };
@@ -788,6 +789,17 @@
 			zcfgFeJsonObjFree(categoryObj);
 		}
 	}
+	IID_INIT(categoryIid);
+	if(json_object_object_get(Jobj, "IGMP")){
+		while(zcfgFeObjJsonGetNext(RDM_OID_LOG_CATEGORY, &categoryIid, &categoryObj) == ZCFG_SUCCESS){
+			if(!strcmp(json_object_get_string(json_object_object_get(categoryObj, "Describe")),"IGMP")){
+				json_object_object_add(categoryObj, "Enable", JSON_OBJ_COPY(json_object_object_get(Jobj, "IGMP")));
+				zcfgFeObjJsonSetWithoutApply(RDM_OID_LOG_CATEGORY, &categoryIid, categoryObj, NULL);
+				zcfgFeJsonObjFree(categoryObj);
+			}
+			zcfgFeJsonObjFree(categoryObj);
+		}
+	}
 
 
 	return ret;	
@@ -896,6 +908,7 @@
 	printf("%-4s %-12s %-10s\n", "", "ACL", json_object_get_string(json_object_object_get(obj, "ACL")));
 	printf("%-4s %-12s %-10s\n", "", "Wireless", json_object_get_string(json_object_object_get(obj, "Wireless")));
 	printf("%-4s %-12s %-10s\n", "", "Voice", json_object_get_string(json_object_object_get(obj, "Voice")));
+	printf("%-4s %-12s %-10s\n", "", "IGMP", json_object_get_string(json_object_object_get(obj, "IGMP")));
 	
 	printf("%-2s Security Log \n","");
 	printf("%-4s %-12s %-10s\n", "", "Account", json_object_get_string(json_object_object_get(obj, "Account")));
