Index: libzcfg_fe_dal-1.0/voip/phone/zcfg_fe_dal_phone_region.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/voip/phone/zcfg_fe_dal_phone_region.c	2019-07-19 18:12:42.121915263 +0800
+++ libzcfg_fe_dal-1.0/voip/phone/zcfg_fe_dal_phone_region.c	2019-07-19 18:14:33.496090900 +0800
@@ -16,6 +16,7 @@
 	{"Region",									dalType_VoipRegion,	0,	0,	NULL,	NULL,	NULL},
 	{"Region_Cap",                              dalType_string,	0,	0,	NULL,	NULL,	NULL},
     {"FxsCIDSigProtocol",                       dalType_string,  0,  0,  NULL,  NULL,  NULL},
+    {"FxsCIDMode",                              dalType_string,  0,  0,  NULL,  NULL,  NULL},
 	{NULL,										0,				0,	0,	NULL},
 };
 
@@ -45,6 +46,7 @@
 	printf("Region : %s \n",json_object_get_string(json_object_object_get(obj, "Region")));
 #ifdef ZYXEL_PHONE_CUSTOMIZED_FXS_CLID_PARAMS_SUPPORT
     printf("Caller ID Signal Type : %s \n", json_object_get_string(json_object_object_get(obj, "FxsCIDSigProtocol")));
+    printf("Caller ID Mode : %s \n", json_object_get_string(json_object_object_get(obj, "FxsCIDMode")));
 #endif
 	printf("Call Service Mode : %s \n",serviceMode);
 
@@ -140,6 +142,7 @@
 	multiJobj = zcfgFeJsonObjNew();
     struct json_object *voiceCommonObj = NULL;
     char *callerIDSignalType;
+    char *callerIDMode;
 
 	if((ret = zcfgFeDalVoipFlashAndCountryEdit(Jobj)) != ZCFG_SUCCESS){
 		printf("%s: Edit prof  obj fail \n", __FUNCTION__);
@@ -161,6 +164,18 @@
             }
         }
 
+        if ((callerIDMode = json_object_get_string(json_object_object_get(Jobj, "FxsCIDMode"))) != NULL)
+        {
+            if (!strcmp(callerIDMode, "Default"))
+            {
+                json_object_object_add(voiceCommonObj, "FxsCIDMode", json_object_new_string(""));
+            }
+            else
+            {
+                json_object_object_add(voiceCommonObj, "FxsCIDMode", json_object_new_string(callerIDMode));
+            }
+        }
+
         if (zcfgFeJsonMultiObjAppend(RDM_OID_VOICE_COMMON, &Iid, multiJobj, voiceCommonObj) == NULL)
         {
             printf("%s:(ERROR) fail to append object oid(%d)\n", __FUNCTION__, RDM_OID_VOICE_COMMON);
@@ -210,6 +225,7 @@
 	struct json_object *voiceregionJobj = NULL;
     struct json_object *voiceCommonObj = NULL;
     char *callerIDSignalType;
+    char *callerIDMode;
 
     paramJobj = json_object_new_object();
 
@@ -236,6 +252,19 @@
                 json_object_object_add(paramJobj, "FxsCIDSigProtocol", json_object_new_string(callerIDSignalType));
             }
         }
+
+        if ((callerIDMode = json_object_get_string(json_object_object_get(voiceCommonObj, "FxsCIDMode"))) != NULL)
+        {
+            if (callerIDMode[0] == '\0')
+            {
+                json_object_object_add(paramJobj, "FxsCIDMode", json_object_new_string("Default"));
+            }
+            else
+            {
+                json_object_object_add(paramJobj, "FxsCIDMode", json_object_new_string(callerIDMode));
+            }
+        }
+
         zcfgFeJsonObjFree(voiceCommonObj);
     }
 
