Index: libzcfg_fe_dal-1.0/network/dns/zcfg_fe_dal_d_dns.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/dns/zcfg_fe_dal_d_dns.c	2019-07-24 14:36:24.729316123 +0800
+++ libzcfg_fe_dal-1.0/network/dns/zcfg_fe_dal_d_dns.c	2019-07-24 14:41:31.591633059 +0800
@@ -14,11 +14,11 @@
 {
 	{"Enable", 			dalType_boolean,	0,	0,		NULL, NULL,		0},
 #ifdef ABOG_CUSTOMIZATION
-	{"ServiceProvider", 		dalType_string, 	1, 	32,		NULL, "dyndns|zoneedit|noip|dtdns|freedns|dynv6|userdefined",		0},
+	{"ServiceProvider", 		dalType_string, 	1, 	32,		NULL, "dyndns|zoneedit|noip|dtdns|freedns|dynv6|dynu|userdefined",		0},
 #elif ZYXEL_USA_PRODUCT
-    {"ServiceProvider",         dalType_string,     1,  32,     NULL, "dyndns|zoneedit|noip|dtdns|freedns|userdefined|dnsomatic",       0},
+	{"ServiceProvider", 		dalType_string, 	1,	32,		NULL, "dyndns|zoneedit|noip|dtdns|freedns|dynu|userdefined|dnsomatic",       0},
 #else
-	{"ServiceProvider", 		dalType_string, 	1, 	32,		NULL, "dyndns|zoneedit|noip|dtdns|freedns|userdefined",		0},
+	{"ServiceProvider", 		dalType_string, 	1, 	32,		NULL, "dyndns|zoneedit|noip|dtdns|freedns|dynu|userdefined",		0},
 #endif
 	{"HostName", 			dalType_string, 	0, 	256,	NULL, NULL,		0},
 	{"UserName", 			dalType_string, 	0, 	256, 	NULL, NULL,		0},
