Index: libzcfg_fe_dal-1.0/network/broadband/zcfg_fe_dal_broadband.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/broadband/zcfg_fe_dal_broadband.c	2019-07-23 20:06:16.184863849 +0800
+++ libzcfg_fe_dal-1.0/network/broadband/zcfg_fe_dal_broadband.c	2019-07-23 20:08:16.445052735 +0800
@@ -3548,7 +3548,7 @@
 {
 	zcfgRet_t ret = ZCFG_SUCCESS;
 	struct json_object *wanLanList = NULL, *wanLanObj = NULL;
-	int i, len;
+	int i = 0, len = 0, wanNum = 0;
 	char type[16] = {0};
 
 	isAdd = true;
@@ -3581,6 +3581,21 @@
 			json_object_put(wanLanList);
 			return ZCFG_REQUEST_REJECT;
 		}
+
+		if(strcmp("WWAN", Jgets(wanLanObj, "Name")) != 0 && !strcmp("WAN", Jgets(wanLanObj, "Type")))
+		{
+			wanNum++;
+		}
+	}
+
+	if(wanNum >= WAN_INTERFACES)
+	{
+		json_object_object_add(Jobj, "__multi_lang_replyMsg", json_object_new_string("zylang.Broadband.Broadband.warning.exceed_wan_limit"));
+		if(replyMsg)
+			strcpy(replyMsg,"The number of WAN interfaces exceeds the limit.");
+
+		json_object_put(wanLanList);
+		return ZCFG_REQUEST_REJECT;
 	}
 
 	if(checkWanInterfaceLimit(Jobj, wanLanList, 1) == false){
