Index: libzcfg_fe_dal-1.0/network/homeNetworking/zcfg_fe_dal_lansetup.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/homeNetworking/zcfg_fe_dal_lansetup.c	2019-08-30 15:55:10.534357417 +0800
+++ libzcfg_fe_dal-1.0/network/homeNetworking/zcfg_fe_dal_lansetup.c	2019-08-30 15:55:10.000000000 +0800
@@ -982,14 +982,17 @@
 	const char * v6PrefixAddress = NULL;
 	bool found = false;
 	const char *ip6LanAdrAssign = NULL;
-
+	const char *ipv6Origin = NULL;
 	IID_INIT(v6prefixIid);
 	while(zcfgFeSubInObjJsonGetNext(RDM_OID_IP_IFACE_V6_PREFIX, &curIpIfaceIid, &v6prefixIid, &v6prefixObj)== ZCFG_SUCCESS){
 		ip6Prefix = json_object_get_string(json_object_object_get(v6prefixObj, "Prefix"));
 		ipfacev6Prefix = json_object_get_string(json_object_object_get(curIpIfaceObj, "X_ZYXEL_IPv6Prefix"));
+		ipv6Origin = json_object_get_string(json_object_object_get(v6prefixObj, "Origin"));
 		if(!strcmp(ip6Prefix, ipfacev6Prefix)){
-			found = true;
-			break;
+			if(!strcmp(ipv6Origin, "Static")){
+				found = true;
+				break;
+			}
 		}
 		zcfgFeJsonObjFree(v6prefixObj);
 	}
