Index: libzcfg_fe_dal-1.0/security/parentalControl/zcfg_fe_dal_parentalControl.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/security/parentalControl/zcfg_fe_dal_parentalControl.c	2019-09-27 14:11:52.504873493 +0800
+++ libzcfg_fe_dal-1.0/security/parentalControl/zcfg_fe_dal_parentalControl.c	2019-09-27 14:30:55.583126700 +0800
@@ -6,6 +6,7 @@
 #include "zcfg_common.h"
 #include "zcfg_fe_rdm_string.h"
 #include "zcfg_fe_rdm_struct.h"
+#include "zcfg_fe_schema.h"
 #include "zcfg_rdm_oid.h"
 #include "zcfg_rdm_obj.h"
 #include "zcfg_debug.h"
@@ -867,10 +868,13 @@
 		}
 	}
 	else{
+		int maxlen = zcfgFeObjQueryMaxInstance(RDM_OID_PAREN_CTL_PROF);
+
 		IID_INIT(objIid);
 		if(zcfgFeObjJsonGetWithoutUpdate(RDM_OID_PAREN_CTL, &objIid, &obj) == ZCFG_SUCCESS){
 			pramJobj = json_object_new_object();
 			replaceParam(pramJobj, "PrentalCtlEnable", obj, "Enable");
+			json_object_object_add(pramJobj, "MaxLenPrentalCtlPrf", json_object_new_int(maxlen));
 			json_object_array_add(Jarray, pramJobj);
 			zcfgFeJsonObjFree(obj);
 		}
