Index: libzcfg_fe_dal-1.0/zcfg_fe_dal.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/zcfg_fe_dal.c	2019-10-02 16:18:05.965635823 +0800
+++ libzcfg_fe_dal-1.0/zcfg_fe_dal.c	2019-10-02 16:20:51.811666131 +0800
@@ -1009,7 +1009,7 @@
 	struct json_object *atmLinkJarray, *atmLinkObj;
 	struct json_object *Jobj = NULL;
 	char ifacePath[32] = {0}, BridgingBrPath[32] = {0};
-	int len, len2, i, j;
+	int len, len2, i, j, X_ZYXEL_ConcurrentWan = 0;
 	unsigned char idx;
 	const char *X_ZYXEL_BridgeName, *X_ZYXEL_SrvName, *X_ZYXEL_ConnectionType, *X_ZYXEL_IfName, *X_ZYXEL_Group_WAN_IpIface, *LowerLayers;
 	const char *Status, *ipIfaceStatus, *IPAddress, *X_ZYXEL_ActiveDefaultGateway, *X_ZYXEL_ActiveV6DefaultGateway;
@@ -1064,6 +1064,7 @@
 		ipIfaceObj = json_object_array_get_idx(ipIfaceJarray, i);
 		X_ZYXEL_SrvName = json_object_get_string(json_object_object_get(ipIfaceObj, "X_ZYXEL_SrvName"));
 		X_ZYXEL_ConnectionType = json_object_get_string(json_object_object_get(ipIfaceObj, "X_ZYXEL_ConnectionType"));
+		X_ZYXEL_ConcurrentWan = json_object_get_int(json_object_object_get(ipIfaceObj, "X_ZYXEL_ConcurrentWan"));
 		LowerLayers = json_object_get_string(json_object_object_get(ipIfaceObj, "LowerLayers"));
 		X_ZYXEL_Group_WAN_IpIface = json_object_get_string(json_object_object_get(ipIfaceObj, "X_ZYXEL_Group_WAN_IpIface"));
 		sprintf(ifacePath, "IP.Interface.%u", i+1);
@@ -1087,6 +1088,7 @@
 		Jaddi(Jobj, "VLANPriority", -1);
 		Jaddi(Jobj, "VLANID", -1);
 		json_object_object_add(Jobj, "ConnectionType", json_object_new_string(X_ZYXEL_ConnectionType));
+		json_object_object_add(Jobj, "ConcurrentWan", json_object_new_int(X_ZYXEL_ConcurrentWan));		
 		json_object_object_add(Jobj, "BindToIntfGrp", json_object_new_boolean(false));
 		json_object_object_add(Jobj, "LowerLayerUp", json_object_new_boolean(false));
 		json_object_object_add(Jobj, "WANConnectionReady", json_object_new_boolean(false));
