Index: libzcfg_fe_dal-1.0/network/wireless/zcfg_fe_dal_wifi_general.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/wireless/zcfg_fe_dal_wifi_general.c	2019-12-13 18:12:29.342199254 +0800
+++ libzcfg_fe_dal-1.0/network/wireless/zcfg_fe_dal_wifi_general.c	2019-12-13 18:18:30.344612000 +0800
@@ -2503,7 +2503,8 @@
 		arrLen = json_object_array_length(Jobj);
 		if(arrLen == 1) {
 			apply = true;
-			ret = zcfgFeDalWifiGeneralEdit(Jobj, NULL, apply);
+			struct json_object *oneObj = json_object_array_get_idx(Jobj, 0);
+			ret = zcfgFeDalWifiGeneralEdit(oneObj, NULL, apply);
 		}
 		else {
 			for (idx = 0; idx < arrLen; idx++) {
