Index: pure-ftpd-1.0.30/src/ftpd.c
===================================================================
--- pure-ftpd-1.0.30.orig/src/ftpd.c	2019-08-30 16:34:08.779575345 +0800
+++ pure-ftpd-1.0.30/src/ftpd.c	2019-08-30 16:37:23.770433780 +0800
@@ -4533,6 +4533,10 @@
             }
         }
     }
+    else if(strstr(workingfolder, "/mnt/usb") != NULL || strstr(workingfolder, "/usb") != NULL)
+    {
+        /*the file is uploading to usb disk, continue*/
+    }
     else
     {
         if (ulhandler->total_uploaded > 4*1024*1024 || *uploaded > 4*1024*1024 || 
