--- miniupnpd-1.6/upnpdescstrings.h	2007-02-09 18:13:21.000000000 +0800
+++ miniupnpd-1.6/upnpdescstrings.h	2012-03-20 18:02:22.566674222 +0800
@@ -10,12 +10,12 @@
 #include "config.h"
 
 /* strings used in the root device xml description */
-#define ROOTDEV_FRIENDLYNAME		OS_NAME " router"
-#define ROOTDEV_MANUFACTURER		OS_NAME
-#define ROOTDEV_MANUFACTURERURL		OS_URL
-#define ROOTDEV_MODELNAME			OS_NAME " router"
-#define ROOTDEV_MODELDESCRIPTION	OS_NAME " router"
-#define ROOTDEV_MODELURL			OS_URL
+#define ROOTDEV_FRIENDLYNAME		"ZyXEL Router"
+#define ROOTDEV_MANUFACTURER		"ZyXEL Communications Corp."
+#define ROOTDEV_MANUFACTURERURL		"http://www.zyxel.com"
+#define ROOTDEV_MODELNAME		"ZyXEL Router"
+#define ROOTDEV_MODELDESCRIPTION	"ZyXEL Router"
+#define ROOTDEV_MODELURL		""
 
 #define WANDEV_FRIENDLYNAME			"WANDevice"
 #define WANDEV_MANUFACTURER			"MiniUPnP"
