Index: miniupnpd-1.6.20110730/netfilter/iptcrdr.c
===================================================================
--- miniupnpd-1.6.20110730.orig/netfilter/iptcrdr.c	2011-07-30 21:15:24.000000000 +0800
+++ miniupnpd-1.6.20110730/netfilter/iptcrdr.c	2019-06-17 15:02:20.427014672 +0800
@@ -625,7 +625,7 @@
 	struct ip_nat_range * range;
 	size_t size;
 
-	size =   IPT_ALIGN(sizeof(struct ipt_entry_target))
+	size =   IPT_ALIGN(sizeof(struct ipt_standard_target))
 	       + IPT_ALIGN(sizeof(struct ip_nat_multi_range));
 	target = calloc(1, size);
 	target->u.target_size = size;

