diff -urN net-snmp-5.8.org/agent/mibgroup/if-mib/data_access/interface_linux.c net-snmp-5.8/agent/mibgroup/if-mib/data_access/interface_linux.c
--- net-snmp-5.8.org/agent/mibgroup/if-mib/data_access/interface_linux.c	2019-04-08 10:56:29.445078431 +0800
+++ net-snmp-5.8/agent/mibgroup/if-mib/data_access/interface_linux.c	2019-04-08 11:17:37.806242303 +0800
@@ -111,12 +111,12 @@
 unsigned long long
 netsnmp_linux_interface_get_if_speed_linux(int fd, const char *name,
         unsigned long long defaultspeed);
+#endif
 #ifdef HAVE_LINUX_ETHTOOL_H
 unsigned long long
 netsnmp_linux_interface_get_if_speed_mii(int fd, const char *name,
         unsigned long long defaultspeed);
 #endif
-#endif
 
 #define PROC_SYS_NET_IPVx_NEIGH_RETRANS_TIME_MS "/proc/sys/net/ipv%d/neigh/%s/retrans_time_ms"
 #define PROC_SYS_NET_IPVx_NEIGH_RETRANS_TIME    "/proc/sys/net/ipv%d/neigh/%s/retrans_time"
@@ -828,6 +828,19 @@
         int nif;
         for(nif = 0; nif < (sizeof(_notShown_if)/sizeof(char*)); nif++){
             if (0 == strncmp(entry->name, _notShown_if[nif], strlen(_notShown_if[nif]))) {
+                break;
+            }
+        }
+        if(nif < (sizeof(_notShown_if)/sizeof(char*))){
+            netsnmp_access_interface_entry_free(entry);
+            continue;
+        }
+#endif
+#if defined(ECONET_PLATFORM) //__ZYXEL__, JSu
+        static char *_notShown_if[] = {"lo","sit", "ip6tnl"};
+        int nif;
+        for(nif = 0; nif < (sizeof(_notShown_if)/sizeof(char*)); nif++){
+            if (0 == strncmp(entry->name, _notShown_if[nif], strlen(_notShown_if[nif]))) {
                 break;
             }
         }
