diff -urN net-snmp-5.8.org/agent/mibgroup/ZYXEL-PRESTIGE-MIB/prestigeLAN.c net-snmp-5.8/agent/mibgroup/ZYXEL-PRESTIGE-MIB/prestigeLAN.c
--- net-snmp-5.8.org/agent/mibgroup/ZYXEL-PRESTIGE-MIB/prestigeLAN.c	2019-04-08 15:54:33.945195400 +0800
+++ net-snmp-5.8/agent/mibgroup/ZYXEL-PRESTIGE-MIB/prestigeLAN.c	2019-04-08 17:53:36.749414155 +0800
@@ -452,6 +452,7 @@
 {
 	int ret = SNMP_ERR_NOERROR, enbl;
 	char ifname[10];
+    bool check_requestvb = false;
 	
 	ZYSNMP_DEBUG("enter\n");
 	
@@ -460,7 +461,13 @@
 		case MODE_GETBULK:
 			sprintf(ifname,"br%d", 0);
 			ret = handleOfLanDhcpRelayParam(ifname, "Enable", requests, ASN_INTEGER , HANDLE_GET);
-			if(ret == SNMP_ERR_NOERROR && *requests->requestvb->val.integer == 1){
+			
+            if(requests->requestvb->val.integer){
+                if(*requests->requestvb->val.integer == 1)
+                   check_requestvb = true;
+            }
+           
+            if(ret == SNMP_ERR_NOERROR && check_requestvb){
 				*requests->requestvb->val.integer = DHCP_RELAY;
 			}
 			else{
