Index: wide-dhcpv6-20080615/dhcp6c.c
===================================================================
--- wide-dhcpv6-20080615.orig/dhcp6c.c	2019-05-21 14:09:59.337098185 +0800
+++ wide-dhcpv6-20080615/dhcp6c.c	2019-05-21 14:13:43.381852632 +0800
@@ -3074,7 +3074,7 @@
 int dhcp6sConfDnsListUpdate(struct dhcp6_optinfo *opt)
 {
 	struct dhcp6_listval *d;
-	char nameserver[128] = {0};
+	char nameserver[192] = {0};
 	char domainName[64] = {0};
 
 	dprintf(LOG_ERR, FNAME, "%s : Enter\n", __FUNCTION__);
@@ -3087,6 +3087,12 @@
 		if (d != NULL) {
 			strcat(nameserver, ",");
 			strcat(nameserver, in6addr2str(&d->val_addr6, 0));
+
+			d = TAILQ_NEXT(d, link);
+			if (d != NULL) {
+				strcat(nameserver, ",");
+				strcat(nameserver, in6addr2str(&d->val_addr6, 0));
+			}
 		}
 	}
 
