Index: wide-dhcpv6-20080615/dhcp6c_script.c
===================================================================
--- wide-dhcpv6-20080615.orig/dhcp6c_script.c	2019-07-08 16:51:51.219009016 +0800
+++ wide-dhcpv6-20080615/dhcp6c_script.c	2019-07-17 08:49:04.519111900 +0800
@@ -433,13 +433,17 @@
 			    "script \"%s\" terminated", scriptpath);
 		}
 	} else {
+#if 1////def ZYXEL_DHCPV6_OPTION64
+		char *argv[3];
+#else
 		char *argv[2];
+#endif
 		int fd;
 
 		argv[0] = scriptpath;
 #if 1////def ZYXEL_DHCPV6_OPTION64
 		argv[1] = "/var/run/dhcpv6DSlite.info";
-		//argv[2] = NULL;
+		argv[2] = NULL;
 #else
 		argv[1] = NULL;
 #endif
