Index: libzcfg_fe_rdm_string-1.0/zcfg_fe_rdm_string.c
===================================================================
--- libzcfg_fe_rdm_string-1.0.orig/zcfg_fe_rdm_string.c	2018-10-11 11:12:26.432377677 +0800
+++ libzcfg_fe_rdm_string-1.0/zcfg_fe_rdm_string.c	2018-10-11 11:15:55.030204405 +0800
@@ -565,7 +565,7 @@
 		case json_type_uint16:
 		case json_type_uint32:
 		case json_type_ulong:
-			setValue = json_object_new_int((uint32_t)atoi(value));
+			setValue = json_object_new_int(strtoul(value, NULL, 10));
 			break;
 		case json_type_string:
 		case json_type_hex:
