Index: linux-3.18.21/init/main.c
===================================================================
--- linux-3.18.21.orig/init/main.c	2018-05-16 12:49:43.744866036 +0800
+++ linux-3.18.21/init/main.c	2018-05-17 00:52:46.977004000 +0800
@@ -969,7 +969,8 @@
 		pr_err("Failed to execute %s (error %d).  Attempting defaults...\n",
 			execute_command, ret);
 	}
-	if (!try_to_run_init_process("/sbin/init") ||
+	if (!try_to_run_init_process("/lib/preinit/preinit") ||
+	    !try_to_run_init_process("/sbin/init") ||
 	    !try_to_run_init_process("/etc/init") ||
 	    !try_to_run_init_process("/bin/init") ||
 	    !try_to_run_init_process("/bin/sh"))
