Index: linux-3.18.21/include/linux/netlink.h
===================================================================
--- linux-3.18.21.orig/include/linux/netlink.h	2018-04-27 16:51:10.094714844 +0800
+++ linux-3.18.21/include/linux/netlink.h	2018-04-30 18:43:56.476543767 +0800
@@ -8,6 +8,8 @@
 #include <net/scm.h>
 #include <uapi/linux/netlink.h>
 
+#define NETLINK_MONITOR 25 /*send events to userspace monitor task*/
+
 struct net;
 
 static inline struct nlmsghdr *nlmsg_hdr(const struct sk_buff *skb)
