Index: linux-2.6.36/include/linux/netfilter/xt_AUTOMAP.h
===================================================================
--- linux-2.6.36.orig/include/linux/netfilter/xt_AUTOMAP.h	2017-04-27 16:29:27.315833318 +0800
+++ linux-2.6.36/include/linux/netfilter/xt_AUTOMAP.h	2017-04-27 16:43:57.646781314 +0800
@@ -27,33 +27,32 @@
 		/*
 		Auto Priority Mapping Table
 
-
-			DSCP	|   Packet Length	| 802.1P	|     Queue     |
-		   ---------------------------------------------
-					|				|   001	|       0 		|
-					|				|		|			|
-					|				|   010	|       1		|
-					|				|		|			|
-			0x00       |       >1100		|   000	|       2		|
-					|				|		|			|
-			0x08       |       250-1100	|   011	|       3		|
-					|				|		|			|
-			0x10       |				|   100	|       4		|
-					|				|		|			|
-			0x18       |       <250		|   101	|       5		|
-					|				|		|			|
-		0x20,0x28	|				|   110	|       6		|
-					|				|		|			|
-		0x30,0x38  	|				|   111	|       7		|
+              DSCP  |  Packet Length  |  802.1P  |   Queue   |   Priority  |
+           ------------------------------------------------------------------
+             0x00   |                 |    001   |     0     |      4      |
+                    |                 |          |           |             |
+             0x08   |       >1100     |    010   |     1     |      4      |
+                    |                 |          |           |             |
+             0x18   |                 |    000   |     2     |      3      |
+                    |                 |          |           |             |
+             0x22   |     250-1100    |    011   |     3     |      3      |
+                    |                 |          |           |             |
+             0x26   |                 |    100   |     4     |      2      |
+                    |                 |          |           |             |
+             0x30   |       <250      |    101   |     5     |      2      |
+                    |                 |          |           |             |
+             0x34   |                 |    110   |     6     |      1      |
+                    |                 |          |           |             |
+             0x38   |                 |    111   |     7     |      1      |
 		*/
 
 
 /* accoding to tr181 8021p to DSCP mapping table(upstream) higher value higher priority */
 unsigned short vlan8021pToDSCP[8]= {0x00,0x00,0x00,0x08,0x10,0x18,0x28,0x38};
-unsigned short dscpPrecedenceTo8021p[8] = {0,3,4,5,6,6,7,7};
+unsigned short dscpPrecedenceTo8021p[8] = {0x40,0x40,0x30,0x30,0x20,0x20,0x10,0x10};
 
-unsigned short vlan8021pToPriorityQueue[8] = {2,0,1,3,4,5,6,7};
-unsigned short dscpPrecedenceToPriorityQueue[8] = {2,3,4,5,6,6,7,7};
+unsigned short vlan8021pToPriorityQueue[8] = {0x30,0x40,0x40,0x30,0x20,0x20,0x10,0x10};
+unsigned short dscpPrecedenceToPriorityQueue[8] = {0x40,0x40,0x30,0x30,0x20,0x20,0x10,0x10};
 
 /* target info */
 struct xt_automap_target_info {
Index: linux-2.6.36/net/netfilter/xt_AUTOMAP.c
===================================================================
--- linux-2.6.36.orig/net/netfilter/xt_AUTOMAP.c	2017-04-27 16:29:27.351851318 +0800
+++ linux-2.6.36/net/netfilter/xt_AUTOMAP.c	2017-04-27 16:43:57.646781314 +0800
@@ -137,16 +137,16 @@
 
 //					if(iph->tot_len > 1100){
 					if(ih->tot_len > 1100){
-						skb->mark|=vlan8021pToPriorityQueue[0]; /* queue priority 2 */
-//                    }else if(iph->tot_len < 250){
-                    }else if(ih->tot_len < 250){
-						skb->mark|=vlan8021pToPriorityQueue[5]; /* queue priority 5 */
-                    }else{  /*250~1100*/
-						skb->mark|=vlan8021pToPriorityQueue[3]; /* queue priority 3 */
-                    }
+						skb->mark|=vlan8021pToPriorityQueue[1]; /* priority 4 */
+//					}else if(iph->tot_len < 250){
+					}else if(ih->tot_len < 250){
+						skb->mark|=vlan8021pToPriorityQueue[6]; /* priority 3 */
+					}else{  /*250~1100*/
+						skb->mark|=vlan8021pToPriorityQueue[3]; /* priority 1 */
+					}
 					break;
 				default:
-					printk("unknown AUTOMAP type: \n");
+					printk("unknown AUTOMAP type\n");
 					break;
 			}
 		}
