Index: linux-3.18.21/drivers/usb/host/Kconfig
===================================================================
--- linux-3.18.21.orig/drivers/usb/host/Kconfig	2018-06-12 10:12:20.285168696 +0800
+++ linux-3.18.21/drivers/usb/host/Kconfig	2018-06-12 10:22:40.463731847 +0800
@@ -32,8 +32,8 @@
        default y
 
 config USB_XHCI_PLATFORM
-	bool "xHCI platform"
-	default y
+	tristate "Generic xHCI driver for a platform device"
+	default n
 
 config USB_EN7512_XHCI_HCD
 	bool "ECONET EN7512 xHCI"
