Index: linux-3.18.21/include/uapi/linux/netfilter/nf_nat.h
===================================================================
--- linux-3.18.21.orig/include/uapi/linux/netfilter/nf_nat.h	2018-06-13 14:59:33.314680877 +0800
+++ linux-3.18.21/include/uapi/linux/netfilter/nf_nat.h	2018-06-13 15:00:16.091522834 +0800
@@ -24,6 +24,14 @@
 	__be32				max_ip;
 	union nf_conntrack_man_proto	min;
 	union nf_conntrack_man_proto	max;
+
+#if 1 /* Support one-to-one port mapping. __TELEFONICA__, ZyXEL Stan Su, 20120307. */
+	/* mappingFlag is 1, if it needs to do one-to-one port mapping. */
+	unsigned int mappingFlag;
+
+	/* Record external port information. */
+	union nf_conntrack_man_proto mappingMin, mappingMax;
+#endif
 };
 
 struct nf_nat_ipv4_multi_range_compat {
