Index: linux-3.18.21/kernel_ext/drivers/mtd/chips/spi_nand_flash.c
===================================================================
--- linux-3.18.21.orig/kernel_ext/drivers/mtd/chips/spi_nand_flash.c	2018-07-26 21:48:08.057843830 +0800
+++ linux-3.18.21/kernel_ext/drivers/mtd/chips/spi_nand_flash.c	2018-07-26 22:24:59.869752324 +0800
@@ -3247,7 +3247,9 @@
 	//_SPI_NAND_PRINTF( "read page %d(block %d and page %d) with len %d \n\n",page_number,bmt_block,bmt_page, len);
 	//_SPI_NAND_PRINTF( "oob size is %d\n", ptr_dev_info_t->oob_size);
 
-	if( spi_nand_read_page(page_number, SPI_NAND_FLASH_READ_SPEED_MODE_SINGLE) == SPI_NAND_FLASH_RTN_NO_ERROR)
+	SPI_NAND_Flash_Clear_Read_Cache_Data();
+
+	if (spi_nand_read_page(page_number, ptr_dev_info_t->read_mode) == SPI_NAND_FLASH_RTN_NO_ERROR)
  	{
 		//memcpy( oob_buf,  &_current_cache_page_oob_mapping[0], len );
 		memcpy( oob_buf,  &_current_cache_page_oob[0], len );
