Index: linux-3.18.21/drivers/mtd/chips/spi_nand_flash.c
===================================================================
--- linux-3.18.21.orig/drivers/mtd/chips/spi_nand_flash.c	2018-11-12 15:19:33.331080000 +0800
+++ linux-3.18.21/drivers/mtd/chips/spi_nand_flash.c	2018-11-12 15:27:02.222172592 +0800
@@ -4130,6 +4130,7 @@
 
 	_SPI_NAND_PRINTF("en7512_nand_mark_badblock: page_num=0x%x\n", page_number);
 
+	spi_nand_erase_block((offset / (ptr_dev_info_t->erase_size)));
 	rtn_status = spi_nand_write_page(page_number, 0, NULL, 0, 0, &buf[0], 8, ptr_dev_info_t->write_mode);
 	
 	if( rtn_status == SPI_NAND_FLASH_RTN_NO_ERROR)
