Index: linux-3.18.21/arch/mips/include/asm/tc3162/TCIfSetQuery_os.h
===================================================================
--- linux-3.18.21.orig/arch/mips/include/asm/tc3162/TCIfSetQuery_os.h	2019-04-18 16:08:06.418096684 +0800
+++ linux-3.18.21/arch/mips/include/asm/tc3162/TCIfSetQuery_os.h	2019-04-18 16:17:38.410600100 +0800
@@ -231,6 +231,10 @@
 #define    		ME_CMD_VDSL2_GINP					0x0E
 #define    		ME_CMD_ADSL2_GINP					0x0F
 #define    		ME_CMD_ADSL2PLUS_GINP					0x10
+#define         ME_CMD_GVECTOR_GINP_US_ONLY         0x11
+#define         ME_CMD_GVECTOR_GINP_DS_ONLY         0x12
+#define         ME_CMD_VDSL2_GINP_US_ONLY           0x13
+#define         ME_CMD_VDSL2_GINP_DS_ONLY           0x14
 
 //xzwang
 #define         ME_CMD_ADSL_ANSI                    0x05
