Index: linux-3.18.21/net/core/ecnt_skb_alloc.c
===================================================================
--- linux-3.18.21.orig/net/core/ecnt_skb_alloc.c	2019-04-19 11:43:35.087695775 +0800
+++ linux-3.18.21/net/core/ecnt_skb_alloc.c	2019-04-19 13:18:06.255385074 +0800
@@ -48,6 +48,8 @@
 int skbmgr_max_alloc_no = 0;
 int skbmgr_alloc_fail = 0;
 
+int skbmgr_2k_truesize = 0;
+int skbmgr_4k_truesize = 0;
 
 int skbmgr_alloc_normal = 0;
 
@@ -119,7 +121,10 @@
 		if (likely(skb)) {
 			skb->skb_recycling_callback = skbmgr_recycling_callback;
 			skb->skb_recycling_ind = SKBMGR_INDICATION;
-
+			if(0 == skbmgr_2k_truesize)
+			{
+				skbmgr_2k_truesize = skb->truesize;
+			}
 			atomic_inc(&skbmgr_alloc_no);
 			alloc_no = atomic_read(&skbmgr_alloc_no);
 			if (alloc_no > skbmgr_max_alloc_no)
@@ -143,7 +148,7 @@
 	list = &skbmgr_pool[SKBMGR_QUEUE_ID].list;	
 
 	if (skb_queue_len(list) < skbmgr_hot_list_len) {
-		if (/*(skb->truesize - sizeof(struct sk_buff) != SKBMGR_RX_BUF_LEN) ||*/
+		if ((skbmgr_2k_truesize != skb->truesize) ||
 			skb_is_nonlinear(skb) ||
 			(skb->fclone != SKB_FCLONE_UNAVAILABLE) ||
 			skb_cloned(skb) ||
@@ -304,6 +309,10 @@
 		if (likely(skb)) {
 			skb->skb_recycling_callback = skbmgr_4k_recycling_callback;
 			skb->skb_recycling_ind = SKBMGR_4K_INDICATION;
+			if(0 == skbmgr_4k_truesize)
+			{
+				skbmgr_4k_truesize = skb->truesize;
+			}
 			atomic_inc(&skbmgr_4k_alloc_no);
 			alloc_no = atomic_read(&skbmgr_4k_alloc_no);
 			if (alloc_no > skbmgr_4k_max_alloc_no)
@@ -327,7 +336,7 @@
 	list = &skbmgr_4k_pool[SKBMGR_QUEUE_ID].list;	
 
 	if (skb_queue_len(list) < skbmgr_4k_hot_list_len) {
-		if (/*(skb->truesize - sizeof(struct sk_buff) != SKBMGR_4K_RX_BUF_LEN) ||*/
+		if ((skbmgr_4k_truesize != skb->truesize) ||
 			(skb->fclone != SKB_FCLONE_UNAVAILABLE) ||
 			skb_cloned(skb) ||
 			(skb_shinfo(skb)->nr_frags) || 
@@ -729,6 +738,9 @@
 	out += sprintf(out, "skbmgr_4k_alloc_no = %d\n", atomic_read(&skbmgr_4k_alloc_no));
 	out += sprintf(out, "skbmgr_4k_max_list_len = %d\n", skbmgr_4k_max_list_len);
 	out += sprintf(out, "skbmgr_4k_alloc_normal = %d\n", skbmgr_4k_alloc_normal);
+	out += sprintf(out, "skbmgr_2k_truesize = %d\n",  skbmgr_2k_truesize);
+	out += sprintf(out, "skbmgr_4k_truesize = %d\n",  skbmgr_4k_truesize);
+	
 #if defined(TC3262_GMAC_SG_MODE) || defined(TC3262_PTM_SG_MODE)
 	out += sprintf(out, "skbmgr_sg_max_list_len = %d\n", skbmgr_sg_max_list_len);
 #endif
