Index: linux-3.18.21/include/linux/netfilter/xt_AUTOMAP.h
===================================================================
--- linux-3.18.21.orig/include/linux/netfilter/xt_AUTOMAP.h	2019-06-14 15:42:24.659450139 +0800
+++ linux-3.18.21/include/linux/netfilter/xt_AUTOMAP.h	2019-06-14 15:51:34.426965500 +0800
@@ -31,9 +31,9 @@
            ------------------------------------------------------------------
              0x00   |                 |    001   |     0     |      4      |
                     |                 |          |           |             |
-             0x08   |       >1100     |    010   |     1     |      4      |
+             0x08   |       >1100     |    000   |     1     |      4      |
                     |                 |          |           |             |
-             0x18   |                 |    000   |     2     |      3      |
+             0x18   |                 |    010   |     2     |      3      |
                     |                 |          |           |             |
              0x22   |     250-1100    |    011   |     3     |      3      |
                     |                 |          |           |             |
@@ -51,7 +51,7 @@
 unsigned short vlan8021pToDSCP[8]= {0x00,0x00,0x00,0x08,0x10,0x18,0x28,0x38};
 unsigned short dscpPrecedenceTo8021p[8] = {0x40,0x40,0x30,0x30,0x20,0x20,0x10,0x10};
 
-unsigned short vlan8021pToPriorityQueue[8] = {0x30,0x40,0x40,0x30,0x20,0x20,0x10,0x10};
+unsigned short vlan8021pToPriorityQueue[8] = {0x40,0x40,0x30,0x30,0x20,0x20,0x10,0x10};
 unsigned short dscpPrecedenceToPriorityQueue[8] = {0x40,0x40,0x30,0x30,0x20,0x20,0x10,0x10};
 
 /* target info */
Index: linux-3.18.21/include/linux/netfilter_bridge/ebt_AUTOMAP.h
===================================================================
--- linux-3.18.21.orig/include/linux/netfilter_bridge/ebt_AUTOMAP.h	2019-06-14 15:42:24.703450139 +0800
+++ linux-3.18.21/include/linux/netfilter_bridge/ebt_AUTOMAP.h	2019-06-14 15:51:21.428465500 +0800
@@ -13,46 +13,33 @@
 
 
 		/*
-              		DSCP  |  Packet Length  |  802.1P  |   Queue   |   Priority  |
+		Auto Priority Mapping Table
 
-           		------------------------------------------------------------------
-
-             		0x00   |                 |    001   |     0     |      4      |
-
-             		       |                 |          |           |             |
-
-             		0x08   |       >1100     |    010   |     1     |      4      |
-
-             		       |                 |          |           |             |
-
-             		0x10   |                 |    000   |     2     |      3      |
-
-             		       |                 |          |           |             |
-
-             		0x18   |     250-1100    |    011   |     3     |      3      |
-
-             		       |                 |          |           |             |
-
-             		0x20   |                 |    100   |     4     |      2      |
-
-             		       |                 |          |           |             |
-
-             		0x28   |       <250      |    101   |     5     |      2      |
-
-             		       |                 |          |           |             |
-
-             		0x30   |                 |    110   |     6     |      1      |
-
-             		       |                 |          |           |             |
-
-             		0x38   |                 |    111   |     7     |      1      |
+              DSCP  |  Packet Length  |  802.1P  |   Queue   |   Priority  |
+           ------------------------------------------------------------------
+             0x00   |                 |    001   |     0     |      4      |
+                    |                 |          |           |             |
+             0x08   |       >1100     |    000   |     1     |      4      |
+                    |                 |          |           |             |
+             0x10   |                 |    010   |     2     |      3      |
+                    |                 |          |           |             |
+             0x18   |     250-1100    |    011   |     3     |      3      |
+                    |                 |          |           |             |
+             0x20   |                 |    100   |     4     |      2      |
+                    |                 |          |           |             |
+             0x28   |       <250      |    101   |     5     |      2      |
+                    |                 |          |           |             |
+             0x30   |                 |    110   |     6     |      1      |
+                    |                 |          |           |             |
+             0x38   |                 |    111   |     7     |      1      |
 		*/
 		
 
 /* accoding to tr181 8021p to DSCP mapping table(upstream), higher value higher priority*/
 unsigned short vlan8021pToDSCP[8]= {0x00,0x00,0x00,0x08,0x10,0x18,0x28,0x38};
 unsigned short dscpPrecedenceTo8021p[8] = {0x40,0x40,0x30,0x30,0x20,0x20,0x10,0x10};
-unsigned short vlan8021pToPriorityQueue[8] = {0x30,0x40,0x40,0x30,0x20,0x20,0x10,0x10};
+
+unsigned short vlan8021pToPriorityQueue[8] = {0x40,0x40,0x30,0x30,0x20,0x20,0x10,0x10};
 unsigned short dscpPrecedenceToPriorityQueue[8] = {0x40,0x40,0x30,0x30,0x20,0x20,0x10,0x10};
 
 /* target info */
