Index: linux-3.18.21/include/uapi/linux/netfilter/nf_nat.h
===================================================================
--- linux-3.18.21.orig/include/uapi/linux/netfilter/nf_nat.h	2015-09-01 06:19:23.000000000 +0800
+++ linux-3.18.21/include/uapi/linux/netfilter/nf_nat.h	2018-05-30 19:23:03.361734552 +0800
@@ -37,6 +37,13 @@
 	union nf_inet_addr		max_addr;
 	union nf_conntrack_man_proto	min_proto;
 	union nf_conntrack_man_proto	max_proto;
+#if 1 /* Support one-to-one port mapping. __TELEFONICA__, ZyXEL Stan Su, 20120307. */
+	/* mappingFlag is 1, if it needs to do one-to-one port mapping. */
+	unsigned int			mappingFlag;
+
+	/* Record external port information. */
+	union nf_conntrack_man_proto	mappingMin, mappingMax;
+#endif
 };
 
 #endif /* _NETFILTER_NF_NAT_H */
Index: linux-3.18.21/net/netfilter/nf_nat_core.c
===================================================================
--- linux-3.18.21.orig/net/netfilter/nf_nat_core.c	2015-09-01 06:19:23.000000000 +0800
+++ linux-3.18.21/net/netfilter/nf_nat_core.c	2018-05-30 19:27:39.515092818 +0800
@@ -342,7 +342,11 @@
 	/* Only bother mapping if it's not already in range and unique */
 	if (!(range->flags & NF_NAT_RANGE_PROTO_RANDOM_ALL)) {
 		if (range->flags & NF_NAT_RANGE_PROTO_SPECIFIED) {
-			if (l4proto->in_range(tuple, maniptype,
+			/* Support one-to-one port mapping. __TELEFONICA__, ZyXEL Stan Su, 20120308. */
+			/* If external port range is 1000-2000, and internal port range is 1500-2500. */
+			/* In this case, in_range() will return trun when the incoming traffic uses port 1555. */
+			/* Therefore, only check in_range() when mappingFlag is false. */
+			if (!range->mappingFlag && l4proto->in_range(tuple, maniptype,
 					      &range->min_proto,
 					      &range->max_proto) &&
 			    (range->min_proto.all == range->max_proto.all ||
Index: linux-3.18.21/net/netfilter/nf_nat_proto_common.c
===================================================================
--- linux-3.18.21.orig/net/netfilter/nf_nat_proto_common.c	2015-09-01 06:19:23.000000000 +0800
+++ linux-3.18.21/net/netfilter/nf_nat_proto_common.c	2018-05-30 19:29:36.648740198 +0800
@@ -45,6 +45,9 @@
 	__be16 *portptr;
 	u_int16_t off;
 
+#if 1 /* Support one-to-one port mapping. __TELEFONICA__, ZyXEL Stan Su, 20120308. */
+	u_int16_t mappingOffset = 0;
+#endif
 	if (maniptype == NF_NAT_MANIP_SRC)
 		portptr = &tuple->src.u.all;
 	else
@@ -84,7 +87,20 @@
 		off = *rover;
 	}
 
+#if 1 /* Support one-to-one port mapping. __TELEFONICA__, ZyXEL Stan Su, 20120308. */
+	mappingOffset = ntohs(tuple->dst.u.all) - ntohs(range->mappingMin.all);
+	for (i = 0; ; ++off, mappingOffset++) {
+#else
 	for (i = 0; ; ++off) {
+#endif
+#if 1 /* Support one-to-one port mapping. __TELEFONICA__, ZyXEL Stan Su, 20120308. */
+		/* if range->flags & IP_NAT_RANGE_PROTO_RANDOM is true, portprt should be counted by off. */
+		if (range->mappingFlag == 1 && !(range->flags & NF_NAT_RANGE_PROTO_RANDOM))
+		{
+			*portptr = htons(min + mappingOffset % range_size);
+		}
+		else
+#endif
 		*portptr = htons(min + off % range_size);
 		if (++i != range_size && nf_nat_used_tuple(tuple, ct))
 			continue;
