Index: linux-3.18.21/net/ipv6/ecnt_net_ipv6.h
===================================================================
--- linux-3.18.21.orig/net/ipv6/ecnt_net_ipv6.h	2018-05-31 14:56:50.221373708 +0800
+++ linux-3.18.21/net/ipv6/ecnt_net_ipv6.h	2018-05-31 14:59:20.362118216 +0800
@@ -250,7 +250,8 @@
 
 static inline int ecnt_addrconf_dev_config_if_inline_hook(struct net_device *dev, struct inet6_dev *idev)
 {
-#ifdef TCSUPPORT_IPV6_ENHANCEMENT
+//#ifdef TCSUPPORT_IPV6_ENHANCEMENT
+#if 0	//Redmine 30553, need to generate link local address according to EUI-64
 		struct in6_addr addr;
 		//Disable lan device add linklocal address,except br0
 			if(is_wan_dev(dev) || (strstr(dev->name, "br") != NULL)){
