Index: linux-3.18.21/net/bridge/br_fdb.c
===================================================================
--- linux-3.18.21.orig/net/bridge/br_fdb.c	2018-05-31 18:24:01.882733074 +0800
+++ linux-3.18.21/net/bridge/br_fdb.c	2018-05-31 18:36:32.189506919 +0800
@@ -572,6 +572,11 @@
 		return;
 
 	fdb = fdb_find_rcu(head, addr, vid);
+	/* if packets come frome WAN with LAN device MAC, don't update brctl MAC table */
+	if(fdb){
+		if((!strncmp(fdb->dst->dev->name, "eth", 3) || !strncmp(fdb->dst->dev->name, "ra", 2)) && (!strncmp(source->dev->name, "nas", 3)))
+			return;
+	}
 	if (likely(fdb)) {
 		/* attempt to update an entry for a local interface */
 		if (unlikely(fdb->is_local)) {
