Index: linux-3.18.21/arch/mips/econet/tcwdog.c
===================================================================
--- linux-3.18.21.orig/arch/mips/econet/tcwdog.c	2018-05-31 18:05:28.755770766 +0800
+++ linux-3.18.21/arch/mips/econet/tcwdog.c	2018-05-31 18:08:16.766627295 +0800
@@ -227,7 +227,7 @@
 		printk("Watchdog cannot be stopped once started. \n");
 	}
   
-	timerSet(5, 2000 * TIMERTICKS_10MS, ENABLE, TIMER_TOGGLEMODE, TIMER_HALTDISABLE);
+	timerSet(5, 6000 * TIMERTICKS_10MS, ENABLE, TIMER_TOGGLEMODE, TIMER_HALTDISABLE);
 	timer_WatchDogConfigure(ENABLE, ENABLE);
 
 	watchdog_enabled=1;
