Index: linux-3.18.21/drivers/net/ppp/ppp_generic.c
===================================================================
--- linux-3.18.21.orig/drivers/net/ppp/ppp_generic.c	2018-07-27 10:55:16.593454772 +0800
+++ linux-3.18.21/drivers/net/ppp/ppp_generic.c	2018-07-30 17:11:14.521799586 +0800
@@ -2771,6 +2771,11 @@
       num[2] = u<<(32-(FIELD2))>>(32-FIELD2);
       if (num[1] == 0)
       {
+	#if defined(CONFIG_ZYXEL_PPTP_SUPPORT)/* __ZyXEL__, MelissaChen, 20171122, Support PPTP VPN.  */
+	 if (num[2] == 11 ){
+	  sprintf(dev->name, "ppp%d", num[2]);
+	 }else
+	#endif
          sprintf(dev->name, "ppp%d", num[0]);
       }
       else if (num[1] == 1) /* vlan mux */
Index: linux-3.18.21/net/netfilter/Kconfig
===================================================================
--- linux-3.18.21.orig/net/netfilter/Kconfig	2018-07-27 10:55:16.785454777 +0800
+++ linux-3.18.21/net/netfilter/Kconfig	2018-07-30 17:11:58.311343879 +0800
@@ -1506,6 +1506,9 @@
 config ZYXEL_NF_SESSION_CTL
         bool "ZyXEL feature - netfilter session control"
 
+config ZYXEL_PPTP_SUPPORT
+        bool "Support PPTP VPN"
+
 endmenu
 
 source "net/netfilter/ipset/Kconfig"
