Index: linux-3.18.21/arch/mips/include/asm/tc3162/ledcetrl.h
===================================================================
--- linux-3.18.21.orig/arch/mips/include/asm/tc3162/ledcetrl.h	2018-07-27 10:22:55.635399151 +0800
+++ linux-3.18.21/arch/mips/include/asm/tc3162/ledcetrl.h	2018-08-01 15:52:38.178176953 +0800
@@ -264,6 +264,11 @@
 #if defined(TCSUPPORT_CPU_EN7516)||defined(TCSUPPORT_CPU_EN7527)
 #define GPIO_WLAN_WPS_5G			(LED_BASE + 104)
 #define LED_WLAN_RADIO_5G			(LED_BASE + 105) 
+#if defined(ZYXEL_CUSTOMIZE_LED_BEHAVIOR) //defined for wifi 5G
+#define LED_WLAN_5G_STATUS          (LED_BASE + 106)
+#define LED_WLAN_5G_ACT_STATUS      (LED_BASE + 107)
+#define LED_WLAN_5G_NOACT_STATUS    (LED_BASE + 108)
+#endif
 #endif
 
 #define mtkled_read_reg_word(reg) 			regRead32(reg)
