Index: linux-3.18.21/net/ipv6/addrconf.c
===================================================================
--- linux-3.18.21.orig/net/ipv6/addrconf.c	2018-08-07 09:59:51.905112000 +0800
+++ linux-3.18.21/net/ipv6/addrconf.c	2018-08-07 15:41:05.786427250 +0800
@@ -429,11 +429,11 @@
 
 	/* Join all-node multicast group */
 	ipv6_dev_mc_inc(dev, &in6addr_linklocal_allnodes);
-
+#if 0 /* ZYXEL workaround, don't send all-routers join to avoid DSLAM VES series drop our RS */
 	/* Join all-router multicast group if forwarding is set */
 	if (ndev->cnf.forwarding && (dev->flags & IFF_MULTICAST))
 		ipv6_dev_mc_inc(dev, &in6addr_linklocal_allrouters);
-
+#endif
 	return ndev;
 
 err_release:
