Index: linux-3.18.21/drivers/mtd/maps/tc3162-flash.c
===================================================================
--- linux-3.18.21.orig/drivers/mtd/maps/tc3162-flash.c	2018-08-17 10:25:20.845116000 +0800
+++ linux-3.18.21/drivers/mtd/maps/tc3162-flash.c	2018-08-17 10:28:00.229116000 +0800
@@ -60,6 +60,13 @@
 static uint64_t tclinux_part_offset = OFFSET_CONTINUOUS;	
 static int kernel_part_index = -1;
 static int kernel_slave_part_index = -1;
+
+#if defined(ZYXEL_FLASH_PARTITION_RUN_TIME_CHECK)
+extern const u32 ZYXEL_FLASH_SIZE_CHECK_128MB;
+extern const u32 ZYXEL_FLASH_SIZE_CHECK_256MB;
+extern u32 flashSize_mtd_partition_parse;
+#endif
+
 #endif
 
 #define WINDOW_ADDR 0x1fc00000
@@ -212,11 +219,6 @@
 #define SIZE_KB_FOR_WWANPKG    0
 #endif
 
-#if defined(ZYXEL_FLASH_PARTITION_RUN_TIME_CHECK)
-extern const u32 ZYXEL_FLASH_SIZE_CHECK_128MB;
-extern const u32 ZYXEL_FLASH_SIZE_CHECK_256MB;
-extern u32 flashSize_mtd_partition_parse;
-#endif
 
 static struct mtd_info *tc3162_mtd_info;
 
@@ -1722,7 +1724,7 @@
     {
         //printk("%s,%d, device size:0x%x \n", __FUNCTION__, __LINE__, flashSize_mtd_partition_parse);
         /* parse command line */
-        if (tc3162_mtdpart_setup(VMG8623_T50A_TCSUPPORT_PARTITIONS_CMDLINE_STR, blocksize) == 0)
+        if (tc3162_mtdpart_setup(FLASH_128MB_TCSUPPORT_PARTITIONS_CMDLINE_STR, blocksize) == 0)
         {
             printk("tc3162_mtdpart_setup fail, 0x8000000\n");
             return 0;
@@ -1732,7 +1734,7 @@
     {
         //printk("%s,%d, device size:0x%x \n", __FUNCTION__, __LINE__, flashSize_mtd_partition_parse);
         /* parse command line */
-        if (tc3162_mtdpart_setup(VMG8825_T50A_TCSUPPORT_PARTITIONS_CMDLINE_STR, blocksize) == 0)
+        if (tc3162_mtdpart_setup(FLASH_256MB_TCSUPPORT_PARTITIONS_CMDLINE_STR, blocksize) == 0)
         {
             printk("tc3162_mtdpart_setup fail, 0x8000000\n");
             return 0;
