Index: linux-3.18.21/drivers/mtd/chips/spi_nand_flash.c
===================================================================
--- linux-3.18.21.orig/drivers/mtd/chips/spi_nand_flash.c	2018-11-13 09:14:36.115080000 +0800
+++ linux-3.18.21/drivers/mtd/chips/spi_nand_flash.c	2018-11-13 09:26:09.036899052 +0800
@@ -188,8 +188,9 @@
 #define _SPI_NAND_MANUFACTURER_ID_TOSHIBA		0x98
 #define _SPI_NAND_MANUFACTURER_ID_MICRON		0x2C
 #define _SPI_NAND_MANUFACTURER_ID_HEYANG		0xC9
-#define _SPI_NAND_MANUFACTURER_ID_XTX			0xA1
+#define _SPI_NAND_MANUFACTURER_ID_PN			0xA1
 #define _SPI_NAND_MANUFACTURER_ID_ATO			0x9B
+#define _SPI_NAND_MANUFACTURER_ID_ATO_2			0xAD
 #define _SPI_NAND_MANUFACTURER_ID_FM			0xA1
 #define _SPI_NAND_MANUFACTURER_ID_FORESEE		0xCD
 #define _SPI_NAND_MANUFACTURER_ID_DS			0xE5
@@ -201,8 +202,10 @@
 #define _SPI_NAND_DEVICE_ID_GD5F1GQ4UAYIG	0xF1
 #define _SPI_NAND_DEVICE_ID_GD5F1GQ4UBYIG	0xD1
 #define _SPI_NAND_DEVICE_ID_GD5F1GQ4UCYIG	0xB1
+#define _SPI_NAND_DEVICE_ID_GD5F1GQ4UEYIS	0xD3
 #define _SPI_NAND_DEVICE_ID_GD5F2GQ4UBYIG	0xD2
 #define _SPI_NAND_DEVICE_ID_GD5F2GQ4UCYIG	0xB2
+#define _SPI_NAND_DEVICE_ID_GD5F2GQ4UE9IS	0xD5
 #define _SPI_NAND_DEVICE_ID_GD5F4GQ4UBYIG	0xD4
 #define _SPI_NAND_DEVICE_ID_GD5F4GQ4UCYIG	0xB4
 #define _SPI_NAND_DEVICE_ID_F50L512M41A		0x20
@@ -232,6 +235,8 @@
 #define _SPI_NAND_DEVICE_ID_PN26G01AWSIUG	0xE1
 #define _SPI_NAND_DEVICE_ID_PN26G02AWSIUG	0xE2
 #define _SPI_NAND_DEVICE_ID_ATO25D1GA		0x12
+#define _SPI_NAND_DEVICE_ID_ATO25D2GA		0xF1
+#define _SPI_NAND_DEVICE_ID_ATO25D2GB		0xDA
 #define _SPI_NAND_DEVICE_ID_FM25G01B		0xD1
 #define _SPI_NAND_DEVICE_ID_FM25G02B		0xD2
 #define _SPI_NAND_DEVICE_ID_FM25G02C		0x92
@@ -414,6 +419,11 @@
 	.oobfree = {{0,128}}
 };
 
+struct spi_nand_flash_ooblayout ooblayout_gigadevice_GD5FXGQ4U = {
+	.oobsize = 52, 
+	.oobfree = {{0,16}, {20,12}, {36,12}, {52,12}}
+};
+
 struct spi_nand_flash_ooblayout ooblayout_zentel = {
 	.oobsize = 36, 
 	.oobfree = {{0,1}, {8,8}, {16,1}, {24,8}, {32,1}, {40,8}, {48,1}, {56,8} }
@@ -460,9 +470,10 @@
 	.oobfree = {{0,8}, {32,8}, {64,8}, {96,8}}
 };
 
-struct spi_nand_flash_ooblayout ooblayout_xtx = {
-	.oobsize = 76, 
-	.oobfree = {{0,6}, {19,2}, {34,2}, {49,2}, {64,64}}
+/* only use user meta data with ECC protected */
+struct spi_nand_flash_ooblayout ooblayout_pn = {
+	.oobsize = 44, 
+	.oobfree = {{0,4}, {4,2}, {19,2}, {34,2}, {49,2}, {96,32}}
 };
 
 /* only use user meta data with ECC protected */
@@ -471,6 +482,19 @@
 	.oobfree = {{0,64}}
 };
 
+/* only use user meta data with ECC protected */
+struct spi_nand_flash_ooblayout ooblayout_ato_25D2GA = {
+	.oobsize = 12, 
+	.oobfree = {{0,3}, {16,3}, {32,3}, {48,3}}
+};
+
+/* only use user meta data with ECC protected */
+struct spi_nand_flash_ooblayout ooblayout_ato_25D2GB = {
+	.oobsize = 48, 
+	.oobfree = {{0,12}, {16,12}, {32,12}, {48,12}}
+};
+
+/* only use user meta data with ECC protected */
 struct spi_nand_flash_ooblayout ooblayout_fm = {
 	.oobsize = 64, 
 	.oobfree = {{0,64}}
@@ -507,16 +531,16 @@
 	.oobfree = {{0,64}}
 };
 
-struct spi_nand_flash_ooblayout ooblayout_foresee = {
-	.oobsize = 64, 
-	.oobfree = {{0,64}}
-};
-
 struct spi_nand_flash_ooblayout ooblayout_tym = {
 	.oobsize = 12, 
 	.oobfree = {{0,3}, {16,3}, {32,3}, {48,3}}
 };
 
+struct spi_nand_flash_ooblayout ooblayout_foresee = {
+	.oobsize = 64, 
+	.oobfree = {{0,64}}
+};
+
 /*****************************[ Notice]******************************/
 /* If new spi nand chip have page size more than 4KB,  or oob size more than 256 bytes,  than*/
 /* it will need to adjust the #define of _SPI_NAND_PAGE_SIZE and _SPI_NAND_OOB_SIZE */
@@ -570,6 +594,21 @@
 
 	{
 		mfr_id:						_SPI_NAND_MANUFACTURER_ID_GIGADEVICE,
+		dev_id: 					_SPI_NAND_DEVICE_ID_GD5F1GQ4UEYIS,
+		ptr_name:					"_SPI_NAND_DEVICE_ID_GD5F1GQ4UEYIS",
+		device_size:				_SPI_NAND_CHIP_SIZE_1GBIT,
+		page_size:					_SPI_NAND_PAGE_SIZE_2KBYTE,
+		oob_size:					_SPI_NAND_OOB_SIZE_128BYTE,
+		erase_size: 				_SPI_NAND_BLOCK_SIZE_128KBYTE,
+		dummy_mode: 				SPI_NAND_FLASH_READ_DUMMY_BYTE_APPEND,
+		read_mode:					SPI_NAND_FLASH_READ_SPEED_MODE_DUAL,
+		write_mode: 				SPI_NAND_FLASH_WRITE_SPEED_MODE_SINGLE,
+		oob_free_layout :			&ooblayout_gigadevice_GD5FXGQ4U,
+		feature:					SPI_NAND_FLASH_FEATURE_NONE,
+	},
+
+	{
+		mfr_id:						_SPI_NAND_MANUFACTURER_ID_GIGADEVICE,
 		dev_id:						_SPI_NAND_DEVICE_ID_GD5F2GQ4UBYIG,
 		ptr_name:					"_SPI_NAND_DEVICE_ID_GD5F2GQ4UB",
 		device_size:				_SPI_NAND_CHIP_SIZE_2GBIT,
@@ -585,6 +624,20 @@
 	
 	{
 		mfr_id: 					_SPI_NAND_MANUFACTURER_ID_GIGADEVICE,
+		dev_id:						_SPI_NAND_DEVICE_ID_GD5F2GQ4UE9IS,
+		ptr_name:					"_SPI_NAND_DEVICE_ID_GD5F2GQ4UE9IS",
+		device_size:				_SPI_NAND_CHIP_SIZE_2GBIT,
+		page_size:					_SPI_NAND_PAGE_SIZE_2KBYTE,
+		oob_size:					_SPI_NAND_OOB_SIZE_128BYTE,
+		erase_size:					_SPI_NAND_BLOCK_SIZE_128KBYTE,
+		dummy_mode:					SPI_NAND_FLASH_READ_DUMMY_BYTE_APPEND,
+		read_mode:					SPI_NAND_FLASH_READ_SPEED_MODE_DUAL,
+		write_mode:					SPI_NAND_FLASH_WRITE_SPEED_MODE_SINGLE,
+		oob_free_layout :			&ooblayout_gigadevice_GD5FXGQ4U,
+		feature:					SPI_NAND_FLASH_FEATURE_NONE,
+	},		
+	{
+		mfr_id: 					_SPI_NAND_MANUFACTURER_ID_GIGADEVICE,
 		dev_id: 					_SPI_NAND_DEVICE_ID_GD5F2GQ4UCYIG,
 		ptr_name:					"_SPI_NAND_DEVICE_ID_GD5F2GQ4UC",
 		device_size:				_SPI_NAND_CHIP_SIZE_2GBIT,
@@ -958,9 +1011,39 @@
 	},
 
 	{
-		mfr_id: 					_SPI_NAND_MANUFACTURER_ID_XTX,
+		mfr_id: 					_SPI_NAND_MANUFACTURER_ID_HEYANG,
+		dev_id: 					_SPI_NAND_DEVICE_ID_HYF1GQ4UDACAE,
+		ptr_name:					"_SPI_NAND_DEVICE_ID_HYF1GQ4UDACAE",
+		device_size:				_SPI_NAND_CHIP_SIZE_1GBIT,
+		page_size:					_SPI_NAND_PAGE_SIZE_2KBYTE,
+		oob_size:					_SPI_NAND_OOB_SIZE_64BYTE,
+		erase_size: 				_SPI_NAND_BLOCK_SIZE_128KBYTE,
+		dummy_mode: 				SPI_NAND_FLASH_READ_DUMMY_BYTE_APPEND,
+		read_mode:					SPI_NAND_FLASH_READ_SPEED_MODE_DUAL,
+		write_mode:					SPI_NAND_FLASH_WRITE_SPEED_MODE_SINGLE,
+		oob_free_layout :			&ooblayout_type1,
+		feature:					SPI_NAND_FLASH_FEATURE_NONE,
+	},
+
+	{
+		mfr_id: 					_SPI_NAND_MANUFACTURER_ID_HEYANG,
+		dev_id: 					_SPI_NAND_DEVICE_ID_HYF2GQ4UDACAE,
+		ptr_name:					"_SPI_NAND_DEVICE_ID_HYF2GQ4UDACAE",
+		device_size:				_SPI_NAND_CHIP_SIZE_2GBIT,
+		page_size:					_SPI_NAND_PAGE_SIZE_2KBYTE,
+		oob_size:					_SPI_NAND_OOB_SIZE_64BYTE,
+		erase_size: 				_SPI_NAND_BLOCK_SIZE_128KBYTE,
+		dummy_mode: 				SPI_NAND_FLASH_READ_DUMMY_BYTE_APPEND,
+		read_mode:					SPI_NAND_FLASH_READ_SPEED_MODE_DUAL,
+		write_mode:					SPI_NAND_FLASH_WRITE_SPEED_MODE_SINGLE,
+		oob_free_layout :			&ooblayout_type1,
+		feature:					SPI_NAND_FLASH_FEATURE_NONE,
+	},
+
+	{
+		mfr_id: 					_SPI_NAND_MANUFACTURER_ID_PN,
 		dev_id: 					_SPI_NAND_DEVICE_ID_PN26G01AWSIUG,
-		ptr_name:					"_SPI_NAND_DEVICE_ID_PN26G01AWSIUG",
+		ptr_name:					"_SPI_NAND_DEVICE_ID_PN26G01A-X",
 		device_size:				_SPI_NAND_CHIP_SIZE_1GBIT,
 		page_size:					_SPI_NAND_PAGE_SIZE_2KBYTE,
 		oob_size:					_SPI_NAND_OOB_SIZE_128BYTE,
@@ -968,14 +1051,14 @@
 		dummy_mode: 				SPI_NAND_FLASH_READ_DUMMY_BYTE_APPEND,
 		read_mode:					SPI_NAND_FLASH_READ_SPEED_MODE_DUAL,
 		write_mode:					SPI_NAND_FLASH_WRITE_SPEED_MODE_SINGLE,
-		oob_free_layout :			&ooblayout_xtx, 	
+		oob_free_layout :			&ooblayout_pn, 	
 		feature:					SPI_NAND_FLASH_FEATURE_NONE,
 	},
 
 	{
-		mfr_id: 					_SPI_NAND_MANUFACTURER_ID_XTX,
+		mfr_id: 					_SPI_NAND_MANUFACTURER_ID_PN,
 		dev_id: 					_SPI_NAND_DEVICE_ID_PN26G02AWSIUG,
-		ptr_name:					"_SPI_NAND_DEVICE_ID_PN26G02AWSIUG",
+		ptr_name:					"_SPI_NAND_DEVICE_ID_PN26G02A-X",
 		device_size:				_SPI_NAND_CHIP_SIZE_2GBIT,
 		page_size:					_SPI_NAND_PAGE_SIZE_2KBYTE,
 		oob_size:					_SPI_NAND_OOB_SIZE_128BYTE,
@@ -983,7 +1066,7 @@
 		dummy_mode: 				SPI_NAND_FLASH_READ_DUMMY_BYTE_APPEND,
 		read_mode:					SPI_NAND_FLASH_READ_SPEED_MODE_DUAL,
 		write_mode:					SPI_NAND_FLASH_WRITE_SPEED_MODE_SINGLE,
-		oob_free_layout :			&ooblayout_xtx, 	
+		oob_free_layout :			&ooblayout_pn, 	
 		feature:					SPI_NAND_FLASH_FEATURE_NONE,
 	},
 
@@ -1003,24 +1086,24 @@
 	},
 
 	{
-		mfr_id: 					_SPI_NAND_MANUFACTURER_ID_FM,
-		dev_id: 					_SPI_NAND_DEVICE_ID_FM25G01B,
-		ptr_name:					"_SPI_NAND_DEVICE_ID_FM25G01B",
-		device_size:				_SPI_NAND_CHIP_SIZE_1GBIT,
+		mfr_id: 					_SPI_NAND_MANUFACTURER_ID_ATO,
+		dev_id: 					_SPI_NAND_DEVICE_ID_ATO25D2GA,
+		ptr_name:					"_SPI_NAND_DEVICE_ID_ATO25D2GA",
+		device_size:				_SPI_NAND_CHIP_SIZE_2GBIT,
 		page_size:					_SPI_NAND_PAGE_SIZE_2KBYTE,
-		oob_size:					_SPI_NAND_OOB_SIZE_128BYTE,
+		oob_size:					_SPI_NAND_OOB_SIZE_64BYTE,
 		erase_size: 				_SPI_NAND_BLOCK_SIZE_128KBYTE,
 		dummy_mode: 				SPI_NAND_FLASH_READ_DUMMY_BYTE_APPEND,
 		read_mode:					SPI_NAND_FLASH_READ_SPEED_MODE_DUAL,
 		write_mode:					SPI_NAND_FLASH_WRITE_SPEED_MODE_SINGLE,
-		oob_free_layout :			&ooblayout_fm, 	
+		oob_free_layout :			&ooblayout_ato_25D2GA,	
 		feature:					SPI_NAND_FLASH_FEATURE_NONE,
 	},
 
 	{
-		mfr_id: 					_SPI_NAND_MANUFACTURER_ID_FM,
-		dev_id: 					_SPI_NAND_DEVICE_ID_FM25G02B,
-		ptr_name:					"_SPI_NAND_DEVICE_ID_FM25G02B",
+		mfr_id: 					_SPI_NAND_MANUFACTURER_ID_ATO_2,
+		dev_id: 					_SPI_NAND_DEVICE_ID_ATO25D2GB,
+		ptr_name:					"_SPI_NAND_DEVICE_ID_ATO25D2GB",
 		device_size:				_SPI_NAND_CHIP_SIZE_2GBIT,
 		page_size:					_SPI_NAND_PAGE_SIZE_2KBYTE,
 		oob_size:					_SPI_NAND_OOB_SIZE_128BYTE,
@@ -1028,37 +1111,37 @@
 		dummy_mode: 				SPI_NAND_FLASH_READ_DUMMY_BYTE_APPEND,
 		read_mode:					SPI_NAND_FLASH_READ_SPEED_MODE_DUAL,
 		write_mode:					SPI_NAND_FLASH_WRITE_SPEED_MODE_SINGLE,
-		oob_free_layout :			&ooblayout_fm, 	
+		oob_free_layout :			&ooblayout_ato_25D2GB,	
 		feature:					SPI_NAND_FLASH_FEATURE_NONE,
 	},
 
 	{
 		mfr_id: 					_SPI_NAND_MANUFACTURER_ID_FM,
-		dev_id: 					_SPI_NAND_DEVICE_ID_FM25G02C,
-		ptr_name:					"_SPI_NAND_DEVICE_ID_FM25G02C",
-		device_size:				_SPI_NAND_CHIP_SIZE_2GBIT,
+		dev_id: 					_SPI_NAND_DEVICE_ID_FM25G01B,
+		ptr_name:					"_SPI_NAND_DEVICE_ID_FM25G01B",
+		device_size:				_SPI_NAND_CHIP_SIZE_1GBIT,
 		page_size:					_SPI_NAND_PAGE_SIZE_2KBYTE,
-		oob_size:					_SPI_NAND_OOB_SIZE_64BYTE,
+		oob_size:					_SPI_NAND_OOB_SIZE_128BYTE,
 		erase_size: 				_SPI_NAND_BLOCK_SIZE_128KBYTE,
 		dummy_mode: 				SPI_NAND_FLASH_READ_DUMMY_BYTE_APPEND,
 		read_mode:					SPI_NAND_FLASH_READ_SPEED_MODE_DUAL,
 		write_mode: 				SPI_NAND_FLASH_WRITE_SPEED_MODE_SINGLE,
-		oob_free_layout :			&ooblayout_fm_32,	
+		oob_free_layout :			&ooblayout_fm, 	
 		feature:					SPI_NAND_FLASH_FEATURE_NONE,
 	},
 
 	{
-		mfr_id: 					_SPI_NAND_MANUFACTURER_ID_FORESEE,
-		dev_id: 					_SPI_NAND_DEVICE_ID_FS35ND01GS1F1,
-		ptr_name:					"_SPI_NAND_DEVICE_ID_FS35ND01GS1F1",
-		device_size:				_SPI_NAND_CHIP_SIZE_1GBIT,
+		mfr_id: 					_SPI_NAND_MANUFACTURER_ID_FM,
+		dev_id: 					_SPI_NAND_DEVICE_ID_FM25G02B,
+		ptr_name:					"_SPI_NAND_DEVICE_ID_FM25G02B",
+		device_size:				_SPI_NAND_CHIP_SIZE_2GBIT,
 		page_size:					_SPI_NAND_PAGE_SIZE_2KBYTE,
-		oob_size:					_SPI_NAND_OOB_SIZE_64BYTE,
+		oob_size:					_SPI_NAND_OOB_SIZE_128BYTE,
 		erase_size: 				_SPI_NAND_BLOCK_SIZE_128KBYTE,
 		dummy_mode: 				SPI_NAND_FLASH_READ_DUMMY_BYTE_APPEND,
 		read_mode:					SPI_NAND_FLASH_READ_SPEED_MODE_DUAL,
 		write_mode: 				SPI_NAND_FLASH_WRITE_SPEED_MODE_SINGLE,
-		oob_free_layout :			&ooblayout_foresee, 	
+		oob_free_layout :			&ooblayout_fm,	
 		feature:					SPI_NAND_FLASH_FEATURE_NONE,
 	},
 
@@ -2205,14 +2288,18 @@
 		((ptr_dev_info_t->dev_id == _SPI_NAND_DEVICE_ID_GD5F1GQ4UAYIG) ||
 		(ptr_dev_info_t->dev_id == _SPI_NAND_DEVICE_ID_GD5F1GQ4UBYIG) ||
 		(ptr_dev_info_t->dev_id == _SPI_NAND_DEVICE_ID_GD5F1GQ4UCYIG) ||
+		(ptr_dev_info_t->dev_id == _SPI_NAND_DEVICE_ID_GD5F1GQ4UEYIS) ||
 		(ptr_dev_info_t->dev_id == _SPI_NAND_DEVICE_ID_GD5F2GQ4UBYIG) ||
+		(ptr_dev_info_t->dev_id == _SPI_NAND_DEVICE_ID_GD5F2GQ4UE9IS) ||
 		(ptr_dev_info_t->dev_id == _SPI_NAND_DEVICE_ID_GD5F2GQ4UCYIG) ||
 		(ptr_dev_info_t->dev_id == _SPI_NAND_DEVICE_ID_GD5F4GQ4UBYIG) ||
 		(ptr_dev_info_t->dev_id == _SPI_NAND_DEVICE_ID_GD5F4GQ4UCYIG)))
 	{
 		if((ptr_dev_info_t->dev_id == _SPI_NAND_DEVICE_ID_GD5F1GQ4UAYIG) ||
 		   (ptr_dev_info_t->dev_id == _SPI_NAND_DEVICE_ID_GD5F1GQ4UBYIG) ||
+		   (ptr_dev_info_t->dev_id == _SPI_NAND_DEVICE_ID_GD5F1GQ4UEYIS) || 
 		   (ptr_dev_info_t->dev_id == _SPI_NAND_DEVICE_ID_GD5F2GQ4UBYIG) ||
+			(ptr_dev_info_t->dev_id == _SPI_NAND_DEVICE_ID_GD5F2GQ4UE9IS) ||
 		   (ptr_dev_info_t->dev_id == _SPI_NAND_DEVICE_ID_GD5F4GQ4UBYIG))
 		{
 			if(((status & 0x30) >> 4) == 0x2)
@@ -2296,24 +2383,24 @@
 			rtn_status = SPI_NAND_FLASH_RTN_DETECTED_BAD_BLOCK;
 		}
 	}
-	else if(((ptr_dev_info_t->mfr_id == _SPI_NAND_MANUFACTURER_ID_XTX) && (ptr_dev_info_t->dev_id == _SPI_NAND_DEVICE_ID_PN26G01AWSIUG)) ||
-		((ptr_dev_info_t->mfr_id == _SPI_NAND_MANUFACTURER_ID_XTX) && (ptr_dev_info_t->dev_id == _SPI_NAND_DEVICE_ID_PN26G02AWSIUG)))
+	else if(((ptr_dev_info_t->mfr_id == _SPI_NAND_MANUFACTURER_ID_PN) && (ptr_dev_info_t->dev_id == _SPI_NAND_DEVICE_ID_PN26G01AWSIUG)) ||
+		((ptr_dev_info_t->mfr_id == _SPI_NAND_MANUFACTURER_ID_PN) && (ptr_dev_info_t->dev_id == _SPI_NAND_DEVICE_ID_PN26G02AWSIUG)))
 	{
 		if(((status & 0x30) >> 4) == 0x2)
 		{
 			rtn_status = SPI_NAND_FLASH_RTN_DETECTED_BAD_BLOCK;
 		}
 	}
-	else if(((ptr_dev_info_t->mfr_id == _SPI_NAND_MANUFACTURER_ID_FM) && (ptr_dev_info_t->dev_id == _SPI_NAND_DEVICE_ID_FM25G01B)) ||
-		((ptr_dev_info_t->mfr_id == _SPI_NAND_MANUFACTURER_ID_FM) && (ptr_dev_info_t->dev_id == _SPI_NAND_DEVICE_ID_FM25G02B)) ||
-		((ptr_dev_info_t->mfr_id == _SPI_NAND_MANUFACTURER_ID_FM) && (ptr_dev_info_t->dev_id == _SPI_NAND_DEVICE_ID_FM25G02C)))
+	else if(((ptr_dev_info_t->mfr_id == _SPI_NAND_MANUFACTURER_ID_ATO) && (ptr_dev_info_t->dev_id == _SPI_NAND_DEVICE_ID_ATO25D2GA)) ||
+		((ptr_dev_info_t->mfr_id == _SPI_NAND_MANUFACTURER_ID_ATO_2) && (ptr_dev_info_t->dev_id == _SPI_NAND_DEVICE_ID_ATO25D2GB)))
 	{
-		if(((status & 0x70) >> 4) == 0x7)
+		if(((status & 0x30) >> 4) == 0x2)
 		{
 			rtn_status = SPI_NAND_FLASH_RTN_DETECTED_BAD_BLOCK;
 		}
 	}
-	else if((ptr_dev_info_t->mfr_id == _SPI_NAND_MANUFACTURER_ID_FORESEE) && (ptr_dev_info_t->dev_id == _SPI_NAND_DEVICE_ID_FS35ND01GS1F1)) 
+	else if(((ptr_dev_info_t->mfr_id == _SPI_NAND_MANUFACTURER_ID_FM) && (ptr_dev_info_t->dev_id == _SPI_NAND_DEVICE_ID_FM25G01B)) ||
+		((ptr_dev_info_t->mfr_id == _SPI_NAND_MANUFACTURER_ID_FM) && (ptr_dev_info_t->dev_id == _SPI_NAND_DEVICE_ID_FM25G02B)))
 	{
 		if(((status & 0x70) >> 4) == 0x7)
 		{
@@ -2337,9 +2424,7 @@
 			rtn_status = SPI_NAND_FLASH_RTN_DETECTED_BAD_BLOCK;
 		}
 	}
-	else if(((ptr_dev_info_t->mfr_id == _SPI_NAND_MANUFACTURER_ID_TYM) && (ptr_dev_info_t->dev_id == _SPI_NAND_DEVICE_ID_TYM25D2GA01)) || 
-		((ptr_dev_info_t->mfr_id == _SPI_NAND_MANUFACTURER_ID_TYM) && (ptr_dev_info_t->dev_id == _SPI_NAND_DEVICE_ID_TYM25D2GA02)) ||
-		((ptr_dev_info_t->mfr_id == _SPI_NAND_MANUFACTURER_ID_TYM) && (ptr_dev_info_t->dev_id == _SPI_NAND_DEVICE_ID_TYM25D1GA03)))
+	else if((ptr_dev_info_t->mfr_id == _SPI_NAND_MANUFACTURER_ID_TYM) && (ptr_dev_info_t->dev_id == _SPI_NAND_DEVICE_ID_TYM25D2GA01)) 
 	{
 		if(((status & 0x30) >> 4) == 0x2)
 		{
@@ -3778,7 +3863,9 @@
 	if(((ptr_device_t->mfr_id == _SPI_NAND_MANUFACTURER_ID_GIGADEVICE) && (ptr_device_t->dev_id == _SPI_NAND_DEVICE_ID_GD5F1GQ4UAYIG)) ||
 		((ptr_device_t->mfr_id == _SPI_NAND_MANUFACTURER_ID_GIGADEVICE) && (ptr_device_t->dev_id == _SPI_NAND_DEVICE_ID_GD5F1GQ4UBYIG)) ||
 		((ptr_device_t->mfr_id == _SPI_NAND_MANUFACTURER_ID_GIGADEVICE) && (ptr_device_t->dev_id == _SPI_NAND_DEVICE_ID_GD5F1GQ4UCYIG)) ||
+		((ptr_device_t->mfr_id == _SPI_NAND_MANUFACTURER_ID_GIGADEVICE) && (ptr_device_t->dev_id == _SPI_NAND_DEVICE_ID_GD5F1GQ4UEYIS)) ||
 		((ptr_device_t->mfr_id == _SPI_NAND_MANUFACTURER_ID_GIGADEVICE) && (ptr_device_t->dev_id == _SPI_NAND_DEVICE_ID_GD5F2GQ4UBYIG)) ||
+		((ptr_device_t->mfr_id == _SPI_NAND_MANUFACTURER_ID_GIGADEVICE) && (ptr_device_t->dev_id == _SPI_NAND_DEVICE_ID_GD5F2GQ4UE9IS)) ||
 		((ptr_device_t->mfr_id == _SPI_NAND_MANUFACTURER_ID_GIGADEVICE) && (ptr_device_t->dev_id == _SPI_NAND_DEVICE_ID_GD5F2GQ4UCYIG)) ||
 		((ptr_device_t->mfr_id == _SPI_NAND_MANUFACTURER_ID_GIGADEVICE) && (ptr_device_t->dev_id == _SPI_NAND_DEVICE_ID_GD5F4GQ4UBYIG)) ||
 		((ptr_device_t->mfr_id == _SPI_NAND_MANUFACTURER_ID_GIGADEVICE) && (ptr_device_t->dev_id == _SPI_NAND_DEVICE_ID_GD5F4GQ4UCYIG)))
@@ -3990,8 +4077,8 @@
 		feature |= 0x1;
 		spi_nand_protocol_set_status_reg_2(feature);
 	}
-	else if(((ptr_device_t->mfr_id == _SPI_NAND_MANUFACTURER_ID_XTX) && (ptr_device_t->dev_id == _SPI_NAND_DEVICE_ID_PN26G01AWSIUG)) ||
-		((ptr_device_t->mfr_id == _SPI_NAND_MANUFACTURER_ID_XTX) && (ptr_device_t->dev_id == _SPI_NAND_DEVICE_ID_PN26G02AWSIUG)))
+	else if(((ptr_device_t->mfr_id == _SPI_NAND_MANUFACTURER_ID_PN) && (ptr_device_t->dev_id == _SPI_NAND_DEVICE_ID_PN26G01AWSIUG)) ||
+		((ptr_device_t->mfr_id == _SPI_NAND_MANUFACTURER_ID_PN) && (ptr_device_t->dev_id == _SPI_NAND_DEVICE_ID_PN26G02AWSIUG)))
 	{
 		/* 1. Unlock All block */
 		spi_nand_protocol_get_status_reg_1(&feature);
@@ -4003,7 +4090,8 @@
 		feature |= 0x1;
 		spi_nand_protocol_set_status_reg_2(feature);
 	}
-	else if( (ptr_device_t->mfr_id) == _SPI_NAND_MANUFACTURER_ID_ATO)
+	else if( ((ptr_device_t->mfr_id) == _SPI_NAND_MANUFACTURER_ID_ATO) ||
+		((ptr_device_t->mfr_id) == _SPI_NAND_MANUFACTURER_ID_ATO_2) )
 	{
 		/* 1. Unlock All block */
 		spi_nand_protocol_get_status_reg_1(&feature);
@@ -4011,8 +4099,7 @@
 		spi_nand_protocol_set_status_reg_1(feature);
 	}
 	else if(((ptr_device_t->mfr_id == _SPI_NAND_MANUFACTURER_ID_FM) && (ptr_device_t->dev_id == _SPI_NAND_DEVICE_ID_FM25G01B)) ||
-		((ptr_device_t->mfr_id == _SPI_NAND_MANUFACTURER_ID_FM) && (ptr_device_t->dev_id == _SPI_NAND_DEVICE_ID_FM25G02B)) ||
-		((ptr_device_t->mfr_id == _SPI_NAND_MANUFACTURER_ID_FM) && (ptr_device_t->dev_id == _SPI_NAND_DEVICE_ID_FM25G02C)))
+			((ptr_device_t->mfr_id == _SPI_NAND_MANUFACTURER_ID_FM) && (ptr_device_t->dev_id == _SPI_NAND_DEVICE_ID_FM25G02B)))
 	{
 		/* 1. Unlock All block */
 		spi_nand_protocol_get_status_reg_1(&feature);
@@ -4058,9 +4145,7 @@
 		spi_nand_protocol_get_status_reg_1(&feature);
 		_SPI_NAND_DEBUG_PRINTF(SPI_NAND_FLASH_DEBUG_LEVEL_1, "After Unlock all block setup, the status register1 =0x%x\n", feature);		
 	}
-	else if(((ptr_device_t->mfr_id == _SPI_NAND_MANUFACTURER_ID_TYM) && (ptr_device_t->dev_id == _SPI_NAND_DEVICE_ID_TYM25D2GA01)) ||
-		((ptr_device_t->mfr_id == _SPI_NAND_MANUFACTURER_ID_TYM) && (ptr_device_t->dev_id == _SPI_NAND_DEVICE_ID_TYM25D2GA02)) ||
-		((ptr_device_t->mfr_id == _SPI_NAND_MANUFACTURER_ID_TYM) && (ptr_device_t->dev_id == _SPI_NAND_DEVICE_ID_TYM25D1GA03)))
+	else if((ptr_device_t->mfr_id == _SPI_NAND_MANUFACTURER_ID_TYM) && (ptr_device_t->dev_id == _SPI_NAND_DEVICE_ID_TYM25D2GA01)) 
 	{
 		/* 1. Unlock All block */
 		spi_nand_protocol_get_status_reg_1(&feature);
