Index: linux-3.18.21/drivers/mtd/chips/spi_nand_flash.c
===================================================================
--- linux-3.18.21.orig/drivers/mtd/chips/spi_nand_flash.c	2018-11-14 14:12:59.555080000 +0800
+++ linux-3.18.21/drivers/mtd/chips/spi_nand_flash.c	2018-11-14 14:26:02.185544358 +0800
@@ -221,8 +221,12 @@
 #define _SPI_NAND_DEVICE_ID_A5U1GA21BWS		0x21
 #define _SPI_NAND_DEVICE_ID_EM73C044SNB		0x11
 #define _SPI_NAND_DEVICE_ID_EM73C044SNA		0x19
+#define _SPI_NAND_DEVICE_ID_EM73C044VCD		0x1C
+#define _SPI_NAND_DEVICE_ID_EM73C044VCA		0x18
 #define _SPI_NAND_DEVICE_ID_EM73D044SNA		0x12
 #define _SPI_NAND_DEVICE_ID_EM73D044SNC		0x0A
+#define _SPI_NAND_DEVICE_ID_EM73D044VCG		0x1F
+#define _SPI_NAND_DEVICE_ID_EM73D044VCH		0x1B
 #define _SPI_NAND_DEVICE_ID_EM73E044SNA		0x03
 #define _SPI_NAND_DEVICE_ID_TC58CVG0S3H		0xC2
 #define _SPI_NAND_DEVICE_ID_TC58CVG1S3H		0xCB
@@ -235,20 +239,20 @@
 #define _SPI_NAND_DEVICE_ID_PN26G01AWSIUG	0xE1
 #define _SPI_NAND_DEVICE_ID_PN26G02AWSIUG	0xE2
 #define _SPI_NAND_DEVICE_ID_ATO25D1GA		0x12
-#define _SPI_NAND_DEVICE_ID_ATO25D2GA		0xF1
-#define _SPI_NAND_DEVICE_ID_ATO25D2GB		0xDA
+#define _SPI_NAND_DEVICE_ID_FM25S01			0xA1
 #define _SPI_NAND_DEVICE_ID_FM25G01B		0xD1
 #define _SPI_NAND_DEVICE_ID_FM25G02B		0xD2
-#define _SPI_NAND_DEVICE_ID_FM25G02C		0x92
-#define _SPI_NAND_DEVICE_ID_FS35ND01GS1F1	0xB1 	
-#define _SPI_NAND_DEVICE_ID_DS35Q2GA		0x72
-#define _SPI_NAND_DEVICE_ID_DS35Q1GA		0x71
-#define _SPI_NAND_DEVICE_ID_CS11G0T0A0AA	0x00
-#define _SPI_NAND_DEVICE_ID_CS11G1T0A0AA	0x01
-#define _SPI_NAND_DEVICE_ID_CS11G0G0A0AA	0x10
-#define _SPI_NAND_DEVICE_ID_TYM25D2GA01		0x01
-#define _SPI_NAND_DEVICE_ID_TYM25D2GA02		0x02
-#define _SPI_NAND_DEVICE_ID_TYM25D1GA03		0x03
+#define _SPI_NAND_DEVICE_ID_FM25G02			0xF2
+#define _SPI_NAND_DEVICE_ID_XT26G02B		0xF2
+#define _SPI_NAND_DEVICE_ID_XT26G01A		0xE1
+#define _SPI_NAND_DEVICE_ID_XT26G02A		0xE2
+#define _SPI_NAND_DEVICE_ID_PSU1GS20BN		0x21
+#define _SPI_NAND_DEVICE_ID_BWJX08U			0xB1
+#define _SPI_NAND_DEVICE_ID_BWET08U			0xB2
+#define _SPI_NAND_DEVICE_ID_FS35ND01GD1F1	0xA1
+#define _SPI_NAND_DEVICE_ID_FS35ND02GS2F1	0xA2
+#define _SPI_NAND_DEVICE_ID_FS35ND02GD1F1	0xB2
+
 
 
 
@@ -847,6 +851,66 @@
 	
 	{
 		mfr_id: 					_SPI_NAND_MANUFACTURER_ID_ETRON,
+		dev_id: 					_SPI_NAND_DEVICE_ID_EM73C044SND,
+		ptr_name:					"_SPI_NAND_DEVICE_ID_EM73C044SND",
+		device_size:				_SPI_NAND_CHIP_SIZE_1GBIT,
+		page_size:					_SPI_NAND_PAGE_SIZE_2KBYTE,
+		oob_size:					_SPI_NAND_OOB_SIZE_64BYTE,
+		erase_size: 				_SPI_NAND_BLOCK_SIZE_128KBYTE,
+		dummy_mode: 				SPI_NAND_FLASH_READ_DUMMY_BYTE_APPEND,
+		read_mode:					SPI_NAND_FLASH_READ_SPEED_MODE_DUAL,
+		write_mode:					SPI_NAND_FLASH_WRITE_SPEED_MODE_SINGLE,
+		oob_free_layout : 			&ooblayout_type1,		
+		feature:					SPI_NAND_FLASH_FEATURE_NONE,
+	},
+
+	{
+		mfr_id: 					_SPI_NAND_MANUFACTURER_ID_ETRON,
+		dev_id: 					_SPI_NAND_DEVICE_ID_EM73C044SNF,
+		ptr_name:					"_SPI_NAND_DEVICE_ID_EM73C044SNF",
+		device_size:				_SPI_NAND_CHIP_SIZE_1GBIT,
+		page_size:					_SPI_NAND_PAGE_SIZE_2KBYTE,
+		oob_size:					_SPI_NAND_OOB_SIZE_128BYTE,
+		erase_size: 				_SPI_NAND_BLOCK_SIZE_128KBYTE,
+		dummy_mode: 				SPI_NAND_FLASH_READ_DUMMY_BYTE_APPEND,
+		read_mode:					SPI_NAND_FLASH_READ_SPEED_MODE_DUAL,
+		write_mode:					SPI_NAND_FLASH_WRITE_SPEED_MODE_SINGLE,
+		oob_free_layout : 			&ooblayout_type10,		
+		feature:					SPI_NAND_FLASH_FEATURE_NONE,
+	},
+
+	{
+		mfr_id: 					_SPI_NAND_MANUFACTURER_ID_ETRON,
+		dev_id: 					_SPI_NAND_DEVICE_ID_EM73C044VCA,
+		ptr_name:					"_SPI_NAND_DEVICE_ID_EM73C044VCA",
+		device_size:				_SPI_NAND_CHIP_SIZE_1GBIT,
+		page_size:					_SPI_NAND_PAGE_SIZE_2KBYTE,
+		oob_size:					_SPI_NAND_OOB_SIZE_64BYTE,
+		erase_size: 				_SPI_NAND_BLOCK_SIZE_128KBYTE,
+		dummy_mode: 				SPI_NAND_FLASH_READ_DUMMY_BYTE_APPEND,
+		read_mode:					SPI_NAND_FLASH_READ_SPEED_MODE_DUAL,
+		write_mode:					SPI_NAND_FLASH_WRITE_SPEED_MODE_SINGLE,
+		oob_free_layout : 			&ooblayout_type1,		
+		feature:					SPI_NAND_FLASH_FEATURE_NONE,
+	},
+
+	{
+		mfr_id: 					_SPI_NAND_MANUFACTURER_ID_ETRON,
+		dev_id: 					_SPI_NAND_DEVICE_ID_EM73C044VCD,
+		ptr_name:					"_SPI_NAND_DEVICE_ID_EM73C044VCD",
+		device_size:				_SPI_NAND_CHIP_SIZE_1GBIT,
+		page_size:					_SPI_NAND_PAGE_SIZE_2KBYTE,
+		oob_size:					_SPI_NAND_OOB_SIZE_64BYTE,
+		erase_size: 				_SPI_NAND_BLOCK_SIZE_128KBYTE,
+		dummy_mode: 				SPI_NAND_FLASH_READ_DUMMY_BYTE_APPEND,
+		read_mode:					SPI_NAND_FLASH_READ_SPEED_MODE_DUAL,
+		write_mode:					SPI_NAND_FLASH_WRITE_SPEED_MODE_SINGLE,
+		oob_free_layout : 			&ooblayout_type1,		
+		feature:					SPI_NAND_FLASH_FEATURE_NONE,
+	},
+
+	{
+		mfr_id: 					_SPI_NAND_MANUFACTURER_ID_ETRON,
 		dev_id: 					_SPI_NAND_DEVICE_ID_EM73D044SNA,
 		ptr_name:					"_SPI_NAND_DEVICE_ID_EM73D044SNA",
 		device_size:				_SPI_NAND_CHIP_SIZE_2GBIT,
@@ -877,6 +941,111 @@
 
 	{
 		mfr_id: 					_SPI_NAND_MANUFACTURER_ID_ETRON,
+		dev_id: 					_SPI_NAND_DEVICE_ID_EM73D044SND,
+		ptr_name:					"_SPI_NAND_DEVICE_ID_EM73D044SND",
+		device_size:				_SPI_NAND_CHIP_SIZE_2GBIT,
+		page_size:					_SPI_NAND_PAGE_SIZE_2KBYTE,
+		oob_size:					_SPI_NAND_OOB_SIZE_64BYTE,
+		erase_size: 				_SPI_NAND_BLOCK_SIZE_128KBYTE,
+		dummy_mode: 				SPI_NAND_FLASH_READ_DUMMY_BYTE_APPEND,
+		read_mode:					SPI_NAND_FLASH_READ_SPEED_MODE_DUAL,
+		write_mode:					SPI_NAND_FLASH_WRITE_SPEED_MODE_SINGLE,
+		oob_free_layout :			&ooblayout_type1,
+		feature:					SPI_NAND_FLASH_FEATURE_NONE,
+	},
+
+	{
+		mfr_id: 					_SPI_NAND_MANUFACTURER_ID_ETRON,
+		dev_id: 					_SPI_NAND_DEVICE_ID_EM73D044SNF,
+		ptr_name:					"_SPI_NAND_DEVICE_ID_EM73D044SNF",
+		device_size:				_SPI_NAND_CHIP_SIZE_2GBIT,
+		page_size:					_SPI_NAND_PAGE_SIZE_2KBYTE,
+		oob_size:					_SPI_NAND_OOB_SIZE_128BYTE,
+		erase_size: 				_SPI_NAND_BLOCK_SIZE_128KBYTE,
+		dummy_mode: 				SPI_NAND_FLASH_READ_DUMMY_BYTE_APPEND,
+		read_mode:					SPI_NAND_FLASH_READ_SPEED_MODE_DUAL,
+		write_mode:					SPI_NAND_FLASH_WRITE_SPEED_MODE_SINGLE,
+		oob_free_layout :			&ooblayout_type10,		
+		feature:					SPI_NAND_FLASH_FEATURE_NONE,
+	},
+
+	{
+		mfr_id: 					_SPI_NAND_MANUFACTURER_ID_ETRON,
+		dev_id: 					_SPI_NAND_DEVICE_ID_EM73D044VCA,
+		ptr_name:					"_SPI_NAND_DEVICE_ID_EM73D044VCA",
+		device_size:				_SPI_NAND_CHIP_SIZE_2GBIT,
+		page_size:					_SPI_NAND_PAGE_SIZE_2KBYTE,
+		oob_size:					_SPI_NAND_OOB_SIZE_128BYTE,
+		erase_size: 				_SPI_NAND_BLOCK_SIZE_128KBYTE,
+		dummy_mode: 				SPI_NAND_FLASH_READ_DUMMY_BYTE_APPEND,
+		read_mode:					SPI_NAND_FLASH_READ_SPEED_MODE_DUAL,
+		write_mode:					SPI_NAND_FLASH_WRITE_SPEED_MODE_SINGLE,
+		oob_free_layout :			&ooblayout_type18,		
+		feature:					SPI_NAND_FLASH_FEATURE_NONE,
+	},
+
+	{
+		mfr_id: 					_SPI_NAND_MANUFACTURER_ID_ETRON,
+		dev_id: 					_SPI_NAND_DEVICE_ID_EM73D044VCB,
+		ptr_name:					"_SPI_NAND_DEVICE_ID_EM73D044VCB",
+		device_size:				_SPI_NAND_CHIP_SIZE_2GBIT,
+		page_size:					_SPI_NAND_PAGE_SIZE_2KBYTE,
+		oob_size:					_SPI_NAND_OOB_SIZE_64BYTE,
+		erase_size: 				_SPI_NAND_BLOCK_SIZE_128KBYTE,
+		dummy_mode: 				SPI_NAND_FLASH_READ_DUMMY_BYTE_APPEND,
+		read_mode:					SPI_NAND_FLASH_READ_SPEED_MODE_DUAL,
+		write_mode:					SPI_NAND_FLASH_WRITE_SPEED_MODE_SINGLE,
+		oob_free_layout :			&ooblayout_type1,		
+		feature:					SPI_NAND_FLASH_FEATURE_NONE,
+	},
+
+	{
+		mfr_id: 					_SPI_NAND_MANUFACTURER_ID_ETRON,
+		dev_id: 					_SPI_NAND_DEVICE_ID_EM73D044VCD,
+		ptr_name:					"_SPI_NAND_DEVICE_ID_EM73D044VCD",
+		device_size:				_SPI_NAND_CHIP_SIZE_2GBIT,
+		page_size:					_SPI_NAND_PAGE_SIZE_2KBYTE,
+		oob_size:					_SPI_NAND_OOB_SIZE_128BYTE,
+		erase_size: 				_SPI_NAND_BLOCK_SIZE_128KBYTE,
+		dummy_mode: 				SPI_NAND_FLASH_READ_DUMMY_BYTE_APPEND,
+		read_mode:					SPI_NAND_FLASH_READ_SPEED_MODE_DUAL,
+		write_mode:					SPI_NAND_FLASH_WRITE_SPEED_MODE_SINGLE,
+		oob_free_layout :			&ooblayout_type10,		
+		feature:					SPI_NAND_FLASH_FEATURE_NONE,
+	},
+
+	{
+		mfr_id: 					_SPI_NAND_MANUFACTURER_ID_ETRON,
+		dev_id: 					_SPI_NAND_DEVICE_ID_EM73D044VCG,
+		ptr_name:					"_SPI_NAND_DEVICE_ID_EM73D044VCG",
+		device_size:				_SPI_NAND_CHIP_SIZE_2GBIT,
+		page_size:					_SPI_NAND_PAGE_SIZE_2KBYTE,
+		oob_size:					_SPI_NAND_OOB_SIZE_64BYTE,
+		erase_size: 				_SPI_NAND_BLOCK_SIZE_128KBYTE,
+		dummy_mode: 				SPI_NAND_FLASH_READ_DUMMY_BYTE_APPEND,
+		read_mode:					SPI_NAND_FLASH_READ_SPEED_MODE_DUAL,
+		write_mode:					SPI_NAND_FLASH_WRITE_SPEED_MODE_SINGLE,
+		oob_free_layout :			&ooblayout_type1,		
+		feature:					SPI_NAND_FLASH_FEATURE_NONE,
+	},
+
+	{
+		mfr_id: 					_SPI_NAND_MANUFACTURER_ID_ETRON,
+		dev_id: 					_SPI_NAND_DEVICE_ID_EM73D044VCH,
+		ptr_name:					"_SPI_NAND_DEVICE_ID_EM73D044VCH",
+		device_size:				_SPI_NAND_CHIP_SIZE_2GBIT,
+		page_size:					_SPI_NAND_PAGE_SIZE_2KBYTE,
+		oob_size:					_SPI_NAND_OOB_SIZE_64BYTE,
+		erase_size: 				_SPI_NAND_BLOCK_SIZE_128KBYTE,
+		dummy_mode: 				SPI_NAND_FLASH_READ_DUMMY_BYTE_APPEND,
+		read_mode:					SPI_NAND_FLASH_READ_SPEED_MODE_DUAL,
+		write_mode:					SPI_NAND_FLASH_WRITE_SPEED_MODE_SINGLE,
+		oob_free_layout :			&ooblayout_type1,		
+		feature:					SPI_NAND_FLASH_FEATURE_NONE,
+	},
+
+	{
+		mfr_id: 					_SPI_NAND_MANUFACTURER_ID_ETRON,
 		dev_id: 					_SPI_NAND_DEVICE_ID_EM73E044SNA,
 		ptr_name:					"_SPI_NAND_DEVICE_ID_EM73E044SNA",
 		device_size:				_SPI_NAND_CHIP_SIZE_4GBIT,
@@ -1086,25 +1255,10 @@
 	},
 
 	{
-		mfr_id: 					_SPI_NAND_MANUFACTURER_ID_ATO,
-		dev_id: 					_SPI_NAND_DEVICE_ID_ATO25D2GA,
-		ptr_name:					"_SPI_NAND_DEVICE_ID_ATO25D2GA",
-		device_size:				_SPI_NAND_CHIP_SIZE_2GBIT,
-		page_size:					_SPI_NAND_PAGE_SIZE_2KBYTE,
-		oob_size:					_SPI_NAND_OOB_SIZE_64BYTE,
-		erase_size: 				_SPI_NAND_BLOCK_SIZE_128KBYTE,
-		dummy_mode: 				SPI_NAND_FLASH_READ_DUMMY_BYTE_APPEND,
-		read_mode:					SPI_NAND_FLASH_READ_SPEED_MODE_DUAL,
-		write_mode:					SPI_NAND_FLASH_WRITE_SPEED_MODE_SINGLE,
-		oob_free_layout :			&ooblayout_ato_25D2GA,	
-		feature:					SPI_NAND_FLASH_FEATURE_NONE,
-	},
-
-	{
-		mfr_id: 					_SPI_NAND_MANUFACTURER_ID_ATO_2,
-		dev_id: 					_SPI_NAND_DEVICE_ID_ATO25D2GB,
-		ptr_name:					"_SPI_NAND_DEVICE_ID_ATO25D2GB",
-		device_size:				_SPI_NAND_CHIP_SIZE_2GBIT,
+		mfr_id: 					_SPI_NAND_MANUFACTURER_ID_FM,
+		dev_id: 					_SPI_NAND_DEVICE_ID_FM25S01,
+		ptr_name:					"_SPI_NAND_DEVICE_ID_FM25S01",
+		device_size:				_SPI_NAND_CHIP_SIZE_1GBIT,
 		page_size:					_SPI_NAND_PAGE_SIZE_2KBYTE,
 		oob_size:					_SPI_NAND_OOB_SIZE_128BYTE,
 		erase_size: 				_SPI_NAND_BLOCK_SIZE_128KBYTE,
@@ -1176,9 +1330,9 @@
 	},
 	
 	{
-		mfr_id: 					_SPI_NAND_MANUFACTURER_ID_FISON,
-		dev_id: 					_SPI_NAND_DEVICE_ID_CS11G0T0A0AA,
-		ptr_name:					"_SPI_NAND_DEVICE_ID_CS11G0T0A0AA",
+		mfr_id: 					_SPI_NAND_MANUFACTURER_ID_XTX,
+		dev_id: 					_SPI_NAND_DEVICE_ID_XT26G01A,
+		ptr_name:					"_SPI_NAND_DEVICE_ID_XT26G01A",
 		device_size:				_SPI_NAND_CHIP_SIZE_1GBIT,
 		page_size:					_SPI_NAND_PAGE_SIZE_2KBYTE,
 		oob_size:					_SPI_NAND_OOB_SIZE_64BYTE,
@@ -1186,14 +1340,14 @@
 		dummy_mode: 				SPI_NAND_FLASH_READ_DUMMY_BYTE_APPEND,
 		read_mode:					SPI_NAND_FLASH_READ_SPEED_MODE_DUAL,
 		write_mode: 				SPI_NAND_FLASH_WRITE_SPEED_MODE_SINGLE,
-		oob_free_layout :			&ooblayout_fison,	
+		oob_free_layout :			&ooblayout_type19, 	
 		feature:					SPI_NAND_FLASH_FEATURE_NONE,
 	},
 	
 	{
-		mfr_id: 					_SPI_NAND_MANUFACTURER_ID_FISON,
-		dev_id: 					_SPI_NAND_DEVICE_ID_CS11G1T0A0AA,
-		ptr_name:					"_SPI_NAND_DEVICE_ID_CS11G1T0A0AA",
+		mfr_id: 					_SPI_NAND_MANUFACTURER_ID_XTX,
+		dev_id: 					_SPI_NAND_DEVICE_ID_XT26G02A,
+		ptr_name:					"_SPI_NAND_DEVICE_ID_XT26G02A",
 		device_size:				_SPI_NAND_CHIP_SIZE_2GBIT,
 		page_size:					_SPI_NAND_PAGE_SIZE_2KBYTE,
 		oob_size:					_SPI_NAND_OOB_SIZE_64BYTE,
@@ -1251,9 +1405,9 @@
 	},
 
 	{
-		mfr_id: 					_SPI_NAND_MANUFACTURER_ID_TYM,
-		dev_id: 					_SPI_NAND_DEVICE_ID_TYM25D1GA03,
-		ptr_name:					"_SPI_NAND_DEVICE_ID_TYM25D1GA03",
+		mfr_id: 					_SPI_NAND_MANUFACTURER_ID_FORESEE,
+		dev_id: 					_SPI_NAND_DEVICE_ID_FS35ND01GD1F1,
+		ptr_name:					"_SPI_NAND_DEVICE_ID_FS35ND01GD1F1",
 		device_size:				_SPI_NAND_CHIP_SIZE_1GBIT,
 		page_size:					_SPI_NAND_PAGE_SIZE_2KBYTE,
 		oob_size:					_SPI_NAND_OOB_SIZE_64BYTE,
@@ -1261,10 +1415,39 @@
 		dummy_mode: 				SPI_NAND_FLASH_READ_DUMMY_BYTE_APPEND,
 		read_mode:					SPI_NAND_FLASH_READ_SPEED_MODE_DUAL,
 		write_mode: 				SPI_NAND_FLASH_WRITE_SPEED_MODE_SINGLE,
-		oob_free_layout :			&ooblayout_tym, 	
+		oob_free_layout :			&ooblayout_type1, 	
 		feature:					SPI_NAND_FLASH_FEATURE_NONE,
 	},
 
+	{
+		mfr_id: 					_SPI_NAND_MANUFACTURER_ID_FORESEE,
+		dev_id: 					_SPI_NAND_DEVICE_ID_FS35ND02GS2F1,
+		ptr_name:					"_SPI_NAND_DEVICE_ID_FS35ND02GS2F1",
+		device_size:				_SPI_NAND_CHIP_SIZE_2GBIT,
+		page_size:					_SPI_NAND_PAGE_SIZE_2KBYTE,
+		oob_size:					_SPI_NAND_OOB_SIZE_128BYTE,
+		erase_size: 				_SPI_NAND_BLOCK_SIZE_128KBYTE,
+		dummy_mode: 				SPI_NAND_FLASH_READ_DUMMY_BYTE_APPEND,
+		read_mode:					SPI_NAND_FLASH_READ_SPEED_MODE_DUAL,
+		write_mode:					SPI_NAND_FLASH_WRITE_SPEED_MODE_SINGLE,
+		oob_free_layout :			&ooblayout_type1, 	
+		feature:					SPI_NAND_FLASH_FEATURE_NONE,
+	},
+
+	{
+		mfr_id: 					_SPI_NAND_MANUFACTURER_ID_FORESEE,
+		dev_id: 					_SPI_NAND_DEVICE_ID_FS35ND02GD1F1,
+		ptr_name:					"_SPI_NAND_DEVICE_ID_FS35ND02GD1F1",
+		device_size:				_SPI_NAND_CHIP_SIZE_2GBIT,
+		page_size:					_SPI_NAND_PAGE_SIZE_2KBYTE,
+		oob_size:					_SPI_NAND_OOB_SIZE_128BYTE,
+		erase_size: 				_SPI_NAND_BLOCK_SIZE_128KBYTE,
+		dummy_mode: 				SPI_NAND_FLASH_READ_DUMMY_BYTE_APPEND,
+		read_mode:					SPI_NAND_FLASH_READ_SPEED_MODE_DUAL,
+		write_mode:					SPI_NAND_FLASH_WRITE_SPEED_MODE_SINGLE,
+		oob_free_layout :			&ooblayout_type1, 	
+		feature:					SPI_NAND_FLASH_FEATURE_NONE,
+	},
 };
 
 /* LOCAL SUBPROGRAM BODIES------------------------------------------------------------ */
@@ -2391,8 +2574,7 @@
 			rtn_status = SPI_NAND_FLASH_RTN_DETECTED_BAD_BLOCK;
 		}
 	}
-	else if(((ptr_dev_info_t->mfr_id == _SPI_NAND_MANUFACTURER_ID_ATO) && (ptr_dev_info_t->dev_id == _SPI_NAND_DEVICE_ID_ATO25D2GA)) ||
-		((ptr_dev_info_t->mfr_id == _SPI_NAND_MANUFACTURER_ID_ATO_2) && (ptr_dev_info_t->dev_id == _SPI_NAND_DEVICE_ID_ATO25D2GB)))
+	else if((ptr_dev_info_t->mfr_id == _SPI_NAND_MANUFACTURER_ID_FM) && (ptr_dev_info_t->dev_id == _SPI_NAND_DEVICE_ID_FM25S01))
 	{
 		if(((status & 0x30) >> 4) == 0x2)
 		{
@@ -2407,8 +2589,21 @@
 			rtn_status = SPI_NAND_FLASH_RTN_DETECTED_BAD_BLOCK;
 		}
 	}
-	else if(((ptr_dev_info_t->mfr_id == _SPI_NAND_MANUFACTURER_ID_DS) && (ptr_dev_info_t->dev_id == _SPI_NAND_DEVICE_ID_DS35Q2GA)) || 
-		((ptr_dev_info_t->mfr_id == _SPI_NAND_MANUFACTURER_ID_DS) && (ptr_dev_info_t->dev_id == _SPI_NAND_DEVICE_ID_DS35Q1GA)))
+	else if(((ptr_dev_info_t->mfr_id == _SPI_NAND_MANUFACTURER_ID_XTX) && (ptr_dev_info_t->dev_id == _SPI_NAND_DEVICE_ID_XT26G02B)))
+	{
+		if(((status & 0x70) >> 4) == 0x7)
+		{
+			rtn_status = SPI_NAND_FLASH_RTN_DETECTED_BAD_BLOCK;
+		}
+	}
+	else if(((ptr_dev_info_t->mfr_id == _SPI_NAND_MANUFACTURER_ID_XTX) && (ptr_dev_info_t->dev_id == _SPI_NAND_DEVICE_ID_XT26G01A)))
+	{
+		if(((status & 0x3C) >> 2) == 0x8 )
+		{
+			rtn_status = SPI_NAND_FLASH_RTN_DETECTED_BAD_BLOCK;
+		}
+	}
+	else if(((ptr_dev_info_t->mfr_id == _SPI_NAND_MANUFACTURER_ID_XTX) && (ptr_dev_info_t->dev_id == _SPI_NAND_DEVICE_ID_XT26G02A)))
 	{
 		if(((status & 0x30) >> 4) == 0x2)
 		{
@@ -2431,6 +2626,16 @@
 			rtn_status = SPI_NAND_FLASH_RTN_DETECTED_BAD_BLOCK;
 		}
 	}
+	else if(((ptr_dev_info_t->mfr_id == _SPI_NAND_MANUFACTURER_ID_FORESEE) && (ptr_dev_info_t->dev_id == _SPI_NAND_DEVICE_ID_FS35ND02GS2F1)) ||
+		((ptr_dev_info_t->mfr_id == _SPI_NAND_MANUFACTURER_ID_FORESEE) && (ptr_dev_info_t->dev_id == _SPI_NAND_DEVICE_ID_FS35ND02GD1F1)) ||
+		((ptr_dev_info_t->mfr_id == _SPI_NAND_MANUFACTURER_ID_FORESEE) && (ptr_dev_info_t->dev_id == _SPI_NAND_DEVICE_ID_FS35ND01GD1F1)))
+	{
+		if(((status & 0x70) >> 4) == 0x7)
+		{
+			rtn_status = SPI_NAND_FLASH_RTN_DETECTED_BAD_BLOCK;
+		}
+	}
+
 	else if(rtn_status == SPI_NAND_FLASH_RTN_DETECTED_BAD_BLOCK)
 	{
 		_SPI_NAND_PRINTF("[spinand_ecc_fail_check] : ECC cannot recover detected !, page=0x%x\n", page_number);
@@ -4098,6 +4303,13 @@
 		feature &= 0xC7;
 		spi_nand_protocol_set_status_reg_1(feature);
 	}
+	else if((ptr_device_t->mfr_id == _SPI_NAND_MANUFACTURER_ID_FM) && (ptr_device_t->dev_id == _SPI_NAND_DEVICE_ID_FM25S01))
+	{
+		/* 1. Unlock All block */
+		spi_nand_protocol_get_status_reg_1(&feature);
+		feature &= 0x87;
+		spi_nand_protocol_set_status_reg_1(feature);
+	}
 	else if(((ptr_device_t->mfr_id == _SPI_NAND_MANUFACTURER_ID_FM) && (ptr_device_t->dev_id == _SPI_NAND_DEVICE_ID_FM25G01B)) ||
 			((ptr_device_t->mfr_id == _SPI_NAND_MANUFACTURER_ID_FM) && (ptr_device_t->dev_id == _SPI_NAND_DEVICE_ID_FM25G02B)))
 	{
@@ -4145,7 +4357,9 @@
 		spi_nand_protocol_get_status_reg_1(&feature);
 		_SPI_NAND_DEBUG_PRINTF(SPI_NAND_FLASH_DEBUG_LEVEL_1, "After Unlock all block setup, the status register1 =0x%x\n", feature);		
 	}
-	else if((ptr_device_t->mfr_id == _SPI_NAND_MANUFACTURER_ID_TYM) && (ptr_device_t->dev_id == _SPI_NAND_DEVICE_ID_TYM25D2GA01)) 
+	else if(((ptr_device_t->mfr_id == _SPI_NAND_MANUFACTURER_ID_FORESEE) && (ptr_device_t->dev_id == _SPI_NAND_DEVICE_ID_FS35ND02GS2F1)) ||
+		((ptr_device_t->mfr_id == _SPI_NAND_MANUFACTURER_ID_FORESEE) && (ptr_device_t->dev_id == _SPI_NAND_DEVICE_ID_FS35ND02GD1F1)) ||
+		((ptr_device_t->mfr_id == _SPI_NAND_MANUFACTURER_ID_FORESEE) && (ptr_device_t->dev_id == _SPI_NAND_DEVICE_ID_FS35ND01GD1F1)))
 	{
 		/* 1. Unlock All block */
 		spi_nand_protocol_get_status_reg_1(&feature);
