Index: linux-3.18.21/net/bridge/ecnt_br.c
===================================================================
--- linux-3.18.21.orig/net/bridge/ecnt_br.c	2019-11-11 21:55:48.458907522 +0800
+++ linux-3.18.21/net/bridge/ecnt_br.c	2019-11-11 21:55:48.594907518 +0800
@@ -508,8 +508,8 @@
 
 	hash = br_ip_hash_for_export(mdb, &group);
 	mp = br_multicast_get_group(br, port, &group, hash);
-
-	if (mp==NULL)
+	if (mp==NULL || PTR_ERR(mp) == -E2BIG || PTR_ERR(mp) == -EEXIST 
+           || PTR_ERR(mp) == -EAGAIN || PTR_ERR(mp) == -ENOMEM || PTR_ERR(mp) == -EINVAL)
 	{
 		IGMP_HWNAT_DEBUG("mp == NULL");
 		return 0;
Index: linux-3.18.21/net/bridge/br_multicast.c
===================================================================
--- linux-3.18.21.orig/net/bridge/br_multicast.c	2019-11-11 21:55:47.858907539 +0800
+++ linux-3.18.21/net/bridge/br_multicast.c	2019-11-11 21:59:17.282901497 +0800
@@ -647,6 +647,10 @@
 		break;
 
 	case -EAGAIN:
+	case -E2BIG:
+	case -EEXIST:
+	case -ENOMEM:
+	case -EINVAL:
 rehash:
 		mdb = rcu_dereference_protected(br->mdb, 1);
 		hash = br_ip_hash(mdb, group);
