Index: udhcp-0.9.8/dhcpc.c
===================================================================
--- udhcp-0.9.8.orig/dhcpc.c	2019-01-14 13:43:52.826583349 +0800
+++ udhcp-0.9.8/dhcpc.c	2019-01-14 13:47:01.229364410 +0800
@@ -91,11 +91,7 @@
 char Forecerenewkey[128]="";
 #endif
 
-#if defined(ECONET_PLATFORM) && !defined(SUPPORT_XPON) //only for en7512 projects
-#define INIT_TIMEOUT 60	//original source code definition
-#else
 #define INIT_TIMEOUT 5
-#endif
 
 #define DEFAULT_SCRIPT	"/usr/share/udhcpc/default.script"
 
