Index: libzcfg_fe_tr98-1.0/wanDev.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/wanDev.c	2018-10-18 15:14:16.477029208 +0800
+++ libzcfg_fe_tr98-1.0/wanDev.c	2018-10-18 17:10:33.893252487 +0800
@@ -1206,7 +1206,7 @@
 
 #ifdef TR98_DSLINTFCFG_STATS_REQ_SUBOBJPARAM_NOTIFY
 // require 'WANDSLInterfaceConfig.Stats' sub object param notify
-zcfgRet_t WANDslIfaceCfgStatsSubObj_RetrieveTr181DslLineStatsSubObj(char *tr98FullPathName, struct json_object **tr98Jobj, struct json_object **dslStatsObj, bool updateFlag)
+zcfgRet_t WANDslIfaceCfgStatsSubObj_RetrieveTr181DslLineStatsSubObj(char *tr98FullPathName, struct json_object **tr98Jobj, struct json_object **dslStatsObj, bool updateFlag, int handler)
 {
 	zcfgRet_t ret = ZCFG_SUCCESS;
 	tr98Parameter_t *paramList = NULL;
@@ -1268,7 +1268,7 @@
 	return ZCFG_SUCCESS;
 }
 #else
-zcfgRet_t WANDslIfaceCfgStatsSubObj_RetrieveTr181DslLineStatsSubObj(char *tr98FullPathName, struct json_object **tr98Jobj, struct json_object **dslStatsObj, bool updateFlag)
+zcfgRet_t WANDslIfaceCfgStatsSubObj_RetrieveTr181DslLineStatsSubObj(char *tr98FullPathName, struct json_object **tr98Jobj, struct json_object **dslStatsObj, bool updateFlag, int handler)
 {
 	zcfgRet_t ret = ZCFG_SUCCESS;
 	tr98Parameter_t *paramList = NULL;
@@ -1377,7 +1377,7 @@
 	zcfgLog(ZCFG_LOG_INFO, "%s : Enter\n", __FUNCTION__);
 
 	*tr98Jobj = NULL;
-	ret = WANDslIfaceCfgStatsSubObj_RetrieveTr181DslLineStatsSubObj(tr98FullPathName, tr98Jobj, &dslStatsObj, updateFlag);
+	ret = WANDslIfaceCfgStatsSubObj_RetrieveTr181DslLineStatsSubObj(tr98FullPathName, tr98Jobj, &dslStatsObj, updateFlag, handler);
 	if(*tr98Jobj)
 		return ret;
 	else if(!dslStatsObj)
