Index: libzcfg_fe_tr98-1.0/voice.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/voice.h	2019-03-11 15:33:58.867417792 +0800
+++ libzcfg_fe_tr98-1.0/voice.h	2019-03-11 15:35:41.667991791 +0800
@@ -564,6 +564,7 @@
 #else
 	{ "X_ZYXEL_LineSelect", PARAMETER_ATTR_WRITE, 256, json_type_string},
 #endif
+	{ "X_ZYXEL_HookState", PARAMETER_ATTR_READONLY, 32, json_type_string},
 	//{ "X_ZYXEL_MsnIncomingSelect", PARAMETER_ATTR_WRITE, 0, json_type_uint32},
 	//{ "X_ZYXEL_MsnOutgoingSelect", PARAMETER_ATTR_WRITE, 0, json_type_uint32},
 	//{ "X_ZYXEL_HowlerToneTimeout", PARAMETER_ATTR_WRITE, 0, json_type_uint32},
