Index: libzcfg_fe_tr98-1.0/wanDev.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/wanDev.c	2019-04-22 14:57:42.812181727 +0800
+++ libzcfg_fe_tr98-1.0/wanDev.c	2019-04-23 16:27:37.982446417 +0800
@@ -6050,6 +6050,7 @@
 	struct json_object *pppIfaceJobj = NULL, *ipIfaceJobj = NULL, *vlanTermJobj = NULL;
 	struct json_object *paramValue = NULL;
 	struct json_object *paramValue2 = NULL;
+	struct json_object *paramValue3 = NULL;
 	struct json_object *tr181ParamValue = NULL;
 	struct json_object *tmpObj = NULL;
 	struct json_object *dnsSrvJobj = NULL, *dnsSrvV6Jobj = NULL, *dnsJobj = NULL, *igmpJobj = NULL;
@@ -6404,7 +6405,15 @@
 				continue;
 			}
 			else if(!strcmp(paramList->name, "X_ZYXEL_IPCPEnable")){
-				json_object_object_add(pppIfaceJobj, "IPCPEnable", JSON_OBJ_COPY(paramValue));
+				paramValue3 = json_object_object_get(ipIfaceJobj, "IPv4Enable");
+				//printf("IPv6CPEnable: pppIfaceJobj para3=%d %d\n", paramValue3, json_object_get_boolean(paramValue3));
+				if(paramValue3)
+				{
+					if (json_object_get_boolean(paramValue3) == true)
+						json_object_object_add(pppIfaceJobj, "IPCPEnable", json_object_new_boolean(true));
+					else
+						json_object_object_add(pppIfaceJobj, "IPCPEnable", json_object_new_boolean(false));
+				}
 				paramList++;
 				continue;
 			}
@@ -6563,7 +6572,15 @@
 			}
 #ifdef IPV6INTERFACE_PROFILE		
 			else if(!strcmp(paramList->name, "X_ZYXEL_IPv6CPEnable")){
-				json_object_object_add(pppIfaceJobj, "IPv6CPEnable", JSON_OBJ_COPY(paramValue));
+				paramValue3 = json_object_object_get(ipIfaceJobj, "IPv6Enable");
+				//printf("IPv6CPEnable: pppIfaceJobj para3=%d %d\n", paramValue3, json_object_get_boolean(paramValue3));
+				if(paramValue3)
+				{
+					if (json_object_get_boolean(paramValue3) == true)
+						json_object_object_add(pppIfaceJobj, "IPv6CPEnable", json_object_new_boolean(true));
+					else
+						json_object_object_add(pppIfaceJobj, "IPv6CPEnable", json_object_new_boolean(false));
+				}
 				paramList++;
 				continue;
 			}
