Index: libzcfg_fe_tr98-1.0/lanDev.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/lanDev.c	2019-04-02 14:10:10.612438595 +0800
+++ libzcfg_fe_tr98-1.0/lanDev.c	2019-04-02 14:12:53.712502331 +0800
@@ -4818,6 +4818,9 @@
 		else if(!strcmp(paramList->name, "X_ZYXEL_SingleSKU")) {
 			paramValue = json_object_object_get(radioJobj, "X_ZYXEL_SingleSKU");
 		}
+		else if(!strcmp(paramList->name, "X_ZYXEL_Wireless_Mode")) {
+			paramValue = json_object_object_get(radioJobj, "X_ZYXEL_Wireless_Mode");
+		}
 		else if(!strcmp(paramList->name, "WEPKeyIndex")) {
 			int defaultKeyIndex = json_object_get_int(json_object_object_get(apSecJobj, "X_ZYXEL_DefaultKeyID"));
 			defaultKeyIndex = (defaultKeyIndex == 0) ? 1 : defaultKeyIndex;
Index: libzcfg_fe_tr98-1.0/lanDev_parameter.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/lanDev_parameter.h	2019-04-02 14:10:10.216438579 +0800
+++ libzcfg_fe_tr98-1.0/lanDev_parameter.h	2019-04-02 14:32:26.255503538 +0800
@@ -308,6 +308,7 @@
 #endif
 	{ "X_ZYXEL_SSID_Priority_Enable", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
 	{ "X_ZYXEL_SSID_Priority", PARAMETER_ATTR_WRITE, 0, json_type_uint32},
+	{ "X_ZYXEL_Wireless_Mode", PARAMETER_ATTR_WRITE, 8, json_type_string},
 	{ NULL, 0, 0, 0}
 };
 
