diff -urN libzcfg_fe_tr98-1.0.org/firewall.c libzcfg_fe_tr98-1.0/firewall.c
--- libzcfg_fe_tr98-1.0.org/firewall.c	2019-05-07 17:28:22.358292100 +0800
+++ libzcfg_fe_tr98-1.0/firewall.c	2019-05-08 11:00:02.091804050 +0800
@@ -216,7 +216,7 @@
 
 zcfgRet_t firewallObjAttrSet(char *tr98FullPathName, int handler, char *paramName, int newNotify, struct json_object *multiAttrJobj)
 {
-	zcfgRet_t ret;
+	zcfgRet_t ret = ZCFG_SUCCESS;
 	char tr181Obj[128] = {0};
 	char mapObj[128] = {0};
 	uint32_t  devInfoOid = 0;
@@ -475,7 +475,7 @@
 
 zcfgRet_t firewallLevelObjAttrSet(char *tr98FullPathName, int handler, char *paramName, int newNotify, struct json_object *multiAttrJobj)
 {
-	zcfgRet_t ret;
+	zcfgRet_t ret = ZCFG_SUCCESS;
 	char tr181Obj[128] = {0};
 	char mapObj[128] = {0};
 	uint32_t  devInfoOid = 0;
@@ -737,7 +737,7 @@
 
 zcfgRet_t firewallChainObjAttrSet(char *tr98FullPathName, int handler, char *paramName, int newNotify, struct json_object *multiAttrJobj)
 {
-	zcfgRet_t ret;
+	zcfgRet_t ret = ZCFG_SUCCESS;
 	char tr181Obj[128] = {0};
 	char mapObj[128] = {0};
 	uint32_t  devInfoOid = 0;
@@ -1001,7 +1001,7 @@
 
 zcfgRet_t firewallRuleObjAttrSet(char *tr98FullPathName, int handler, char *paramName, int newNotify, struct json_object *multiAttrJobj)
 {
-	zcfgRet_t ret;
+	zcfgRet_t ret = ZCFG_SUCCESS;
 	char tr181Obj[128] = {0};
 	char mapObj[128] = {0};
 	uint32_t  devInfoOid = 0;
diff -urN libzcfg_fe_tr98-1.0.org/l2bridge.c libzcfg_fe_tr98-1.0/l2bridge.c
--- libzcfg_fe_tr98-1.0.org/l2bridge.c	2019-05-07 17:28:23.682292100 +0800
+++ libzcfg_fe_tr98-1.0/l2bridge.c	2019-05-08 11:04:59.831088293 +0800
@@ -245,7 +245,7 @@
 
 zcfgRet_t l2BridingObjAttrSet(char *tr98FullPathName, int handler, char *paramName, int newNotify, struct json_object *multiAttrJobj)
 {
-	zcfgRet_t ret;
+	zcfgRet_t ret = ZCFG_SUCCESS;
 	char tr181Obj[128] = {0};
 	char mapObj[128] = {0};
 	uint32_t  devInfoOid = 0;
@@ -722,7 +722,7 @@
 
 zcfgRet_t l2BrObjAttrSet(char *tr98FullPathName, int handler, char *paramName, int newNotify, struct json_object *multiAttrJobj)
 {
-	zcfgRet_t ret;
+	zcfgRet_t ret = ZCFG_SUCCESS;
 	char tr181Obj[128] = {0};
 	char mapObj[128] = {0};
 	uint32_t  devInfoOid = 0;
@@ -1302,7 +1302,7 @@
 
 zcfgRet_t l2BrFilterObjAttrSet(char *tr98FullPathName, int handler, char *paramName, int newNotify, struct json_object *multiAttrJobj)
 {
-	zcfgRet_t ret;
+	zcfgRet_t ret = ZCFG_SUCCESS;
 	char tr181Obj[128] = {0};
 	char mapObj[128] = {0};
 	uint32_t  devInfoOid = 0;
@@ -1538,7 +1538,7 @@
 
 zcfgRet_t l2BrAvailableIntfObjAttrSet(char *tr98FullPathName, int handler, char *paramName, int newNotify, struct json_object *multiAttrJobj)
 {
-	zcfgRet_t ret;
+	zcfgRet_t ret = ZCFG_SUCCESS;
 	char tr181Obj[128] = {0};
 	char mapObj[128] = {0};
 	uint32_t  devInfoOid = 0;
@@ -2401,7 +2401,7 @@
 
 zcfgRet_t l2BrVlanObjAttrSet(char *tr98FullPathName, int handler, char *paramName, int newNotify, struct json_object *multiAttrJobj)
 {
-	zcfgRet_t ret;
+	zcfgRet_t ret = ZCFG_SUCCESS;
 	char tr181Obj[128] = {0};
 	char mapObj[128] = {0};
 	uint32_t  devInfoOid = 0;
diff -urN libzcfg_fe_tr98-1.0.org/lanConfSec.c libzcfg_fe_tr98-1.0/lanConfSec.c
--- libzcfg_fe_tr98-1.0.org/lanConfSec.c	2019-05-07 17:28:11.550292000 +0800
+++ libzcfg_fe_tr98-1.0/lanConfSec.c	2019-05-08 11:05:23.347117852 +0800
@@ -185,7 +185,7 @@
 
 zcfgRet_t lanConfSecObjAttrSet(char *tr98FullPathName, int handler, char *paramName, int newNotify, struct json_object *multiAttrJobj)
 {
-	zcfgRet_t ret;
+	zcfgRet_t ret = ZCFG_SUCCESS;
 	char tr181Obj[128] = {0};
 	char mapObj[128] = {0};
 	uint32_t  devInfoOid = 0;
diff -urN libzcfg_fe_tr98-1.0.org/lanDev.c libzcfg_fe_tr98-1.0/lanDev.c
--- libzcfg_fe_tr98-1.0.org/lanDev.c	2019-05-07 17:28:24.662292100 +0800
+++ libzcfg_fe_tr98-1.0/lanDev.c	2019-05-08 11:05:59.289252101 +0800
@@ -3166,7 +3166,7 @@
 
 zcfgRet_t lanDhcpStaticAddrObjAttrSet(char *tr98FullPathName, int handler, char *paramName, int newNotify, struct json_object *multiAttrJobj)
 {
-	zcfgRet_t ret;
+	zcfgRet_t ret =ZCFG_SUCCESS;
 	char tr181Obj[128] = {0};
 	char mapObj[128] = {0};
 	uint32_t  devInfoOid = 0;
@@ -3535,7 +3535,7 @@
 
 zcfgRet_t lanDhcpOptObjAttrSet(char *tr98FullPathName, int handler, char *paramName, int newNotify, struct json_object *multiAttrJobj)
 {
-	zcfgRet_t ret;
+	zcfgRet_t ret = ZCFG_SUCCESS;
 	char tr181Obj[128] = {0};
 	char mapObj[128] = {0};
 	uint32_t  devInfoOid = 0;
@@ -4463,7 +4463,7 @@
 
 zcfgRet_t lanEthIntfConfObjAttrSet(char *tr98FullPathName, int handler, char *paramName, int newNotify, struct json_object *multiAttrJobj)
 {
-	zcfgRet_t ret;
+	zcfgRet_t ret = ZCFG_SUCCESS;
 	char tr181Obj[128] = {0};
 	char mapObj[128] = {0};
 	uint32_t  devInfoOid = 0;
diff -urN libzcfg_fe_tr98-1.0.org/mgmtSrv.c libzcfg_fe_tr98-1.0/mgmtSrv.c
--- libzcfg_fe_tr98-1.0.org/mgmtSrv.c	2019-05-07 17:28:24.618292100 +0800
+++ libzcfg_fe_tr98-1.0/mgmtSrv.c	2019-05-08 11:07:57.095715704 +0800
@@ -265,7 +265,7 @@
 /* all use the batch set, only tr069 use this action*/
 zcfgRet_t mgmtSrvObjAttrSet(char *tr98FullPathName, int handler, char *paramName, int newNotify, struct json_object *multiAttrJobj)
 {
-	zcfgRet_t ret;
+	zcfgRet_t ret = ZCFG_SUCCESS;
 	char tr181Obj[128] = {0};
 	char mapObj[128] = {0};
 	uint32_t  mgmtSrvOid = 0;
diff -urN libzcfg_fe_tr98-1.0.org/qos.c libzcfg_fe_tr98-1.0/qos.c
--- libzcfg_fe_tr98-1.0.org/qos.c	2019-05-07 17:28:20.306292100 +0800
+++ libzcfg_fe_tr98-1.0/qos.c	2019-05-08 11:08:38.669104540 +0800
@@ -282,7 +282,7 @@
 
 zcfgRet_t qMgmtAttrSet(char *tr98FullPathName, int handler, char *paramName, int newNotify, struct json_object *multiAttrJobj)
 {
-	zcfgRet_t ret;
+	zcfgRet_t ret = ZCFG_SUCCESS;
 	char tr181Obj[128] = {0};
 	char mapObj[128] = {0};
 	uint32_t  devInfoOid = 0;
@@ -694,7 +694,7 @@
 
 zcfgRet_t qMgmtQueAttrSet(char *tr98FullPathName, int handler, char *paramName, int newNotify, struct json_object *multiAttrJobj)
 {
-	zcfgRet_t ret;
+	zcfgRet_t ret = ZCFG_SUCCESS;
 	char tr181Obj[128] = {0};
 	char mapObj[128] = {0};
 	uint32_t  devInfoOid = 0;
@@ -1336,7 +1336,7 @@
 
 zcfgRet_t qMgmtClsAttrSet(char *tr98FullPathName, int handler, char *paramName, int newNotify, struct json_object *multiAttrJobj)
 {
-	zcfgRet_t ret;
+	zcfgRet_t ret = ZCFG_SUCCESS;
 	char tr181Obj[128] = {0};
 	char mapObj[128] = {0};
 	uint32_t  devInfoOid = 0;
@@ -1645,7 +1645,7 @@
 
 zcfgRet_t qMgmtShaperAttrSet(char *tr98FullPathName, int handler, char *paramName, int newNotify, struct json_object *multiAttrJobj)
 {
-	zcfgRet_t ret;
+	zcfgRet_t ret = ZCFG_SUCCESS;
 	char tr181Obj[128] = {0};
 	char mapObj[128] = {0};
 	uint32_t  devInfoOid = 0;
