diff -urN libzcfg_fe_tr98-1.0.org/lanDev.c libzcfg_fe_tr98-1.0/lanDev.c
--- libzcfg_fe_tr98-1.0.org/lanDev.c	2019-05-08 11:11:00.907243300 +0800
+++ libzcfg_fe_tr98-1.0/lanDev.c	2019-05-08 15:23:51.051000222 +0800
@@ -3284,7 +3284,7 @@
 	objIndex_t dhcpOptIid;
 	struct json_object *dhcpOptJobj = NULL;
 	struct json_object *paramValue = NULL;
-	struct json_object *tr181ParamValue = NULL;
+	// struct json_object *tr181ParamValue = NULL;
 	//struct json_object *tmpObj = NULL;
 	tr98Parameter_t *paramList = NULL;
 	
@@ -6093,13 +6093,14 @@
 	objIndex_t apIid;
 	struct json_object *apJobj = NULL, *apWpsJobj = NULL, *apSecJobj = NULL;
 	struct json_object *paramValue = NULL;
-	struct json_object *tr181ParamValue = NULL;
+	// struct json_object *tr181ParamValue = NULL;
 	struct json_object *tmpObj = NULL;
 	tr98Parameter_t *paramList = NULL;
 	//bool tr98WpsEnable, tr181WpsEnable;
 	bool tr98WpsEnable;
 	//int tr98WpsMthSupport = 0;
-	const char *tr98WpsMthEnabledStr = NULL, *tr98WpsMthSupportStr = NULL;
+	// const char *tr98WpsMthEnabledStr = NULL;
+	const char *tr98WpsMthSupportStr = NULL;
 	//bool wpsConfigured = true, wpsOldConfigured = true;
 	bool wpsConfigured = true;
 	//int tr98WpsMthEnabled = 0;
@@ -6224,7 +6225,7 @@
 
 	tr98WpsEnable = json_object_get_boolean(json_object_object_get(tr98Jobj, "Enable"));
 	tr98WpsMthSupportStr =	json_object_get_string(json_object_object_get(tr98Jobj, "ConfigMethodsSupported"));
-	tr98WpsMthEnabledStr =	json_object_get_string(json_object_object_get(tr98Jobj, "ConfigMethodsEnabled"));
+	// tr98WpsMthEnabledStr =	json_object_get_string(json_object_object_get(tr98Jobj, "ConfigMethodsEnabled"));
 
 #if 0
 	tr181WpsEnable = json_object_get_boolean(json_object_object_get(apWpsJobj, "Enable"));
@@ -6307,7 +6308,7 @@
 		}
 #ifdef WPS_ACCESSPOINT_PIN
 		else if(!strcmp(paramList->name, "X_ZYXEL_WPS_Configured")) {
-			tr181ParamValue = json_object_object_get(apWpsJobj, "X_ZYXEL_WPS_Configured");
+			// tr181ParamValue = json_object_object_get(apWpsJobj, "X_ZYXEL_WPS_Configured");
 #if 0 //don't allow ACS trigger WPS method
 			wpsOldConfigured = json_object_get_boolean(tr181ParamValue);
 #endif
diff -urN libzcfg_fe_tr98-1.0.org/qos.c libzcfg_fe_tr98-1.0/qos.c
--- libzcfg_fe_tr98-1.0.org/qos.c	2019-05-08 11:11:00.907243300 +0800
+++ libzcfg_fe_tr98-1.0/qos.c	2019-05-08 14:58:40.365192311 +0800
@@ -820,7 +820,8 @@
 	zcfgRet_t ret = ZCFG_SUCCESS;
 	objIndex_t objIid;
 	struct json_object *qPolicerObj = NULL;
-	struct json_object *paramValue = NULL, *tr98ParamValue = NULL;
+	// struct json_object *paramValue = NULL;
+	struct json_object *tr98ParamValue = NULL;
 	struct json_object *tmpObj = NULL;
 	tr98Parameter_t *paramList = NULL;
 
@@ -849,11 +850,13 @@
 
 		if(!strcmp(paramList->name, "PolicerEnable"))
 		{
-			if((paramValue = json_object_object_get(qPolicerObj, "Enable")))
+			// if((paramValue = json_object_object_get(qPolicerObj, "Enable")))
+			if((json_object_object_get(qPolicerObj, "Enable")))
 				json_object_object_add(qPolicerObj, "Enable", JSON_OBJ_COPY(tr98ParamValue));
 		}
 		else{
-			if((paramValue = json_object_object_get(qPolicerObj, paramList->name))){
+			// if((paramValue = json_object_object_get(qPolicerObj, paramList->name))){
+			if((json_object_object_get(qPolicerObj, paramList->name))){
 				json_object_object_add(qPolicerObj, paramList->name, JSON_OBJ_COPY(tr98ParamValue));
 			}
 		}
diff -urN libzcfg_fe_tr98-1.0.org/wanDev.c libzcfg_fe_tr98-1.0/wanDev.c
--- libzcfg_fe_tr98-1.0.org/wanDev.c	2019-05-08 11:11:00.819243300 +0800
+++ libzcfg_fe_tr98-1.0/wanDev.c	2019-05-08 15:16:54.882211575 +0800
@@ -3476,7 +3476,7 @@
 				continue;
 			}
 			else if(!strcmp(paramList->name, "X_ZYXEL_FullConeEnabled")){
-				const char *cmd = json_object_get_string(paramValue);
+				// const char *cmd = json_object_get_string(paramValue);
 				json_object_object_add(natIntfJobj, paramList->name, JSON_OBJ_COPY(paramValue));
 				paramList++;
 				continue;
@@ -9215,7 +9215,7 @@
 	objIndex_t atmF5LoIid;
 	struct json_object *atmF5LoJobj = NULL;
 	struct json_object *paramValue = NULL;
-	struct json_object *ifValue = NULL;
+	// struct json_object *ifValue = NULL;
 	tr98Parameter_t *paramList = NULL;	
 	
 	ztrdbg(ZTR_DL_DEBUG, "%s : Enter\n", __FUNCTION__);
@@ -9244,7 +9244,7 @@
 	/*fill up tr98 WANATMF5LoopbackDiagnostics object*/
 	*tr98Jobj = json_object_new_object();
 	paramList = tr98Obj[handler].parameter;
-	ifValue = json_object_object_get(atmF5LoJobj, "Interface");
+	// ifValue = json_object_object_get(atmF5LoJobj, "Interface");
 
 	while(paramList->name != NULL) {
 		paramValue = json_object_object_get(atmF5LoJobj, paramList->name);
@@ -9410,7 +9410,7 @@
 	objIndex_t atmF4LoIid;
 	struct json_object *atmF4LoJobj = NULL;
 	struct json_object *paramValue = NULL;
-	struct json_object *ifValue = NULL;
+	// struct json_object *ifValue = NULL;
 	tr98Parameter_t *paramList = NULL;	
 	
 	ztrdbg(ZTR_DL_DEBUG, "%s : Enter\n", __FUNCTION__);
@@ -9439,7 +9439,7 @@
 	/*fill up tr98 WANATMF4LoopbackDiagnostics object*/
 	*tr98Jobj = json_object_new_object();
 	paramList = tr98Obj[handler].parameter;
-	ifValue = json_object_object_get(atmF4LoJobj, "Interface");
+	// ifValue = json_object_object_get(atmF4LoJobj, "Interface");
 
 	while(paramList->name != NULL) {
 		paramValue = json_object_object_get(atmF4LoJobj, paramList->name);
