Index: libzcfg_fe_tr98-1.0/diag.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/diag.c	2019-05-31 18:00:41.032303885 +0800
+++ libzcfg_fe_tr98-1.0/diag.c	2019-05-31 18:02:07.039295287 +0800
@@ -626,7 +626,7 @@
 	return ZCFG_SUCCESS;
 }
 
-#define DOWNLOADDIAG_SET_PARM_NUM 5
+#define DOWNLOADDIAG_SET_PARM_NUM 6
 /*
  *   TR98 Object Name : InternetGatewayDevice.DownloadDiagnostics
  *
@@ -847,7 +847,7 @@
 	return ZCFG_SUCCESS;
 }
 
-#define UPLOADDIAG_SET_PARM_NUM 6
+#define UPLOADDIAG_SET_PARM_NUM 7
 /*
  *   TR98 Object Name : InternetGatewayDevice.UploadDiagnostics
  *
Index: libzcfg_fe_tr98-1.0/diag_parameter.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/diag_parameter.h	2019-05-31 18:00:38.720303874 +0800
+++ libzcfg_fe_tr98-1.0/diag_parameter.h	2019-05-31 18:02:29.020751127 +0800
@@ -48,6 +48,7 @@
 	{ "DownloadURL", PARAMETER_ATTR_WRITE, 257, json_type_string},
 	{ "DSCP", PARAMETER_ATTR_WRITE, 0, json_type_uint32},
 	{ "EthernetPriority", PARAMETER_ATTR_WRITE, 0, json_type_uint32},
+	{ "X_ZYXEL_NumOfThreads", PARAMETER_ATTR_WRITE, 0, json_type_uint32},
 	{ "ROMTime", PARAMETER_ATTR_READONLY, 27, json_type_time},
 	{ "BOMTime", PARAMETER_ATTR_READONLY, 27, json_type_time},
 	{ "EOMTime", PARAMETER_ATTR_READONLY, 27, json_type_time},
@@ -71,6 +72,7 @@
 	{ "DSCP", PARAMETER_ATTR_WRITE, 0, json_type_uint32},
 	{ "EthernetPriority", PARAMETER_ATTR_WRITE, 0, json_type_uint32},
 	{ "TestFileLength", PARAMETER_ATTR_WRITE, 0, json_type_uint32},
+	{ "X_ZYXEL_NumOfThreads", PARAMETER_ATTR_WRITE, 0, json_type_uint32},
 	{ "ROMTime", PARAMETER_ATTR_READONLY, 27, json_type_time},
 	{ "BOMTime", PARAMETER_ATTR_READONLY, 27, json_type_time},
 	{ "EOMTime", PARAMETER_ATTR_READONLY, 27, json_type_time},
