Index: libzcfg_fe_tr98-1.0/voice.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/voice.h	2019-06-09 11:58:47.965831735 +0800
+++ libzcfg_fe_tr98-1.0/voice.h	2019-06-17 08:47:16.712935266 +0800
@@ -596,9 +596,11 @@
 	//{ "DialPlan", PARAMETER_ATTR_WRITE, 257, json_type_string},
 	//{ "DialMethod", PARAMETER_ATTR_WRITE, 257, json_type_string},
 	//{ "X_ZYXEL_Pound", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
-	{ "VoipIOPFlags", PARAMETER_ATTR_WRITE, 0, json_type_uint32},
-	{ "P_AccessNetworkInfo", PARAMETER_ATTR_WRITE, 128, json_type_string},
+	//{ "VoipIOPFlags", PARAMETER_ATTR_WRITE, 0, json_type_uint32}, //Michael.20190616.001: Remarked to disabled the parameter '~.Services.VoiceService.{i}.X_ZYXEL_Common.VoipIOPFlags' from being accessed via the ACS/TR-069.
+	{ "P_AccessNetworkInfo", PARAMETER_ATTR_WRITE, 129, json_type_string},
 	{ "UserAgent_strReplaceRule", PARAMETER_ATTR_WRITE, 257, json_type_string},
+	{ "FxsCIDMode", PARAMETER_ATTR_WRITE, 17, json_type_string}, //Michael.20190516.002: Add two parameters (1)'FxsCIDMode' and (2)'FxsCIDSigProtocol' in the TR-104 data model branch '~.Services.VoiceService.{i}.X_ZYXEL_Common.' to support the feature done by 'Michael.20190516.001' to Manually-Configure/Customize the Caller-ID (CLID) Operation Mode.
+	{ "FxsCIDSigProtocol", PARAMETER_ATTR_WRITE, 9, json_type_string}, //Michael.20190516.002: Add two parameters (1)'FxsCIDMode' and (2)'FxsCIDSigProtocol' in the TR-104 data model branch '~.Services.VoiceService.{i}.X_ZYXEL_Common.' to support the feature done by 'Michael.20190516.001' to Manually-Configure/Customize the Caller-ID (CLID) Operation Mode.
 	{ NULL, 0, 0, 0}
 };
 tr98Parameter_t para_VoicePhoneBook[] = {
