Index: libzcfg_fe_tr98-1.0/lanDev.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/lanDev.c	2019-08-29 16:39:15.616925786 +0800
+++ libzcfg_fe_tr98-1.0/lanDev.c	2019-08-29 16:54:15.056570700 +0800
@@ -7341,10 +7341,14 @@
 	if((ret = zcfgFeObjJsonGet(RDM_OID_WIFI_DIAGNOSTIC, &objIid, &jobj)) != ZCFG_SUCCESS)
 		return ret;
 
+	tr98Jobj = zcfgFeRetrieveSpv(tr98FullPathName);
+
 	if(multiJobj){
 		tmpObj = jobj;
 		jobj = NULL;
 		jobj = zcfgFeJsonMultiObjAppend(RDM_OID_WIFI_DIAGNOSTIC, &objIid, multiJobj, tmpObj);
+		zcfgFeJsonMultiObjAddParam(multiJobj, RDM_OID_WIFI_DIAGNOSTIC, &objIid, "partialset", json_object_new_boolean(true));
+		json_object_put(tmpObj);
 	}
 
 	paramList = tr98Obj[handler].parameter;
@@ -7353,11 +7357,13 @@
 		paramValue = json_object_object_get(tr98Jobj, paramList->name);
 		if(paramValue != NULL) {
 			if(!strcmp(paramList->name, "DiagnosticsState")) {
-				if(!strcmp(json_object_get_string(paramValue), "Requested")){
-					if(zcfg_msg_eidGet() == ZCFG_EID_TR69){
-						json_object_object_add(jobj, "X_ZYXEL_Creator", json_object_new_string(CREATOR_ACS));
-					}
+				if(strcmp(json_object_get_string(paramValue), "Requested"))
+					return ZCFG_INVALID_PARAM_VALUE;
+
+				if(zcfg_msg_eidGet() == ZCFG_EID_TR69){
+					json_object_object_add(jobj, "X_ZYXEL_Creator", json_object_new_string(CREATOR_ACS));
 				}
+
 			}
 
 			tr181ParamValue = json_object_object_get(jobj, paramList->name);
@@ -7374,7 +7380,7 @@
 	}
 
 	if(multiJobj){
-		json_object_put(tmpObj);
+		//json_object_put(tmpObj);
 	}
 	else{
 		if((ret = zcfgFeObjJsonSet(RDM_OID_WIFI_DIAGNOSTIC, &objIid, jobj, NULL)) != ZCFG_SUCCESS ) {
