Index: libzyutil-1.0/ECONET/libzyutil_econet_wrapper.h
===================================================================
--- libzyutil-1.0.orig/ECONET/libzyutil_econet_wrapper.h	2019-02-18 18:52:34.443841260 +0800
+++ libzyutil-1.0/ECONET/libzyutil_econet_wrapper.h	2019-02-18 18:52:34.463841260 +0800
@@ -33,4 +33,6 @@
 int endian_check();
 
 #define ECONET_BOARDFEATURES_PROC_PATH "/proc/tc3162/prtype"
+#define ECONET_BOARDFEATURES_PRNAME_PATH "/proc/tc3162/prname"
+
 #endif
Index: libzyutil-1.0/ECONET/libzyutil_wrapper.c
===================================================================
--- libzyutil-1.0.orig/ECONET/libzyutil_wrapper.c	2019-02-18 18:52:34.447841260 +0800
+++ libzyutil-1.0/ECONET/libzyutil_wrapper.c	2019-02-18 18:52:34.463841260 +0800
@@ -209,6 +209,16 @@
 	/*Read MRD info to buffer*/
 #endif
 	fread(mrd, 1, (uint32_t)MRDINFO_SIZE, fptr);
+#ifdef ZYXEL_DETECT_PRODUCT_TYPE
+	char prname[MRD_PRODUCT_NAME_LEN] = {0};
+	int prnameLen = MRD_PRODUCT_NAME_LEN;
+	if(zyUtilAskProductName(prname, prnameLen) > 0){
+		if(strcmp(mrd->PRODUCT_NAME, prname)){
+			strcpy(mrd->PRODUCT_NAME, prname);
+			zyUtilSetMrdInfo(mrd);
+		}
+	}
+#endif
 	dumpMRDinfo(mrd);
 	fclose(fptr);
 	unlink(MRDINFO_FILE);
@@ -872,6 +882,20 @@
    return n;
 }
 
+int zyUtilAskProductName(char *prname, int prnameLen)
+{
+   FILE *pf = fopen(ECONET_BOARDFEATURES_PRNAME_PATH, "r");
+
+   if(!pf || !prname){
+     return -1;
+   }
+
+   int n = fread(prname, 1, prnameLen-1, pf);
+
+   fclose(pf);
+
+   return n;
+}
 
 /*
 000  (2 port FXS) VMG8623-T50B
Index: libzyutil-1.0/zyutil.h
===================================================================
--- libzyutil-1.0.orig/zyutil.h	2019-02-18 18:52:34.447841260 +0800
+++ libzyutil-1.0/zyutil.h	2019-02-18 18:52:34.463841260 +0800
@@ -57,12 +57,14 @@
 
 #ifdef ZYXEL_DETECT_PRODUCT_TYPE
 int zyUtilAskBoardFeaturesStringBits(char *, int );
+int zyUtilAskProductName(char *prname, int prnameLen);
 int zyUtilAskBoardDslSupport(const char *);
 int zyUtilAskBoardVoIPSupport(const char *);
 #else
 #define zyUtilAskBoardFeaturesStringBits(bits, bitsLen) (-1)
+#define zyUtilAskProductName(prname, prnameLen) (-1)
 #define zyUtilAskBoardDslSupport(bits) (0)
-#define zyUtilAskBoardVoIPSupport(bits) {0}
+#define zyUtilAskBoardVoIPSupport(bits) (0)
 #endif
 
 zcfgRet_t zyUtilHandleRecvdMsgDebugCfg(const char *msg);
