Index: libzyutil-1.0/ECONET/libzyutil_wrapper.c
===================================================================
--- libzyutil-1.0.orig/ECONET/libzyutil_wrapper.c	2019-03-15 11:39:37.212818000 +0800
+++ libzyutil-1.0/ECONET/libzyutil_wrapper.c	2019-03-15 11:42:23.156818000 +0800
@@ -489,6 +489,16 @@
 	memset(bootflag, 0x0, 2);
 	fread(bootflag, 1, (uint32_t)BOOT_FLAG_SIZE, fptr);
 	*boot_flag = bootflag[0] - '0';
+
+    /*
+     * ref. bootrom/bootram/init/main.c:: readBootFlagFromFlash() 
+     * this byte may be "FF"" when NAND flash just burned
+     */
+    if (*boot_flag != 0 && *boot_flag != 1)
+    {
+        *boot_flag = 0;
+    }
+
 	//printf("%s : Current boot flag is %s(%d)\n", __FUNCTION__, bootflag, *boot_flag);
 	fclose(fptr);
 	unlink(BOOT_FLAG_FILE);
