Index: libzyutil-1.0/BROADCOM/libzyutil_brcm_wrapper.h
===================================================================
--- libzyutil-1.0.orig/BROADCOM/libzyutil_brcm_wrapper.h	2019-04-11 11:22:20.131707000 +0800
+++ libzyutil-1.0/BROADCOM/libzyutil_brcm_wrapper.h	2019-04-11 11:23:48.127707000 +0800
@@ -18,4 +18,5 @@
 int nvramDataWrite(NVRAM_DATA *nvramData);
 int nvramDataGet(NVRAM_DATA *nvramData);
 int zyFwInfoGet(ZY_FW_INFO *zyFwInfo);
-#endif
\ No newline at end of file
+int zyFwInfoNonBootGet(ZY_FW_INFO *zyFwInfo);
+#endif
Index: libzyutil-1.0/BROADCOM/libzyutil_wrapper.c
===================================================================
--- libzyutil-1.0.orig/BROADCOM/libzyutil_wrapper.c	2019-04-11 11:22:20.419707000 +0800
+++ libzyutil-1.0/BROADCOM/libzyutil_wrapper.c	2019-04-11 11:23:56.751707000 +0800
@@ -167,6 +167,42 @@
 	return 0;
 }
 
+int zyFwInfoNonBootGet(ZY_FW_INFO *zyFwInfo)
+{
+#if defined(BCM_SDK_502L05)
+	int boardFd = 0;
+	int rc = 0;
+	unsigned int offset = 0;
+	BOARD_IOCTL_PARMS ioctlParms;
+
+	boardFd = open(BOARD_DEVICE_NAME, O_RDWR);
+
+	if(boardFd != -1) {
+		ioctlParms.string = zyFwInfo;
+		ioctlParms.strLen = sizeof(ZY_FW_INFO);
+		ioctlParms.offset = offset;
+		ioctlParms.action = 0;
+		ioctlParms.buf    = NULL;
+		ioctlParms.result = -1;
+
+		rc = ioctl(boardFd, BOARD_IOCTL_GET_NONBOOTZYFWINFO, &ioctlParms);
+		close(boardFd);
+
+		if(rc < 0) {
+			printf("%s Get ZY_FW_INFO, NONBOOT Failure\n", __FUNCTION__);
+			return -1;
+		}
+	}
+	else {
+		return -1;
+	}
+
+#else //defined(BCM_SDK_502L05)
+    printf("zyFwInfoNonBootGet() only support BRCM SDK 502L05, you need to add version check in this API\n");
+#endif //defined(BCM_SDK_502L05)
+	return 0;
+}
+
 /*-------------------------------------------------
  * eraseRomdPartition() :
  * Here, the second parameter of boardIoctl()
