Index: libzyutil-1.0/zyutil_cofw.h
===================================================================
--- libzyutil-1.0.orig/zyutil_cofw.h	2019-10-14 19:17:01.468986294 +0800
+++ libzyutil-1.0/zyutil_cofw.h	2019-10-14 19:20:06.368991389 +0800
@@ -16,7 +16,9 @@
     MODEL_VMG9827_B50A,
     MODEL_EMG6726_B10A,
     MODEL_PX7501_B0,
-    MODEL_AX7501_B0
+    MODEL_AX7501_B0,
+    MODEL_VMG8825_B50B,
+    MODEL_AM3100_B0
 } model_name_t;
 
 /*!
Index: libzyutil-1.0/zyutil_cofw.c
===================================================================
--- libzyutil-1.0.orig/zyutil_cofw.c	2019-10-14 19:17:01.468986294 +0800
+++ libzyutil-1.0/zyutil_cofw.c	2019-10-15 15:25:44.270984503 +0800
@@ -44,6 +44,10 @@
 			return MODEL_PX7501_B0;
 		}else if (strstr(modelName,"AX7501-B0") != NULL){
 			return MODEL_AX7501_B0;
+		}else if (strstr(modelName,"VMG8825-B50B") != NULL){
+			return MODEL_VMG8825_B50B;
+		}else if (strstr(modelName,"AM3100-B0") != NULL){
+			return MODEL_AM3100_B0;
 		}
 	}
 
