Index: libzyutil-1.0/common.c
===================================================================
--- libzyutil-1.0.orig/common.c	2019-12-09 11:15:12.843620737 +0800
+++ libzyutil-1.0/common.c	2019-12-09 11:16:53.759748752 +0800
@@ -1394,7 +1394,10 @@
 
     memset(accountInfo, 0, sizeof(accountInfo));
     while(fgets(accountInfo, sizeof(accountInfo), fp) != NULL){
-        if(strncmp(username, accountInfo, strlen(username)) == 0){
+        if((strncmp(username, accountInfo, strlen(username)) == 0) &&
+           (accountInfo[strlen(username)] == 0x3A ) // 0x3A = ":"
+          )
+        {
             fclose(fp);
             return accountInfo;
         }
