Index: linux-3.18.21/kernel_ext/drivers/mtd/chips/spi_nand_flash.c
===================================================================
--- linux-3.18.21.orig/kernel_ext/drivers/mtd/chips/spi_nand_flash.c	2018-05-21 20:36:31.966096682 +0800
+++ linux-3.18.21/kernel_ext/drivers/mtd/chips/spi_nand_flash.c	2018-05-21 20:57:51.052439329 +0800
@@ -3227,6 +3227,37 @@
 		return -1;
 	}
 }
+
+/*----------------------------------------------------------------------------------
+ *	en7512_nand_exec_read_page_zyxel()
+ *  For ZyXEL atcommand "ATDS",dump data of spare area in page Y of block X.
+ *	INPUT : bmt_block, bmt_page, len
+ *	OUTPUT:  *oob_buf
+ *	RETURN: 0-SUCCESS, 1-ERROR
+ *----------------------------------------------------------------------------------*/
+int en7512_nand_exec_read_page_zyxel(u32 bmt_block, u32 bmt_page, u8 *oob_buf, u32 len)
+{
+	u32 page_number;
+	struct SPI_NAND_FLASH_INFO_T   *ptr_dev_info_t;
+	ptr_dev_info_t  = _SPI_NAND_GET_DEVICE_INFO_PTR;
+	u32 pages_per_blk = (ptr_dev_info_t->erase_size / ptr_dev_info_t->page_size);
+	page_number = (bmt_block*pages_per_blk)+bmt_page;
+
+	//_SPI_NAND_PRINTF( "[DEBUG] en7512_nand_exec_read_page_zyxel------------------>>>>\n");
+	//_SPI_NAND_PRINTF( "read page %d(block %d and page %d) with len %d \n\n",page_number,bmt_block,bmt_page, len);
+	//_SPI_NAND_PRINTF( "oob size is %d\n", ptr_dev_info_t->oob_size);
+
+	if( spi_nand_read_page(page_number, SPI_NAND_FLASH_READ_SPEED_MODE_SINGLE) == SPI_NAND_FLASH_RTN_NO_ERROR)
+ 	{
+		//memcpy( oob_buf,  &_current_cache_page_oob_mapping[0], len );
+		memcpy( oob_buf,  &_current_cache_page_oob[0], len );
+ 		return 0;
+ 	}else{
+ 		_SPI_NAND_PRINTF( "en7512_nand_exec_read_page_zyxel: read error!!!!!!\n");
+ 		return -1;
+	}
+} /* en7512_nand_exec_read_page_zyxel */
+
 int en7512_nand_check_block_bad(u32 offset, u32 bmt_block)
 {
 	u32								page_number;
