Index: linux-2.6.36/arch/mips/pci/pci-tc3162u.c
===================================================================
--- linux-2.6.36.orig/arch/mips/pci/pci-tc3162u.c	2017-04-24 20:50:59.057259146 +0800
+++ linux-2.6.36/arch/mips/pci/pci-tc3162u.c	2017-04-26 01:05:08.959321782 +0800
@@ -2877,7 +2877,11 @@
 			{
 				//set GPIO share scheme reg for PERST output
 				tmp = regRead32(0xbfb00860);
+				#if defined(ZYXEL_DISABLE_SECOND_PCI)
+				regWrite32(0xbfb00860, (tmp | (1<<19)));
+				#else
 				regWrite32(0xbfb00860, (tmp | ((1<<19) | (1<<20))));
+				#endif
 			}
 #endif
 			/*first reset to default*/
