Index: linux-2.6.36/arch/mips/pci/pci-tc3162u.c
===================================================================
--- linux-2.6.36.orig/arch/mips/pci/pci-tc3162u.c	2017-04-26 01:15:52.981914782 +0800
+++ linux-2.6.36/arch/mips/pci/pci-tc3162u.c	2017-04-26 01:22:01.725914782 +0800
@@ -2835,6 +2835,10 @@
 			printk("PCI-E RC0 & RC1 can not link up\n");
 			return -1;
 		}
+		#if defined(ZYXEL_PCIE_RESET1_PIN_AS_GPIO_31)
+		tmp = regRead32(0xbfa20104);
+		regWrite32(0xbfa20104, (tmp | (1<<21)));
+		#endif
 		mt7512_pcie_set_mac();
 
 	}else	if(isRT63365 || isMT751020 || isMT7505){
