Index: linux-3.18.21/include/linux/foe_hook.h
===================================================================
--- linux-3.18.21.orig/include/linux/foe_hook.h	2018-05-30 15:12:13.000000000 +0800
+++ linux-3.18.21/include/linux/foe_hook.h	2018-05-30 16:20:48.421922332 +0800
@@ -225,7 +225,7 @@
 extern int (*ra_sw_nat_hook_xfer) (struct sk_buff *skb, const struct sk_buff *prev_p);
 extern int (*ra_sw_nat_hook_foeentry) (void * inputvalue,int operation);
 extern int (*ra_sw_nat_hook_is_alive_pkt)(unsigned int crsn);
-#if defined(TCSUPPORT_CPU_MT7520) || defined(TCSUPPORT_CPU_EN7512)
+#if 1 //ZYXEL for port map defined(TCSUPPORT_CPU_MT7520) || defined(TCSUPPORT_CPU_EN7512)
 extern int (*MT7530LanPortMap2Switch_hook)(int port); 
 #endif
 
Index: linux-3.18.21/net/bridge/ecnt_br.c
===================================================================
--- linux-3.18.21.orig/net/bridge/ecnt_br.c	2018-05-30 16:19:15.057459362 +0800
+++ linux-3.18.21/net/bridge/ecnt_br.c	2018-05-30 16:19:34.941557962 +0800
@@ -580,7 +580,7 @@
         /*lan port*/
         if(port >= 0 && port <= 5)
         {
-#if defined(TCSUPPORT_CPU_MT7520) || defined(TCSUPPORT_CPU_EN7512)
+#if 1 //ZYXEL for port map defined(TCSUPPORT_CPU_MT7520) || defined(TCSUPPORT_CPU_EN7512)
             if(MT7530LanPortMap2Switch_hook)
             {
                 switch_port = MT7530LanPortMap2Switch_hook(port);
Index: linux-3.18.21/net/bridge/ecnt_br.h
===================================================================
--- linux-3.18.21.orig/net/bridge/ecnt_br.h	2018-05-30 15:12:13.697518515 +0800
+++ linux-3.18.21/net/bridge/ecnt_br.h	2018-05-30 16:15:00.624197696 +0800
@@ -749,7 +749,7 @@
                 /*lan port*/
                 if(port >= 0 && port <= 5)
                 {
-#if defined(TCSUPPORT_CPU_MT7520) || defined(TCSUPPORT_CPU_EN7512)
+#if 1 //ZYXEL for port map defined(TCSUPPORT_CPU_MT7520) || defined(TCSUPPORT_CPU_EN7512)
                     if(MT7530LanPortMap2Switch_hook)
                     {
                         switch_port = MT7530LanPortMap2Switch_hook(port);
