Index: linux-3.18.21/kernel/workqueue.c
===================================================================
--- linux-3.18.21.orig/kernel/workqueue.c	2015-09-01 06:19:23.000000000 +0800
+++ linux-3.18.21/kernel/workqueue.c	2018-05-31 19:23:28.674739396 +0800
@@ -1378,6 +1378,7 @@
 
 	spin_unlock(&pwq->pool->lock);
 }
+EXPORT_SYMBOL(queue_work);
 
 /**
  * queue_work_on - queue work on specific cpu
@@ -2576,7 +2577,7 @@
 out_unlock:
 	mutex_unlock(&wq->mutex);
 }
-EXPORT_SYMBOL_GPL(flush_workqueue);
+EXPORT_SYMBOL(flush_workqueue);
 
 /**
  * drain_workqueue - drain a workqueue
@@ -4140,7 +4141,7 @@
 	destroy_workqueue(wq);
 	return NULL;
 }
-EXPORT_SYMBOL_GPL(__alloc_workqueue_key);
+EXPORT_SYMBOL(__alloc_workqueue_key);
 
 /**
  * destroy_workqueue - safely terminate a workqueue
@@ -4221,7 +4222,7 @@
 		put_pwq_unlocked(pwq);
 	}
 }
-EXPORT_SYMBOL_GPL(destroy_workqueue);
+EXPORT_SYMBOL(destroy_workqueue);
 
 /**
  * workqueue_set_max_active - adjust max_active of a workqueue
