Index: linux-3.18.21/arch/mips/include/asm/tc3162/ledcetrl.h
===================================================================
--- linux-3.18.21.orig/arch/mips/include/asm/tc3162/ledcetrl.h	2018-06-11 10:23:47.459507000 +0800
+++ linux-3.18.21/arch/mips/include/asm/tc3162/ledcetrl.h	2018-06-11 10:25:00.643507000 +0800
@@ -257,9 +257,13 @@
 #define LED_VOIP_REG_STATUS	(LED_BASE + 101)
 #define LED_PHY_VCC_DISABLE		(LED_BASE + 102)
 
+#if defined(ZYXEL_CUSTOMIZE_LED_BEHAVIOR) && defined(TCSUPPORT_CPU_EN7516)
+#define LED_GE_ETH_WAN_STATUS		(LED_BASE + 103)
+#endif
+
 #if defined(TCSUPPORT_CPU_EN7516)||defined(TCSUPPORT_CPU_EN7527)
-#define GPIO_WLAN_WPS_5G	(LED_BASE + 104)
-#define LED_WLAN_RADIO_5G	(LED_BASE + 105) 
+#define GPIO_WLAN_WPS_5G			(LED_BASE + 104)
+#define LED_WLAN_RADIO_5G			(LED_BASE + 105) 
 #endif
 
 #define mtkled_read_reg_word(reg) 			regRead32(reg)
