Index: linux-3.18.21/net/core/dev.c
===================================================================
--- linux-3.18.21.orig/net/core/dev.c	2018-07-17 18:12:31.498093000 +0800
+++ linux-3.18.21/net/core/dev.c	2018-07-17 18:40:42.702093000 +0800
@@ -3711,9 +3711,29 @@
 
 	if (skb->protocol == cpu_to_be16(ETH_P_8021Q) ||
 	    skb->protocol == cpu_to_be16(ETH_P_8021AD)) {
+#if defined(ZYXEL_KF_TMS)
+        /************************************
+            TMS Module 6.86 will handle VLAN manipulation,
+            if remvoe VLAN tag, will cause TMS module receive 1ag packet fail=>1ag loopback fail,
+            so bybass skb_vlan_untag() setting for 1ag packet.
+
+            Trevor, 20180717
+         ***********************************/
+        struct vlan_hdr *vh = (struct vlan_hdr *)skb->data;
+        if (vh->h_vlan_encapsulated_proto != htons(ETH_P_8021AG))
+        {
+            skb = skb_vlan_untag(skb);
+            if (unlikely(!skb))
+            {
+                goto unlock;
+            }
+        }
+#else /*defined(ZYXEL_KF_TMS)*/
+        /* original code */
 		skb = skb_vlan_untag(skb);
 		if (unlikely(!skb))
 			goto unlock;
+#endif /*defined(ZYXEL_KF_TMS)*/
 	}
 
 #ifdef CONFIG_NET_CLS_ACT
