Index: linux-3.18.21/arch/mips/econet/prom.c
===================================================================
--- linux-3.18.21.orig/arch/mips/econet/prom.c	2018-06-15 10:26:37.387951259 +0800
+++ linux-3.18.21/arch/mips/econet/prom.c	2018-06-15 13:44:31.707555655 +0800
@@ -529,6 +529,7 @@
 }
 #endif
 #if defined(TCSUPPORT_DYING_GASP) && (defined(CONFIG_MIPS_RT65168) || defined(CONFIG_MIPS_RT63365))
+extern void (*cpu_wait)(void);
 __IMEM
 void dying_gasp_setup_mem_cpu(void){
 #ifdef CONFIG_MIPS_RT65168	
