Index: linux-3.18.21/net/bridge/ecnt_br.c
===================================================================
--- linux-3.18.21.orig/net/bridge/ecnt_br.c	2018-07-04 10:48:56.073684203 +0800
+++ linux-3.18.21/net/bridge/ecnt_br.c	2018-07-12 21:52:15.677353860 +0800
@@ -187,7 +187,7 @@
 				continue;
 			}
 			bpg = f->ports;
-			while(bpg){
+			while(bpg&&(num<maxnum)){
 				if (has_expired(br, bpg)){
 					bpg = bpg->next;
 					continue;
