Index: linux-3.18.21/include/uapi/linux/ipv6.h
===================================================================
--- linux-3.18.21.orig/include/uapi/linux/ipv6.h	2015-09-01 06:19:23.000000000 +0800
+++ linux-3.18.21/include/uapi/linux/ipv6.h	2018-07-12 21:55:05.669353860 +0800
@@ -164,6 +164,13 @@
 	DEVCONF_MLDV2_UNSOLICITED_REPORT_INTERVAL,
 	DEVCONF_SUPPRESS_FRAG_NDISC,
 	DEVCONF_ACCEPT_RA_FROM_LOCAL,
+/* add by ecnt */
+	DEVCONF_SLAAC_ADDR,
+	DEVCONF_CHILD_PREFIX_ORIGN,
+	DEVCONF_CHILD_PREFIX,
+	DEVCONF_PARENT_PD_PREFIX,
+	DEVCONF_SLAAC_PREFIX,
+/* add by ecnt */
 	DEVCONF_MAX
 };
 
