Index: linux-3.18.21/arch/mips/econet/tcwdog.c
===================================================================
--- linux-3.18.21.orig/arch/mips/econet/tcwdog.c	2018-08-06 14:22:10.180619000 +0800
+++ linux-3.18.21/arch/mips/econet/tcwdog.c	2018-08-06 14:35:29.812655252 +0800
@@ -160,8 +160,10 @@
 }
 EXPORT_SYMBOL(attack_protect_set);
 
+int iswatchDogReset = 0;
 void watchDogReset(void)
 {
+	iswatchDogReset = 1;
 #ifdef CONFIG_TC3162_ADSL
     /* stop adsl */
 	if (adsl_dev_ops)
@@ -204,13 +206,16 @@
 	VPint(CR_DRAMC_CONF) &= ~(0x1<<2);
 #endif
 
-	timerSet(5, TIMERTICKS_1MS, ENABLE, TIMER_TOGGLEMODE, TIMER_HALTDISABLE);				
 	while(1);
 //	_machine_restart();
 }
 
 void tc3162wdog_kick(void)
 {
+
+	if(iswatchDogReset){
+		return;
+	}
       /* Because kernel 3.18 don't support the type before 63365*/
 	regWrite32(CR_WDOG_RLD, 0x1); 
 }
