Index: linux-3.18.21/net/bridge/br_fdb.c
===================================================================
--- linux-3.18.21.orig/net/bridge/br_fdb.c	2018-10-15 15:36:45.482812159 +0800
+++ linux-3.18.21/net/bridge/br_fdb.c	2018-10-18 13:54:24.754469539 +0800
@@ -558,7 +558,7 @@
 		   const unsigned char *addr, u16 vid, bool added_by_user)
 {
 	struct hlist_head *head = &br->hash[br_mac_hash(addr, vid)];
-	struct net_bridge_fdb_entry *fdb;
+	struct net_bridge_fdb_entry *fdb = NULL;
 	bool fdb_modified = false;
 	struct net_data_s net_data;
 
@@ -570,13 +570,32 @@
 	if (!(source->state == BR_STATE_LEARNING ||
 	      source->state == BR_STATE_FORWARDING))
 		return;
-
+#if 1 /* <JoannaSu>, <20181015>, <Access invalid address>*/
+    fdb = NULL;
+#endif
 	fdb = fdb_find_rcu(head, addr, vid);
 	/* if packets come frome WAN with LAN device MAC, don't update brctl MAC table */
-	if(fdb){
+	if(likely(fdb)){
+#if 1 /* <JoannaSu>, <20181015>, <Access invalid address>*/
+        if(fdb->dst != NULL)
+        {
+            if(fdb->dst->dev != NULL)
+            {
+                if(fdb->dst->dev->name != NULL)
+                {
+                     if((!strncmp(fdb->dst->dev->name, "eth", 3) || !strncmp(fdb->dst->dev->name, "ra", 2)) && (!strncmp(source->dev->name, "nas", 3)))
+                     {
+                         return;
+                     }
+                }
+            }
+        }
+#else
 		if((!strncmp(fdb->dst->dev->name, "eth", 3) || !strncmp(fdb->dst->dev->name, "ra", 2)) && (!strncmp(source->dev->name, "nas", 3)))
 			return;
+#endif
 	}
+
 	if (likely(fdb)) {
 		/* attempt to update an entry for a local interface */
 		if (unlikely(fdb->is_local)) {
@@ -596,6 +615,7 @@
 				fdb->dst = source;
 				fdb_modified = true;
 			}
+
 			fdb->updated = jiffies;
 			if (unlikely(added_by_user))
 				fdb->added_by_user = 1;
@@ -605,6 +625,9 @@
 	} else {
 		spin_lock(&br->hash_lock);
 		if (likely(!fdb_find(head, addr, vid))) {
+#if 1 /* <JoannaSu>, <20181015>, <Access invalid address>*/
+			fdb = NULL;
+#endif
 			fdb = fdb_create(head, source, addr, vid);
 			if (fdb) {
 				if (unlikely(added_by_user))
