Index: linux-3.18.21/net/core/ecnt_net_core.h
===================================================================
--- linux-3.18.21.orig/net/core/ecnt_net_core.h	2018-10-24 14:09:24.230412851 +0800
+++ linux-3.18.21/net/core/ecnt_net_core.h	2018-10-24 15:45:16.296549090 +0800
@@ -369,9 +369,17 @@
 		ret = smux_pkt_recv_hook(skb, skb->dev, orig_dev);		  
 #if ((defined(TCSUPPORT_WAN_ETHER) || defined(TCSUPPORT_WAN_PTM)) && defined(TCSUPPORT_MULTISERVICE_ON_WAN) ) || defined(TCSUPPORT_WAN_GPON) || defined(TCSUPPORT_WAN_EPON)
 #ifdef TCSUPPORT_VLAN_TAG
+/*+++ Kide@zyxel:
+ * The function 'smux_pkt_recv_hook' will handle the skb and then invoke 'netif_receive_skb' for further processing.
+ * It means the skb could be forwarded or dropped/freed in it. In case, the skb was dropped in it and then set VLAN_TAG_FROM_INDEV
+ * here. The skb was polluted as allocated by others.
+ */
+#if 0
 		  if (skb) {
 			skb->vlan_tag_flag |= VLAN_TAG_FROM_INDEV;
 		  }
+#endif
+/*---*/
 #endif			  
 #endif	
 		return ECNT_RETURN;
