Index: linux-3.18.21/ipc/mqueue.c
===================================================================
--- linux-3.18.21.orig/ipc/mqueue.c	2015-09-01 06:19:23.000000000 +0800
+++ linux-3.18.21/ipc/mqueue.c	2019-03-18 13:46:42.062643450 +0800
@@ -1239,8 +1239,10 @@
 
 			timeo = MAX_SCHEDULE_TIMEOUT;
 			ret = netlink_attachskb(sock, nc, &timeo, NULL);
-			if (ret == 1)
+			if (ret == 1) {
+				sock = NULL;
 				goto retry;
+			}
 			if (ret) {
 				sock = NULL;
 				nc = NULL;
