Index: zoneinfo-2018d/Makefile
===================================================================
--- zoneinfo-2018d.orig/Makefile	2018-03-21 07:10:33.000000000 +0800
+++ zoneinfo-2018d/Makefile	2018-12-28 09:23:46.372667412 +0800
@@ -537,7 +537,9 @@
 		cp -f $(TABDATA) '$(DESTDIR)$(TZDIR)/.'
 		cp tzselect '$(DESTDIR)$(BINDIR)/.'
 		cp zdump '$(DESTDIR)$(ZDUMPDIR)/.'
+		cp zdump_arm '$(DESTDIR)$(ZDUMPDIR)/.'
 		cp zic '$(DESTDIR)$(ZICDIR)/.'
+		cp zic_arm '$(DESTDIR)$(ZICDIR)/.'
 		cp libtz.a '$(DESTDIR)$(LIBDIR)/.'
 		$(RANLIB) '$(DESTDIR)$(LIBDIR)/libtz.a'
 		cp -f newctime.3 newtzset.3 '$(DESTDIR)$(MANDIR)/man3/.'
@@ -578,9 +580,16 @@
 
 zdump:		$(TZDOBJS)
 		$(CC) -o $@ $(CFLAGS) $(LDFLAGS) $(TZDOBJS) $(LDLIBS)
+		$(CC2) $(CPPFLAGS) -c -o zdump_arm.o zdump.c
+		$(CC2) $(CPPFLAGS) -c -o localtime1.o localtime.c
+		$(CC2) $(CPPFLAGS) -c -o asctime1.o asctime.c
+		$(CC2) $(CPPFLAGS) -c -o strftime1.o strftime.c
+		$(CC2) -o zdump_arm $(CFLAGS) $(LDFLAGS) zdump_arm.o localtime1.o asctime1.o strftime1.o $(LDLIBS)
 
 zic:		$(TZCOBJS)
 		$(CC) -o $@ $(CFLAGS) $(LDFLAGS) $(TZCOBJS) $(LDLIBS)
+		$(CC2) $(CPPFLAGS) -c -o zic_arm.o zic.c
+		$(CC2) -o zic_arm $(CFLAGS) $(LDFLAGS) zic_arm.o $(LDLIBS)
 
 yearistype:	yearistype.sh
 		cp yearistype.sh yearistype
@@ -958,12 +967,17 @@
 		@$(AWK) '/^Z/ { print $$2 } /^L/ { print $$3 }' tzdata.zi
 
 asctime.o:	private.h tzfile.h
+asctime1.o:	private.h tzfile.h
 date.o:		private.h
 difftime.o:	private.h
 localtime.o:	private.h tzfile.h
+localtime1.o:	private.h tzfile.h
 strftime.o:	private.h tzfile.h
+strftime1.o:	private.h tzfile.h
 zdump.o:	version.h
+zdump_arm.o:	version.h
 zic.o:		private.h tzfile.h version.h
+zic_arm.o:		private.h tzfile.h version.h
 
 .KEEP_STATE:
 
