Index: busybox-1_20_1/loginutils/login.c
===================================================================
--- busybox-1_20_1.orig/loginutils/login.c	2019-06-20 15:31:03.466418329 +0800
+++ busybox-1_20_1/loginutils/login.c	2019-06-20 15:38:37.015246800 +0800
@@ -814,7 +814,9 @@
 							syslog(LOG_INFO, "Account privilege:  %s Priviege = %s !!!!", username, userPriviege);
 						}
 #endif
+#ifndef ZYXEL_REMOTE_ACCOUNT_ACCESS_PRIVILEGE
 						free(accountObj);
+#endif
 						break;
 					}
 					free(accountObj);
@@ -827,13 +829,20 @@
 				logGpObjIid.idx[GROUP_IID] = logAccountObjIid.idx[GROUP_IID];
 				logGpObjIid.idx[ACCOUNT_IID] = 0;
 				if(zcfgFeObjStructGet(RDM_OID_ZY_LOG_CFG_GP, &logGpObjIid, (void **) &logGpObj) == ZCFG_SUCCESS) {
-					if (strstr(logGpObj->GP_Privilege, "telnet") == NULL){
+						if (strstr(logGpObj->GP_Privilege, "telnet") == NULL
+#ifdef ZYXEL_REMOTE_ACCOUNT_ACCESS_PRIVILEGE
+					    || strstr(accountObj->Privilege, "telnet") == NULL
+#endif
+					    ){
 						snprintf(logStr, sizeof(logStr), "Account: '%s' TELNET permission denied.", username);
 						puts(logStr);
 						syslog(LOG_INFO, "Account:'%s' TELNET permission denied.", username);
 						free(logGpObj);
 						return EXIT_FAILURE;
 					}
+#ifdef ZYXEL_REMOTE_ACCOUNT_ACCESS_PRIVILEGE
+					free(accountObj);
+#endif
 					free(logGpObj);
 				}
 #ifdef ZYXEL_REMOMGMT_ONLY_HTTPS_ALLOWS_FOR_WLAN_CONNECT
