Index: busybox-1_20_1/networking/wget.c
===================================================================
--- busybox-1_20_1.orig/networking/wget.c	2019-10-16 14:18:00.909317401 +0800
+++ busybox-1_20_1/networking/wget.c	2019-10-16 14:21:36.517606895 +0800
@@ -42,7 +42,7 @@
 #include "zcfg_msg.h"
 #endif
 
-#if defined(ZYXEL_DIAGNOSTIC_SPEEDTEST) && defined(PLATFORM_ECONET) && defined(TC_SDK_7_3_207_4)
+#if defined(ZYXEL_DIAGNOSTIC_SPEEDTEST) && defined(PLATFORM_ECONET)
 #define TCP_WINDOW_SIZE_VARIABLE "/tmp/download_variables"
 #define DEFAULT_TCP_SCALE 7
 #define DEFAULT_TCP_WINDOW 50000
@@ -93,7 +93,7 @@
 	char *ifName;
 	char *dscp;
 	char *ethernetPriority;
-#if defined(ZYXEL_DIAGNOSTIC_SPEEDTEST) && defined(PLATFORM_ECONET) && defined(TC_SDK_7_3_207_4)
+#if defined(ZYXEL_DIAGNOSTIC_SPEEDTEST) && defined(PLATFORM_ECONET)
 	char *speedTestEnable;
 #endif //PLATFORM_ECONET
 #endif
@@ -165,7 +165,7 @@
 	WGET_OPT_INTERFACE  = (1 << 8),
 	WGET_OPT_DSCP       = (1 << 9),
 	WGET_OPT_ETNERPRIORITY = (1 << 10),
-#if defined(ZYXEL_DIAGNOSTIC_SPEEDTEST) && defined(PLATFORM_ECONET) && defined(TC_SDK_7_3_207_4)
+#if defined(ZYXEL_DIAGNOSTIC_SPEEDTEST) && defined(PLATFORM_ECONET)
 	WGET_OPT_SPEEDTEST  = (1 << 11),
 	WGET_OPT_RETRIES    = (1 << 12),
 	WGET_OPT_PASSIVE    = (1 << 13),
@@ -269,7 +269,7 @@
 
 	json_object_put(diag_result);
 	set_priorityMark(0);
-#if defined(ZYXEL_DIAGNOSTIC_SPEEDTEST) && defined(PLATFORM_ECONET) && defined(TC_SDK_7_3_207_4)
+#if defined(ZYXEL_DIAGNOSTIC_SPEEDTEST) && defined(PLATFORM_ECONET)
 	set_qdmaRateLimit(1);
 #endif
 
@@ -327,7 +327,7 @@
 		system(cmd);
 	}
 }
-#if defined(ZYXEL_DIAGNOSTIC_SPEEDTEST) && defined(PLATFORM_ECONET) && defined(TC_SDK_7_3_207_4)
+#if defined(ZYXEL_DIAGNOSTIC_SPEEDTEST) && defined(PLATFORM_ECONET)
 void set_qdmaRateLimit(int act)
 {
 	char cmd[128];
@@ -387,7 +387,7 @@
 #endif
 }
 
-#if defined(ZYXEL_DIAGNOSTIC_SPEEDTEST) && defined(PLATFORM_ECONET) && defined(TC_SDK_7_3_207_4)
+#if defined(ZYXEL_DIAGNOSTIC_SPEEDTEST) && defined(PLATFORM_ECONET)
 /* New EOMTime = BOMTime + speedtest diff time */
 void calculate_EOMTime_speedtest(char *buf, unsigned long diff_stime, unsigned long diff_ustime)
 {
@@ -480,7 +480,7 @@
                 return 0;
 }
 
-#if defined(ZYXEL_DIAGNOSTIC_SPEEDTEST) && defined(PLATFORM_ECONET) && defined(TC_SDK_7_3_207_4)
+#if defined(ZYXEL_DIAGNOSTIC_SPEEDTEST) && defined(PLATFORM_ECONET)
 void execute_speedtest(char *s, char *host_info, char *target_path, int *rev_bytes, int *send_bytes)
 {
 	FILE *fp, *fp2;
@@ -863,7 +863,7 @@
 		return;
 	}
 
-#if defined(ZYXEL_DIAGNOSTIC_SPEEDTEST) && defined(PLATFORM_ECONET) && defined(TC_SDK_7_3_207_4)
+#if defined(ZYXEL_DIAGNOSTIC_SPEEDTEST) && defined(PLATFORM_ECONET)
 	is_ipv6 = true;
 #endif
 
@@ -1427,7 +1427,7 @@
 	target.allocated = NULL;
 	server.user = NULL;
 	target.user = NULL;
-#if defined(ZYXEL_DIAGNOSTIC_SPEEDTEST) && defined(PLATFORM_ECONET) && defined(TC_SDK_7_3_207_4)
+#if defined(ZYXEL_DIAGNOSTIC_SPEEDTEST) && defined(PLATFORM_ECONET)
 	int rev_bytes = 0, send_bytes = 0;
 #endif
 
@@ -1508,7 +1508,7 @@
 #if defined(ZCFG_SUPPORT) //TR143
     strcpy(dstIP, inet_ntoa(lsa->u.sin.sin_addr));
 	set_priorityMark(1);
-#if defined(ZYXEL_DIAGNOSTIC_SPEEDTEST) && defined(PLATFORM_ECONET) && defined(TC_SDK_7_3_207_4)
+#if defined(ZYXEL_DIAGNOSTIC_SPEEDTEST) && defined(PLATFORM_ECONET)
 	set_qdmaRateLimit(0);
 #endif
 #endif
@@ -1543,7 +1543,7 @@
 			outputDiagDlData();
         }
 		getDateTime(TCPOpenResponseTime);
-#if defined(ZYXEL_DIAGNOSTIC_SPEEDTEST) && defined(PLATFORM_ECONET) && defined(TC_SDK_7_3_207_4)
+#if defined(ZYXEL_DIAGNOSTIC_SPEEDTEST) && defined(PLATFORM_ECONET)
 		if(!strcmp(G.speedTestEnable, "1") && !is_ipv6)
 		{
 			goto speed_test;
@@ -1830,7 +1830,7 @@
 	fclose(sfp);
 
 #if defined(ZCFG_SUPPORT) //TR143:DownloadDiagnstic
-#if defined(ZYXEL_DIAGNOSTIC_SPEEDTEST) && defined(PLATFORM_ECONET) && defined(TC_SDK_7_3_207_4)
+#if defined(ZYXEL_DIAGNOSTIC_SPEEDTEST) && defined(PLATFORM_ECONET)
 speed_test:
 	if(!strcmp(G.speedTestEnable, "1") && !is_ipv6)
 	{
@@ -1877,7 +1877,7 @@
 #ifdef ZYXEL_DL_DIAG_CALCULATION
    	EOMTimeMs = getDateTime(EOMTime);
 #else
-#if defined(ZYXEL_DIAGNOSTIC_SPEEDTEST) && defined(PLATFORM_ECONET) && defined(TC_SDK_7_3_207_4)
+#if defined(ZYXEL_DIAGNOSTIC_SPEEDTEST) && defined(PLATFORM_ECONET)
 	if(!strcmp(G.speedTestEnable, "0") && !is_ipv6)
 	{
 		//Get user-space time
@@ -1926,7 +1926,7 @@
         "interfaceName"      Required_argument "i"
         "DSCP"               Required_argument "d"
         "ethernetPriority"   Required_argument "M"
-#if defined(ZYXEL_DIAGNOSTIC_SPEEDTEST) && defined(PLATFORM_ECONET) && defined(TC_SDK_7_3_207_4)
+#if defined(ZYXEL_DIAGNOSTIC_SPEEDTEST) && defined(PLATFORM_ECONET)
         "speedTestEnable"    Required_argument "E"
 #endif //PLATFORM_ECONET
 #endif
@@ -1952,7 +1952,7 @@
 #endif
 	opt_complementary = "-1" IF_FEATURE_WGET_TIMEOUT(":T+") IF_FEATURE_WGET_LONG_OPTIONS(":\xfe::");
 #if defined(ZCFG_SUPPORT) //TR143
-#if defined(ZYXEL_DIAGNOSTIC_SPEEDTEST) && defined(PLATFORM_ECONET) && defined(TC_SDK_7_3_207_4)
+#if defined(ZYXEL_DIAGNOSTIC_SPEEDTEST) && defined(PLATFORM_ECONET)
 	getopt32(argv, "csqO:P:Y:U:T:i:wd:M:E:R:" /*ignored:*/ "t:",
 #else
 	getopt32(argv, "csqO:P:Y:U:T:i:wd:M:R:" /*ignored:*/ "t:",
@@ -1961,7 +1961,7 @@
 		&G.proxy_flag, &G.user_agent,
 		IF_FEATURE_WGET_TIMEOUT(&G.timeout_seconds) IF_NOT_FEATURE_WGET_TIMEOUT(NULL),
 		&G.ifName, &G.dscp, &G.ethernetPriority,
-#if defined(ZYXEL_DIAGNOSTIC_SPEEDTEST) && defined(PLATFORM_ECONET) && defined(TC_SDK_7_3_207_4)
+#if defined(ZYXEL_DIAGNOSTIC_SPEEDTEST) && defined(PLATFORM_ECONET)
 		&G.speedTestEnable,
 #endif //PLATFORM_ECONET
 		NULL /* -t RETRIES */
@@ -1971,7 +1971,7 @@
 
 	ipDscp = (G.dscp) ? xatou_range(G.dscp, 0, 255) : 0;
 	printf("Dscp: %u\n", ipDscp);
-#if defined(ZYXEL_DIAGNOSTIC_SPEEDTEST) && defined(PLATFORM_ECONET) && defined(TC_SDK_7_3_207_4)
+#if defined(ZYXEL_DIAGNOSTIC_SPEEDTEST) && defined(PLATFORM_ECONET)
 	speedTestEnable = (G.speedTestEnable) ? xatou_range(G.speedTestEnable, 0, 255) : 0;
 	printf("speedTestEnable: %u\n", speedTestEnable);
 #endif //PLATFORM_ECONET
