Index: pure-ftpd-1.0.30/src/ftpd.c
===================================================================
--- pure-ftpd-1.0.30.orig/src/ftpd.c	2019-06-20 15:53:36.950379282 +0800
+++ pure-ftpd-1.0.30/src/ftpd.c	2019-06-20 16:05:14.205954662 +0800
@@ -2015,8 +2015,10 @@
 				strcpy(userPriviege, accountObj->RemoteAccessPrivilege);
 				logfile(LOG_WARNING, "Account privilege:  %s Priviege = %s !!!!", account, userPriviege);
 			}
-#endif			
+#endif		
+#ifndef ZYXEL_REMOTE_ACCOUNT_ACCESS_PRIVILEGE
 			free(accountObj);
+#endif
 			break;
 		}
 		free(accountObj);
@@ -2031,13 +2033,20 @@
 		logGpObjIid.idx[GROUP_IID] = logAccountObjIid.idx[GROUP_IID];
 		logGpObjIid.idx[ACCOUNT_IID] = 0;
 		if (zcfgFeObjStructGet(RDM_OID_ZY_LOG_CFG_GP, &logGpObjIid, (void **) &logGpObj) == ZCFG_SUCCESS) {
-			if (strstr(logGpObj->GP_Privilege,"ftp")== NULL){
+			if (strstr(logGpObj->GP_Privilege,"ftp")== NULL
+#ifdef ZYXEL_REMOTE_ACCOUNT_ACCESS_PRIVILEGE				
+			|| strstr(accountObj->Privilege,"ftp")== NULL
+#endif				
+			){
 				snprintf(logStr, sizeof(logStr), "Account:'%s' FTP permission denied.", account);
 				puts(logStr);
 				logfile(LOG_WARNING, "Account:'%s' FTP permission denied.", account);
 				free(logGpObj);
 				return;
 			}
+#ifdef ZYXEL_REMOTE_ACCOUNT_ACCESS_PRIVILEGE
+			free(accountObj);
+#endif
 			free(logGpObj);
 		}
 #ifdef ZYXEL_REMOMGMT_ONLY_HTTPS_ALLOWS_FOR_WLAN_CONNECT
