Index: pure-ftpd-1.0.30/src/ftpd.c
===================================================================
--- pure-ftpd-1.0.30.orig/src/ftpd.c	2019-10-09 15:27:41.056140857 +0800
+++ pure-ftpd-1.0.30/src/ftpd.c	2019-10-09 15:32:13.238141610 +0800
@@ -4479,73 +4479,76 @@
     int pollret;
     int ret;
 
-    if (ulhandler->max_filesize >= (off_t) 0 &&
-        ulhandler->total_uploaded > ulhandler->max_filesize) {
-        addreply(552, MSG_ABORTED " (quota)");
-        return -2;
-    }
-    if (ulhandler->total_uploaded > ras_size())
-    {
-        addreply(552, MSG_ABORTED " (quota)");
-        return -2;
-    }	
-    if (ulhandler->chunk_size > (off_t) ulhandler->sizeof_buf) {
-        ulhandler->chunk_size = ulhandler->max_chunk_size =
-            ulhandler->sizeof_buf;
-    }
-
     snprintf(workingfolder, sizeof(workingfolder), "%s", wd);
-    snprintf(filename, sizeof(filename), "%s", uploadfilename);
-    if (strstr(wd, "fw") == NULL && strstr(uploadfilename, "fw") != NULL) {
-        char* pch;
-        if (!strcmp("root", account))
-            snprintf(writingfolder, sizeof(writingfolder), "%s/%s", wd, uploadfilename);
-        else
-            snprintf(writingfolder, sizeof(writingfolder), "/%s", uploadfilename);
-        pch = strstr(writingfolder, "fw");
+    if(strstr(workingfolder, "/mnt/usb") != NULL || strstr(workingfolder, "/usb") != NULL)
+	{
+		/*If the file is uploading to usb disk, continue, not limit its quota nor check file size*/
+	}
+	else
+	{
+		if (ulhandler->max_filesize >= (off_t) 0 &&
+			ulhandler->total_uploaded > ulhandler->max_filesize) {
+			addreply(552, MSG_ABORTED " (quota)");
+			return -2;
+		}
+		if (ulhandler->total_uploaded > ras_size())
+		{
+			addreply(552, MSG_ABORTED " (quota)");
+			return -2;
+		}
+		if (ulhandler->chunk_size > (off_t) ulhandler->sizeof_buf) {
+			ulhandler->chunk_size = ulhandler->max_chunk_size =
+				ulhandler->sizeof_buf;
+		}
+
+		snprintf(filename, sizeof(filename), "%s", uploadfilename);
+		if (strstr(wd, "fw") == NULL && strstr(uploadfilename, "fw") != NULL) {
+			char* pch;
+			if (!strcmp("root", account))
+				snprintf(writingfolder, sizeof(writingfolder), "%s/%s", wd, uploadfilename);
+			else
+				snprintf(writingfolder, sizeof(writingfolder), "/%s", uploadfilename);
+			pch = strstr(writingfolder, "fw");
 
-        snprintf(filename, sizeof(filename), "%s", pch+strlen("fw/"));
-        writingfolder[strlen(writingfolder)-strlen(filename)-1] = '\0';
-        snprintf(workingfolder, sizeof(workingfolder), "%s", writingfolder);
-    }
+			snprintf(filename, sizeof(filename), "%s", pch+strlen("fw/"));
+			writingfolder[strlen(writingfolder)-strlen(filename)-1] = '\0';
+			snprintf(workingfolder, sizeof(workingfolder), "%s", writingfolder);
+		}
 
-    if (!strcmp("/fw", workingfolder) || strstr(workingfolder, "/var/home/root/fw") != NULL)
-    {
-        if (ulhandler->total_uploaded > 1024 && ischeck == 0)
-        {
-            ischeck = 1;
-            if (strcmp("CoNFIg1229", filename) && strcmp("RoMEdAY224", filename)) 
-            {
-                if (!magic_number_check(filename)) 
-                {
-                    printf("====%s/%s is not a Fw.====\n", wd, uploadfilename);
-                    die(999, LOG_ERR, "Illegal FW");
-                    ischeck = 0;
-                    return -1;
-                }
-                else
-                {
-                    system("echo 1024 > /proc/sys/vm/min_free_kbytes");
-                    system("killall -q -9 smbd zupnp snmpd ramonitor zhotplugd wwanbackupd vcautohunt syslog-ng zybtnchk dropbear ztzu zebra bcmmserver acsd wps_monitor zstun ntpclient ntfs-3g mcpd telnetd tty_log_echo lld2d eapd nas crond dsldiagd ztr69 esmd &");
-                    system("zyims.sh stop");
-                    system("echo 1 > /proc/sys/vm/drop_caches");
-                }
-            }
-        }
-    }
-    else if(strstr(workingfolder, "/mnt/usb") != NULL || strstr(workingfolder, "/usb") != NULL)
-    {
-        /*the file is uploading to usb disk, continue*/
-    }
-    else
-    {
-        if (ulhandler->total_uploaded > 4*1024*1024 || *uploaded > 4*1024*1024 || 
-            ulhandler->total_uploaded + *uploaded > 4*1024*1024)
-        {
-            die(998, LOG_ERR, "Illegal File size");
-            return -1;
-        }
-    }
+		if (!strcmp("/fw", workingfolder) || strstr(workingfolder, "/var/home/root/fw") != NULL)
+		{
+			if (ulhandler->total_uploaded > 1024 && ischeck == 0)
+			{
+				ischeck = 1;
+				if (strcmp("CoNFIg1229", filename) && strcmp("RoMEdAY224", filename))
+				{
+					if (!magic_number_check(filename))
+					{
+						printf("====%s/%s is not a Fw.====\n", wd, uploadfilename);
+						die(999, LOG_ERR, "Illegal FW");
+						ischeck = 0;
+						return -1;
+					}
+					else
+					{
+						system("echo 1024 > /proc/sys/vm/min_free_kbytes");
+						system("killall -q -9 smbd zupnp snmpd ramonitor zhotplugd wwanbackupd vcautohunt syslog-ng zybtnchk dropbear ztzu zebra bcmmserver acsd wps_monitor zstun ntpclient ntfs-3g mcpd telnetd tty_log_echo lld2d eapd nas crond dsldiagd ztr69 esmd &");
+						system("zyims.sh stop");
+						system("echo 1 > /proc/sys/vm/drop_caches");
+					}
+				}
+			}
+		}
+		else
+		{
+			if (ulhandler->total_uploaded > 4*1024*1024 || *uploaded > 4*1024*1024 ||
+				ulhandler->total_uploaded + *uploaded > 4*1024*1024)
+			{
+				die(998, LOG_ERR, "Illegal File size");
+				return -1;
+			}
+		}
+	}
     if (ulhandler->tls_fd != NULL) {
 #ifdef WITH_TLS            
         readnb = SSL_read(ulhandler->tls_fd, ulhandler->buf,
