diff -urN net-snmp-5.8.org/agent/mibgroup/if-mib/data_access/interface_ioctl.c net-snmp-5.8/agent/mibgroup/if-mib/data_access/interface_ioctl.c
--- net-snmp-5.8.org/agent/mibgroup/if-mib/data_access/interface_ioctl.c	2019-04-08 14:50:19.079411326 +0800
+++ net-snmp-5.8/agent/mibgroup/if-mib/data_access/interface_ioctl.c	2019-04-08 16:25:38.810982474 +0800
@@ -307,6 +307,7 @@
          */
         if(ifentry->os_flags & IFF_UP) {
             ifentry->admin_status = IFADMINSTATUS_UP;
+#if 0 //20140117 Mark, because wifi performance down
 #if defined(BROADCOM) //__ZYXEL__, Mark
             if(strncmp(ifentry->name, "eth", 3) == 0){
                 strncpy(ifname, ifentry->name, sizeof(ifname));
@@ -342,6 +343,7 @@
             }
             else
 #endif
+#endif
             if(ifentry->os_flags & IFF_RUNNING)
                 ifentry->oper_status = IFOPERSTATUS_UP;
             else
diff -urN net-snmp-5.8.org/agent/mibgroup/if-mib/data_access/interface_linux.c net-snmp-5.8/agent/mibgroup/if-mib/data_access/interface_linux.c
--- net-snmp-5.8.org/agent/mibgroup/if-mib/data_access/interface_linux.c	2019-04-08 15:05:42.266054263 +0800
+++ net-snmp-5.8/agent/mibgroup/if-mib/data_access/interface_linux.c	2019-04-08 16:26:55.858419234 +0800
@@ -1040,6 +1040,7 @@
 netsnmp_linux_interface_get_if_speed(int fd, const char *name,
             unsigned long long defaultspeed)
 {
+#if 0 //20140117 Mark, because wifi performance down
    char cmd[256], ifname[50];
    FILE *pfd;
    unsigned long long speed = defaultspeed;
@@ -1086,7 +1087,9 @@
        speed = 1000;
        return speed * 1000LL * 1000LL;
    }
-   else return netsnmp_linux_interface_get_if_speed_linux(fd, name, defaultspeed);
+   else
+#endif
+   return netsnmp_linux_interface_get_if_speed_linux(fd, name, defaultspeed);
 }
 #endif
 
