Index: libzcfg_fe_tr98-1.0/voice.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/voice.h	2019-02-25 18:15:35.823956792 +0800
+++ libzcfg_fe_tr98-1.0/voice.h	2019-03-04 17:14:48.915521746 +0800
@@ -570,6 +570,8 @@
 	//{ "X_ZYXEL_MsnIncomingSelect", PARAMETER_ATTR_WRITE, 0, json_type_uint32},
 	//{ "X_ZYXEL_MsnOutgoingSelect", PARAMETER_ATTR_WRITE, 0, json_type_uint32},
 	//{ "X_ZYXEL_HowlerToneTimeout", PARAMETER_ATTR_WRITE, 0, json_type_uint32},
+	{ "X_ZYXEL_Pound", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
+	{ "X_ZYXEL_HookFlashIgnore", PARAMETER_ATTR_WRITE, 0, json_type_boolean}, //Michael.20190301.002: Add to Support to Disable/Remove the Phone Hook Flash function by Using a (PhyInterface-level dedicated) Parameter 'X_ZYXEL_HookFlashIgnore' to Ignore the Phone Hook Flash Event. Related Tickets: [eITS#170700466][#39697][HTC], [eITS#180901009][Altibox] and [eITS#181100069][#68892][EU Generic].
 	{ NULL, 0, 0, 0}
 };
 tr98Parameter_t para_VoicePstn[] = {
