Index: libzcfg_fe_tr98-1.0/Makefile
===================================================================
--- libzcfg_fe_tr98-1.0.orig/Makefile	2019-01-09 17:46:35.416414553 +0800
+++ libzcfg_fe_tr98-1.0/Makefile	2019-01-09 17:47:16.117655500 +0800
@@ -170,6 +170,11 @@
 CFLAGS += -DISKON_CUSTOMIZATION
 endif
 
+#TISCALI	
+ifeq ($(strip $(CONFIG_TISCALI_CUSTOMIZATION)),y)
+    CFLAGS += -DTISCALI_CUSTOMIZATION
+endif
+
 #ELISA
 ifeq ($(CONFIG_ELISA_TR69_WANDSL_CUSTOMIZATION),y)
 CFLAGS += -DELISA_TR69_WANDSL_CUSTOMIZATION
Index: libzcfg_fe_tr98-1.0/zyExt.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zyExt.c	2019-01-09 17:46:35.444414552 +0800
+++ libzcfg_fe_tr98-1.0/zyExt.c	2019-01-09 17:46:54.606423400 +0800
@@ -4121,7 +4121,7 @@
 	struct json_object *loginCfgGpAccoutObj = NULL;
 	struct json_object *paramValue = NULL;
 	tr98Parameter_t *paramList = NULL;
-
+	char *default_password = NULL;
 	zcfgLog(ZCFG_LOG_DEBUG, "Enter %s\n", __FUNCTION__);
 
 	IID_INIT(loginCfgGpAccoutIid);
@@ -4141,6 +4141,16 @@
 	paramList = tr98Obj[handler].parameter;
 	while(paramList->name != NULL) {
 		paramValue = json_object_object_get(loginCfgGpAccoutObj, paramList->name);
+#ifdef TISCALI_CUSTOMIZATION
+		if( loginCfgGpAccoutIid.idx[0] == 1 && loginCfgGpAccoutIid.idx[1] == 2){
+			if(!strcmp(paramList->name, "Password")){
+				default_password = json_object_to_json_string(paramValue);
+				if(strlen(default_password)==2){
+					paramValue = json_object_object_get(loginCfgGpAccoutObj, "DefaultPassword");
+				}
+			}
+		}
+#endif
 		if(paramValue != NULL) {
 			json_object_object_add(*tr98Jobj, paramList->name, JSON_OBJ_COPY(paramValue));
 			paramList++;
