Index: libzcfg_fe_tr98-1.0/zcfg_fe_tr98.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zcfg_fe_tr98.c	2019-09-09 11:36:29.093590296 +0800
+++ libzcfg_fe_tr98-1.0/zcfg_fe_tr98.c	2019-09-09 11:41:53.521580936 +0800
@@ -2218,13 +2216,13 @@
 	}
 
 	paramObj = json_object_object_get(queryObj, "Value");
-	if(paramObj && (paramValueStr = json_object_get_string(paramObj))){
+	if(paramObj && (paramValueStr = json_object_get_string(paramObj)) && strlen(paramValueStr)){
 		char hexStr[256*2+1] = {0};
 
 		printf("%s: param value: %s\n", __FUNCTION__, paramValueStr);
 		if(ascStr2HexStr(paramValueStr, hexStr) >= 0){
 			json_object_object_add(queryObj, "Value", json_object_new_string(hexStr));
-			rst = ZCFG_OBJECT_UPDATED;
+			rst = ZCFG_SUCCESS;
 			printf("%s: hex str: %s\n", __FUNCTION__, hexStr);
 		}else{
 			rst = ZCFG_REQUEST_REJECT;
