Index: libzyutil-1.0/common.c
===================================================================
--- libzyutil-1.0.orig/common.c	2019-07-17 15:40:16.632366430 +0800
+++ libzyutil-1.0/common.c	2019-07-17 15:42:49.766367703 +0800
@@ -1538,3 +1538,39 @@
 
     return ret;
 }
+
+/*!
+*  replace string
+*  @param[IN/OUT]	source
+*  @param[IN]	find_str //keyword
+*  @param[IN]	rep_str //replace
+*/
+void str_replace (char *source, char* find_str, char*rep_str){
+	char result[65] = {0};
+	char* str_temp = NULL;
+	char* str_location = NULL;
+	int rep_str_len = 0;
+	int find_str_len = 0;
+	int gap = 0;
+
+	strcpy(result, source);
+	find_str_len = strlen(find_str);
+	rep_str_len = strlen(rep_str);
+	str_temp = source;
+	str_location = strstr(str_temp, find_str);
+
+	while(str_location != NULL){
+		gap += (str_location - str_temp);
+		result[gap] = '\0';
+		strcat(result, rep_str);
+		gap += rep_str_len;
+		str_temp = str_location + find_str_len;
+		strcat(result, str_temp);
+		str_location = strstr(str_temp, find_str);
+	}
+
+	result[strlen(result)] = '\0';
+	memset(source, 0 , strlen(source));
+	strcpy(source, result);
+}
+
