Index: libzyutil-1.0/ECONET/libzyutil_wrapper.c
===================================================================
--- libzyutil-1.0.orig/ECONET/libzyutil_wrapper.c	2019-08-13 15:11:48.011036606 +0800
+++ libzyutil-1.0/ECONET/libzyutil_wrapper.c	2019-08-13 15:18:42.757477900 +0800
@@ -187,13 +187,15 @@
     }else{
         zyUtilReqSendWithResponse(ZCFG_MSG_FAKEROOT_COMMAND_REPLY, ZCFG_EID_SYS_CMD, ZCFG_EID_ESMD, strlen(sysCmd), sysCmd, (char**)&sysCmd_reply);
 
-        if((sysCmd_reply == NULL) && strcmp(sysCmd_reply, "1") != 0){
+        //if((sysCmd_reply == NULL) && strcmp(sysCmd_reply, "1") != 0){
+        if(!sysCmd_reply || (sysCmd_reply && strcmp(sysCmd_reply, "1") != 0)){
             printf("%s: Update Fail!!! \n", __FUNCTION__);        
         }
 	    sprintf(sysCmd,"chmod 666 %s | chown %s %s",filename, username, filename);
         zyUtilReqSendWithResponse(ZCFG_MSG_FAKEROOT_COMMAND_REPLY, ZCFG_EID_SYS_CMD, ZCFG_EID_ESMD, strlen(sysCmd), sysCmd, (char**)&sysCmd_reply);
 
-        if((sysCmd_reply == NULL) && strcmp(sysCmd_reply, "1") != 0){
+        //if((sysCmd_reply == NULL) && strcmp(sysCmd_reply, "1") != 0){
+        if(!sysCmd_reply || (sysCmd_reply && strcmp(sysCmd_reply, "1") != 0)){
             printf("%s: Update Fail!!! \n", __FUNCTION__);        
         }
     }
@@ -256,7 +258,8 @@
 
 	zyUtilReqSendWithResponse(ZCFG_MSG_FAKEROOT_COMMAND_REPLY, ZCFG_EID_SYS_CMD, ZCFG_EID_ESMD, strlen(sysCmd), sysCmd, (char**)&sysCmd_reply);
 
-    if((sysCmd_reply == NULL) && strcmp(sysCmd_reply, "1") != 0){
+    //if((sysCmd_reply == NULL) && strcmp(sysCmd_reply, "1") != 0){
+    if(!sysCmd_reply || (sysCmd_reply && strcmp(sysCmd_reply, "1") != 0)){
     	printf("%s: Update Fail!!! \n", __FUNCTION__);        
     }
 
@@ -530,7 +533,8 @@
 
 	zyUtilReqSendWithResponse(ZCFG_MSG_FAKEROOT_COMMAND_REPLY, ZCFG_EID_SYS_CMD, ZCFG_EID_ESMD_GROUND, strlen(sysCmd), sysCmd, (char**)&sysCmd_reply);
 
-    if((sysCmd_reply == NULL) && strcmp(sysCmd_reply, "1") != 0){
+    //if((sysCmd_reply == NULL) && strcmp(sysCmd_reply, "1") != 0){
+    if(!sysCmd_reply || (sysCmd_reply && strcmp(sysCmd_reply, "1") != 0)){
     	printf("%s: Update Fail!!! \n", __FUNCTION__);        
     }
 
