Index: libzyutil-1.0/flash.c
===================================================================
--- libzyutil-1.0.orig/flash.c	2019-12-06 16:56:00.278901291 +0800
+++ libzyutil-1.0/flash.c	2019-12-06 16:59:27.706392000 +0800
@@ -626,12 +626,6 @@
 		return -1;
 	}
 
-	/* No ecc*/
-	if(ioctl(devFd, MTDFILEMODE, (void *) MTD_FILE_MODE_RAW) != 0) {
-		perror("MTDFILEMODE");
-		goto closeall;
-	}
-
 	/* Initialize start/end addresses and block size */
 	// partition size
 	if(buf_size > mtd.size)
