Index: libzyutil-1.0/BROADCOM/libzyutil_wrapper.c
===================================================================
--- libzyutil-1.0.orig/BROADCOM/libzyutil_wrapper.c	2019-12-13 18:09:37.158204222 +0800
+++ libzyutil-1.0/BROADCOM/libzyutil_wrapper.c	2019-12-13 18:11:07.302728337 +0800
@@ -420,6 +420,22 @@
 	return ret;
 }
 
+int zyUtilSetGPIOoff(int gpioNum)
+{
+	int ret = 0;
+	ret = boardIoctl(BOARD_IOCTL_SET_GPIO, 0, "", gpioNum, 1, "");
+	return ret;
+}
+
+int zyUtilSetGPIOon(int gpioNum)
+{
+	int ret = 0;
+	ret = boardIoctl(BOARD_IOCTL_SET_GPIO, 0, "", gpioNum, 0, "");
+	return ret;
+}
+
+
+
 zcfgRet_t zyUtilIGetDefaultGponSnPw(char *serianumber, char *password)
 {
 	NVRAM_DATA nvramData;
Index: libzyutil-1.0/ECONET/libzyutil_wrapper.c
===================================================================
--- libzyutil-1.0.orig/ECONET/libzyutil_wrapper.c	2019-12-13 18:09:37.174204221 +0800
+++ libzyutil-1.0/ECONET/libzyutil_wrapper.c	2019-12-13 18:10:55.044596928 +0800
@@ -615,6 +615,19 @@
 	return 0;
 }
 
+int zyUtilSetGPIOoff(int gpioNum)
+{
+	/*This function is only used to check VMG3926-B10A and VMG8924-B10B, so Cross Platform is not need*/
+	return 0;
+}
+
+int zyUtilSetGPIOon(int gpioNum)
+{
+	/*This function is only used to check VMG3926-B10A and VMG8924-B10B, so Cross Platform is not need*/
+	return 0;
+}
+
+
 float transferOpticalTemperature(int input_value)
 {
 	int i;
