Index: ppp-2.4.3/pppd/ipcp.c
===================================================================
--- ppp-2.4.3.orig/pppd/ipcp.c	2018-11-20 18:05:51.269728104 +0800
+++ ppp-2.4.3/pppd/ipcp.c	2018-11-20 18:06:38.065960154 +0800
@@ -1758,6 +1758,9 @@
 
     json_object_put(ppp_arg);
 #endif
+#ifdef ZYXEL_FAKE_IP_169_254_X_X
+	pppConnRound = 0;
+#endif
     return 1;
 }
 
@@ -2027,6 +2030,9 @@
 
     json_object_put(ppp_arg);
 #endif
+#ifdef ZYXEL_FAKE_IP_169_254_X_X
+	pppConnRound = 0;
+#endif
 }
 
 
Index: ppp-2.4.3/pppd/main.c
===================================================================
--- ppp-2.4.3.orig/pppd/main.c	2018-11-20 18:05:51.353728520 +0800
+++ ppp-2.4.3/pppd/main.c	2018-11-20 18:05:51.397728738 +0800
@@ -388,6 +388,36 @@
 }
 #endif
 
+#ifdef ZYXEL_FAKE_IP_169_254_X_X
+zcfgRet_t sendFakeIP(void){
+	zcfgRet_t ret;
+	struct json_object *ppp_arg = NULL;
+	const char *payload = NULL;
+	int payloadLen = 0;
+	char ip[16] = {0};
+
+	ppp_arg = json_object_new_object();
+
+	srand(time(NULL));
+	sprintf(ip, "169.254.%d.%d", rand() % 256, rand() % 254 + 1);
+	json_object_object_add(ppp_arg, "ip", json_object_new_string(ip));
+	json_object_object_add(ppp_arg, "remote", json_object_new_string(""));
+	json_object_object_add(ppp_arg, "dns", json_object_new_string(""));
+	json_object_object_add(ppp_arg, "linkname", json_object_new_string(linkname));
+	json_object_object_add(ppp_arg, "type", json_object_new_string("ipv4"));
+	json_object_object_add(ppp_arg, "state", json_object_new_string("up"));
+
+	notice("Json PPP %s", json_object_to_json_string(ppp_arg));
+	payload = json_object_to_json_string(ppp_arg);
+	payloadLen = strlen(payload) + 1;
+	ret = pppdMsgSend(ZCFG_MSG_PPPD_EVENT , ZCFG_EID_ESMD, payloadLen, payload);
+
+	json_object_put(ppp_arg);
+
+	return ret;
+}
+#endif
+
 int
 main(argc, argv)
     int argc;
@@ -672,6 +702,13 @@
 	    }
 	}
 
+#ifdef ZYXEL_FAKE_IP_169_254_X_X
+	if(++pppConnRound >= 2){
+		holdoff = 0x7fffffff;
+		sendFakeIP();
+	}
+#endif
+
 #ifdef ZyXEL_WEB_REDIRECT /* __ZyXEL__, Albert, 20150416, Web redirect   */
 
         switch(status){
Index: ppp-2.4.3/pppd/options.c
===================================================================
--- ppp-2.4.3.orig/pppd/options.c	2018-11-20 18:05:51.381728659 +0800
+++ ppp-2.4.3/pppd/options.c	2018-11-20 18:05:51.397728738 +0800
@@ -100,6 +100,9 @@
 #ifdef ABOG_CUSTOMIZATION
 int count_sendPADI = 0;	/* count how many packets sending*/
 #endif
+#ifdef ZYXEL_FAKE_IP_169_254_X_X
+int pppConnRound = 0;	/* count how many round that PPP try to connect with server */
+#endif
 bool	holdoff_specified;	/* true if a holdoff value has been given */
 int	log_to_fd = 1;		/* send log messages to this fd too */
 bool	log_default = 1;	/* log_to_fd is default (stdout) */
Index: ppp-2.4.3/pppd/pppd.h
===================================================================
--- ppp-2.4.3.orig/pppd/pppd.h	2018-11-20 18:05:51.385728679 +0800
+++ ppp-2.4.3/pppd/pppd.h	2018-11-20 18:05:51.401728758 +0800
@@ -329,6 +329,9 @@
 extern bool	dump_options;	/* print out option values */
 extern bool	dryrun;		/* check everything, print options, exit */
 extern int	child_wait;	/* # seconds to wait for children at end */
+#ifdef ZYXEL_FAKE_IP_169_254_X_X
+extern int	pppConnRound;	/* count how many round that PPP try to connect with server */
+#endif
 
 #ifdef MAXOCTETS
 extern unsigned int maxoctets;	     /* Maximum octetes per session (in bytes) */
