Index: libzcfg_fe_dal-1.0/network/broadband/zcfg_fe_dal_broadband.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/broadband/zcfg_fe_dal_broadband.c	2019-03-27 16:29:30.558431356 +0800
+++ libzcfg_fe_dal-1.0/network/broadband/zcfg_fe_dal_broadband.c	2019-03-27 16:30:31.369546158 +0800
@@ -23,6 +23,7 @@
 dal_param_t WAN_param[]={
 	{"Index",				dalType_int,	0,	0,	NULL, NULL, dal_Add_ignore, dal_Edit_ignore},
 	{"ipIfacePath",			dalType_string,	1,	32,	NULL, NULL, dalcmd_Forbid}, //read only
+	{"X_ZYXEL_IfName",		dalType_string,	0,	32,	NULL, NULL, dalcmd_Forbid}, //read only
 	{"Name",				dalType_string,	1,	32,	NULL, NULL, dal_Add|dal_Edit|dal_Delete},
 	{"Enable",				dalType_boolean,	0,	0,	NULL},
 	{"Type",				dalType_string,	3,	4,	NULL, "ATM|PTM|ETH|GPON", dal_Add | dal_Edit_ignore},
@@ -79,9 +80,11 @@
 	{"IPAddress",			dalType_v4Addr,	0,	0,	NULL},
 	{"SubnetMask",			dalType_v4Mask,	0,	0,	NULL},
 	{"GatewayIPAddress",	dalType_v4Addr,	0,	0,	NULL},
+	{"CurIPAddress",		dalType_v4Addr,	0,	0,	NULL, NULL, dalcmd_Forbid}, //read only
 	{"ip6Static",			dalType_boolean,	0,	0,	NULL},
 	{"IP6Address",			dalType_v6AddrPrefix,	0,	0,	NULL},
 	{"NextHop",				dalType_v6Addr,	0,	0,	NULL},
+	{"CurIP6Address",		dalType_v6AddrPrefix,	0,	0,	NULL, NULL, dalcmd_Forbid}, //read only
      /* __ZyXEL__, Albert, 20180209,Support IPv6 option IA_NA and IA_PD  */
 	{"IPv6Manual_IANA",		dalType_boolean,	0,	0,	NULL},
 	{"IPv6Manual_IAPD",		dalType_boolean,	0,	0,	NULL},	
@@ -4237,12 +4240,15 @@
 	if(ret == ZCFG_SUCCESS && v4AddrObj){
 		AddressingType = json_object_get_string(json_object_object_get(v4AddrObj, "AddressingType"));
 		Enable = json_object_get_boolean(json_object_object_get(v4AddrObj, "Enable"));
-		if(Enable == true && !strcmp(AddressingType, "Static") && strstr(CurrEncapsulation, "IP")){
-			ipStatic = true;
-			json_object_object_add(pramJobj, "ipStatic", json_object_new_boolean(true));
-			json_object_object_add(pramJobj, "IPAddress", JSON_OBJ_COPY(json_object_object_get(v4AddrObj, "IPAddress")));
-			json_object_object_add(pramJobj, "SubnetMask", JSON_OBJ_COPY(json_object_object_get(v4AddrObj, "SubnetMask")));
-		}
+        if(Enable == true){
+            replaceParam(pramJobj, "CurIPAddress", v4AddrObj, "IPAddress");
+    		if(!strcmp(AddressingType, "Static") && strstr(CurrEncapsulation, "IP")){
+    			ipStatic = true;
+    			json_object_object_add(pramJobj, "ipStatic", json_object_new_boolean(true));
+    			json_object_object_add(pramJobj, "IPAddress", JSON_OBJ_COPY(json_object_object_get(v4AddrObj, "IPAddress")));
+    			json_object_object_add(pramJobj, "SubnetMask", JSON_OBJ_COPY(json_object_object_get(v4AddrObj, "SubnetMask")));
+    		}
+        }
 	}
 
 	if(ipStatic){
@@ -4324,10 +4330,13 @@
 	if(ret == ZCFG_SUCCESS && v6AddrObj){
 		Origin = json_object_get_string(json_object_object_get(v6AddrObj, "Origin"));
 		Enable = json_object_get_boolean(json_object_object_get(v6AddrObj, "Enable"));
-		if(Enable == true && !strcmp(Origin, "Static")){
-			ip6Static = true;
-			json_object_object_add(pramJobj, "ip6Static", json_object_new_boolean(true));
-			json_object_object_add(pramJobj, "IP6Address", JSON_OBJ_COPY(json_object_object_get(v6AddrObj, "IPAddress")));
+		if(Enable == true){
+            replaceParam(pramJobj, "CurIP6Address", v6AddrObj, "IPAddress");
+            if(!strcmp(Origin, "Static")){
+			    ip6Static = true;
+			    json_object_object_add(pramJobj, "ip6Static", json_object_new_boolean(true));
+			    json_object_object_add(pramJobj, "IP6Address", JSON_OBJ_COPY(json_object_object_get(v6AddrObj, "IPAddress")));
+            }
 		}
 	}
 
@@ -4477,6 +4486,7 @@
 			count++;
 			json_object_object_add(pramJobj, "index", json_object_new_int(count));
 			Jadds(pramJobj, "ipIfacePath", ipIfacePath);
+            replaceParam(pramJobj, "X_ZYXEL_IfName", ipIfObj, "X_ZYXEL_IfName");
 
 #if defined(SAME_FW_FOR_TWO_BOARD)
             if (strcmp(CurrType, "ETH") == 0)
