Index: libzcfg_fe_dal-1.0/js/zcfg_fe_dal_networkmap.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/js/zcfg_fe_dal_networkmap.c	2018-12-20 11:06:11.974605037 +0800
+++ libzcfg_fe_dal-1.0/js/zcfg_fe_dal_networkmap.c	2018-12-20 11:12:28.284978428 +0800
@@ -18,6 +18,13 @@
 dal_param_t NETWORK_MAP_param[] = 
 {
 	{"Internet_Blocking_Enable",	    dalType_boolean,	0,	0,	NULL},
+#ifdef ZYXEL_HOME_CYBER_SECURITY_FSECURE	
+	{"BrowsingProtection",	    dalType_boolean,	0,	0,	NULL},
+	{"TrackingProtection",	    dalType_boolean,	0,	0,	NULL},
+	{"X_ZYXEL_FirstSeen",			dalType_int,		0,	0,	NULL},
+	{"X_ZYXEL_LastSeen",			dalType_int,		0,	0,	NULL},
+	{"SourceVendorClassID",			dalType_string,		0,	0,	NULL},	
+#endif	
 	{"HostName",						dalType_string,		0,	0,	NULL},
 	{"IPAddress",						dalType_string,		0,	0,	NULL},
 	{"IPLinkLocalAddress6",				dalType_string,		0,	0,	NULL},
@@ -1088,6 +1095,13 @@
 		Jadds(obj, "WiFiname", "");
 		Jadds(obj, "DeviceIcon", "");
 		Jaddb(obj, "Internet_Blocking_Enable", false);
+#ifdef ZYXEL_HOME_CYBER_SECURITY_FSECURE		
+		Jaddb(obj, "BrowsingProtection", false);
+		Jaddb(obj, "TrackingProtection", false);
+		Jaddi(obj, "X_ZYXEL_FirstSeen", 0);
+		Jaddi(obj, "X_ZYXEL_LastSeen", 0);
+		Jadds(obj, "SourceVendorClassID", "");		
+#endif		
 		replaceParam(obj, "DeviceName", obj, "HostName");
 		Jadds(obj, "curHostName", "");
 		
@@ -1166,6 +1180,13 @@
 			if(isSameMac(PhysAddress, MacAddress)){
 				replaceParam(obj, "DeviceIcon", iconJobj, "DeviceIcon");
 				replaceParam(obj, "Internet_Blocking_Enable", iconJobj, "Internet_Blocking_Enable");
+#ifdef ZYXEL_HOME_CYBER_SECURITY_FSECURE				
+				replaceParam(obj, "BrowsingProtection", iconJobj, "BrowsingProtection");
+				replaceParam(obj, "TrackingProtection", iconJobj, "TrackingProtection");	
+				replaceParam(obj, "X_ZYXEL_FirstSeen", iconJobj, "X_ZYXEL_FirstSeen");	
+				replaceParam(obj, "X_ZYXEL_LastSeen", iconJobj, "X_ZYXEL_LastSeen");	
+				replaceParam(obj, "SourceVendorClassID", iconJobj, "SourceVendorClassID");					
+#endif				
 				if(strcmp("", Jgets(iconJobj, "HostName")))
 					replaceParam(obj, "curHostName", iconJobj, "HostName");
 			}
@@ -1253,14 +1274,22 @@
 	zcfgRet_t ret = ZCFG_SUCCESS;
 	struct json_object *obj = NULL;
 	objIndex_t Iid = {0};
-	const char *macAddr = NULL, *curmacAddr = NULL, *GUIicon = NULL, *hostName = NULL;
+	const char *macAddr = NULL, *curmacAddr = NULL, *GUIicon = NULL, *hostName = NULL, *SourceVendorClassID = NULL;
 	char icon[16] = {0};
-	bool changeIconRuleExist = false, Internet_Blocking_Enable = false;
+	bool changeIconRuleExist = false, Internet_Blocking_Enable = false,BrowsingProtection = false,TrackingProtection = false;
+	unsigned int X_ZYXEL_FirstSeen =0, X_ZYXEL_LastSeen = 0;
+	
 	macAddr = json_object_get_string(json_object_object_get(Jobj,"MacAddress"));
 	GUIicon = json_object_get_string(json_object_object_get(Jobj,"DeviceIcon"));
 	hostName = json_object_get_string(json_object_object_get(Jobj,"HostName"));
 	Internet_Blocking_Enable = json_object_get_boolean(json_object_object_get(Jobj,"Internet_Blocking_Enable"));
-
+#ifdef ZYXEL_HOME_CYBER_SECURITY_FSECURE	
+	BrowsingProtection = json_object_get_boolean(json_object_object_get(Jobj,"BrowsingProtection"));
+	TrackingProtection = json_object_get_boolean(json_object_object_get(Jobj,"TrackingProtection"));
+	X_ZYXEL_FirstSeen = json_object_get_int(json_object_object_get(Jobj,"X_ZYXEL_FirstSeen"));
+	X_ZYXEL_LastSeen = json_object_get_int(json_object_object_get(Jobj,"X_ZYXEL_LastSeen"));
+	SourceVendorClassID = json_object_get_string(json_object_object_get(Jobj,"SourceVendorClassID"));	
+#endif
 	if(!strcmp(GUIicon, "ipc")){
 		strcat(icon, "iPC");
 	}else if(!strcmp(GUIicon, "igame")){
@@ -1302,6 +1331,23 @@
 			if(json_object_object_get(Jobj, "Internet_Blocking_Enable")){
 				json_object_object_add(obj, "Internet_Blocking_Enable", json_object_new_boolean(Internet_Blocking_Enable));
 			}
+#ifdef ZYXEL_HOME_CYBER_SECURITY_FSECURE			
+			if(json_object_object_get(Jobj, "BrowsingProtection")){
+				json_object_object_add(obj, "BrowsingProtection", json_object_new_boolean(BrowsingProtection));
+			}
+			if(json_object_object_get(Jobj, "TrackingProtection")){
+				json_object_object_add(obj, "TrackingProtection", json_object_new_boolean(TrackingProtection));
+			}			
+			if(json_object_object_get(Jobj, "X_ZYXEL_FirstSeen")){
+				json_object_object_add(obj, "X_ZYXEL_FirstSeen", json_object_new_int(X_ZYXEL_FirstSeen));
+			}			
+			if(json_object_object_get(Jobj, "X_ZYXEL_LastSeen")){
+				json_object_object_add(obj, "X_ZYXEL_LastSeen", json_object_new_int(X_ZYXEL_LastSeen));
+			}
+			if(json_object_object_get(Jobj, "SourceVendorClassID")){
+				json_object_object_add(obj, "SourceVendorClassID", json_object_new_string(SourceVendorClassID));
+			}			
+#endif			
 			zcfgFeObjJsonSet(RDM_OID_CHANGE_ICON_NAME, &Iid, obj, NULL);
 			zcfgFeJsonObjFree(obj);
 			break;			
@@ -1325,6 +1371,23 @@
 			if(json_object_object_get(Jobj, "Internet_Blocking_Enable")){
 				json_object_object_add(obj, "Internet_Blocking_Enable", json_object_new_boolean(Internet_Blocking_Enable));
 			}
+#ifdef ZYXEL_HOME_CYBER_SECURITY_FSECURE			
+			if(json_object_object_get(Jobj, "BrowsingProtection")){
+				json_object_object_add(obj, "BrowsingProtection", json_object_new_boolean(BrowsingProtection));
+			}
+			if(json_object_object_get(Jobj, "TrackingProtection")){
+				json_object_object_add(obj, "TrackingProtection", json_object_new_boolean(TrackingProtection));
+			}		
+			if(json_object_object_get(Jobj, "X_ZYXEL_FirstSeen")){
+				json_object_object_add(obj, "X_ZYXEL_FirstSeen", json_object_new_int(X_ZYXEL_FirstSeen));
+			}	
+			if(json_object_object_get(Jobj, "X_ZYXEL_LastSeen")){
+				json_object_object_add(obj, "X_ZYXEL_LastSeen", json_object_new_int(X_ZYXEL_LastSeen));
+			}	
+			if(json_object_object_get(Jobj, "SourceVendorClassID")){
+				json_object_object_add(obj, "SourceVendorClassID", json_object_new_string(SourceVendorClassID));
+			}				
+#endif			
 			zcfgFeObjJsonSet(RDM_OID_CHANGE_ICON_NAME, &Iid, obj, NULL);
 			zcfgFeJsonObjFree(obj);
 		}
