Index: libzcfg_fe_dal-1.0/network/wireless/zcfg_fe_dal_wifi_wps.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/wireless/zcfg_fe_dal_wifi_wps.c	2019-02-19 11:12:49.579837317 +0800
+++ libzcfg_fe_dal-1.0/network/wireless/zcfg_fe_dal_wifi_wps.c	2019-02-19 11:17:37.409406703 +0800
@@ -250,6 +250,7 @@
 	}
 	
 	// wps apply check
+#if 0	/* Austin, 2019-02-19 10:58:36, fail to disable 5G WPS when hide ssid was enabled */
 	if((ret = zcfgFeObjJsonGet(RDM_OID_WIFI_ACCESS_POINT, &objIid, &tmpObj)) != ZCFG_SUCCESS){
 		return ret;
 	}
@@ -261,6 +262,7 @@
 	else{
 		zcfgFeJsonObjFree(tmpObj);
 	}
+#endif
 	
 	if((ret = zcfgFeObjJsonGet(RDM_OID_WIFI_ACCESS_POINT_WPS, &objIid, &WifiApWpsJobj)) != ZCFG_SUCCESS) {
 		return ret;
