Index: libzcfg_fe_dal-1.0/js/zcfg_fe_dal_networkmap.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/js/zcfg_fe_dal_networkmap.c	2019-02-25 18:54:50.846352850 +0800
+++ libzcfg_fe_dal-1.0/js/zcfg_fe_dal_networkmap.c	2019-02-25 18:56:29.751168333 +0800
@@ -794,6 +794,23 @@
 		json_object_array_add(voipLineJarray, voipLineObj);
 	}
 
+    if (zyUtil_is_sfp_present() != 0)
+    {
+        json_object_object_add(gponstatusObj, "sfp_presence",json_object_new_boolean(true));
+    }
+    else
+    {
+        json_object_object_add(gponstatusObj, "sfp_presence",json_object_new_boolean(false));
+    }
+    if (zyUtil_bosa_exist())
+    {
+        json_object_object_add(gponstatusObj, "bosa_presence",json_object_new_boolean(true));
+    }
+    else
+    {
+        json_object_object_add(gponstatusObj, "bosa_presence",json_object_new_boolean(false));
+    }
+
 	getWifiInfo(&ssidJarray);
 	getWanLanList_ext(&wanLanJarray);
 	
