diff -ruN a/network/broadband/zcfg_fe_dal_broadband.c b/network/broadband/zcfg_fe_dal_broadband.c
--- a/network/broadband/zcfg_fe_dal_broadband.c	2019-04-30 10:00:14.000000000 +0800
+++ b/network/broadband/zcfg_fe_dal_broadband.c	2019-05-02 16:28:37.800355005 +0800
@@ -3497,9 +3497,17 @@
 		}
 		
 		if(!strcmp(LinkType, Jgets(wanLanObj, "LinkType")) && (VLANID == Jgeti(wanLanObj, "VLANID"))){
-			if(strcmp(LinkType, "ATM"))
-				return true;
-			else if(!strcmp(VpiVci, Jgets(wanLanObj, "VpiVci")))
+#ifdef ZYXEL_USA_PRODUCT
+			if(strcmp(LinkType, "ATM")&& (VLANID !=-1))
+#else
+ 			if(strcmp(LinkType, "ATM"))
+#endif
+ 				return true;
+#ifdef ZYXEL_USA_PRODUCT
+			else if(!strcmp(VpiVci, Jgets(wanLanObj, "VpiVci"))&& strcmp(VpiVci,""))
+#else
+ 			else if(!strcmp(VpiVci, Jgets(wanLanObj, "VpiVci")))
+#endif	
 				return true;
 		}
 	}
