diff -urN libzcfg_fe_dal-1.0.org/js/zcfg_fe_dal_quickstart.c libzcfg_fe_dal-1.0/js/zcfg_fe_dal_quickstart.c
--- libzcfg_fe_dal-1.0.org/js/zcfg_fe_dal_quickstart.c	2019-05-08 11:27:51.255258400 +0800
+++ libzcfg_fe_dal-1.0/js/zcfg_fe_dal_quickstart.c	2019-05-08 15:45:02.966463492 +0800
@@ -526,7 +526,7 @@
 	objIndex_t dnsdeleteIid = {0};
 	const char *originterface = NULL;
 	const char *interface = NULL;
-	const char *type = NULL;
+	// const char *type = NULL;
 	int index = 0;
 	IID_INIT(dnsdeleteIid);
 	reqObject = json_object_array_get_idx(Jobj, 9);
@@ -534,7 +534,7 @@
 		if(!strcmp(iptype,"PPP dynamic")){
 			reqObject1 = json_object_array_get_idx(Jobj, 1);
 			interface = json_object_get_string(json_object_object_get(reqObject1, "Interface"));
-			type = json_object_get_string(json_object_object_get(reqObject1, "X_ZYXEL_Type"));
+			// type = json_object_get_string(json_object_object_get(reqObject1, "X_ZYXEL_Type"));
 			while((ret = zcfgFeObjJsonGetNext(RDM_OID_DNS_CLIENT_SRV, &dnsdeleteIid, &dnsdeleteObj)) == ZCFG_SUCCESS){
 				originterface = json_object_get_string(json_object_object_get(dnsdeleteObj, "Interface"));
 				if(!strcmp(interface,originterface)){
diff -urN libzcfg_fe_dal-1.0.org/maintenance/mgmt/zcfg_fe_dal_mgmt_srv.c libzcfg_fe_dal-1.0/maintenance/mgmt/zcfg_fe_dal_mgmt_srv.c
--- libzcfg_fe_dal-1.0.org/maintenance/mgmt/zcfg_fe_dal_mgmt_srv.c	2019-05-08 11:27:49.607258400 +0800
+++ libzcfg_fe_dal-1.0/maintenance/mgmt/zcfg_fe_dal_mgmt_srv.c	2019-05-08 15:48:01.661595358 +0800
@@ -222,7 +222,7 @@
 
 			IID_INIT(mgmtSrvIid);
 			index = json_object_get_int(json_object_object_get(reqObject, "Index"));
-			protocol = json_object_get_string(json_object_object_get(reqObject, "Protocol"));
+			// protocol = json_object_get_string(json_object_object_get(reqObject, "Protocol"));
 			mgmtSrvIid.idx[0] = index;
 			mgmtSrvIid.level = 1;
 			ret = zcfgFeObjJsonGetWithoutUpdate(RDM_OID_REMO_SRV, &mgmtSrvIid, &mgmtSrvObj);
diff -urN libzcfg_fe_dal-1.0.org/network/broadband/zcfg_fe_dal_broadband.c libzcfg_fe_dal-1.0/network/broadband/zcfg_fe_dal_broadband.c
--- libzcfg_fe_dal-1.0.org/network/broadband/zcfg_fe_dal_broadband.c	2019-05-08 11:27:51.215258400 +0800
+++ libzcfg_fe_dal-1.0/network/broadband/zcfg_fe_dal_broadband.c	2019-05-08 16:36:37.876727105 +0800
@@ -1268,8 +1268,9 @@
 	bool ipStatic = false, ip6Static = false;
 	const char *IPAddress, *SubnetMask, *GatewayIPAddress;
 	const char *CurrIPAddress, *CurrSubnetMask, *CurrGatewayIPAddress;
-	const char *IP6Address, *NextHop;
-	const char *CurrIP6Address, *CurrNextHop;
+	const char *IP6Address;
+	const char *CurrIP6Address;
+	// const char *NextHop, *CurrNextHop;
 	bool ipdnsmodify = false;
 
 	pramJobj = json_object_object_get(Jobj, "ipStatic");
@@ -1400,9 +1401,9 @@
 	if(CurrIp6Static && ip6Static){ //enable to enable
 		//v6 addr
 		IP6Address = json_object_get_string(json_object_object_get(Jobj, "IP6Address"));
-		NextHop = json_object_get_string(json_object_object_get(Jobj, "NextHop"));
+		// NextHop = json_object_get_string(json_object_object_get(Jobj, "NextHop"));
 		CurrIP6Address = json_object_get_string(json_object_object_get(v6AddrObj, "IPAddress"));
-		CurrNextHop = json_object_get_string(json_object_object_get(v6FwdObj, "NextHop"));
+		// CurrNextHop = json_object_get_string(json_object_object_get(v6FwdObj, "NextHop"));
 		
 		if(strcmp(IP6Address, CurrIP6Address)){
 			json_object_object_add(v6AddrObj, "Enable", json_object_new_boolean(true));
@@ -3284,7 +3285,7 @@
 		IID_INIT(v6FwdIid);
 		while(zcfgFeSubInObjJsonGetNextWithoutUpdate(RDM_OID_ROUTING_ROUTER_V6_FWD, &routerIid, &v6FwdIid, &v6FwdObj) == ZCFG_SUCCESS){
 			Interface = json_object_get_string(json_object_object_get(v6FwdObj, "Interface"));
-			Origin = json_object_get_string(json_object_object_get(v6FwdObj, "Origin"));
+			// Origin = json_object_get_string(json_object_object_get(v6FwdObj, "Origin"));
 			//if(!strcmp(Interface, ipIfacePath) && !strcmp(Origin, "Static")){
 			if(!strcmp(Interface, ipIfacePath)){
 				DestIPPrefix = json_object_get_string(json_object_object_get(v6FwdObj, "DestIPPrefix"));
@@ -4393,7 +4394,7 @@
 	if(v6FwdObj) zcfgFeJsonObjFree(v6FwdObj);
 	while(zcfgFeSubInObjJsonGetNextWithoutUpdate(RDM_OID_ROUTING_ROUTER_V6_FWD, &routerIid, &v6FwdIid, &v6FwdObj) == ZCFG_SUCCESS){
 		Interface = json_object_get_string(json_object_object_get(v6FwdObj, "Interface"));
-		Origin = json_object_get_string(json_object_object_get(v6FwdObj, "Origin"));
+		// Origin = json_object_get_string(json_object_object_get(v6FwdObj, "Origin"));
 		DestIPPrefix = Jgets(v6FwdObj, "DestIPPrefix");
 		if(strstr(CurrEncapsulation, "IP")){
 			if(!strcmp(Interface, ipIfacePath) && (!strcmp(DestIPPrefix, "::/0") || !strcmp(DestIPPrefix, ""))){
diff -urN libzcfg_fe_dal-1.0.org/network/homeNetworking/zcfg_fe_dal_addnl_subnet.c libzcfg_fe_dal-1.0/network/homeNetworking/zcfg_fe_dal_addnl_subnet.c
--- libzcfg_fe_dal-1.0.org/network/homeNetworking/zcfg_fe_dal_addnl_subnet.c	2019-05-08 11:27:50.831258400 +0800
+++ libzcfg_fe_dal-1.0/network/homeNetworking/zcfg_fe_dal_addnl_subnet.c	2019-05-08 16:20:35.047618043 +0800
@@ -147,12 +147,18 @@
 
 zcfgRet_t setPublicLan(struct json_object *Jobj, struct json_object *ipIntfObj, objIndex_t *ipIntfIid){
 	zcfgRet_t ret = ZCFG_SUCCESS;
+#if 0
 	bool pubLan = false, curPubLan = false;
 	bool pubLanDHCP = false, curPubDHCP = false;
 	bool pubLanARP = false, curPubARP = false;
 	const char *pubLanIP = NULL, *curPubLanIP = NULL;
 	const char *pubLanMask = NULL, *curPubLanMask = NULL;
-	
+#endif
+	bool pubLan = false;
+	bool pubLanDHCP = false;
+	bool pubLanARP = false;
+	const char *pubLanIP = NULL;
+	const char *pubLanMask = NULL;
 	pubLan = json_object_get_boolean(json_object_object_get(Jobj, "PubLan_Enable"));
 	pubLanIP = json_object_get_string(json_object_object_get(Jobj, "PubLan_IP"));
 	pubLanMask = json_object_get_string(json_object_object_get(Jobj, "PubLan_SubnetMask"));
@@ -160,11 +166,13 @@
 	pubLanARP = json_object_get_boolean(json_object_object_get(Jobj, "PubLan_ARP"));
 
 	if(ipIntfObj){
+#if 0
 		curPubLan = json_object_get_boolean(json_object_object_get(ipIntfObj, "X_ZYXEL_PubLan"));
 		curPubLanIP = json_object_get_string(json_object_object_get(ipIntfObj, "X_ZYXEL_PubLan_IP"));
 		curPubLanMask = json_object_get_string(json_object_object_get(ipIntfObj, "X_ZYXEL_PubLan_Mask"));
 		curPubDHCP = json_object_get_boolean(json_object_object_get(ipIntfObj, "X_ZYXEL_PubLan_DHCP"));
 		curPubARP = json_object_get_boolean(json_object_object_get(ipIntfObj, "X_ZYXEL_PubLan_ARP"));
+#endif
 		if(json_object_object_get(Jobj, "PubLan_Enable"))
 			json_object_object_add(ipIntfObj, "X_ZYXEL_PubLan", json_object_new_boolean(pubLan));
 		if(json_object_object_get(Jobj, "PubLan_IP") && pubLan)
diff -urN libzcfg_fe_dal-1.0.org/network/interfaceGrouping/zcfg_fe_dal_intfGrp.c libzcfg_fe_dal-1.0/network/interfaceGrouping/zcfg_fe_dal_intfGrp.c
--- libzcfg_fe_dal-1.0.org/network/interfaceGrouping/zcfg_fe_dal_intfGrp.c	2019-05-08 11:27:51.003258400 +0800
+++ libzcfg_fe_dal-1.0/network/interfaceGrouping/zcfg_fe_dal_intfGrp.c	2019-05-08 16:21:42.818001523 +0800
@@ -1589,7 +1589,7 @@
 
 	int i, len = 0;
 	struct json_object *obj;
-	const char *ipver = NULL;
+	// const char *ipver = NULL;
 
 
 	if(json_object_get_type(Jarray) != json_type_array){
@@ -1618,7 +1618,7 @@
 		if(!strcmp(wanInterface,""))
 			wanInterface = "Any WAN";
 		
-		ipver = json_object_get_string(json_object_object_get(obj, "ipver"));
+		// ipver = json_object_get_string(json_object_object_get(obj, "ipver"));
 			printf("%-6s %-12s %-35s %-55s %-35s \n",
 					json_object_get_string(json_object_object_get(obj, "Index")),
 				    json_object_get_string(json_object_object_get(obj, "Group_Name")),
diff -urN libzcfg_fe_dal-1.0.org/network/tunnelSetting/zcfg_fe_dal_tunnel_Setting_entry.c libzcfg_fe_dal-1.0/network/tunnelSetting/zcfg_fe_dal_tunnel_Setting_entry.c
--- libzcfg_fe_dal-1.0.org/network/tunnelSetting/zcfg_fe_dal_tunnel_Setting_entry.c	2019-05-08 11:27:49.879258400 +0800
+++ libzcfg_fe_dal-1.0/network/tunnelSetting/zcfg_fe_dal_tunnel_Setting_entry.c	2019-05-08 16:25:30.719943216 +0800
@@ -197,7 +197,7 @@
 	char *greBridgeName = NULL;  //raghu
 	const char *lanIntf = NULL;
 	const char *modify_str = NULL;
-	const char *org_lanIntf;
+	// const char *org_lanIntf;
 	int shapingRate = 0;
 	int downstreamShapingRate = 0;
 	int dscp = 0;
@@ -342,7 +342,7 @@
 
 		org_obj = json_object_object_get(Jobj, "org");
 		if(org_obj) {
-			org_lanIntf = json_object_get_string(json_object_object_get(org_obj, "lanIntf"));
+			// org_lanIntf = json_object_get_string(json_object_object_get(org_obj, "lanIntf"));
 		}
 		else {
 			printf("%s(%d)\n\r", __FUNCTION__, __LINE__);
diff -urN libzcfg_fe_dal-1.0.org/network/usbService/zcfg_fe_dal_usb_filesharing.c libzcfg_fe_dal-1.0/network/usbService/zcfg_fe_dal_usb_filesharing.c
--- libzcfg_fe_dal-1.0.org/network/usbService/zcfg_fe_dal_usb_filesharing.c	2019-05-08 11:27:50.671258400 +0800
+++ libzcfg_fe_dal-1.0/network/usbService/zcfg_fe_dal_usb_filesharing.c	2019-05-08 16:42:33.778151881 +0800
@@ -24,23 +24,26 @@
 };
 void zcfgFeDalShowUsbFilesharing(struct json_object *Jarray){
 	struct json_object *obj = NULL;
-	struct json_object *servconfobj = NULL;
-	struct json_object *accountobj = NULL, *accounttmpobj = NULL;
-	struct json_object *sambaobj = NULL, *sambatmpobj = NULL;
-	struct json_object *usbinfoobj = NULL, *usbinfotmpobj = NULL;
-	int len_accountobj = 0, len_sambaobj = 0, len_usbinfoobj = 0, num_usbinfoobj;
-	int num_accountobj, num_sambaobj;
+	// struct json_object *servconfobj = NULL;
+	// struct json_object *accountobj = NULL, *accounttmpobj = NULL;
+	// struct json_object *sambaobj = NULL, *sambatmpobj = NULL;
+	struct json_object *usbinfoobj = NULL;
+	struct json_object *usbinfotmpobj = NULL;
+	// int len_accountobj = 0;
+	int len_usbinfoobj = 0, num_usbinfoobj;
+	// int len_sambaobj = 0;
+	// int num_accountobj, num_sambaobj;
 	if(json_object_get_type(Jarray) != json_type_array){
 		printf("wrong Jobj format!\n");
 	return;
 }
 	obj = json_object_array_get_idx(Jarray, 0);
-	servconfobj = json_object_object_get(obj, "Service Conf");
-	accountobj = json_object_object_get(obj, "Account");
-	sambaobj = json_object_object_get(obj, "Samba");
+	// servconfobj = json_object_object_get(obj, "Service Conf");
+	// accountobj = json_object_object_get(obj, "Account");
+	// sambaobj = json_object_object_get(obj, "Samba");
 	usbinfoobj = json_object_object_get(obj, "Usb Info");
-	len_accountobj = json_object_array_length(accountobj);
-	len_sambaobj = json_object_array_length(sambaobj);
+	// len_accountobj = json_object_array_length(accountobj);
+	// len_sambaobj = json_object_array_length(sambaobj);
 	len_usbinfoobj = json_object_array_length(usbinfoobj);
 	
 	if(len_usbinfoobj != 0){
diff -urN libzcfg_fe_dal-1.0.org/zcfg_fe_dal.c libzcfg_fe_dal-1.0/zcfg_fe_dal.c
--- libzcfg_fe_dal-1.0.org/zcfg_fe_dal.c	2019-05-08 11:27:51.259258400 +0800
+++ libzcfg_fe_dal-1.0/zcfg_fe_dal.c	2019-05-08 16:32:41.362329112 +0800
@@ -1804,14 +1804,14 @@
 	struct json_object *voipObj = NULL, *obj = NULL;
 	char voipline[64] = {0}, sipnumber[64] = {0}, sip[8] = {0}, boundList[64] = {0};
 	int count = 1;
-	const char *directoryNumber = NULL;
+	// const char *directoryNumber = NULL;
 
 	*VoipMultiObj = json_object_new_array();
 	
 	IID_INIT(voipIid);
 	while(zcfgFeObjJsonGetNextWithoutUpdate(RDM_OID_VOICE_LINE, &voipIid, &voipObj) == ZCFG_SUCCESS){
 		obj = json_object_new_object();
-		directoryNumber = json_object_get_string(json_object_object_get(voipObj, "DirectoryNumber"));
+		// directoryNumber = json_object_get_string(json_object_object_get(voipObj, "DirectoryNumber"));
 		sprintf(sip,"SIP%d",count);
 		//sprintf(voipline,"VoiceProfile.%d.Line.%d",voipIid.idx[1],voipIid.idx[2]);
 		//sprintf(boundList,"Device.Services.VoiceService.1.VoiceProfile.%d.Line.%d",voipIid.idx[1],voipIid.idx[2]);
