Index: libzcfg_fe_dal-1.0/js/zcfg_fe_dal_loginprivilege.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/js/zcfg_fe_dal_loginprivilege.c	2019-04-26 09:40:19.387312702 +0800
+++ libzcfg_fe_dal-1.0/js/zcfg_fe_dal_loginprivilege.c	2019-04-26 09:40:36.001194015 +0800
@@ -178,6 +178,9 @@
 		json_object_object_add(paramJobj, "Username", JSON_OBJ_COPY(json_object_object_get(loginPrivilegeObj, "Username")));
 		json_object_object_add(paramJobj, "Password", JSON_OBJ_COPY(json_object_object_get(loginPrivilegeObj, "Password")));
 		json_object_object_add(paramJobj, "EnableQuickStart", JSON_OBJ_COPY(json_object_object_get(loginPrivilegeObj, "EnableQuickStart")));
+
+		json_object_object_add(paramJobj, "Privilege", JSON_OBJ_COPY(json_object_object_get(loginPrivilegeObj, "Privilege")));
+
 		json_object_array_add(Jarray, paramJobj);
 		zcfgFeJsonObjFree(loginPrivilegeObj);
 	}

