Index: libzcfg_fe_dal-1.0/Makefile
===================================================================
--- libzcfg_fe_dal-1.0.orig/Makefile	2019-07-04 16:10:01.918219502 +0800
+++ libzcfg_fe_dal-1.0/Makefile	2019-07-04 16:12:05.696833818 +0800
@@ -177,6 +177,11 @@
 CFLAGS += -DZYXEL_REMOTE_ACCESS_PRIVILEGE
 endif
 
+#ZYXEL_REMOMGMT_SEPARATE_LAN_WLAN_PRIVILEGE
+ifeq ($(CONFIG_ZYXEL_REMOMGMT_SEPARATE_LAN_WLAN_PRIVILEGE),y)
+CFLAGS += -DZYXEL_REMOMGMT_SEPARATE_LAN_WLAN_PRIVILEGE
+endif
+
 #ZYXEL_REMOTE_ACCOUNT_ACCESS_PRIVILEGE
 ifeq ($(CONFIG_ZYXEL_REMOTE_ACCOUNT_ACCESS_PRIVILEGE),y)
 CFLAGS += -DZYXEL_REMOTE_ACCOUNT_ACCESS_PRIVILEGE
Index: libzcfg_fe_dal-1.0/maintenance/mgmt/zcfg_fe_dal_mgmt_srv.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/maintenance/mgmt/zcfg_fe_dal_mgmt_srv.c	2019-07-04 16:10:01.558219513 +0800
+++ libzcfg_fe_dal-1.0/maintenance/mgmt/zcfg_fe_dal_mgmt_srv.c	2019-07-04 16:12:39.921177752 +0800
@@ -17,7 +17,11 @@
 	{"Name", 					dalType_string,	0,	0,		NULL, NULL, 				dalcmd_Forbid},
 	{"WAN_Interface", 			dalType_string,	0,	0,		NULL, "Any_WAN|Multi_WAN", 				0},
 	{"BoundInterfaceList", 		dalType_MultiRtWanWWan,		0,	0,		NULL, NULL, 				0},
-	{"Mode", 					dalType_string,	0,	0,		NULL, "LAN_ONLY|WAN_ONLY|LAN_WAN|Trust_Dm|LAN_TstDm|None", 	0},
+#ifdef ZYXEL_REMOMGMT_SEPARATE_LAN_WLAN_PRIVILEGE	
+	{"Mode",					dalType_string, 0,	0,		NULL, "LAN_WL|LAN_WL_WAN|LAN_WL_TstDm|WL_WAN|WL_Trust_Dm|WL_ONLY|LAN_ONLY|WAN_ONLY|LAN_WAN|Trust_Dm|LAN_TstDm|None",	0},
+#else	
+	{"Mode",					dalType_string, 0,	0,		NULL, "LAN_ONLY|WAN_ONLY|LAN_WAN|Trust_Dm|LAN_TstDm|None",	0},
+#endif
 	{"Port", 	 				dalType_int,	1,	65535,	NULL, NULL, 				0},
 	{"Index", 	 				dalType_int,	0,	0,		NULL, NULL, 				0},
 	{"Protocol", 				dalType_string,	0,	0,		NULL, NULL, 				dalcmd_Forbid},
