Index: libzcfg_fe_dal-1.0/js/zcfg_fe_dal_networkmap.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/js/zcfg_fe_dal_networkmap.c	2019-07-18 16:11:15.000000000 +0800
+++ libzcfg_fe_dal-1.0/js/zcfg_fe_dal_networkmap.c	2019-07-18 16:11:15.000000000 +0800
@@ -657,6 +657,7 @@
     char WanRate_RX[8] = {0};
     char WanRate_TX[8] = {0};
     const char *optStatus = NULL;
+    int X_ZYXEL_InterfaceUpTime = 0;
 
 	IID_INIT(iid);
 	if((ret = zcfgFeObjJsonGet(RDM_OID_DEV_INFO, &iid, &devInfoObj)) != ZCFG_SUCCESS)
@@ -686,6 +687,9 @@
 	IID_INIT(iid);
 	while(zcfgFeObjJsonGetNext(RDM_OID_ETH_IFACE, &iid, &ethIfaceObj) == ZCFG_SUCCESS) {
 		json_object_array_add(ethIfaceJarray, ethIfaceObj);
+		if( true == json_object_get_boolean(json_object_object_get(ethIfaceObj,"Upstream"))) {
+			X_ZYXEL_InterfaceUpTime = json_object_get_int(json_object_object_get(ethIfaceObj,"X_ZYXEL_UpTime"));
+		}
 	}
 
 	dslChannelJarray = json_object_new_array();
@@ -829,6 +833,7 @@
     Jadds (statusJobj, "RdpaWanType", RdpaWanType);
     Jadds (statusJobj, "WanRate_RX", WanRate_RX);
     Jadds (statusJobj, "WanRate_TX", WanRate_TX);
+    Jaddi (statusJobj, "X_ZYXEL_InterfaceUpTime", X_ZYXEL_InterfaceUpTime);
 	
 	return ZCFG_SUCCESS;
 
