Index: libzcfg_fe_dal-1.0/maintenance/mgmt/zcfg_fe_dal_mgmt_srv.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/maintenance/mgmt/zcfg_fe_dal_mgmt_srv.c	2019-07-19 14:44:50.331477259 +0800
+++ libzcfg_fe_dal-1.0/maintenance/mgmt/zcfg_fe_dal_mgmt_srv.c	2019-07-19 14:47:11.411383180 +0800
@@ -18,7 +18,7 @@
 	{"WAN_Interface", 			dalType_string,	0,	0,		NULL, "Any_WAN|Multi_WAN", 				0},
 	{"BoundInterfaceList", 		dalType_MultiRtWanWWan,		0,	0,		NULL, NULL, 				0},
 #ifdef ZYXEL_REMOMGMT_SEPARATE_LAN_WLAN_PRIVILEGE	
-	{"Mode",					dalType_string, 0,	0,		NULL, "LAN_WL|LAN_WL_WAN|LAN_WL_TstDm|WL_WAN|WL_Trust_Dm|WL_ONLY|LAN_ONLY|WAN_ONLY|LAN_WAN|Trust_Dm|LAN_TstDm|None",	0},
+	{"Mode",					dalType_string, 0,	0,		NULL, "LAN_withoutWL|LAN_withoutWL_WAN|LAN_withoutWL_TstDm|WL_WAN|WL_Trust_Dm|WL_ONLY|LAN_ONLY|WAN_ONLY|LAN_WAN|Trust_Dm|LAN_TstDm|None",	0},
 #else	
 	{"Mode",					dalType_string, 0,	0,		NULL, "LAN_ONLY|WAN_ONLY|LAN_WAN|Trust_Dm|LAN_TstDm|None",	0},
 #endif
