Index: libzcfg_fe_dal-1.0/network/broadband/zcfg_fe_dal_broadband.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/broadband/zcfg_fe_dal_broadband.c	2019-09-05 20:36:54.850625103 +0800
+++ libzcfg_fe_dal-1.0/network/broadband/zcfg_fe_dal_broadband.c	2019-09-05 20:43:25.850613823 +0800
@@ -2383,6 +2383,8 @@
 		}
 	}
 
+	memset(newInterface, 0, sizeof(newInterface));
+
 	Interface = json_object_get_string(json_object_object_get(routerObj, "X_ZYXEL_ActiveV6DefaultGateway"));
 	if(CurrV6GwEnable && !v6GwEnable){ ////enable to disable
 		if(genNewInterface(",", newInterface, Interface, ipIfacePath, false)){
