Index: libzcfg_fe_dal-1.0/security/schedulerRule/zcfg_fe_dal_schedulerRule.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/security/schedulerRule/zcfg_fe_dal_schedulerRule.c	2019-09-06 22:03:44.447982245 +0800
+++ libzcfg_fe_dal-1.0/security/schedulerRule/zcfg_fe_dal_schedulerRule.c	2019-09-06 22:08:37.691070000 +0800
@@ -25,6 +25,12 @@
 	{NULL,				0,					0,	0,	NULL,	NULL,	0}
 };
 
+#ifdef ZYXEL_HOME_CYBER_SECURITY
+static int hcbSecur = true;
+#else
+static int hcbSecur = false;
+#endif
+
 int schedule_days_string_to_int(char *str){
 	int ret = 0;
 	if(strcasestr(str,"MON"))
@@ -233,11 +239,8 @@
 	char path[32] = {0};
 	struct json_object *retJobj = NULL;
 
-#ifdef ZYXEL_HOME_CYBER_SECURITY
-#if 0
-#endif
 	// Name validation
-	if(!ISSET("DontCheckDupName") && json_object_object_get(Jobj, "Name")){
+	if(!hcbSecur && !ISSET("DontCheckDupName") && json_object_object_get(Jobj, "Name")){
 		name = json_object_get_string(json_object_object_get(Jobj, "Name"));
 		ret = dalcmdParamCheck(NULL,name, RDM_OID_SCHEDULE, "Name", "Name", msg);
 		if(ret != ZCFG_SUCCESS){
@@ -246,9 +249,8 @@
 			return ret; 		
 		}
 	}
-#ifdef ZYXEL_HOME_CYBER_SECURITY	
-#endif
-#endif
+
+
 	// Time autofill
 	if(!json_object_object_get(Jobj, "TimeStartHour") || !json_object_object_get(Jobj, "TimeStartMin")){
 		json_object_object_add(Jobj, "TimeStartHour", json_object_new_int(0));
