Index: busybox-1_20_1/loginutils/login.c
===================================================================
--- busybox-1_20_1.orig/loginutils/login.c	2018-09-03 22:20:24.831396390 +0800
+++ busybox-1_20_1/loginutils/login.c	2018-09-03 22:30:26.072598027 +0800
@@ -796,7 +796,7 @@
 		if (correct_password(pw)){
 #ifdef ZCFG_SUPPORT
 			reset_authorize_fail_cnt(cur_user_idx);
-#endif
+
 			if (authGpPrivilege){
 				objIndex_t logAccountObjIid;
 				rdm_ZyLogCfgGpAccount_t *accountObj = NULL;
@@ -926,7 +926,7 @@
 
 #endif
 			}
-
+#endif //#ifdef ZCFG_SUPPORT
 			break;
 		}
 #endif /* ENABLE_PAM */
Index: busybox-1_20_1/modutils/depmod.c
===================================================================
--- busybox-1_20_1.orig/modutils/depmod.c	2018-09-03 22:21:08.863397101 +0800
+++ busybox-1_20_1/modutils/depmod.c	2018-09-03 22:33:44.606694592 +0800
@@ -14,7 +14,7 @@
 #include "modutils.h"
 #include <sys/utsname.h> /* uname() */
 
-#if 1 /* <JoannaSu>, <20180420>, <modify buffer for filepath to avoid segfault> */
+#ifdef ZCFG_SUPPORT /* <JoannaSu>, <20180420>, <modify buffer for filepath to avoid segfault> */
 #include "zos_api.h"
 #include "zlog_api.h"
 #define   ZYXEL_TMP_BASE_PATH "/var/"
@@ -129,7 +129,7 @@
 
 static void xfreopen_write(const char *file, FILE *f)
 {
-#if 1 //due to depmod can't generate file at read-only file system, so modify the path that generate file
+#ifdef ZCFG_SUPPORT //due to depmod can't generate file at read-only file system, so modify the path that generate file
     #if 1 /* <JoannaSu>, <20180420>, <modify buffer for filepath to avoid segfault> */
     char *modulePath = NULL;
 
