Index: libnetfilter_queue-1.0.2/include/libnetfilter_queue/linux_nfnetlink_queue.h
===================================================================
--- libnetfilter_queue-1.0.2.orig/include/libnetfilter_queue/linux_nfnetlink_queue.h	2012-08-06 20:47:29.548983945 +0800
+++ libnetfilter_queue-1.0.2/include/libnetfilter_queue/linux_nfnetlink_queue.h	2019-06-04 15:56:16.333390743 +0800
@@ -38,6 +38,9 @@
 	NFQA_PACKET_HDR,
 	NFQA_VERDICT_HDR,		/* nfqnl_msg_verdict_hrd */
 	NFQA_MARK,			/* u_int32_t nfmark */
+#ifdef ZYXEL_EXT_MARK
+	NFQA_ZEXTMARK,			/* u_int32_t nfmark */
+#endif
 	NFQA_TIMESTAMP,			/* nfqnl_msg_packet_timestamp */
 	NFQA_IFINDEX_INDEV,		/* u_int32_t ifindex */
 	NFQA_IFINDEX_OUTDEV,		/* u_int32_t ifindex */
Index: libnetfilter_queue-1.0.2/include/linux/netfilter/nfnetlink_queue.h
===================================================================
--- libnetfilter_queue-1.0.2.orig/include/linux/netfilter/nfnetlink_queue.h	2012-08-06 20:47:40.420983359 +0800
+++ libnetfilter_queue-1.0.2/include/linux/netfilter/nfnetlink_queue.h	2019-06-04 15:56:19.057414565 +0800
@@ -35,6 +35,9 @@
 	NFQA_PACKET_HDR,
 	NFQA_VERDICT_HDR,		/* nfqnl_msg_verdict_hrd */
 	NFQA_MARK,			/* __u32 nfmark */
+#ifdef ZYXEL_EXT_MARK
+	NFQA_ZEXTMARK,			/* u_int32_t nfmark */
+#endif
 	NFQA_TIMESTAMP,			/* nfqnl_msg_packet_timestamp */
 	NFQA_IFINDEX_INDEV,		/* __u32 ifindex */
 	NFQA_IFINDEX_OUTDEV,		/* __u32 ifindex */
