diff -urN clinkc.org/src/cybergarage/http/chttp_val_list.c clinkc/src/cybergarage/http/chttp_val_list.c
--- clinkc.org/src/cybergarage/http/chttp_val_list.c	2019-05-06 14:57:44.892868300 +0800
+++ clinkc/src/cybergarage/http/chttp_val_list.c	2019-05-07 17:20:40.286887350 +0800
@@ -136,5 +136,6 @@
 		value = cg_http_val_getvalue(val);
 		printf("CgHttpVal: %s = %s\n", name, value);
 	}
+	return NULL;
 }
 #endif
\ No newline at end of file
