Index: libzcfg_fe_tr98-1.0/zyExt.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zyExt.c	2018-12-26 19:35:58.324177236 +0800
+++ libzcfg_fe_tr98-1.0/zyExt.c	2018-12-26 19:41:10.406383300 +0800
@@ -3118,6 +3118,7 @@
 
 	printf("%s : Enter\n", __FUNCTION__);
 
+	return ZCFG_REQUEST_REJECT;
 	IID_INIT(objIid);
 	ret = zcfgFeObjStructAdd(RDM_OID_REMO_SRV, &objIid, NULL);
 	*idx = objIid.idx[0];
@@ -3131,6 +3132,8 @@
 
 	printf("%s : Enter\n", __FUNCTION__);
 
+	return ZCFG_DELETE_REJECT;
+
 	IID_INIT(objIid);
 	objIid.level = 1;
 	sscanf(tr98FullPathName, "InternetGatewayDevice.X_ZYXEL_EXT.RemoteManagement.Service.%hhu", &objIid.idx[0]);
@@ -3393,6 +3396,21 @@
 }
 #endif
 
+
+zcfgRet_t zyExtSpRemoteMgmtSrvObjAdd(char *tr98FullPathName, int *idx)
+{
+	return ZCFG_REQUEST_REJECT;
+
+
+}
+
+zcfgRet_t zyExtSpRemoteMgmtSrvObjDelete(char *tr98FullPathName)
+{
+
+	return ZCFG_DELETE_REJECT;
+
+}
+
 /*
  *   TR98 Object Name : InternetGatewayDevice.X_ZYXEL_EXT.RemoteManagement.SpService.i
  *
Index: libzcfg_fe_tr98-1.0/zcfg_fe_tr98_handle.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zcfg_fe_tr98_handle.h	2018-12-26 19:35:58.320177236 +0800
+++ libzcfg_fe_tr98-1.0/zcfg_fe_tr98_handle.h	2018-12-26 19:42:31.681383300 +0800
@@ -662,7 +662,7 @@
 {TR98_REMO_MGMT,				0,				NULL,					NULL,					 NULL,					 NULL,				NULL,			NULL},
 #endif
 {TR98_REMO_SRV, 			    ATTR_INDEX_CREA,para_RemoSrv,		    zyExtRemoteMgmtSrvObjGet,		 zyExtRemoteMgmtSrvObjSet,		 zyExtRemoteMgmtSrvObjAdd, zyExtRemoteMgmtSrvObjDel, zyExtRemoteMgmtSrvObjNotify, zyExtRemoteMgmtSrvObjAttrGet, zyExtRemoteMgmtSrvObjAttrSet},
-{TR98_SP_REMO_SRV, 			    ATTR_INDEX_CREA,para_SpRemoSrv,		    zyExtSpRemoteMgmtSrvObjGet,		 zyExtSpRemoteMgmtSrvObjSet,		 NULL, NULL, NULL, NULL, NULL},
+{TR98_SP_REMO_SRV, 			    ATTR_INDEX_CREA,para_SpRemoSrv,		    zyExtSpRemoteMgmtSrvObjGet,		 zyExtSpRemoteMgmtSrvObjSet,	zyExtSpRemoteMgmtSrvObjAdd, zyExtSpRemoteMgmtSrvObjDelete, NULL, NULL, NULL},
 {TR98_REMO_TRUSTDOMAIN,         ATTR_INDEX_CREA,para_RemoTrustDomain,   zyExtRemoteMgmtTrustDomainObjGet,                    zyExtRemoteMgmtTrustDomainObjSet,                   zyExtRemoteMgmtTrustDomainObjAdd,            zyExtRemoteMgmtTrustDomainObjDel,            NULL},
 {TR98_SP_TRUSTDOMAIN,         ATTR_INDEX_CREA,para_SpTrustDomain,   zyExtSpRemoteMgmtTrustDomainObjGet,                    zyExtSpRemoteMgmtTrustDomainObjSet,                   zyExtSpRemoteMgmtTrustDomainObjAdd,            zyExtSpRemoteMgmtTrustDomainObjDel,            NULL},
 {TR98_IEEE8021AG,               0,              para_8021ag,	        zyExt8021agObjGet, 	     zyExt8021agObjSet,		NULL,             NULL,            zyExt8021agObjNotify, zyExt8021agObjAttrGet, zyExt8021agObjAttrSet},
Index: libzcfg_fe_tr98-1.0/zyExt.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zyExt.h	2018-12-26 19:35:58.324177236 +0800
+++ libzcfg_fe_tr98-1.0/zyExt.h	2018-12-26 19:41:38.429383300 +0800
@@ -193,6 +193,8 @@
 extern zcfgRet_t zyExtRemoteMgmtSrvObjAttrSet(char *, int, char *, int , struct json_object *);
 
 /*InternetGatewayDevice.X_ZYXEL_EXT.RemoteManagement.SpService.i*/
+extern zcfgRet_t zyExtSpRemoteMgmtSrvObjAdd(char *tr98FullPathName, int *idx);
+extern zcfgRet_t zyExtSpRemoteMgmtSrvObjDelete(char *tr98FullPathName);
 extern zcfgRet_t zyExtSpRemoteMgmtSrvObjGet(char *tr98FullPathName, int handler, struct json_object **tr98Jobj, bool updateFlag);
 extern zcfgRet_t zyExtSpRemoteMgmtSrvObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj);
 
