Index: libzcfg_fe_tr98-1.0/voice.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/voice.h	2018-06-04 20:33:10.696234794 -0700
+++ libzcfg_fe_tr98-1.0/voice.h	2018-06-07 03:12:04.088122140 -0700
@@ -124,6 +124,8 @@
 	//{ "X_ZYXEL_SPNum", PARAMETER_ATTR_WRITE, 0, json_type_uint32},
 	//{ "X_ZYXEL_V18_Support", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
 	{ "X_ZYXEL_VoipIOPFlags", PARAMETER_ATTR_WRITE, 0, json_type_uint32},
+	{ "X_ZYXEL_VoipSigIOPFlags", PARAMETER_ATTR_WRITE, 0, json_type_uint32},
+	{ "X_ZYXEL_VoipMediaIOPFlags", PARAMETER_ATTR_WRITE, 0, json_type_uint32},
 	//{ "X_ZYXEL_P_AccessNetworkInfo", PARAMETER_ATTR_WRITE, 128, json_type_string},
 	//{ "X_ZYXEL_BoundIfName", PARAMETER_ATTR_WRITE, 33, json_type_string},
 	//{ "X_ZYXEL_BoundIfList", PARAMETER_ATTR_WRITE, 257, json_type_string},
