Index: libzcfg_fe_tr98-1.0/lanDev.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/lanDev.c	2019-04-16 20:31:39.990792763 +0800
+++ libzcfg_fe_tr98-1.0/lanDev.c	2019-04-16 20:33:57.810408346 +0800
@@ -969,7 +969,6 @@
 		}
 		else if(!strcmp(paramList->name, "InterfaceType")) {
 			paramValue = json_object_object_get(hostJobj, "X_ZYXEL_ConnectionType");
-#ifdef ZYXEL_WIND_ITALY_CUSTOMIZATION
 			const char *conntection = json_object_get_string(json_object_object_get(hostJobj, "X_ZYXEL_ConnectionType"));
 			if (strstr(conntection, "Wi-Fi") != NULL){
 				json_object_object_add(*tr98Jobj, paramList->name, json_object_new_string("802.11"));
@@ -977,9 +976,6 @@
 			else{
 				json_object_object_add(*tr98Jobj, paramList->name, JSON_OBJ_COPY(paramValue));
 			}
-#else
-			json_object_object_add(*tr98Jobj, paramList->name, JSON_OBJ_COPY(paramValue));
-#endif
 			paramList++;
 			continue;
 		}
