Index: libzcfg_fe_tr98-1.0/voice.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/voice.h	2019-06-04 07:31:51.917137003 +0800
+++ libzcfg_fe_tr98-1.0/voice.h	2019-06-05 04:58:25.169986110 +0800
@@ -88,6 +88,7 @@
 	{ "DTMFMethod", PARAMETER_ATTR_WRITE, 33, json_type_string},
 	//{ "X_ZYXEL_HookFlashMethod", PARAMETER_ATTR_WRITE, 33, json_type_string},
 	{ "Region", PARAMETER_ATTR_WRITE, 3, json_type_string},
+	{ "X_ZYXEL_Region_IntlPhoneNumCallingCode", PARAMETER_ATTR_WRITE, 17, json_type_string}, //Michael.20190604.001: Add to support a Manually-Configurable/Customized International PhoneNumber Calling Code parameter for the PlusSign(+)-beginning (International) PhoneNumber CLID Display Process Logic. Related Tickets: [#190400265][#105337][VMG8825-B60B][Generic][DE], [#190401061][#105665][VMG8825-T50K][T-Mobile-NL].
 #ifdef USE_OLD_VOIP_PREFIX
 	//{ "X_ZyXEL_COM_Countrycode", PARAMETER_ATTR_WRITE, 0, json_type_uint32},
 #else
