Index: libzcfg_fe_tr98-1.0/wanDev.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/wanDev.c	2019-09-25 20:07:01.950744545 +0800
+++ libzcfg_fe_tr98-1.0/wanDev.c	2019-09-25 20:11:05.092665400 +0800
@@ -376,6 +376,12 @@
 	objIndex_t objIid;
 	objIndex_t ipAddrIid;
 	bool enableForInternet = false;
+	uint64_t BytesSent = 0;
+	uint64_t BytesReceived = 0;
+	uint64_t PacketsSent = 0;
+	uint64_t PacketsReceived = 0;
+
+
 
 	zcfgLog(ZCFG_LOG_INFO, "%s : Enter\n", __FUNCTION__);
 
@@ -524,17 +530,12 @@
 	json_object_object_add(*tr98Jobj, "PhysicalLinkStatus", json_object_new_string("Unavailable"));
 	json_object_object_add(*tr98Jobj, "Layer1UpstreamMaxBitRate", json_object_new_int(0));
 	json_object_object_add(*tr98Jobj, "Layer1DownstreamMaxBitRate", json_object_new_int(0));
-	
 	IID_INIT(objIid);
 	objIid.level = 1;
 
 	if(strstr(tr181Obj, "DSL") != NULL) {
 		rdm_DslChannel_t *dslChannel = NULL;
 		rdm_DslChannelSt_t *dslChannelSt = NULL;
-		uint64_t BytesSent = 0;
-		uint64_t BytesReceived = 0;
-		uint64_t PacketsSent = 0;
-		uint64_t PacketsReceived = 0;
 		char highLayer[257] = {0};
 
 #ifdef ZYXEL_WIND_ITALY_CUSTOMIZATION
@@ -654,11 +655,13 @@
 	else if(strstr(tr181Obj, "USB") != NULL) {
 		rdm_UsbIntf_t *usbIface = NULL;
 		rdm_UsbIntfStat_t *usbIfaceSt = NULL;
+		char usbIfaceStatus[20] = {0};
 		
 		json_object_object_add(*tr98Jobj, "WANAccessType", json_object_new_string("3G"));
 		sscanf(tr181Obj, "USB.Interface.%hhu", &objIid.idx[0]);
 
 		if(feObjStructGet(RDM_OID_USB_INTF, &objIid, (void **)&usbIface, updateFlag) == ZCFG_SUCCESS) {
+			strcpy(usbIfaceStatus, usbIface->Status);
 			if(!strcmp(usbIface->Status, "Up") || !strcmp(usbIface->Status, "Down")) {
 				json_object_object_add(*tr98Jobj, "PhysicalLinkStatus", json_object_new_string(usbIface->Status));
 			}
@@ -666,13 +669,20 @@
 			zcfgFeObjStructFree(usbIface);
 		}
 
-		if(feObjStructGet(RDM_OID_USB_INTF_STAT, &objIid, (void **)&usbIfaceSt, updateFlag) == ZCFG_SUCCESS) {
-			json_object_object_add(*tr98Jobj, "TotalBytesSent", json_object_new_int(usbIfaceSt->BytesSent));
-			json_object_object_add(*tr98Jobj, "TotalBytesReceived", json_object_new_int(usbIfaceSt->BytesReceived));
-			json_object_object_add(*tr98Jobj, "TotalPacketsSent", json_object_new_int(usbIfaceSt->PacketsSent));
-			json_object_object_add(*tr98Jobj, "TotalPacketsReceived", json_object_new_int(usbIfaceSt->PacketsReceived));
-			zcfgFeObjStructFree(usbIfaceSt);
+		if(!strcmp(usbIfaceStatus, "Up")){
+			if(feObjStructGet(RDM_OID_USB_INTF_STAT, &objIid, (void **)&usbIfaceSt, updateFlag) == ZCFG_SUCCESS) {
+				BytesSent = usbIfaceSt->BytesSent;
+				BytesReceived = usbIfaceSt->BytesReceived;
+				PacketsSent = usbIfaceSt->PacketsSent;
+				PacketsReceived = usbIfaceSt->PacketsReceived;
+				zcfgFeObjStructFree(usbIfaceSt);
+			}
 		}
+		json_object_object_add(*tr98Jobj, "TotalBytesSent", json_object_new_int(BytesSent));
+		json_object_object_add(*tr98Jobj, "TotalBytesReceived", json_object_new_int(BytesReceived));
+		json_object_object_add(*tr98Jobj, "TotalPacketsSent", json_object_new_int(PacketsSent));
+		json_object_object_add(*tr98Jobj, "TotalPacketsReceived", json_object_new_int(PacketsReceived));
+
 	}
 	else {
 		zcfgLog(ZCFG_LOG_ERR, "Unknown Object %s\n", tr181Obj);
