Index: libzcfg_fe_tr98-1.0/lanDev.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/lanDev.c	2019-11-04 21:51:51.806920634 +0800
+++ libzcfg_fe_tr98-1.0/lanDev.c	2019-11-04 22:04:20.831898168 +0800
@@ -8560,7 +8560,7 @@
 	const char *prefixPtr = NULL;
 	uint8_t prefixInstance = 0;
 	char *token = NULL, *token_tmp = NULL;
-	char tr98ObjName[TR98_MAX_OBJ_NAME_LENGTH];
+	char tr98ObjName[TR98_MAX_OBJ_NAME_LENGTH] = {0};
 
 	zcfgLog(ZCFG_LOG_INFO, "%s : Enter\n", __FUNCTION__);
 	strcpy(tr98ConnObjName, tr98FullPathName);
