Index: dropbear-2018.76/Makefile.in
===================================================================
--- dropbear-2018.76.orig/Makefile.in	2019-06-20 15:41:57.086399472 +0800
+++ dropbear-2018.76/Makefile.in	2019-06-20 15:46:14.627053853 +0800
@@ -105,6 +105,10 @@
 	CFLAGS += -DZYXEL_REMOTE_ACCESS_PRIVILEGE
 endif
 
+ifeq ($(CONFIG_ZYXEL_REMOTE_ACCOUNT_ACCESS_PRIVILEGE),y)
+	CFLAGS += -DZYXEL_REMOTE_ACCOUNT_ACCESS_PRIVILEGE
+endif
+
 ifeq ($(CONFIG_ZYXEL_REMOMGMT_ONLY_HTTPS_ALLOWS_FOR_WLAN_CONNECT),y)
 	CFLAGS += -DZYXEL_REMOMGMT_ONLY_HTTPS_ALLOWS_FOR_WLAN_CONNECT
 endif
Index: dropbear-2018.76/svr-auth.c
===================================================================
--- dropbear-2018.76.orig/svr-auth.c	2019-06-20 15:41:57.126399471 +0800
+++ dropbear-2018.76/svr-auth.c	2019-06-20 15:45:22.010678582 +0800
@@ -465,20 +465,29 @@
 					//syslog(LOG_INFO, "Account privilege:  %s Priviege = %s !!!!", username, userPriviege);
 				}
 #endif				
+#ifndef ZYXEL_REMOTE_ACCOUNT_ACCESS_PRIVILEGE
 				free(accountObj);
+#endif
 				break ; 
 			}
 			free(accountObj);
 		}
 		
 		if(zcfgFeObjStructGet(RDM_OID_ZY_LOG_CFG_GP, &logGpObjIid, (void **) &logGpObj) == ZCFG_SUCCESS) {
-			if (strstr(logGpObj->GP_Privilege, SSH) == NULL){
+			if (strstr(logGpObj->GP_Privilege, SSH) == NULL
+#ifdef ZYXEL_REMOTE_ACCOUNT_ACCESS_PRIVILEGE				
+				|| strstr(accountObj->Privilege, SSH)== NULL
+#endif
+			    ){
 				syslog(LOG_INFO, "Account:'%s' SSH permission denied.", username);
 				TRACE(("leave check username:'%s' SSH permission denied.", username))
 				free(logGpObj);
 				send_msg_userauth_failure(0, 1);
 				return DROPBEAR_FAILURE;
 			}
+#ifdef ZYXEL_REMOTE_ACCOUNT_ACCESS_PRIVILEGE
+			free(accountObj);
+#endif
 			free(logGpObj);
 		}
 #ifdef ZYXEL_REMOMGMT_ONLY_HTTPS_ALLOWS_FOR_WLAN_CONNECT
