Index: p910nd-0.97/p910nd.c
===================================================================
--- p910nd-0.97.orig/p910nd.c	2014-01-16 17:04:51.536786053 +0800
+++ p910nd-0.97/p910nd.c	2019-07-24 10:17:40.581831398 +0800
@@ -253,6 +253,7 @@
 	char lockname[sizeof(LOCKFILE)];
 	struct flock lplock;
 
+	system("mkdir /var/lock/subsys/");
 	(void)snprintf(lockname, sizeof(lockname), LOCKFILE, lpnumber);
 	if ((lockfd = open(lockname, O_CREAT | O_RDWR, 0666)) < 0) {
 		dolog(LOGOPTS, "%s: %m\n", lockname);
