Index: atftp-0.7.1/argz.h
===================================================================
--- atftp-0.7.1.orig/argz.h	2011-03-11 14:51:24.000000000 +0800
+++ atftp-0.7.1/argz.h	2018-11-01 13:51:25.178883244 +0800
@@ -177,7 +177,7 @@
 extern char *argz_next (__const char *__restrict __argz, size_t __argz_len,
 			__const char *__restrict __entry) __THROW;
 
-#ifdef __USE_EXTERN_INLINES
+#if 0
 extern inline char *
 __argz_next (__const char *__argz, size_t __argz_len,
 	     __const char *__entry)
