Index: busybox-1_20_1/networking/ntpd.c
===================================================================
--- busybox-1_20_1.orig/networking/ntpd.c	2012-05-28 08:48:55.000000000 +0800
+++ busybox-1_20_1/networking/ntpd.c	2018-12-27 19:01:12.460077267 +0800
@@ -1835,6 +1835,13 @@
 		goto bail;
 	}
 
+	/* Respond only to client and symmetric active packets */
+	if ((msg.m_status & MODE_MASK) != MODE_CLIENT
+	&& (msg.m_status & MODE_MASK) != MODE_SYM_ACT
+	) {
+		goto bail;
+	}
+
 	query_status = msg.m_status;
 	query_xmttime = msg.m_xmttime;
 
Index: busybox-1_20_1/testsuite/mdev.tests
===================================================================
--- busybox-1_20_1.orig/testsuite/mdev.tests	2012-05-28 08:48:55.000000000 +0800
+++ busybox-1_20_1/testsuite/mdev.tests	2018-12-27 19:08:08.878547354 +0800
@@ -11,7 +11,7 @@
 # cut: remove size+date
 FILTER_LS2="grep -v '^total ' | sed -e 's/,  */,/g' -e 's/  */ /g' | cut -d' ' -f 1-4,9-"
 
-# testing "test name" "options" "expected result" "file input" "stdin"
+# testing "test name" "commands" "expected result" "file input" "stdin"
 
 rm -rf mdev.testdir
 mkdir mdev.testdir
@@ -127,6 +127,26 @@
 SKIP=
 
 # continuing to use directory structure from prev test
+rm -rf mdev.testdir/dev/*
+echo "sda 0:0 444 =disk/sd/a" >mdev.testdir/etc/mdev.conf
+optional STATIC FEATURE_MDEV_CONF FEATURE_MDEV_RENAME FEATURE_LS_RECURSIVE FEATURE_LS_TIMESTAMPS FEATURE_LS_USERNAME
+testing "mdev move rule '=bar/baz/fname'" \
+   "env - PATH=$PATH ACTION=add DEVPATH=/block/sda chroot mdev.testdir /mdev 2>&1;
+   ls -lnR mdev.testdir/dev | $FILTER_LS2" \
+"\
+mdev.testdir/dev:
+drwxr-xr-x 3 0 0 disk
+
+mdev.testdir/dev/disk:
+drwxr-xr-x 2 0 0 sd
+
+mdev.testdir/dev/disk/sd:
+br--r--r-- 1 0 0 a
+" \
+   "" ""
+SKIP=
+
+# continuing to use directory structure from prev test
 rm -rf mdev.testdir/dev/*
 # here we complicate things by having non-matching group 1 and using %0
 echo "s([0-9])*d([a-z]+) 0:0 644 >sd/%2_%0" >mdev.testdir/etc/mdev.conf
