Index: ebtables-v2.0.10-4/extensions/ebt_ip.c
===================================================================
--- ebtables-v2.0.10-4.orig/extensions/ebt_ip.c	2018-11-07 20:50:33.098545956 +0800
+++ ebtables-v2.0.10-4/extensions/ebt_ip.c	2018-11-07 21:04:04.905423654 +0800
@@ -342,7 +342,8 @@
 static void
 parse_dhcp_opt125(struct cfgopt *cfg125, char *EnterpriseNum, char *VendorClassData, char *NextArg)
 {
-        int i=0, data_len=0;
+        int i=0;
+        size_t data_len=0;
         char cfg[255],data[8];
 
         if(NextArg!=NULL && *NextArg != '-')
@@ -361,7 +362,7 @@
         memset(cfg, 0, sizeof(cfg));
         memset(data, 0, sizeof(data));
 
-        sprintf(cfg, "%s%02x%s", EnterpriseNum, strlen(VendorClassData)/2,VendorClassData);
+        sprintf(cfg, "%s%02zx%s", EnterpriseNum, strlen(VendorClassData)/2,VendorClassData);
 
         data_len = strlen(cfg)/2; /* Length is enterprise number + data length + vendor class data */
 
