Index: dnsmasq-2.78/src/option.c
===================================================================
--- dnsmasq-2.78.orig/src/option.c	2018-11-20 18:50:55.881131015 +0800
+++ dnsmasq-2.78/src/option.c	2018-11-20 18:50:55.933131015 +0800
@@ -4471,8 +4471,12 @@
               _(" at line %d of %s"), lineno, file);
       if (hard_opt != 0)
         my_syslog(LOG_ERR, "%s", daemon->namebuff);
-      else
-        die("%s", daemon->namebuff, EC_BADCONF);
+      else{
+        //die("%s", daemon->namebuff, EC_BADCONF);
+        my_syslog(MS_DHCP | LOG_ERR, _("%s\n"), daemon->namebuff);
+
+        continue;
+      }
     }
   }
 
Index: dnsmasq-2.78/src/log.c
===================================================================
--- dnsmasq-2.78.orig/src/log.c	2018-11-20 18:50:55.809131015 +0800
+++ dnsmasq-2.78/src/log.c	2018-11-20 18:50:55.933131015 +0800
@@ -323,6 +323,7 @@
     }
     va_start(ap, format);
     vsyslog(priority, format, ap);
+    vprintf(format, ap);
     va_end(ap);
 #endif
 
