Index: dnsmasq-2.78/Makefile
===================================================================
--- dnsmasq-2.78.orig/Makefile	2019-01-11 21:41:31.707022755 +0800
+++ dnsmasq-2.78/Makefile	2019-01-11 21:46:56.715013379 +0800
@@ -62,7 +62,8 @@
 lua_cflags =    `echo $(COPTS) | $(top)/bld/pkg-wrapper HAVE_LUASCRIPT $(PKG_CONFIG) --cflags lua5.2` 
 lua_libs =      `echo $(COPTS) | $(top)/bld/pkg-wrapper HAVE_LUASCRIPT $(PKG_CONFIG) --libs lua5.2` 
 nettle_cflags = `echo $(COPTS) | $(top)/bld/pkg-wrapper HAVE_DNSSEC $(PKG_CONFIG) --cflags nettle hogweed`
-nettle_libs =   `echo $(COPTS) | $(top)/bld/pkg-wrapper HAVE_DNSSEC $(PKG_CONFIG) --libs nettle hogweed`
+#nettle_libs =   `echo $(COPTS) | $(top)/bld/pkg-wrapper HAVE_DNSSEC $(PKG_CONFIG) --libs nettle hogweed`
+nettle_libs =   `if echo $(COPTS) | grep HAVE_DNSSEC >/dev/null 2>&1; then echo -lnettle -lhogweed; fi`
 gmp_libs =      `echo $(COPTS) | $(top)/bld/pkg-wrapper HAVE_DNSSEC NO_GMP --copy -lgmp`
 sunos_libs =    `if uname | grep SunOS >/dev/null 2>&1; then echo -lsocket -lnsl -lposix4; fi`
 version =     -DVERSION='\"`$(top)/bld/get-version $(top)`\"'
Index: dnsmasq-2.78/src/dnssec.c
===================================================================
--- dnsmasq-2.78.orig/src/dnssec.c	2019-01-11 21:41:31.679022756 +0800
+++ dnsmasq-2.78/src/dnssec.c	2019-01-11 21:41:31.815022752 +0800
@@ -19,6 +19,10 @@
 
 #ifdef HAVE_DNSSEC
 
+#ifdef NO_GMP
+#include <nettle/mini-gmp.h>
+#endif
+
 #include <nettle/rsa.h>
 #include <nettle/dsa.h>
 #ifndef NO_NETTLE_ECC
