Index: libzcfg_fe_dal-1.0/network/interfaceGrouping/zcfg_fe_dal_intfGrp.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/interfaceGrouping/zcfg_fe_dal_intfGrp.c	2018-09-26 11:17:15.425544305 +0800
+++ libzcfg_fe_dal-1.0/network/interfaceGrouping/zcfg_fe_dal_intfGrp.c	2018-09-26 11:19:44.293550911 +0800
@@ -298,7 +298,7 @@
 			}
 			zcfgFeJsonObjFree(bridgeportObj);
 		}
-		if(!strcmp(LAN,"")){
+		{
 			int vlanGroupIndexLen = 0;
 			char *vlanGroupIdx = vlanGroupIndex;
 			IID_INIT(vlanIid);
