diff -ruN a/js/zcfg_fe_dal_networkmap.c b/js/zcfg_fe_dal_networkmap.c
--- a/js/zcfg_fe_dal_networkmap.c	2018-12-03 16:38:46.181001000 +0800
+++ b/js/zcfg_fe_dal_networkmap.c	2018-12-07 15:13:28.928810683 +0800
@@ -836,7 +836,12 @@
 	IID_INIT(iid);
 	while(zcfgFeObjJsonGetNext(RDM_OID_ETH_IFACE, &iid, &ethIfaceObj) == ZCFG_SUCCESS) {
 		lanport = json_object_get_string (json_object_object_get(ethIfaceObj, "X_ZYXEL_LanPort"));
+
+#if (defined(ZCFG_VMG4927_B50A_Generic) || defined (ZCFG_EMG6726_B10A_Generic) )
+		if(!strncmp (lanport, "LAN1", 4)) {
+#else
 		if(!strncmp (lanport, "LAN", 3)) {
+#endif
 			lanMacAddr = json_object_get_string (json_object_object_get(ethIfaceObj, "MACAddress"));
 			Jadds (ethnewObj, "LanMac", lanMacAddr);
 		}
diff -ruN a/Makefile b/Makefile
--- a/Makefile	2018-12-03 16:38:46.225001000 +0800
+++ b/Makefile	2018-12-07 15:08:27.552810684 +0800
@@ -255,10 +255,17 @@
 CFLAGS += -DZYXEL_FEATURE_BITS
 endif
 
+ifeq ($(CONFIG_ZCFG_VMG4927_B50A_Generic),y)
+CFLAGS += -DZCFG_VMG4927_B50A_Generic
+endif
+
+ifeq ($(CONFIG_ZCFG_EMG6726_B10A_Generic),y)
+CFLAGS += -DZCFG_EMG6726_B10A_Generic
+endif
+
 LDFLAGS += $(LIB_PATH) $(LINK_LIB)
 CFLAGS += $(INC_PATH) -std=gnu99
 
-
 .PHONY : libzcfg_fe_dal
 
 all: clean libzcfg_fe_dal
