Index: libzcfg_fe_dal-1.0/network/wireless/zcfg_fe_dal_wifi_easymesh.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/wireless/zcfg_fe_dal_wifi_easymesh.c	2018-12-28 16:52:53.608734900 +0800
+++ libzcfg_fe_dal-1.0/network/wireless/zcfg_fe_dal_wifi_easymesh.c	2018-12-28 16:56:04.773733881 +0800
@@ -14,6 +14,16 @@
 	{NULL,		0,	0,	0,	NULL,	NULL, 0}
 };
 
+void zcfgFeDalShowWifiEasyMesh(struct json_object *Jarray){
+	struct json_object *obj;
+	if(json_object_get_type(Jarray) != json_type_array){
+		printf("wrong Jobj format!\n");
+		return;
+	}
+	obj = json_object_array_get_idx(Jarray, 0);
+	printf("%-20s %-10s \n", "EasyMesh Setup:", json_object_get_string(json_object_object_get(obj, "EasyMeshEnable")));
+}
+
 zcfgRet_t zcfgFeDalWifiEasyMeshGet(struct json_object *Jobj, struct json_object *Jarray, char *replyMsg){
 
 	zcfgRet_t ret = ZCFG_SUCCESS;
Index: libzcfg_fe_dal-1.0/zcfg_fe_dal.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/zcfg_fe_dal.c	2018-12-28 16:52:53.608734900 +0800
+++ libzcfg_fe_dal-1.0/zcfg_fe_dal.c	2018-12-28 16:56:16.237304757 +0800
@@ -286,7 +286,9 @@
 extern void zcfgFeDalShowWlanScheduler(struct json_object *Jarray);
 extern void zcfgFeDalShowWlanSchAccess(struct json_object *Jarray);
 extern void zcfgFeDalShowPORT_MIRROR(struct json_object *Jarray);
-
+#ifdef ZYXEL_EASYMESH
+extern void zcfgFeDalShowWifiEasyMesh(struct json_object *Jarray);
+#endif
 
 dal_handler_t dalHandler[] = {
 {"wan",				WAN_param,			zcfgFeDalWan,			zcfgFeDalShowWan,			"get|add|edit|delete",	"Broadband"},
@@ -331,7 +333,7 @@
 {"WIFI_SONIQ",		WIFI_SONIQ_param, 	zcfgFeDalWifiSoniq,		NULL,						"edit|get", 			"Wireless"},//Wireless
 #endif
 #ifdef ZYXEL_EASYMESH
-{"WIFI_EasyMesh",		WIFI_EASYMESH_param, 	zcfgFeDalWifiEasyMesh,		NULL,						"edit|get", 			"Wireless"},//Wireless
+{"wifi_easy_mesh",		WIFI_EASYMESH_param, 	zcfgFeDalWifiEasyMesh,	zcfgFeDalShowWifiEasyMesh,	"edit|get",		"Wireless"},//Wireless
 #endif
 {"wifi_others",		WIFI_OTHERS_param,	zcfgFeDalWifiOthers,	zcfgFeDalShowWifiOthers,	"edit|get",				"Wireless"},//
 {"wlan_scheduler",	WLAN_SCHEDULER_param, zcfgFeDalWlanScheduler,	zcfgFeDalShowWlanScheduler,	"add|edit|delete|get",	"Wireless"},
