Index: libzcfg_fe_dal-1.0/Makefile
===================================================================
--- libzcfg_fe_dal-1.0.orig/Makefile	2018-12-14 14:01:24.587045574 +0800
+++ libzcfg_fe_dal-1.0/Makefile	2018-12-14 14:05:03.213727294 +0800
@@ -276,6 +276,11 @@
 CFLAGS += -DCONFIG_ZLOG_USE_DEBUG
 endif
 
+#cyberSecurity
+ifeq ($(CONFIG_ZYXEL_HOME_CYBER_SECURITY_FSECURE),y)
+CFLAGS += -DZYXEL_HOME_CYBER_SECURITY_FSECURE
+endif
+
 LDFLAGS += $(LIB_PATH) $(LINK_LIB)
 CFLAGS += $(INC_PATH) -std=gnu99
 
