Index: libzcfg_fe_dal-1.0/zcfg_fe_dal.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/zcfg_fe_dal.c	2019-01-24 17:10:13.680587102 +0800
+++ libzcfg_fe_dal-1.0/zcfg_fe_dal.c	2019-01-24 17:16:03.955637102 +0800
@@ -2758,13 +2758,29 @@
 						stringPtr = json_object_get_string(pramJobj);
 						if(stringPtr == NULL || !strcmp(stringPtr, ""))
 							break;
-						
+
 						if(paramList[i].enumeration != NULL){
 							if(findInList(paramList[i].enumeration, stringPtr, "|") == true)
 								break;
 						}
-						if(validateIPv6(stringPtr) == false){
-							printf("Wrong parameter value <%s> of %s\n", stringPtr, paramList[i].paraName);
+						strcpy(buf, stringPtr);
+						ptr = strchr(buf, '/');
+
+						if (ptr != NULL)
+						{
+							*ptr = '\0';
+							intValue = atoi(ptr+1);
+
+							if (intValue < 0 || intValue > 128)
+							{
+								printf("Wrong intValue range 0 to 128 <%s> of %s   %s\n", stringPtr,paramList[i].type, paramList[i].paraName);
+								return false;
+							}
+						}
+
+						if (validateIPv6(buf) == false)
+						{
+							printf("Wrong IPv6 address format <%s> of %s   %s\n", stringPtr, paramList[i].type, paramList[i].paraName);
 							return false;
 						}
 						break;
@@ -2772,7 +2788,7 @@
 						stringPtr = json_object_get_string(pramJobj);
 						if(stringPtr == NULL || !strcmp(stringPtr, ""))
 							break;
-						
+
 						if(paramList[i].enumeration != NULL){
 							if(findInList(paramList[i].enumeration, stringPtr, "|") == true)
 								break;
@@ -2782,11 +2798,11 @@
 						*ptr = '\0';
 						intValue = atoi(ptr+1);
 						if(intValue<0 || intValue>128){
-							printf("Wrong parameter value <%s> of %s\n", stringPtr, paramList[i].paraName);
+							printf("Wrong intValue range 0 to 128 <%s> of %s   %s\n", stringPtr,paramList[i].type, paramList[i].paraName);
 							return false;
 						}
 						if(validateIPv6(buf) == false){
-							printf("Wrong parameter value <%s> of %s\n", stringPtr, paramList[i].paraName);
+							printf("Wrong IPv6 address format <%s> of %s   %s\n", stringPtr, paramList[i].type, paramList[i].paraName);
 							return false;
 						}
 						break;
