Index: libzcfg_fe_dal-1.0/js/zcfg_fe_dal_networkmap.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/js/zcfg_fe_dal_networkmap.c	2019-03-13 14:11:35.130503635 +0800
+++ libzcfg_fe_dal-1.0/js/zcfg_fe_dal_networkmap.c	2019-03-13 14:18:33.580600437 +0800
@@ -21,6 +21,7 @@
 #ifdef ZYXEL_HOME_CYBER_SECURITY_FSECURE	
 	{"BrowsingProtection",	    dalType_boolean,	0,	0,	NULL},
 	{"TrackingProtection",	    dalType_boolean,	0,	0,	NULL},
+	{"IOTtection",	    dalType_boolean,	0,	0,	NULL},
 	{"X_ZYXEL_FirstSeen",			dalType_int,		0,	0,	NULL},
 	{"X_ZYXEL_LastSeen",			dalType_int,		0,	0,	NULL},
 	{"SourceVendorClassID",			dalType_string,		0,	0,	NULL},	
@@ -1103,6 +1104,7 @@
 #ifdef ZYXEL_HOME_CYBER_SECURITY_FSECURE		
 		Jaddb(obj, "BrowsingProtection", false);
 		Jaddb(obj, "TrackingProtection", false);
+		Jaddb(obj, "IOTProtection", false);		
 		Jaddi(obj, "X_ZYXEL_FirstSeen", 0);
 		Jaddi(obj, "X_ZYXEL_LastSeen", 0);
 		Jadds(obj, "SourceVendorClassID", "");		
@@ -1188,6 +1190,7 @@
 #ifdef ZYXEL_HOME_CYBER_SECURITY_FSECURE				
 				replaceParam(obj, "BrowsingProtection", iconJobj, "BrowsingProtection");
 				replaceParam(obj, "TrackingProtection", iconJobj, "TrackingProtection");	
+				replaceParam(obj, "IOTProtection", iconJobj, "IOTProtection");					
 				replaceParam(obj, "X_ZYXEL_FirstSeen", iconJobj, "X_ZYXEL_FirstSeen");	
 				replaceParam(obj, "X_ZYXEL_LastSeen", iconJobj, "X_ZYXEL_LastSeen");	
 				replaceParam(obj, "SourceVendorClassID", iconJobj, "SourceVendorClassID");					
@@ -1281,7 +1284,7 @@
 	objIndex_t Iid = {0};
 	const char *macAddr = NULL, *curmacAddr = NULL, *GUIicon = NULL, *hostName = NULL, *SourceVendorClassID = NULL;
 	char icon[16] = {0};
-	bool changeIconRuleExist = false, Internet_Blocking_Enable = false,BrowsingProtection = false,TrackingProtection = false;
+	bool changeIconRuleExist = false, Internet_Blocking_Enable = false,BrowsingProtection = false,TrackingProtection = false, IOTProtection = false;
 	unsigned int X_ZYXEL_FirstSeen =0, X_ZYXEL_LastSeen = 0;
 	
 	macAddr = json_object_get_string(json_object_object_get(Jobj,"MacAddress"));
@@ -1291,6 +1294,7 @@
 #ifdef ZYXEL_HOME_CYBER_SECURITY_FSECURE	
 	BrowsingProtection = json_object_get_boolean(json_object_object_get(Jobj,"BrowsingProtection"));
 	TrackingProtection = json_object_get_boolean(json_object_object_get(Jobj,"TrackingProtection"));
+	IOTProtection = json_object_get_boolean(json_object_object_get(Jobj,"IOTProtection"));	
 	X_ZYXEL_FirstSeen = json_object_get_int(json_object_object_get(Jobj,"X_ZYXEL_FirstSeen"));
 	X_ZYXEL_LastSeen = json_object_get_int(json_object_object_get(Jobj,"X_ZYXEL_LastSeen"));
 	SourceVendorClassID = json_object_get_string(json_object_object_get(Jobj,"SourceVendorClassID"));	
@@ -1343,6 +1347,9 @@
 			if(json_object_object_get(Jobj, "TrackingProtection")){
 				json_object_object_add(obj, "TrackingProtection", json_object_new_boolean(TrackingProtection));
 			}			
+			if(json_object_object_get(Jobj, "IOTProtection")){
+				json_object_object_add(obj, "IOTProtection", json_object_new_boolean(IOTProtection));
+			}						
 			if(json_object_object_get(Jobj, "X_ZYXEL_FirstSeen")){
 				json_object_object_add(obj, "X_ZYXEL_FirstSeen", json_object_new_int(X_ZYXEL_FirstSeen));
 			}			
@@ -1383,6 +1390,9 @@
 			if(json_object_object_get(Jobj, "TrackingProtection")){
 				json_object_object_add(obj, "TrackingProtection", json_object_new_boolean(TrackingProtection));
 			}		
+			if(json_object_object_get(Jobj, "IOTProtection")){
+				json_object_object_add(obj, "IOTProtection", json_object_new_boolean(IOTProtection));
+			}				
 			if(json_object_object_get(Jobj, "X_ZYXEL_FirstSeen")){
 				json_object_object_add(obj, "X_ZYXEL_FirstSeen", json_object_new_int(X_ZYXEL_FirstSeen));
 			}	
