Index: libzcfg_fe_dal-1.0/zcfg_fe_dal_rdmacl.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/zcfg_fe_dal_rdmacl.c	2019-04-11 14:21:02.890227999 +0800
+++ libzcfg_fe_dal-1.0/zcfg_fe_dal_rdmacl.c	2019-04-11 14:24:22.346449999 +0800
@@ -21,8 +21,8 @@
 const char *rdmACLBridgingBrParams[] = {"X_ZYXEL_BridgeName", NULL};
 const char *rdmACLBridgingFltParams[] = {"Bridge", "SourceMACAddressFilterList", "SourceMACFromVendorClassIDFilter",
 		"SourceMACFromVendorClassIDMode", "SourceMACFromClientIDFilter", "X_ZYXEL_SourceMACFromVSIFilter", NULL};
-
-
+const char *rdmACLClassifyParams[] = {"Name", NULL};
+const char *rdmACLCategoryParams[] = {"Name", NULL};
 //char *rdmACLObjectIid[] = {"1,2", "2"}; //[1,2,0,0,0,0], [2,0,0,0,0,0]
 const char *rdmACLBridgingBrIid[] = {"1", NULL};
 
@@ -32,6 +32,8 @@
 	{RDM_OID_VLAN_GROUP, ZHTTP_RDM_ACL_METHOD_QUERY, NULL, rdmACLVlanGroupParams},
 	{RDM_OID_BRIDGING_BR, ZHTTP_RDM_ACL_METHOD_QUERY, NULL, rdmACLBridgingBrParams},
 	{RDM_OID_BRIDGING_FILTER, ZHTTP_RDM_ACL_METHOD_QUERY, NULL, rdmACLBridgingFltParams},
+	{RDM_OID_LOG_CATEGORY, ZHTTP_RDM_ACL_METHOD_QUERY, NULL, rdmACLCategoryParams},
+	{RDM_OID_LOG_CLASSIFY, ZHTTP_RDM_ACL_METHOD_QUERY, NULL, rdmACLClassifyParams},
 	{0, 0, NULL, NULL}
 };
 
