Index: libzcfg_fe_dal-1.0/js/zcfg_fe_dal_networkmap.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/js/zcfg_fe_dal_networkmap.c	2019-04-14 17:15:12.770846358 +0800
+++ libzcfg_fe_dal-1.0/js/zcfg_fe_dal_networkmap.c	2019-04-14 17:16:05.553051707 +0800
@@ -1387,8 +1387,6 @@
 #ifdef ELISA_NETWORKMAP_CUSTOMIZATION
 			sprintf(X_ZYXEL_ConnectionType, "%s %s", Jgets(obj, "X_ZYXEL_ConnectionType"), SupportedFrequencyBands);
 			Jadds(obj, "X_ZYXEL_ConnectionType", X_ZYXEL_ConnectionType);
-#else
-			Jadds(obj, "X_ZYXEL_ConnectionType", "Wi-Fi");
 #endif
 		}
 #ifdef ELISA_NETWORKMAP_CUSTOMIZATION
