diff -urN libzcfg_fe_dal-1.0.org/js/zcfg_fe_dal_networkmap.c libzcfg_fe_dal-1.0/js/zcfg_fe_dal_networkmap.c
--- libzcfg_fe_dal-1.0.org/js/zcfg_fe_dal_networkmap.c	2019-05-06 15:17:54.668886300 +0800
+++ libzcfg_fe_dal-1.0/js/zcfg_fe_dal_networkmap.c	2019-05-07 17:31:27.602540011 +0800
@@ -67,6 +67,7 @@
 	json_object_array_add(*iid, json_object_new_int(iid5));
 	json_object_array_add(*iid, json_object_new_int(iid6));
 	
+	return true;
 }
 
 void getDHCPStatus(char *status, char *tr181Path, const char *type, const char *ipIfacePath, struct json_object *dhcp4PoolJarray, struct json_object *dhcp4ClientJarray){
diff -urN libzcfg_fe_dal-1.0.org/js/zcfg_fe_dal_quickstart.c libzcfg_fe_dal-1.0/js/zcfg_fe_dal_quickstart.c
--- libzcfg_fe_dal-1.0.org/js/zcfg_fe_dal_quickstart.c	2019-05-06 15:17:52.368886300 +0800
+++ libzcfg_fe_dal-1.0/js/zcfg_fe_dal_quickstart.c	2019-05-07 17:33:57.502821201 +0800
@@ -515,6 +515,7 @@
 		zcfgFeJsonObjFree(dhcpv4deleteObj);
 		return ret;
 	}
+	return ret;
 }
 
 zcfgRet_t DNSDelete(struct json_object *Jobj, const char *iptype){
@@ -542,17 +543,18 @@
 			}
 		}
 		else{
-		index = json_object_get_int(json_object_object_get(reqObject, "DeleteIdx"));
-		dnsdeleteIid.idx[0] = index;
-		dnsdeleteIid.level = 1;
-		ret = zcfgFeObjJsonGet(RDM_OID_DNS_CLIENT_SRV, &dnsdeleteIid, &dnsdeleteObj);
-		ret = zcfgFeObjJsonDel(RDM_OID_DNS_CLIENT_SRV, &dnsdeleteIid, NULL);
+			index = json_object_get_int(json_object_object_get(reqObject, "DeleteIdx"));
+			dnsdeleteIid.idx[0] = index;
+			dnsdeleteIid.level = 1;
+			ret = zcfgFeObjJsonGet(RDM_OID_DNS_CLIENT_SRV, &dnsdeleteIid, &dnsdeleteObj);
+			ret = zcfgFeObjJsonDel(RDM_OID_DNS_CLIENT_SRV, &dnsdeleteIid, NULL);
 		}
 		zcfgFeJsonObjFree(reqObject);
 		zcfgFeJsonObjFree(reqObject1);
 		zcfgFeJsonObjFree(dnsdeleteObj);
 		return ret;
 	}
+	return ret;
 }
 
 zcfgRet_t zcfgFeDal_QuickStart_Edit(struct json_object *Jobj, char *replyMsg){
diff -urN libzcfg_fe_dal-1.0.org/monitor/zcfg_fe_dal_voipmonitor.c libzcfg_fe_dal-1.0/monitor/zcfg_fe_dal_voipmonitor.c
--- libzcfg_fe_dal-1.0.org/monitor/zcfg_fe_dal_voipmonitor.c	2019-05-06 15:17:53.964886300 +0800
+++ libzcfg_fe_dal-1.0/monitor/zcfg_fe_dal_voipmonitor.c	2019-05-07 17:35:25.559118313 +0800
@@ -102,6 +102,7 @@
 zcfgFeDalVoIPStatusGetFail:	
 	json_object_put(devInfoObj);
 	
+	return ret;
 }
 zcfgRet_t zcfgFeDalVoIPStatusEdit(struct json_object *Jobj, char *replyMsg)
 {
diff -urN libzcfg_fe_dal-1.0.org/network/broadband/zcfg_fe_dal_advance.c libzcfg_fe_dal-1.0/network/broadband/zcfg_fe_dal_advance.c
--- libzcfg_fe_dal-1.0.org/network/broadband/zcfg_fe_dal_advance.c	2019-05-06 15:17:54.392886300 +0800
+++ libzcfg_fe_dal-1.0/network/broadband/zcfg_fe_dal_advance.c	2019-05-07 17:39:47.857247551 +0800
@@ -69,6 +69,7 @@
 		json_object_object_add(JsonCfg, key2, json_object_new_boolean(true));
 	else
 		json_object_object_add(JsonCfg, key2, json_object_new_boolean(false));
+	return ZCFG_SUCCESS;
 }
 
 zcfgRet_t dalcmdpraseAdslModCfg(char *key1, char *key2, char *key3, struct json_object *Jobj, struct json_object *JsonCfg){
@@ -78,7 +79,8 @@
 	else
 		tmp = json_object_get_boolean(json_object_object_get(JsonCfg, key1));
 	if(tmp)
-			strcat(key3, key2);
+		strcat(key3, key2);
+	return ZCFG_SUCCESS;
 }		
 
 void zcfgFeDalShowWanAdvance(struct json_object *Jarray){
diff -urN libzcfg_fe_dal-1.0.org/network/broadband/zcfg_fe_dal_ethwanlan.c libzcfg_fe_dal-1.0/network/broadband/zcfg_fe_dal_ethwanlan.c
--- libzcfg_fe_dal-1.0.org/network/broadband/zcfg_fe_dal_ethwanlan.c	2019-05-06 15:17:52.812886300 +0800
+++ libzcfg_fe_dal-1.0/network/broadband/zcfg_fe_dal_ethwanlan.c	2019-05-07 17:41:18.902969394 +0800
@@ -256,6 +256,8 @@
 		else
 			return true;
 	}
+	else
+		return false;
 }
 
 zcfgRet_t zcfgFeDalEthWanLan(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg){
diff -urN libzcfg_fe_dal-1.0.org/security/firewall/zcfg_fe_dal_firewall_generic.c libzcfg_fe_dal-1.0/security/firewall/zcfg_fe_dal_firewall_generic.c
--- libzcfg_fe_dal-1.0.org/security/firewall/zcfg_fe_dal_firewall_generic.c	2019-05-06 15:17:53.408886300 +0800
+++ libzcfg_fe_dal-1.0/security/firewall/zcfg_fe_dal_firewall_generic.c	2019-05-07 17:46:50.134852461 +0800
@@ -38,6 +38,7 @@
 };
 
 zcfgRet_t zcfgFeDalShowFirewall(struct json_object *Jarray, char *replyMsg){
+	zcfgRet_t ret = ZCFG_SUCCESS;
 	struct json_object *obj = NULL;
 	const char *filewalllevel = NULL;
 	const char *level = NULL;
@@ -45,7 +46,7 @@
 	const char *lantowan = NULL;
 	if(json_object_get_type(Jarray) != json_type_array){
 		printf("wrong Jobj format!\n");
-		return;
+		return ZCFG_INTERNAL_ERROR;
 	}
 	obj = json_object_array_get_idx(Jarray, 0);
 	filewalllevel = json_object_get_string(json_object_object_get(obj, "Level_GUI"));
@@ -72,6 +73,7 @@
 	printf("%-20s  %-10s \n", "Wan to LAN:", wantolan);
 	printf("\n%-20s %-10s \n","DoS Protection Blocking :",json_object_get_string(json_object_object_get(obj, "enableDos")));
 	
+	return ret;
 }
 
 
diff -urN libzcfg_fe_dal-1.0.org/zcfg_fe_dal.c libzcfg_fe_dal-1.0/zcfg_fe_dal.c
--- libzcfg_fe_dal-1.0.org/zcfg_fe_dal.c	2019-05-06 15:17:54.660886300 +0800
+++ libzcfg_fe_dal-1.0/zcfg_fe_dal.c	2019-05-07 17:47:38.703977337 +0800
@@ -2051,6 +2051,7 @@
 		radioObj = NULL;
 	}
 
+	return ZCFG_SUCCESS;
 }
 
 zcfgRet_t getwifiSSIDObj(struct json_object **ssidObj){
