Index: libzcfg_fe_dal-1.0/network/broadband/zcfg_fe_dal_broadband.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/broadband/zcfg_fe_dal_broadband.c	2019-05-14 17:31:08.409802720 +0800
+++ libzcfg_fe_dal-1.0/network/broadband/zcfg_fe_dal_broadband.c	2019-05-14 17:32:46.327562856 +0800
@@ -1238,6 +1238,7 @@
 		}
 		json_object_object_add(v6FwdObj, "Enable", json_object_new_boolean(true));
 		json_object_object_add(v6FwdObj, "Interface", json_object_new_string(ipIfacePath));
+        json_object_object_add(v6FwdObj, "DestIPPrefix", json_object_new_string("::/0"));
 		replaceParam(v6FwdObj, "NextHop", Jobj, "NextHop");
 		//if((ret = zcfgFeObjJsonSetWithoutApply(oid, &v6FwdIid, v6FwdObj, NULL)) != ZCFG_SUCCESS)
 		//	printf("set oid:%d fail, line=%d, ret=%d\n", oid, __LINE__, ret);
@@ -2765,7 +2766,19 @@
 	if(v4ClientObj)zcfgFeObjJsonSet(RDM_OID_DHCPV4_CLIENT, &v4ClientIid, v4ClientObj, NULL);
 	if(igmpObj)zcfgFeObjJsonSet(RDM_OID_ZY_IGMP, &igmpIid, igmpObj, NULL);
 	if(mldObj)zcfgFeObjJsonSet(RDM_OID_ZY_MLD, &mldIid, mldObj, NULL);
-	if(routerObj)zcfgFeObjJsonBlockedSet(RDM_OID_ROUTING_ROUTER, &routerIid, routerObj, NULL);
+	if(routerObj)
+	{
+	  if(isAdd && Enable_6RD)
+      {
+             //Avoid to overwrite the v6 default gw
+             //For 6RD case, update the default gw via zcfg_be
+             zcfgFeObjJsonSetWithoutApply(RDM_OID_ROUTING_ROUTER, &routerIid, routerObj, NULL);
+      }
+      else
+      {
+          zcfgFeObjJsonBlockedSet(RDM_OID_ROUTING_ROUTER, &routerIid, routerObj, NULL);
+      }
+    }
 	if(dnsObj)zcfgFeObjJsonSet(RDM_OID_DNS, &dnsIid, dnsObj, NULL);
 	
 	return ret;
