Index: libzcfg_fe_dal-1.0/network/nat/zcfg_fe_dal_nat_port_mapping.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/nat/zcfg_fe_dal_nat_port_mapping.c	2019-05-21 18:50:04.660034477 +0800
+++ libzcfg_fe_dal-1.0/network/nat/zcfg_fe_dal_nat_port_mapping.c	2019-05-21 18:51:57.600851380 +0800
@@ -301,6 +301,9 @@
 		if(!strcmp(interface, "")){
 			autoDetectWanStatus = true;
 		}
+		else{
+			autoDetectWanStatus = false;
+		}
 #endif
 		originatingipaddress = json_object_get_string(json_object_object_get(Jobj, "OriginatingIpAddress"));
 		externalport = json_object_get_int(json_object_object_get(Jobj, "ExternalPortStart"));
