diff -ruN a/network/homeNetworking/zcfg_fe_dal_lansetup.c b/network/homeNetworking/zcfg_fe_dal_lansetup.c
--- a/network/homeNetworking/zcfg_fe_dal_lansetup.c	2019-06-05 10:11:54.000000000 +0800
+++ b/network/homeNetworking/zcfg_fe_dal_lansetup.c	2019-06-05 13:28:39.590533059 +0800
@@ -1506,19 +1506,27 @@
 					}
 					else
 						json_object_object_add(paramJobj, "IPv6_LanAddrAssign", json_object_new_string("Stateless"));
-						v6DNSServer = json_object_get_string(json_object_object_get(dhcpv6srvObj, "X_ZYXEL_DNSServers"));
-						v6DNSList = json_object_get_string(json_object_object_get(dhcpv6srvObj, "X_ZYXEL_DNSList")); 
-						
-						strcpy(v6DNSServerTmp,v6DNSServer);
-						strcpy(v6DNSListTmp,v6DNSList);
-						v6dnstype1 = strtok_r(v6DNSListTmp, ",", &v6dnstypetmp);
-						v6dnstype2 = strtok_r(NULL, ",", &v6dnstypetmp);
-						v6dnstype3 = strtok_r(NULL, ",", &v6dnstypetmp);
-						v6dnsServer1=v6DNSServerTmp;
+					
+					v6DNSServer = json_object_get_string(json_object_object_get(dhcpv6srvObj, "X_ZYXEL_DNSServers"));
+					v6DNSList = json_object_get_string(json_object_object_get(dhcpv6srvObj, "X_ZYXEL_DNSList")); 							
+					strcpy(v6DNSServerTmp,v6DNSServer);
+					strcpy(v6DNSListTmp,v6DNSList);
+					v6dnstype1 = strtok_r(v6DNSListTmp, ",", &v6dnstypetmp);
+					v6dnstype2 = strtok_r(NULL, ",", &v6dnstypetmp);
+					v6dnstype3 = strtok_r(NULL, ",", &v6dnstypetmp);
+					v6dnsServer1=v6DNSServerTmp;
+
+					if(strlen(v6dnsServer1) >0)
+					{
+
 						*(ptr = strchr(v6dnsServer1, ',')) = '\0';
 						v6dnsServer2=ptr+1;
 						*(ptr = strchr(v6dnsServer2, ',')) = '\0';
 						v6dnsServer3=ptr+1;
+					}
+					
+					if(strlen(v6DNSListTmp) >0)
+					{
 						if(!strcmp(v6dnstype1, "None") || !strcmp(v6dnstype1, "ISP") ||!strcmp(v6dnstype1, "Proxy"))
 							json_object_object_add(paramJobj, "Ipv6_DNSServer1", json_object_new_string(v6dnstype1));
 						else
@@ -1531,9 +1539,16 @@
 							json_object_object_add(paramJobj, "Ipv6_DNSServer3", json_object_new_string(v6dnstype3));
 						else
 							json_object_object_add(paramJobj, "Ipv6_DNSServer3", json_object_new_string(v6dnsServer3));
-						zcfgFeJsonObjFree(dhcpv6srvObj);
-						break;
 					}
+					else
+					{
+						json_object_object_add(paramJobj, "Ipv6_DNSServer1", json_object_new_string(""));
+						json_object_object_add(paramJobj, "Ipv6_DNSServer2", json_object_new_string(""));
+						json_object_object_add(paramJobj, "Ipv6_DNSServer3", json_object_new_string(""));
+					}
+					zcfgFeJsonObjFree(dhcpv6srvObj);
+					break;
+				}
 					zcfgFeJsonObjFree(dhcpv6srvObj);
 				}
 				if(zcfgFeObjJsonGet(RDM_OID_DNS, &dnsIid, &dnsObj) == ZCFG_SUCCESS){
diff -ruN a/zcfg_fe_dal.c b/zcfg_fe_dal.c
--- a/zcfg_fe_dal.c	2019-06-05 10:11:54.000000000 +0800
+++ b/zcfg_fe_dal.c	2019-06-05 13:42:46.398533059 +0800
@@ -2857,7 +2857,12 @@
 								break;
 						}
 						strcpy(buf, stringPtr);
-						ptr = strchr(buf, '/');
+						if((ptr = strchr(buf, '/')) == NULL)
+						{
+							printf("Wrong parameter value <%s> of %s\n", stringPtr, paramList[i].paraName);
+							return false;
+						}
+						
 						*ptr = '\0';
 						intValue = atoi(ptr+1);
 						if(intValue<0 || intValue>128){
