Index: libzcfg_fe_dal-1.0/Makefile
===================================================================
--- libzcfg_fe_dal-1.0.orig/Makefile	2019-06-20 16:21:46.502330540 +0800
+++ libzcfg_fe_dal-1.0/Makefile	2019-06-20 16:23:10.667808865 +0800
@@ -177,6 +177,11 @@
 CFLAGS += -DZYXEL_REMOTE_ACCESS_PRIVILEGE
 endif
 
+#ZYXEL_REMOTE_ACCOUNT_ACCESS_PRIVILEGE
+ifeq ($(CONFIG_ZYXEL_REMOTE_ACCOUNT_ACCESS_PRIVILEGE),y)
+CFLAGS += -DZYXEL_REMOTE_ACCOUNT_ACCESS_PRIVILEGE
+endif
+
 #WLAN_SCHEDULER_AUTO_SWITCH_RADIO_ON
 ifeq ($(CONFIG_ZYXEL_WLAN_SCHEDULER_AUTO_SWITCH_RADIO_ON),y)
 CFLAGS += -DZYXEL_WLAN_SCHEDULER_AUTO_SWITCH_RADIO_ON
Index: libzcfg_fe_dal-1.0/maintenance/useraccount/zcfg_fe_dal_user_account.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/maintenance/useraccount/zcfg_fe_dal_user_account.c	2019-06-20 16:21:44.542330596 +0800
+++ libzcfg_fe_dal-1.0/maintenance/useraccount/zcfg_fe_dal_user_account.c	2019-06-20 16:25:03.639904023 +0800
@@ -165,6 +165,9 @@
 	#ifdef ZYXEL_REMOTE_ACCESS_PRIVILEGE
 			replaceParam(pramJobj, "RemoteAccessPrivilege", obj, "RemoteAccessPrivilege");
 	#endif
+	#ifdef ZYXEL_REMOTE_ACCOUNT_ACCESS_PRIVILEGE	
+			replaceParam(pramJobj, "Privilege", obj, "Privilege");
+	#endif
 			if(iid.idx[0]==highGpIdx){//should not happen for current design
 				json_object_object_add(pramJobj, "group",json_object_new_string("Root"));
 			}
@@ -283,7 +286,9 @@
 #ifdef ZYXEL_REMOTE_ACCESS_PRIVILEGE
 		replaceParam(accountObj, "RemoteAccessPrivilege", Jobj, "RemoteAccessPrivilege");
 #endif
-		
+#ifdef ZYXEL_REMOTE_ACCOUNT_ACCESS_PRIVILEGE	
+		replaceParam(accountObj, "Privilege", Jobj, "Privilege");
+#endif		
 		ret = zcfgFeObjJsonBlockedSet(RDM_OID_ZY_LOG_CFG_GP_ACCOUNT, &accountIid, accountObj, NULL);
 	}
 	
@@ -387,7 +392,9 @@
 #ifdef ZYXEL_REMOTE_ACCESS_PRIVILEGE	
 		replaceParam(accountObj, "RemoteAccessPrivilege", Jobj, "RemoteAccessPrivilege");
 #endif
-			
+#ifdef ZYXEL_REMOTE_ACCOUNT_ACCESS_PRIVILEGE	
+		replaceParam(accountObj, "Privilege", Jobj, "Privilege");
+#endif			
         ret = zcfgFeObjJsonBlockedSet(RDM_OID_ZY_LOG_CFG_GP_ACCOUNT, &accountIid, accountObj, NULL);
 		zcfgFeJsonObjFree(accountObj);
 	}
