Index: libzcfg_fe_dal-1.0/network/broadband/zcfg_fe_dal_broadband.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/broadband/zcfg_fe_dal_broadband.c	2019-07-04 08:51:15.441655205 +0800
+++ libzcfg_fe_dal-1.0/network/broadband/zcfg_fe_dal_broadband.c	2019-07-04 08:54:07.136352559 +0800
@@ -1144,11 +1144,13 @@
 	if(CurrIp6DnsStatic && ip6DnsStatic){ //enable to enable
 		DNSServer = json_object_get_string(json_object_object_get(Jobj, "DNS6Server"));
 		CurrDNSServer = json_object_get_string(json_object_object_get(dnsClientSrv6Obj, "DNSServer"));
-		if(strcmp(DNSServer, CurrDNSServer)){
-			json_object_object_add(dnsClientSrv6Obj, "DNSServer", json_object_new_string(DNSServer));
-		}else{
+		
+		if(DNSServer == NULL || CurrDNSServer ==NULL || !strcmp(DNSServer, CurrDNSServer)){
 			if(dnsClientSrv6Obj) {zcfgFeJsonObjFree(dnsClientSrv6Obj);}
 		}
+		else if(strcmp(DNSServer, CurrDNSServer)){
+			json_object_object_add(dnsClientSrv6Obj, "DNSServer", json_object_new_string(DNSServer));
+		}
 	}
 	else if(CurrIp6DnsStatic && !ip6DnsStatic){ ////enable to disable
 		//json_object_object_add(dnsClientSrv6Obj, "Enable", json_object_new_boolean(false));
