Index: libzcfg_fe_dal-1.0/Makefile
===================================================================
--- libzcfg_fe_dal-1.0.orig/Makefile	2019-07-05 16:17:46.283826479 +0800
+++ libzcfg_fe_dal-1.0/Makefile	2019-07-05 16:17:46.327826480 +0800
@@ -310,6 +310,10 @@
 CFLAGS += -DZYXEL_FEATURE_BITS
 endif
 
+ifeq ($(CONFIG_ZYXEL_LED_SWITCH_ACTIVE),y)
+CFLAGS += -DZYXEL_LED_SWITCH_ACTIVE
+endif
+
 #zlog
 ifeq ($(CONFIG_ZLOG_USE_DEBUG),y)
 CFLAGS += -DCONFIG_ZLOG_USE_DEBUG
Index: libzcfg_fe_dal-1.0/maintenance/zcfg_fe_dal_snmp.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/maintenance/zcfg_fe_dal_snmp.c	2019-07-05 16:17:41.943826396 +0800
+++ libzcfg_fe_dal-1.0/maintenance/zcfg_fe_dal_snmp.c	2019-07-05 17:27:05.591602485 +0800
@@ -22,6 +22,9 @@
 /* set RDM_OID_SYSTEM_INFO object */
 {"HostName", 	dalType_string, 	0, 30, NULL},
 {"DomainName",	dalType_string, 	0, 30, NULL},	// system info
+#ifdef ZYXEL_LED_SWITCH_ACTIVE
+{"ledAllEnable",dalType_boolean,	0,	0,	NULL},
+#endif
 {"Location", 	dalType_string,		0, 15, NULL},
 {"Contact", 	dalType_string,		0, 50, NULL},
 /* set RDM_OID_TRAP_ADDR object */
@@ -72,12 +75,16 @@
 	const char *Location = NULL;
 	const char *Contact = NULL;
 	const char *domainName = NULL;
-
+#ifdef ZYXEL_LED_SWITCH_ACTIVE
+	bool ledAllEnable =true;
+#endif
 	HostName = json_object_get_string(json_object_object_get(Jobj, "HostName"));
 	Location = json_object_get_string(json_object_object_get(Jobj, "Location"));
 	Contact = json_object_get_string(json_object_object_get(Jobj, "Contact"));
 	domainName = json_object_get_string(json_object_object_get(Jobj, "DomainName"));
-
+#ifdef ZYXEL_LED_SWITCH_ACTIVE	
+	ledAllEnable = json_object_get_boolean(json_object_object_get(Jobj, "ledAllEnable"));
+#endif
 	IID_INIT(SysInfoIid);
 	ret = zcfgFeObjJsonGetWithoutUpdate(RDM_OID_SYSTEM_INFO, &SysInfoIid, &SysInfoObj);
 
@@ -90,7 +97,11 @@
 		json_object_object_add(SysInfoObj, "Contact", json_object_new_string(Contact));
 		if(json_object_object_get(Jobj, "DomainName"))
 			json_object_object_add(SysInfoObj, "DomainName", json_object_new_string(domainName));
-
+#ifdef ZYXEL_LED_SWITCH_ACTIVE		
+		if(json_object_object_get(Jobj, "ledAllEnable")){
+			json_object_object_add(SysInfoObj, "ledAllEnable", json_object_new_boolean(ledAllEnable));
+		}
+#endif
 		zcfgFeObjJsonSet(RDM_OID_SYSTEM_INFO, &SysInfoIid, SysInfoObj, NULL);
 	}
 
@@ -204,7 +215,9 @@
 	const char *Location = NULL;
 	const char *Contact = NULL;
 	const char *domainname = NULL;
-
+#ifdef ZYXEL_LED_SWITCH_ACTIVE	
+	bool ledAllEnable = true;
+#endif
 	/* RDM_OID_TRAP_ADDR */
 	struct json_object *trapObj = NULL;
 	objIndex_t trapIid = {0};
@@ -236,8 +249,14 @@
 		Location = json_object_get_string(json_object_object_get(SysInfoObj, "Location"));
 		Contact = json_object_get_string(json_object_object_get(SysInfoObj, "Contact"));
 		domainname = json_object_get_string(json_object_object_get(SysInfoObj, "DomainName"));
+#ifdef ZYXEL_LED_SWITCH_ACTIVE
+		ledAllEnable = json_object_get_boolean(json_object_object_get(SysInfoObj, "ledAllEnable"));
+#endif
 		json_object_object_add(retObj, "HostName", json_object_new_string(HostName));
 		json_object_object_add(retObj, "DomainName", json_object_new_string(domainname));
+#ifdef ZYXEL_LED_SWITCH_ACTIVE
+		json_object_object_add(retObj, "ledAllEnable", json_object_new_boolean(ledAllEnable));
+#endif
 		json_object_object_add(retObj, "Location", json_object_new_string(Location));
 		json_object_object_add(retObj, "Contact", json_object_new_string(Contact));
 		zcfgFeJsonObjFree(SysInfoObj);
