Index: libzcfg_fe_dal-1.0/network/wireless/zcfg_fe_dal_wifi_general.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/wireless/zcfg_fe_dal_wifi_general.c	2019-07-09 18:21:12.336454047 +0800
+++ libzcfg_fe_dal-1.0/network/wireless/zcfg_fe_dal_wifi_general.c	2019-07-09 18:21:12.380454047 +0800
@@ -42,6 +42,7 @@
  	{"X_ZYXEL_AX160_Channels",	dalType_string,		0,	0,	NULL,	NULL,	dalcmd_Forbid},
 	{"X_ZYXEL_Wireless_Mode",	dalType_string,		0,	0,	NULL,	NULL,	dalcmd_Forbid},
 	{"X_ZYXEL_OBSS_coex",		dalType_boolean,	0,	0,	NULL,	NULL,	dalcmd_Forbid},
+	{"X_ZYXEL_InActiveChannels",	dalType_string,		0,	0,	NULL,	NULL,	dalcmd_Forbid},
 	//WiFi Access Point
 	{"MaxAssociatedDevices",dalType_int,		1,	64,	NULL,		NULL,			0},
 	{"wlHide",				dalType_boolean, 	0,	0,	NULL,		NULL,			0},
@@ -301,6 +302,7 @@
 			json_object_object_add(pramJobj, "X_ZYXEL_AX160_Channels", JSON_OBJ_COPY(json_object_object_get(obj2, "X_ZYXEL_AX160_Channels")));
 			json_object_object_add(pramJobj, "X_ZYXEL_Wireless_Mode", JSON_OBJ_COPY(json_object_object_get(obj2, "X_ZYXEL_Wireless_Mode")));
 			json_object_object_add(pramJobj, "X_ZYXEL_OBSS_coex", JSON_OBJ_COPY(json_object_object_get(obj2, "X_ZYXEL_OBSS_coex")));
+			json_object_object_add(pramJobj, "X_ZYXEL_InActiveChannels", JSON_OBJ_COPY(json_object_object_get(obj2, "X_ZYXEL_InActiveChannels")));
 			{
 			        char *band = json_object_get_string(json_object_object_get(obj2, "OperatingFrequencyBand"));;
 			        char *chList = json_object_get_string(json_object_object_get(obj2, "PossibleChannels"));
