Index: libzcfg_fe_dal-1.0/network/dns/zcfg_fe_dal_d_dns.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/dns/zcfg_fe_dal_d_dns.c	2019-07-18 12:27:43.322262610 +0800
+++ libzcfg_fe_dal-1.0/network/dns/zcfg_fe_dal_d_dns.c	2019-07-18 12:29:30.653008443 +0800
@@ -29,7 +29,7 @@
 	{"Offline", 			dalType_boolean,	0,	0,		NULL, NULL,		0},
 	{"UpdateURL", 		dalType_string, 	0, 	256,	NULL, NULL,		0},
 	{"ConnectionType", 	dalType_string, 	0, 	12,		NULL, "HTTP|HTTPS",		0},
-#ifdef CODETEL_CUSTOMIZATION
+#if (defined(CODETEL_CUSTOMIZATION) || defined(CLARO_CUSTOMIZATION))
 	{"Interface",                    dalType_string,         0,      32,    NULL, NULL,             0},
 #endif
 	{NULL, 					0, 					0, 	0, 		NULL, NULL,		0}
@@ -79,7 +79,7 @@
 		json_object_object_add(paramJobj, "AuthenticationResult", JSON_OBJ_COPY(json_object_object_get(obj, "AuthenticationResult")));
 		json_object_object_add(paramJobj, "UpdatedTime", JSON_OBJ_COPY(json_object_object_get(obj, "UpdatedTime")));
 		json_object_object_add(paramJobj, "DynamicIP", JSON_OBJ_COPY(json_object_object_get(obj, "DynamicIP")));
-#ifdef CODETEL_CUSTOMIZATION
+#if (defined(CODETEL_CUSTOMIZATION) || defined(CLARO_CUSTOMIZATION))
 		json_object_object_add(paramJobj, "Interface", JSON_OBJ_COPY(json_object_object_get(obj, "Interface")));
 #endif
 		json_object_array_add(Jarray, paramJobj);
@@ -102,7 +102,7 @@
 	int ipaddrpolicy = 0;
 	const char *updateurl = NULL;
 	const char *connectiontype = NULL;
-#ifdef CODETEL_CUSTOMIZATION
+#if (defined(CODETEL_CUSTOMIZATION) || defined(CLARO_CUSTOMIZATION))
 	const char *interface = NULL;
 #endif
 
@@ -120,7 +120,7 @@
 	offline = json_object_get_boolean(json_object_object_get(Jobj, "Offline"));
 	updateurl = json_object_get_string(json_object_object_get(Jobj, "UpdateURL"));
 	connectiontype = json_object_get_string(json_object_object_get(Jobj, "ConnectionType"));
-#ifdef CODETEL_CUSTOMIZATION
+#if (defined(CODETEL_CUSTOMIZATION) || defined(CLARO_CUSTOMIZATION))
 	interface = json_object_get_string(json_object_object_get(Jobj, "Interface"));
 #endif
 
@@ -144,7 +144,7 @@
 			json_object_object_add(ddnsJobj, "UpdateURL", json_object_new_string(updateurl));
 		if(json_object_object_get(Jobj, "ConnectionType"))
 			json_object_object_add(ddnsJobj, "ConnectionType", json_object_new_string(connectiontype));
-#ifdef CODETEL_CUSTOMIZATION
+#if (defined(CODETEL_CUSTOMIZATION) || defined(CLARO_CUSTOMIZATION))
 		if(json_object_object_get(Jobj, "Interface"))
 			json_object_object_add(ddnsJobj, "Interface", json_object_new_string(interface));
 #endif
