Index: radvd-1.8/radvd.c
===================================================================
--- radvd-1.8.orig/radvd.c	2019-10-25 13:56:08.865941485 +0800
+++ radvd-1.8/radvd.c	2019-10-25 13:57:16.342317724 +0800
@@ -980,7 +980,7 @@
 			FILE* pFile = NULL;
 			pFile = fopen("/tmp/radvd_oldprefix", "rb");
 			if(!pFile) {
-				system("echo $(date +%H:%M:%S)'  radvd.bin  set_prefix_to_old()  tmpradvd_oldprefix could not open' >> /tmp/dd");
+				system("echo $(date +%H:%M:%S)'  radvd.bin  set_prefix_to_old()  tmpradvd_oldprefix could not open' > /tmp/dd");
 			}else{
 				// obtain file size:
 				fseek (pFile , 0 , SEEK_END);
@@ -1026,7 +1026,7 @@
 			FILE* pFile = NULL;
 			pFile = fopen("/tmp/radvd_newprefix", "rb");
 			if(!pFile) {
-				system("echo $(date +%H:%M:%S)'  radvd.bin  set_prefix_to_new()  tmpradvd_newprefix could not open' >> /tmp/dd");
+				system("echo $(date +%H:%M:%S)'  radvd.bin  set_prefix_to_new()  tmpradvd_newprefix could not open' > /tmp/dd");
 			}else{
 				// obtain file size:
 				fseek (pFile , 0 , SEEK_END);
@@ -1073,7 +1073,7 @@
             FILE* fp = NULL;
             fp = fopen("/tmp/radvd_oldprefix", "wb");
             if(!fp) {
-                system("echo $(date +%H:%M:%S)'  radvd.bin  backup_old_prefix()  tmpradvd_oldprefix could not open' >> /tmp/dd");
+                system("echo $(date +%H:%M:%S)'  radvd.bin  backup_old_prefix()  tmpradvd_oldprefix could not open' > /tmp/dd");
             }else{
                 fwrite(buff, sizeof(unsigned char), sizeof(buff), fp);
                 fclose(fp);
@@ -1108,7 +1108,7 @@
             FILE* fp = NULL;
             fp = fopen("/tmp/radvd_newprefix", "wb");
             if(!fp) {
-                system("echo $(date +%H:%M:%S)'  radvd.bin  set_prefix_to_old()  tmpradvd_newprefix could not open' >> /tmp/dd");
+                system("echo $(date +%H:%M:%S)'  radvd.bin  set_prefix_to_old()  tmpradvd_newprefix could not open' > /tmp/dd");
             }else{
                 fwrite(buff, sizeof(unsigned char), sizeof(buff), fp);
                 fclose(fp);
@@ -1143,7 +1143,7 @@
     FILE* pFile = NULL;
     pFile = fopen("/tmp/radvd_newprefix", "rb");
     if(!pFile) {
-        system("echo $(date +%H:%M:%S)'  radvd.bin  is_diff()  tmpradvd_newprefix could not open' >> /tmp/dd");
+        system("echo $(date +%H:%M:%S)'  radvd.bin  is_diff()  tmpradvd_newprefix could not open' > /tmp/dd");
     }else{
         // obtain file size:
         fseek (pFile , 0 , SEEK_END);
@@ -1162,7 +1162,7 @@
     FILE* pFileold = NULL;
     pFileold = fopen("/tmp/radvd_oldprefix", "rb");
     if(!pFileold) {
-        system("echo $(date +%H:%M:%S)'  radvd.bin  is_diff()  tmpradvd_oldprefix could not open' >> /tmp/dd");
+        system("echo $(date +%H:%M:%S)'  radvd.bin  is_diff()  tmpradvd_oldprefix could not open' > /tmp/dd");
     }else{
         // obtain file size:
         fseek (pFileold , 0 , SEEK_END);
