Index: igmpproxy-0.2.1/src/request.c
===================================================================
--- igmpproxy-0.2.1.orig/src/request.c	2018-02-14 03:21:42.000000000 +0800
+++ igmpproxy-0.2.1/src/request.c	2019-11-27 10:47:40.186186513 +0800
@@ -80,8 +80,8 @@
     // We have a IF so check that it's an downstream IF.
     if(sourceVif->state == IF_STATE_DOWNSTREAM) {
 
-        my_log(LOG_DEBUG, 0, "Should insert group %s (from: %s) to route table. Vif Ix : %d",
-            inetFmt(group,s1), inetFmt(src,s2), sourceVif->index);
+        my_log(LOG_DEBUG, 0, "Should insert group %s (from: %s) to route table. Vif Ix : %d name: %s",
+            inetFmt(group,s1), inetFmt(src,s2), sourceVif->index, sourceVif->Name);
 
         // If we don't have a whitelist we insertRoute and done
         if(sourceVif->allowedgroups == NULL)
Index: igmpproxy-0.2.1/src/syslog.c
===================================================================
--- igmpproxy-0.2.1.orig/src/syslog.c	2018-02-14 03:21:42.000000000 +0800
+++ igmpproxy-0.2.1/src/syslog.c	2019-11-27 10:47:48.181452971 +0800
@@ -34,7 +34,7 @@
 
 #include "igmpproxy.h"
 
-int LogLevel = LOG_WARNING;
+int LogLevel = LOG_DEBUG;//LOG_WARNING;
 bool Log2Stderr = false;
 
 void my_log( int Severity, int Errno, const char *FmtSt, ... )
